<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<meta name="description" content="Package designed to aid in classifying cells from single-cell RNA sequencing data
    using external reference data (e.g., bulk RNA-seq, scRNA-seq, microarray, gene lists). A 
    variety of correlation based methods and gene list enrichment methods are provided to assist cell
    type assignment.">
<title>Classifier for Single-cell RNA-seq Using Cell Clusters â€¢ clustifyr</title>
<!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="favicon-16x16.png">
<link rel="icon" type="image/png" sizes="32x32" href="favicon-32x32.png">
<link rel="apple-touch-icon" type="image/png" sizes="180x180" href="apple-touch-icon.png">
<link rel="apple-touch-icon" type="image/png" sizes="120x120" href="apple-touch-icon-120x120.png">
<link rel="apple-touch-icon" type="image/png" sizes="76x76" href="apple-touch-icon-76x76.png">
<link rel="apple-touch-icon" type="image/png" sizes="60x60" href="apple-touch-icon-60x60.png">
<script src="deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="deps/bootstrap-5.1.3/bootstrap.min.css" rel="stylesheet">
<script src="deps/bootstrap-5.1.3/bootstrap.bundle.min.js"></script><link href="deps/Source_Sans_Pro-0.4.2/font.css" rel="stylesheet">
<link href="deps/Source_Code_Pro-0.4.2/font.css" rel="stylesheet">
<!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- bootstrap-toc --><script src="https://cdn.rawgit.com/afeld/bootstrap-toc/v1.0.1/dist/bootstrap-toc.min.js"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="pkgdown.js"></script><meta property="og:title" content="Classifier for Single-cell RNA-seq Using Cell Clusters">
<meta property="og:description" content="Package designed to aid in classifying cells from single-cell RNA sequencing data
    using external reference data (e.g., bulk RNA-seq, scRNA-seq, microarray, gene lists). A 
    variety of correlation based methods and gene list enrichment methods are provided to assist cell
    type assignment.">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--><script defer data-domain="rnabioco.github.io" src="https://plausible.io/js/plausible.js"></script>
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-dark navbar-expand-lg bg-dark"><div class="container">
    
    <a class="navbar-brand me-2" href="index.html">clustifyr</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="Released version">1.11.0</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item">
  <a class="nav-link" href="articles/clustifyr.html">Get started</a>
</li>
<li class="nav-item">
  <a class="nav-link" href="reference/index.html">Reference</a>
</li>
<li class="nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-articles">Articles</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-articles">
    <a class="dropdown-item" href="articles/geo-annotations.html">Improving NCBI GEO submissions of scRNA-seq data</a>
  </div>
</li>
      </ul>
<form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="search.json" id="search-input" placeholder="Search for" autocomplete="off">
</form>

      <ul class="navbar-nav">
<li class="nav-item">
  <a class="external-link nav-link" href="https://github.com/rnabioco/clustifyr/" aria-label="github">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>

    
  </div>
</nav><div class="container template-home">
<div class="row">
  <main id="main" class="col-md-9"><div class="section level1">
<div class="page-header"><h1 id="clustifyr">clustifyr<a class="anchor" aria-label="anchor" href="#clustifyr"></a>
</h1></div>
<!-- badges: start -->

<!-- badges: end -->
<p>clustifyr classifies cells and clusters in single-cell RNA sequencing experiments using reference bulk RNA-seq data sets, sorted microarray expression data, single-cell gene signatures, or lists of marker genes.</p>
<div class="section level2">
<h2 id="installation">Installation<a class="anchor" aria-label="anchor" href="#installation"></a>
</h2>
<p>Install the Bioconductor version with:</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw">if</span> <span class="op">(</span><span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/ns-load.html" class="external-link">requireNamespace</a></span><span class="op">(</span><span class="st">"BiocManager"</span>, quietly <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span><span class="op">)</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html" class="external-link">install.packages</a></span><span class="op">(</span><span class="st">"BiocManager"</span><span class="op">)</span></span>
<span></span>
<span><span class="fu">BiocManager</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/BiocManager/man/install.html" class="external-link">install</a></span><span class="op">(</span><span class="st">"clustifyr"</span><span class="op">)</span></span></code></pre></div>
<p>Install the development version with:</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># install.packages("remotes")</span></span>
<span><span class="fu">remotes</span><span class="fu">::</span><span class="fu"><a href="https://remotes.r-lib.org/reference/install_github.html" class="external-link">install_github</a></span><span class="op">(</span><span class="st">"rnabioco/clustifyr"</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="example-usage">Example usage<a class="anchor" aria-label="anchor" href="#example-usage"></a>
</h2>
<p>In this example we use the following built-in input data:</p>
<ul>
<li>an expression matrix of single cell RNA-seq data (<code>pbmc_matrix_small</code>)</li>
<li>a metadata data.frame (<code>pbmc_meta</code>), with cluster information stored (<code>"classified"</code>)</li>
<li>a vector of variable genes (<code>pbmc_vargenes</code>)</li>
<li>a matrix of mean normalized scRNA-seq UMI counts by cell type (<code>cbmc_ref</code>)</li>
</ul>
<p>We then calculate correlation coefficients and plot them on a pre-calculated projection (stored in <code>pbmc_meta</code>).</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/rnabioco/clustifyr" class="external-link">clustifyr</a></span><span class="op">)</span></span>
<span></span>
<span><span class="co"># calculate correlation</span></span>
<span><span class="va">res</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/clustify.html">clustify</a></span><span class="op">(</span></span>
<span>  input <span class="op">=</span> <span class="va">pbmc_matrix_small</span>,</span>
<span>  metadata <span class="op">=</span> <span class="va">pbmc_meta</span><span class="op">$</span><span class="va">classified</span>,</span>
<span>  ref_mat <span class="op">=</span> <span class="va">cbmc_ref</span>,</span>
<span>  query_genes <span class="op">=</span> <span class="va">pbmc_vargenes</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="co"># print assignments</span></span>
<span><span class="fu"><a href="reference/cor_to_call.html">cor_to_call</a></span><span class="op">(</span><span class="va">res</span><span class="op">)</span></span>
<span><span class="co">#&gt; # A tibble: 9 Ã— 3</span></span>
<span><span class="co">#&gt; # Groups:   cluster [9]</span></span>
<span><span class="co">#&gt;   cluster      type           r</span></span>
<span><span class="co">#&gt;   &lt;chr&gt;        &lt;chr&gt;      &lt;dbl&gt;</span></span>
<span><span class="co">#&gt; 1 B            B          0.909</span></span>
<span><span class="co">#&gt; 2 CD14+ Mono   CD14+ Mono 0.915</span></span>
<span><span class="co">#&gt; 3 FCGR3A+ Mono CD16+ Mono 0.929</span></span>
<span><span class="co">#&gt; 4 Memory CD4 T CD4 T      0.861</span></span>
<span><span class="co">#&gt; 5 Naive CD4 T  CD4 T      0.889</span></span>
<span><span class="co">#&gt; 6 DC           DC         0.849</span></span>
<span><span class="co">#&gt; 7 Platelet     Mk         0.732</span></span>
<span><span class="co">#&gt; 8 CD8 T        NK         0.826</span></span>
<span><span class="co">#&gt; 9 NK           NK         0.894</span></span>
<span></span>
<span><span class="co"># plot assignments on a projection</span></span>
<span><span class="fu"><a href="reference/plot_best_call.html">plot_best_call</a></span><span class="op">(</span></span>
<span>  cor_mat <span class="op">=</span> <span class="va">res</span>,</span>
<span>  metadata <span class="op">=</span> <span class="va">pbmc_meta</span>,</span>
<span>  cluster_col <span class="op">=</span> <span class="st">"classified"</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p><img src="reference/figures/readme_example-1.png"><!-- --></p>
<p><code><a href="reference/clustify.html">clustify()</a></code> can take a clustered <code>SingleCellExperiment</code> or <code>seurat</code> object (both v2 and v3) and assign identities.</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># for SingleCellExperiment</span></span>
<span><span class="fu"><a href="reference/clustify.html">clustify</a></span><span class="op">(</span></span>
<span>  input <span class="op">=</span> <span class="va">sce_small</span>,          <span class="co"># an SCE object</span></span>
<span>  ref_mat <span class="op">=</span> <span class="va">cbmc_ref</span>,         <span class="co"># matrix of RNA-seq expression data for each cell type</span></span>
<span>  cluster_col <span class="op">=</span> <span class="st">"cell_type1"</span>, <span class="co"># name of column in meta.data containing cell clusters</span></span>
<span>  obj_out <span class="op">=</span> <span class="cn">TRUE</span>              <span class="co"># output SCE object with cell type inserted as "type" column</span></span>
<span><span class="op">)</span> </span>
<span><span class="co">#&gt; class: SingleCellExperiment </span></span>
<span><span class="co">#&gt; dim: 200 200 </span></span>
<span><span class="co">#&gt; metadata(0):</span></span>
<span><span class="co">#&gt; assays(2): counts logcounts</span></span>
<span><span class="co">#&gt; rownames(200): SGIP1 AZIN2 ... TAF12 SNHG3</span></span>
<span><span class="co">#&gt; rowData names(10): feature_symbol is_feature_control ... total_counts</span></span>
<span><span class="co">#&gt;   log10_total_counts</span></span>
<span><span class="co">#&gt; colnames(200): AZ_A1 AZ_A10 ... HP1502401_E18 HP1502401_E19</span></span>
<span><span class="co">#&gt; colData names(35): cell_quality cell_type1 ... type r</span></span>
<span><span class="co">#&gt; reducedDimNames(0):</span></span>
<span><span class="co">#&gt; mainExpName: NULL</span></span>
<span><span class="co">#&gt; altExpNames(0):</span></span>
<span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://satijalab.org/seurat" class="external-link">Seurat</a></span><span class="op">)</span></span>
<span><span class="co"># for Seurat3/4</span></span>
<span><span class="fu"><a href="reference/clustify.html">clustify</a></span><span class="op">(</span></span>
<span>  input <span class="op">=</span> <span class="va">s_small3</span>,</span>
<span>  cluster_col <span class="op">=</span> <span class="st">"RNA_snn_res.1"</span>,</span>
<span>  ref_mat <span class="op">=</span> <span class="va">cbmc_ref</span>,</span>
<span>  seurat_out <span class="op">=</span> <span class="cn">TRUE</span></span>
<span><span class="op">)</span></span>
<span><span class="co">#&gt; An object of class Seurat </span></span>
<span><span class="co">#&gt; 230 features across 80 samples within 1 assay </span></span>
<span><span class="co">#&gt; Active assay: RNA (230 features, 20 variable features)</span></span>
<span><span class="co">#&gt;  2 dimensional reductions calculated: pca, tsne</span></span>
<span></span>
<span><span class="co"># New output option, directly as a vector (in the order of the metadata), which can then be inserted into metadata dataframes and other workflows</span></span>
<span><span class="fu"><a href="reference/clustify.html">clustify</a></span><span class="op">(</span></span>
<span>  input <span class="op">=</span> <span class="va">s_small3</span>,</span>
<span>  cluster_col <span class="op">=</span> <span class="st">"RNA_snn_res.1"</span>,</span>
<span>  ref_mat <span class="op">=</span> <span class="va">cbmc_ref</span>,</span>
<span>  vec_out <span class="op">=</span> <span class="cn">TRUE</span></span>
<span><span class="op">)</span></span>
<span><span class="co">#&gt;  [1] "Mk"         "Mk"         "Mk"         "Mk"         "Mk"        </span></span>
<span><span class="co">#&gt;  [6] "Mk"         "Mk"         "Mk"         "Mk"         "Mk"        </span></span>
<span><span class="co">#&gt; [11] "B"          "B"          "B"          "B"          "B"         </span></span>
<span><span class="co">#&gt; [16] "B"          "B"          "B"          "B"          "B"         </span></span>
<span><span class="co">#&gt; [21] "CD16+ Mono" "CD16+ Mono" "CD16+ Mono" "CD16+ Mono" "CD16+ Mono"</span></span>
<span><span class="co">#&gt; [26] "CD16+ Mono" "CD16+ Mono" "CD16+ Mono" "CD16+ Mono" "CD16+ Mono"</span></span>
<span><span class="co">#&gt; [31] "Mk"         "B"          "Mk"         "Mk"         "Mk"        </span></span>
<span><span class="co">#&gt; [36] "Mk"         "Mk"         "Mk"         "Mk"         "Mk"        </span></span>
<span><span class="co">#&gt; [41] "Mk"         "B"          "Mk"         "Mk"         "B"         </span></span>
<span><span class="co">#&gt; [46] "B"          "Mk"         "Mk"         "Mk"         "Mk"        </span></span>
<span><span class="co">#&gt; [51] "CD16+ Mono" "CD16+ Mono" "B"          "CD16+ Mono" "CD16+ Mono"</span></span>
<span><span class="co">#&gt; [56] "CD16+ Mono" "CD16+ Mono" "CD16+ Mono" "CD16+ Mono" "Mk"        </span></span>
<span><span class="co">#&gt; [61] "B"          "CD16+ Mono" "B"          "CD16+ Mono" "B"         </span></span>
<span><span class="co">#&gt; [66] "CD16+ Mono" "CD16+ Mono" "CD16+ Mono" "CD16+ Mono" "B"         </span></span>
<span><span class="co">#&gt; [71] "Mk"         "Mk"         "Mk"         "Mk"         "Mk"        </span></span>
<span><span class="co">#&gt; [76] "Mk"         "Mk"         "Mk"         "Mk"         "CD16+ Mono"</span></span></code></pre></div>
<p>New reference matrix can be made directly from <code>SingleCellExperiment</code> and <code>Seurat</code> objects as well. Other scRNAseq experiment object types are supported as well.</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># make reference from SingleCellExperiment objects</span></span>
<span><span class="va">sce_ref</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/object_ref.html">object_ref</a></span><span class="op">(</span></span>
<span>  input <span class="op">=</span> <span class="va">sce_small</span>,               <span class="co"># SCE object</span></span>
<span>  cluster_col <span class="op">=</span> <span class="st">"cell_type1"</span>       <span class="co"># name of column in colData containing cell identities</span></span>
<span><span class="op">)</span></span>
<span><span class="co">#&gt; The following clusters have less than 10 cells for this analysis: co-expression, ductal, endothelial, epsilon, MHC class II, PSC. Classification is likely inaccurate.</span></span>
<span></span>
<span><span class="co"># make reference from seurat objects</span></span>
<span><span class="va">s_ref</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/seurat_ref.html">seurat_ref</a></span><span class="op">(</span></span>
<span>  seurat_object <span class="op">=</span> <span class="va">s_small3</span>,</span>
<span>  cluster_col <span class="op">=</span> <span class="st">"RNA_snn_res.1"</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">s_ref</span><span class="op">)</span></span>
<span><span class="co">#&gt;                 0        1        2</span></span>
<span><span class="co">#&gt; MS4A1    0.000000 1.126047 5.151065</span></span>
<span><span class="co">#&gt; CD79B    2.469341 2.920407 5.031316</span></span>
<span><span class="co">#&gt; CD79A    0.000000 2.535151 5.375681</span></span>
<span><span class="co">#&gt; HLA-DRA  3.640368 6.008446 7.055386</span></span>
<span><span class="co">#&gt; TCL1A    0.000000 1.495867 4.963367</span></span>
<span><span class="co">#&gt; HLA-DQB1 1.603068 3.836290 5.137422</span></span></code></pre></div>
<p><code><a href="reference/clustify_lists.html">clustify_lists()</a></code> handles identity assignment of matrix or <code>SingleCellExperiment</code> and <code>seurat</code> objects based on marker gene lists.</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="reference/clustify_lists.html">clustify_lists</a></span><span class="op">(</span></span>
<span>  input <span class="op">=</span> <span class="va">pbmc_matrix_small</span>,</span>
<span>  metadata <span class="op">=</span> <span class="va">pbmc_meta</span>,</span>
<span>  cluster_col <span class="op">=</span> <span class="st">"classified"</span>,</span>
<span>  marker <span class="op">=</span> <span class="va">pbmc_markers</span>,</span>
<span>  marker_inmatrix <span class="op">=</span> <span class="cn">FALSE</span></span>
<span><span class="op">)</span></span>
<span><span class="co">#&gt;                      0        1        2         3         4        5        6</span></span>
<span><span class="co">#&gt; Naive CD4 T  1.5639055 20.19469 31.77095  8.664074 23.844992 19.06931 19.06931</span></span>
<span><span class="co">#&gt; Memory CD4 T 1.5639055 20.19469 31.77095 10.568007 23.844992 17.97875 19.06931</span></span>
<span><span class="co">#&gt; CD14+ Mono   0.9575077 14.70716 76.21353 17.899569 11.687739 49.86699 16.83210</span></span>
<span><span class="co">#&gt; B            0.6564777 12.70976 31.77095 26.422929 13.536295 20.19469 13.53630</span></span>
<span><span class="co">#&gt; CD8 T        1.0785353 17.97875 31.82210 12.584823 31.822099 22.71234 40.45383</span></span>
<span><span class="co">#&gt; FCGR3A+ Mono 0.6564777 13.63321 72.43684 17.899569  9.726346 56.48245 14.61025</span></span>
<span><span class="co">#&gt; NK           0.6564777 14.61025 31.82210  7.757206 31.822099 22.71234 45.05072</span></span>
<span><span class="co">#&gt; DC           0.6564777 15.80598 63.34978 19.069308 13.758144 40.56298 17.97875</span></span>
<span><span class="co">#&gt; Platelet     0.5428889 13.34769 59.94938 14.215244 15.158755 46.92861 19.49246</span></span>
<span><span class="co">#&gt;                      7          8</span></span>
<span><span class="co">#&gt; Naive CD4 T   6.165348  0.6055118</span></span>
<span><span class="co">#&gt; Memory CD4 T  6.165348  0.9575077</span></span>
<span><span class="co">#&gt; CD14+ Mono   25.181595  1.0785353</span></span>
<span><span class="co">#&gt; B            17.899569  0.1401901</span></span>
<span><span class="co">#&gt; CD8 T         7.882145  0.3309153</span></span>
<span><span class="co">#&gt; FCGR3A+ Mono 21.409177  0.3309153</span></span>
<span><span class="co">#&gt; NK            5.358651  0.3309153</span></span>
<span><span class="co">#&gt; DC           45.101877  0.1401901</span></span>
<span><span class="co">#&gt; Platelet     19.492465 59.9493793</span></span>
<span></span>
<span><span class="fu"><a href="reference/clustify_lists.html">clustify_lists</a></span><span class="op">(</span></span>
<span>  input <span class="op">=</span> <span class="va">s_small3</span>,</span>
<span>  marker <span class="op">=</span> <span class="va">pbmc_markers</span>,</span>
<span>  marker_inmatrix <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  cluster_col <span class="op">=</span> <span class="st">"RNA_snn_res.1"</span>,</span>
<span>  seurat_out <span class="op">=</span> <span class="cn">TRUE</span></span>
<span><span class="op">)</span></span>
<span><span class="co">#&gt; An object of class Seurat </span></span>
<span><span class="co">#&gt; 230 features across 80 samples within 1 assay </span></span>
<span><span class="co">#&gt; Active assay: RNA (230 features, 20 variable features)</span></span>
<span><span class="co">#&gt;  2 dimensional reductions calculated: pca, tsne</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="additional-resources">Additional resources<a class="anchor" aria-label="anchor" href="#additional-resources"></a>
</h2>
<ul>
<li><p><a href="https://github.com/rnabioco/clustifyrdata/blob/master/inst/run_clustifyr.R" class="external-link">Script</a> for benchmarking, compatible with <a href="https://github.com/tabdelaal/scRNAseq_Benchmark" class="external-link"><code>scRNAseq_Benchmark</code></a></p></li>
<li><p>Additional reference data (including tabula muris, immgen, etc) are available in a supplemental package <a href="https://github.com/rnabioco/clustifyrdatahub" class="external-link"><code>clustifyrdatahub</code></a>. Also see <a href="https://rnabioco.github.io/clustifyrdata/articles/download_refs.html" class="external-link">list</a> for individual downloads.</p></li>
<li><p>See the <a href="https://github.com/rnabioco/clustifyr/wiki/Frequently-asked-questions" class="external-link">FAQ</a> for more details.</p></li>
</ul>
</div>
</div>
  </main><aside class="col-md-3"><div class="links">
<h2 data-toc-skip>Links</h2>
<ul class="list-unstyled">
<li><a href="https://www.bioconductor.org/packages/clustifyr" class="external-link">View on Bioconductor</a></li>
<li><a href="https://github.com/rnabioco/clustifyr/" class="external-link">Browse source code</a></li>
<li><a href="https://github.com/rnabioco/clustifyr/issues" class="external-link">Report a bug</a></li>
<li><a href="http://github.com/rnabioco/clustifyrdata" class="external-link">External data sets</a></li>
</ul>
</div>

<div class="license">
<h2 data-toc-skip>License</h2>
<ul class="list-unstyled">
<li>
<a href="https://opensource.org/licenses/mit-license.php" class="external-link">MIT</a> + file <a href="LICENSE-text.html">LICENSE</a>
</li>
</ul>
</div>

<div class="community">
<h2 data-toc-skip>Community</h2>
<ul class="list-unstyled">
<li><a href="CONTRIBUTING.html">Contributing guide</a></li>
<li><a href="CODE_OF_CONDUCT.html">Code of conduct</a></li>
</ul>
</div>

<div class="citation">
<h2 data-toc-skip>Citation</h2>
<ul class="list-unstyled">
<li><a href="authors.html#citation">Citing clustifyr</a></li>
</ul>
</div>

<div class="developers">
<h2 data-toc-skip>Developers</h2>
<ul class="list-unstyled">
<li>
<a href="https://github.com/raysinensis" class="external-link">Rui Fu</a> <br><small class="roles"> Author </small>  </li>
<li>
<a href="http://github.com/kriemo" class="external-link">Kent Riemondy</a> <br><small class="roles"> Maintainer, author </small>  </li>
<li>
<a href="https://medschool.cuanschutz.edu/rbi" class="external-link"><img src="https://rnabioco.github.io/rbitemplate/rbi-pkgdown-logo.svg" alt="RBI" width="150"></a> <br><small class="roles"> Funder, copyright holder </small>  </li>
<li><a href="authors.html">More about authors...</a></li>
</ul>
</div>



  </aside>
</div>


    <footer><div class="pkgdown-footer-left">
  <p></p>
<p>An R package developed by the <a href="https://medschool.cuanschutz.edu/rbi" class="external-link">RNA Bioscience Initiative</a></p>
</div>

<div class="pkgdown-footer-right">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.6.</p>
</div>

    </footer>
</div>

  

  

  </body>
</html>
