<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Classifier for Single-cell RNA-seq Using Cell Clusters • clustifyr</title>
<!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="favicon-16x16.png">
<link rel="icon" type="image/png" sizes="32x32" href="favicon-32x32.png">
<link rel="apple-touch-icon" type="image/png" sizes="180x180" href="apple-touch-icon.png">
<link rel="apple-touch-icon" type="image/png" sizes="120x120" href="apple-touch-icon-120x120.png">
<link rel="apple-touch-icon" type="image/png" sizes="76x76" href="apple-touch-icon-76x76.png">
<link rel="apple-touch-icon" type="image/png" sizes="60x60" href="apple-touch-icon-60x60.png">
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="bootstrap-toc.css">
<script src="bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="pkgdown.css" rel="stylesheet">
<script src="pkgdown.js"></script><meta property="og:title" content="Classifier for Single-cell RNA-seq Using Cell Clusters">
<meta property="og:description" content="Package designed to aid in classifying cells from single-cell RNA sequencing data
    using external reference data (e.g., bulk RNA-seq, scRNA-seq, microarray, gene lists). A 
    variety of correlation based methods and gene list enrichment methods are provided to assist cell
    type assignment.">
<meta property="og:image" content="http://rnabioco.github.io/clustifyr/logo.svg">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    <div class="container template-home">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="index.html">clustifyr</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">1.5.1</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="index.html">
    <span class="fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="articles/clustifyR.html">Introduction to clustifyr</a>
    </li>
    <li>
      <a href="articles/geo-annotations.html">Improving NCBI GEO submissions of scRNA-seq data</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/rnabioco/clustifyr/">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="contents col-md-9">
<div id="clustifyr-" class="section level1">
<div class="page-header"><h1 class="hasAnchor">
<a href="#clustifyr-" class="anchor"></a>clustifyr <img src="reference/figures/logo.png" align="right">
</h1></div>
<!-- badges: start -->

<!-- badges: end -->
<ul>
<li>
<a href="#clustifyr-">clustifyr</a>
<ul>
<li><a href="#installation">Installation</a></li>
<li><a href="#additional-info">Additional info</a></li>
<li><a href="#example-usage">Example usage</a></li>
<li><a href="#frequently-asked-questions">Frequently Asked Questions</a></li>
<li><a href="#code-of-conduct">Code of Conduct</a></li>
</ul>
</li>
</ul>
<p>clustifyr classifies cells and clusters in single-cell RNA sequencing experiments using reference bulk RNA-seq data sets, sorted microarray expression data, single-cell gene signatures, or lists of marker genes.</p>
<p>Single cell transcriptomes are difficult to annotate without knowledge of the underlying biology. Even with this knowledge, accurate identification can be challenging due to the lack of detectable expression of common marker genes. clustifyr solves this problem by automatically annotating single cells or clusters of cells using single-cell RNA-seq, bulk RNA-seq data, microarray, or marker gene lists. Additional functions enable exploratory analysis of similarities between single cell RNA-seq datasets and reference data.</p>
<div id="installation" class="section level2">
<h2 class="hasAnchor">
<a href="#installation" class="anchor"></a>Installation</h2>
<p>Install the Bioconductor version with:</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw">if</span> <span class="op">(</span><span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/ns-load.html">requireNamespace</a></span><span class="op">(</span><span class="st">"BiocManager"</span>, quietly <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span><span class="op">)</span>
    <span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html">install.packages</a></span><span class="op">(</span><span class="st">"BiocManager"</span><span class="op">)</span>

<span class="fu">BiocManager</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/BiocManager/man/install.html">install</a></span><span class="op">(</span><span class="st">"clustifyr"</span><span class="op">)</span></code></pre></div>
<p>Install the development version with:</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># install.packages("remotes")</span>
<span class="fu">remotes</span><span class="fu">::</span><span class="fu"><a href="https://remotes.r-lib.org/reference/install_github.html">install_github</a></span><span class="op">(</span><span class="st">"rnabioco/clustifyr"</span><span class="op">)</span></code></pre></div>
</div>
<div id="additional-info" class="section level2">
<h2 class="hasAnchor">
<a href="#additional-info" class="anchor"></a>Additional info</h2>
<p><a href="https://rnabioco.github.io/clustifyr/articles/clustifyR.html">Intro</a> tutorial</p>
<p><a href="https://rnabioco.github.io/clustifyrdata/articles/otherformats.html">Additional</a> tutorials</p>
<p><a href="https://github.com/rnabioco/clustifyrdata/blob/master/inst/run_clustifyr.R">Script</a> for benchmarking, compatible with <a href="https://github.com/tabdelaal/scRNAseq_Benchmark"><code>scRNAseq_Benchmark</code></a></p>
<p>More reference data (including tabula muris, immgen, etc) are available at supplement package <a href="https://github.com/rnabioco/clustifyrdatahub"><code>clustifyrdatahub</code></a>. Also see <a href="https://rnabioco.github.io/clustifyrdata/articles/download_refs.html">list</a> for individual downloads.</p>
<p><a href="https://f1000research.com/articles/9-223/v2">Publication</a> with parameter and usage discussions has passed peer review on F1000Research.</p>
</div>
<div id="example-usage" class="section level2">
<h2 class="hasAnchor">
<a href="#example-usage" class="anchor"></a>Example usage</h2>
<p>In this example we use the following built-in input data:</p>
<ul>
<li>an expression matrix of single cell RNA-seq data (<code>pbmc_matrix_small</code>)</li>
<li>a metadata data.frame (<code>pbmc_meta</code>), with cluster information stored (<code>"classified"</code>)</li>
<li>a vector of variable genes (<code>pbmc_vargenes</code>)</li>
<li>a matrix of mean normalized scRNA-seq UMI counts by cell type (<code>cbmc_ref</code>):</li>
</ul>
<p>We then calculate correlation coefficients and plot them on a pre-calculated projection (stored in <code>pbmc_meta</code>).</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="http://github.com/rnabioco/clustifyr#readme">clustifyr</a></span><span class="op">)</span>

<span class="co"># calculate correlation</span>
<span class="va">res</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/clustify.html">clustify</a></span><span class="op">(</span>
  input <span class="op">=</span> <span class="va">pbmc_matrix_small</span>,
  metadata <span class="op">=</span> <span class="va">pbmc_meta</span><span class="op">$</span><span class="va">classified</span>,
  ref_mat <span class="op">=</span> <span class="va">cbmc_ref</span>,
  query_genes <span class="op">=</span> <span class="va">pbmc_vargenes</span>
<span class="op">)</span>

<span class="co"># print assignments</span>
<span class="fu"><a href="reference/cor_to_call.html">cor_to_call</a></span><span class="op">(</span><span class="va">res</span><span class="op">)</span>
<span class="co">#&gt; # A tibble: 9 x 3</span>
<span class="co">#&gt; # Groups:   cluster [9]</span>
<span class="co">#&gt;   cluster      type           r</span>
<span class="co">#&gt;   &lt;chr&gt;        &lt;chr&gt;      &lt;dbl&gt;</span>
<span class="co">#&gt; 1 B            B          0.909</span>
<span class="co">#&gt; 2 CD14+ Mono   CD14+ Mono 0.915</span>
<span class="co">#&gt; 3 FCGR3A+ Mono CD16+ Mono 0.929</span>
<span class="co">#&gt; 4 Memory CD4 T CD4 T      0.861</span>
<span class="co">#&gt; 5 Naive CD4 T  CD4 T      0.889</span>
<span class="co">#&gt; 6 DC           DC         0.849</span>
<span class="co">#&gt; 7 Platelet     Mk         0.732</span>
<span class="co">#&gt; 8 CD8 T        NK         0.826</span>
<span class="co">#&gt; 9 NK           NK         0.894</span>

<span class="co"># plot assignments on a projection</span>
<span class="fu"><a href="reference/plot_best_call.html">plot_best_call</a></span><span class="op">(</span>
  cor_mat <span class="op">=</span> <span class="va">res</span>,
  metadata <span class="op">=</span> <span class="va">pbmc_meta</span>,
  cluster_col <span class="op">=</span> <span class="st">"classified"</span>
<span class="op">)</span></code></pre></div>
<p><img src="reference/figures/readme_example-1.png"><!-- --></p>
<p><code><a href="reference/clustify.html">clustify()</a></code> can also take a clustered <code>SingleCellExperiment</code> or <code>seurat</code> object (both v2 and v3) and assign identities.</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># for SingleCellExperiment</span>
<span class="fu"><a href="reference/clustify.html">clustify</a></span><span class="op">(</span>
  input <span class="op">=</span> <span class="va">sce_small</span>,          <span class="co"># an SCE object</span>
  ref_mat <span class="op">=</span> <span class="va">cbmc_ref</span>,         <span class="co"># matrix of RNA-seq expression data for each cell type</span>
  cluster_col <span class="op">=</span> <span class="st">"cell_type1"</span>, <span class="co"># name of column in meta.data containing cell clusters</span>
  obj_out <span class="op">=</span> <span class="cn">TRUE</span>              <span class="co"># output SCE object with cell type inserted as "type" column</span>
<span class="op">)</span> 
<span class="co">#&gt; class: SingleCellExperiment </span>
<span class="co">#&gt; dim: 200 200 </span>
<span class="co">#&gt; metadata(0):</span>
<span class="co">#&gt; assays(2): counts logcounts</span>
<span class="co">#&gt; rownames(200): SGIP1 AZIN2 ... TAF12 SNHG3</span>
<span class="co">#&gt; rowData names(10): feature_symbol is_feature_control ... total_counts</span>
<span class="co">#&gt;   log10_total_counts</span>
<span class="co">#&gt; colnames(200): AZ_A1 AZ_A10 ... HP1502401_E18 HP1502401_E19</span>
<span class="co">#&gt; colData names(35): cell_quality cell_type1 ... type r</span>
<span class="co">#&gt; reducedDimNames(0):</span>
<span class="co">#&gt; mainExpName: NULL</span>
<span class="co">#&gt; altExpNames(0):</span>

<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va">Seurat</span><span class="op">)</span>
<span class="co"># for Seurat3/4</span>
<span class="fu"><a href="reference/clustify.html">clustify</a></span><span class="op">(</span>
  input <span class="op">=</span> <span class="va">s_small3</span>,
  cluster_col <span class="op">=</span> <span class="st">"RNA_snn_res.1"</span>,
  ref_mat <span class="op">=</span> <span class="va">cbmc_ref</span>,
  seurat_out <span class="op">=</span> <span class="cn">TRUE</span>
<span class="op">)</span>
<span class="co">#&gt; An object of class Seurat </span>
<span class="co">#&gt; 230 features across 80 samples within 1 assay </span>
<span class="co">#&gt; Active assay: RNA (230 features, 20 variable features)</span>
<span class="co">#&gt;  2 dimensional reductions calculated: pca, tsne</span></code></pre></div>
<p>New reference matrix can be made directly from <code>SingleCellExperiment</code> and <code>seurat</code> objects as well. Other scRNAseq experiment object types are supported as well.</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># make reference from SingleCellExperiment objects</span>
<span class="va">sce_ref</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/object_ref.html">object_ref</a></span><span class="op">(</span>
  input <span class="op">=</span> <span class="va">sce_small</span>,               <span class="co"># SCE object</span>
  cluster_col <span class="op">=</span> <span class="st">"cell_type1"</span>       <span class="co"># name of column in colData containing cell identities</span>
<span class="op">)</span>
<span class="co">#&gt; The following clusters have less than 10 cells for this analysis: co-expression, ductal, endothelial, epsilon, MHC class II, PSC. Classification is likely inaccurate.</span>

<span class="co"># make reference from seurat objects</span>
<span class="va">s_ref</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/seurat_ref.html">seurat_ref</a></span><span class="op">(</span>
  seurat_object <span class="op">=</span> <span class="va">s_small3</span>,
  cluster_col <span class="op">=</span> <span class="st">"RNA_snn_res.1"</span>
<span class="op">)</span>

<span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span><span class="op">(</span><span class="va">s_ref</span><span class="op">)</span>
<span class="co">#&gt;                 0        1        2</span>
<span class="co">#&gt; MS4A1    0.000000 1.126047 5.151065</span>
<span class="co">#&gt; CD79B    2.469341 2.920407 5.031316</span>
<span class="co">#&gt; CD79A    0.000000 2.535151 5.375681</span>
<span class="co">#&gt; HLA-DRA  3.640368 6.008446 7.055386</span>
<span class="co">#&gt; TCL1A    0.000000 1.495867 4.963367</span>
<span class="co">#&gt; HLA-DQB1 1.603068 3.836290 5.137422</span></code></pre></div>
<p><code><a href="reference/clustify_lists.html">clustify_lists()</a></code> handles identity assignment of matrix or <code>SingleCellExperiment</code> and <code>seurat</code> objects based on marker gene lists.</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="reference/clustify_lists.html">clustify_lists</a></span><span class="op">(</span>
  input <span class="op">=</span> <span class="va">pbmc_matrix_small</span>,
  metadata <span class="op">=</span> <span class="va">pbmc_meta</span>,
  cluster_col <span class="op">=</span> <span class="st">"classified"</span>,
  marker <span class="op">=</span> <span class="va">pbmc_markers</span>,
  marker_inmatrix <span class="op">=</span> <span class="cn">FALSE</span>
<span class="op">)</span>
<span class="co">#&gt;                      0        1        2         3         4        5        6</span>
<span class="co">#&gt; Naive CD4 T  1.5639055 20.19469 31.77095  8.664074 23.844992 19.06931 19.06931</span>
<span class="co">#&gt; Memory CD4 T 1.5639055 20.19469 31.77095 10.568007 23.844992 17.97875 19.06931</span>
<span class="co">#&gt; CD14+ Mono   0.9575077 14.70716 76.21353 17.899569 11.687739 49.86699 16.83210</span>
<span class="co">#&gt; B            0.6564777 12.70976 31.77095 26.422929 13.536295 20.19469 13.53630</span>
<span class="co">#&gt; CD8 T        1.0785353 17.97875 31.82210 12.584823 31.822099 22.71234 40.45383</span>
<span class="co">#&gt; FCGR3A+ Mono 0.6564777 13.63321 72.43684 17.899569  9.726346 56.48245 14.61025</span>
<span class="co">#&gt; NK           0.6564777 14.61025 31.82210  7.757206 31.822099 22.71234 45.05072</span>
<span class="co">#&gt; DC           0.6564777 15.80598 63.34978 19.069308 13.758144 40.56298 17.97875</span>
<span class="co">#&gt; Platelet     0.5428889 13.34769 59.94938 14.215244 15.158755 46.92861 19.49246</span>
<span class="co">#&gt;                      7          8</span>
<span class="co">#&gt; Naive CD4 T   6.165348  0.6055118</span>
<span class="co">#&gt; Memory CD4 T  6.165348  0.9575077</span>
<span class="co">#&gt; CD14+ Mono   25.181595  1.0785353</span>
<span class="co">#&gt; B            17.899569  0.1401901</span>
<span class="co">#&gt; CD8 T         7.882145  0.3309153</span>
<span class="co">#&gt; FCGR3A+ Mono 21.409177  0.3309153</span>
<span class="co">#&gt; NK            5.358651  0.3309153</span>
<span class="co">#&gt; DC           45.101877  0.1401901</span>
<span class="co">#&gt; Platelet     19.492465 59.9493793</span>

<span class="fu"><a href="reference/clustify_lists.html">clustify_lists</a></span><span class="op">(</span>
  input <span class="op">=</span> <span class="va">s_small3</span>,
  marker <span class="op">=</span> <span class="va">pbmc_markers</span>,
  marker_inmatrix <span class="op">=</span> <span class="cn">FALSE</span>,
  cluster_col <span class="op">=</span> <span class="st">"RNA_snn_res.1"</span>,
  seurat_out <span class="op">=</span> <span class="cn">TRUE</span>
<span class="op">)</span>
<span class="co">#&gt; An object of class Seurat </span>
<span class="co">#&gt; 230 features across 80 samples within 1 assay </span>
<span class="co">#&gt; Active assay: RNA (230 features, 20 variable features)</span>
<span class="co">#&gt;  2 dimensional reductions calculated: pca, tsne</span></code></pre></div>
</div>
<div id="frequently-asked-questions" class="section level2">
<h2 class="hasAnchor">
<a href="#frequently-asked-questions" class="anchor"></a>Frequently Asked Questions</h2>
<ol>
<li><p><strong>What types of data can be used as reference?</strong> <code>clustifyr</code> uses gene(row)-by-celltype(column) expression matrices. This means bulk RNA-seq and microarray data can be directly used. For scRNA-seq data, we have <code><a href="reference/average_clusters.html">average_clusters()</a></code> to convert matrix data and metadata. For Seurat and SCE objects, we provide wrapper function <code><a href="reference/object_ref.html">object_ref()</a></code>.</p></li>
<li><p><strong>Can I directly make references from online scRNA-seq datasets?</strong> Yes, with the caveat that metadata containing cell type assignments must be available, which is frustratingly uncommon (see our quantification/monitoring of the issue <a href="https://github.com/rnabioco/someta">here</a>). We now have a Shiny app <code><a href="reference/run_clustifyr_app.html">run_clustifyr_app()</a></code> that can directly preview and use GEO files, and <code><a href="reference/get_ucsc_reference.html">get_ucsc_reference()</a></code> to build reference from a <a href="https://cells.ucsc.edu/" class="uri">https://cells.ucsc.edu/</a> link.</p></li>
<li><p><strong>Should the input/reference data be normalized?</strong> The default metric for <code>clustifyr</code> is ranked correlation, so it does tolerate mixed raw/normalized expression fairly well. Still, we recommend matching the input and ref matrices to the same normalization method if possible. The object wrappers are taking log-normalized data for downstream steps. It should be noted that <strong>data slot from SCtransform obfuscates the original gene expression ranking, and is not ideal for <code>clustifyr</code></strong> - in this case we recommend going directly from raw counts.</p></li>
<li><p><strong>How should I determine parameters?</strong> Please see our published <a href="https://f1000research.com/articles/9-223/v2">manuscript</a> with parameter and usage discussions. In general, default settings are satisfactory in our own internal usage/testing. However, you might want to inspect the correlation matrix and call results, instead of just the final result (use <code>obj_out = FALSE</code> in <code><a href="reference/clustify.html">clustify()</a></code>).</p></li>
<li><p><strong>How many variable genes should I provide?</strong> While this of course greatly depends on the datasets in question, we generally have good results with ~500-1000 variable genes. This is why we recommend running <code>M3Drop</code> for this step. It should be noted that Seurat V3 onwards automatically stores 2000 by default, which may be too many (if the result correlation matrix shows high and similar values for too many cell types). Currently, by default <code><a href="reference/clustify.html">clustify()</a></code> on Seurat objects will use top 1000 genes.</p></li>
<li><p><strong>I have “CLASH” in many of my final calls, why is that?</strong> “CLASH” indicates ties in the correlation values. In practice, this should be very rare unless the amount of query genes is very (dangerously) low (use <code>verbose = TRUE</code> in <code><a href="reference/clustify.html">clustify()</a></code> for more information). Query genes take the intersection of provided gene list (or autodetected from Seurat objects) and genes in the reference.</p></li>
<li><p><strong>I need help troubleshooting unknown errors in my reference building/clustifying.</strong> As we try to provide better error messaging, it is still important to note that, in general, the most error-prone step is at designating the column in the metadata that contains clustering information. This is generally the <code>cluster_col</code> argument.</p></li>
<li><p><strong>What if I only have marker gene lists instead of full transcriptome references?</strong> Please see <code><a href="reference/clustify_lists.html">clustify_lists()</a></code>, which implements several simple methods. In particular, if both positive and negative markers are available, set argument <code>metric = "posneg"</code>.</p></li>
<li><p><strong>Do I need to have equal number of marker genes per cell type?</strong> Better support will be our next focus. Currently <code>metric = "posneg"</code> works with uneven numbers of markers. An alternative workflow can be used with argument <code>input_markers = TRUE</code>:</p></li>
</ol>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># pbmc_markers as FindAllMarkers output gene list</span>
<span class="va">pbmc_input</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/matrixize_markers.html">matrixize_markers</a></span><span class="op">(</span><span class="va">pbmc_markers</span><span class="op">)</span>
<span class="co"># reference gene list that is uneven length</span>
<span class="va">pbmc_ref_mm</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/pos_neg_marker.html">pos_neg_marker</a></span><span class="op">(</span>
  <span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span><span class="op">(</span>B <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"CD79A"</span>, <span class="st">"CD79B"</span>, <span class="st">"MS4A1"</span><span class="op">)</span>, 
       NK <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"GZMB"</span>, <span class="st">"GNLY"</span><span class="op">)</span><span class="op">)</span>
<span class="op">)</span>

<span class="co"># reverse input and reference</span>
<span class="va">res</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/clustify_lists.html">clustify_lists</a></span><span class="op">(</span>
  <span class="va">pbmc_ref_mm</span>,
  <span class="va">pbmc_input</span>,
  metric <span class="op">=</span> <span class="st">"jaccard"</span>,
  input_markers <span class="op">=</span> <span class="cn">TRUE</span>
<span class="op">)</span></code></pre></div>
<ol>
<li><p><strong>Why is the default setting <code>per_cell = FALSE</code>?</strong> While doing classification on per cell level is available, it is slow and not very accurate. Default settings are also not optimized for per-cell classification. <code>clustifyr</code> is mainly focused on leveraging results from clustering techniques. As other aspects of scRNA-seq analysis is often focused on clusters, we have set our focus on this resolution as well. This does mean that improper clustering of either the query or ref datasets will lead to issues, as well as cases of continuous cellular transitions where discrete clusters are not present. From benchmarking, even 15 cells per cluster is still performing well, and in our internal usage we would intentionally overcluster the data and check if <code><a href="reference/clustify.html">clustify()</a></code> results are stable (see also <code><a href="reference/overcluster_test.html">overcluster_test()</a></code>).</p></li>
<li><p><strong>Can I use multiple references in the same clustify run?</strong> Yes, simply adding columns to a reference matrix works to expand it. We also provide <code><a href="reference/build_atlas.html">build_atlas()</a></code>, which can be run along the lines of <code><a href="reference/build_atlas.html">build_atlas(matrix_objs = list(reference1, reference2, reference3, ...), genes_fn = clustifyr::human_genes_10x)</a></code>.</p></li>
<li><p><strong>Does clustifyr work for spatial scRNA-seq data?</strong> It works decently on the Seurat tutorial data. See short <a href="https://github.com/rnabioco/clustifyr/issues/370">example</a> for both <code><a href="reference/clustify.html">clustify()</a></code>(correlation) and <code><a href="reference/clustify_lists.html">clustify_lists()</a></code>(gene list enrichment) approaches. (Note, as mentioned above, we recommend avoiding SCtransform data, and opting for using raw data directly instead. This can now be directly handled by Seurat wrapper.)</p></li>
<li><p><strong>Can I pull out additional information on what gene signatures don’t match the reference clusters?</strong> Please add arguments <code>organism = "hsapiens", plot_name = "rank_diffs"</code> to <code><a href="reference/clustify.html">clustify()</a></code>. This saves a “rank_diffs.pdf”, comparing gene expression of the queried clusters versus the assigned reference cell gene signature. Highlighted in red are genes expressed (ranked) higher in query data, and in blue gene expressed (ranked) lower than the reference. Top 10 GO-BP terms are also included. See the function <code><a href="reference/assess_rank_bias.html">assess_rank_bias()</a></code> for step-by-step generation of the plot outside of the <code><a href="reference/clustify.html">clustify()</a></code> wrapper.</p></li>
<li><p><strong>How do I cite <code>clustifyr</code>?</strong></p></li>
</ol>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/utils/citation.html">citation</a></span><span class="op">(</span><span class="st">"clustifyr"</span><span class="op">)</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; Fu R, Gillen A, Sheridan RM, Tian C, Daya M, Hao Y, Hesselberth JR,</span>
<span class="co">#&gt; Riemondy KA (2019). "clustifyr: An R package for automated single-cell</span>
<span class="co">#&gt; RNA sequencing cluster classification." _F1000 Research_. doi:</span>
<span class="co">#&gt; 10.12688/f1000research.22969.1 (URL:</span>
<span class="co">#&gt; https://doi.org/10.12688/f1000research.22969.1).</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; A BibTeX entry for LaTeX users is</span>
<span class="co">#&gt; </span>
<span class="co">#&gt;   @Article{,</span>
<span class="co">#&gt;     title = {clustifyr: An R package for automated single-cell RNA sequencing cluster classification},</span>
<span class="co">#&gt;     year = {2019},</span>
<span class="co">#&gt;     author = {Rui Fu and Austin Gillen and Ryan M. Sheridan and Chengzhe Tian and Michelle Daya and Yue Hao and Jay R. Hesselberth and Kent A. Riemondy},</span>
<span class="co">#&gt;     journal = {F1000 Research},</span>
<span class="co">#&gt;     doi = {10.12688/f1000research.22969.1},</span>
<span class="co">#&gt;   }</span></code></pre></div>
</div>
<div id="code-of-conduct" class="section level2">
<h2 class="hasAnchor">
<a href="#code-of-conduct" class="anchor"></a>Code of Conduct</h2>
<p>Please note that the clustifyr project is released with a <a href="http://rnabioco.github.io/clustifyr/CODE_OF_CONDUCT.html">Contributor Code of Conduct</a>. By contributing to this project, you agree to abide by its terms.</p>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <div class="links">
<h2>Links</h2>
<ul class="list-unstyled">
<li>Download from BIOC at <br><a href="https://www.bioconductor.org/packages/clustifyr">https://​www.bioconductor.org/​packages/​clustifyr</a>
</li>
<li>Browse source code at <br><a href="https://github.com/rnabioco/clustifyr/">https://​github.com/​rnabioco/​clustifyr/​</a>
</li>
<li>Report a bug at <br><a href="https://github.com/rnabioco/clustifyr/issues">https://​github.com/​rnabioco/​clustifyr/​issues</a>
</li>
<li>External data sets at <br><a href="http://github.com/rnabioco/clustifyrdata">http://​github.com/​rnabioco/​clustifyrdata</a>
</li>
</ul>
</div>
<div class="license">
<h2>License</h2>
<ul class="list-unstyled">
<li>
<a href="https://opensource.org/licenses/mit-license.php">MIT</a> + file <a href="LICENSE-text.html">LICENSE</a>
</li>
</ul>
</div>
<div class="community">
<h2>Community</h2>
<ul class="list-unstyled">
<li><a href="CONTRIBUTING.html">Contributing guide</a></li>
<li><a href="CODE_OF_CONDUCT.html">Code of conduct</a></li>
</ul>
</div>
<div class="citation">
<h2>Citation</h2>
<ul class="list-unstyled">
<li><a href="authors.html">Citing clustifyr</a></li>
</ul>
</div>
<div class="developers">
<h2>Developers</h2>
<ul class="list-unstyled">
<li>Rui Fu <br><small class="roles"> Author, maintainer </small>  </li>
<li>Kent Riemondy <br><small class="roles"> Author </small>  </li>
<li>
<a href="https://medschool.cuanschutz.edu/rbi">RNA Bioscience Initiative</a> <br><small class="roles"> Funder </small>  </li>
<li><a href="authors.html">All authors...</a></li>
</ul>
</div>

  <div class="dev-status">
<h2>Dev status</h2>
<ul class="list-unstyled">
<li><a href="https://github.com/rnabioco/clustifyr/actions"><img src="https://github.com/rnabioco/clustifyr/workflows/R-CMD-check/badge.svg" alt="R build status"></a></li>
<li><a href="https://codecov.io/gh/rnabioco/clustifyr"><img src="https://codecov.io/gh/rnabioco/clustifyr/branch/master/graph/badge.svg" alt="codecov"></a></li>
<li><a href="https://bioconductor.org/packages/release/bioc/html/clustifyr.html"><img src="https://bioconductor.org/shields/availability/release/clustifyr.svg" alt="platforms"></a></li>
<li><a href="https://bioconductor.org/packages/release/bioc/html/clustifyr.html"><img src="https://bioconductor.org/shields/years-in-bioc/clustifyr.svg" alt="bioc"></a></li>
<li><a href="https://bioconductor.org/packages/stats/bioc/clustifyr/clustifyr_stats.tab"><img src="https://img.shields.io/badge/%23%20downloads-2840-brightgreen" alt="#downloads"></a></li>
</ul>
</div>
</div>
</div>


      <footer><div class="copyright">
  <p>Developed by Rui Fu, Kent Riemondy, <a href="https://medschool.cuanschutz.edu/rbi">RNA Bioscience Initiative</a>.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.6.1.</p>
</div>

      </footer>
</div>

  


  </body>
</html>
