<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Classifier for Single-cell RNA-seq Using Cell Clusters • clustifyr</title>
<!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="favicon-16x16.png">
<link rel="icon" type="image/png" sizes="32x32" href="favicon-32x32.png">
<link rel="apple-touch-icon" type="image/png" sizes="180x180" href="apple-touch-icon.png">
<link rel="apple-touch-icon" type="image/png" sizes="120x120" href="apple-touch-icon-120x120.png">
<link rel="apple-touch-icon" type="image/png" sizes="76x76" href="apple-touch-icon-76x76.png">
<link rel="apple-touch-icon" type="image/png" sizes="60x60" href="apple-touch-icon-60x60.png">
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.3.1/jquery.min.js" integrity="sha256-FgpCb/KJQlLNfOu91ta32o/NMZxltwRo8QtmkMRdAu8=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/css/bootstrap.min.css" integrity="sha256-916EbMg70RQy9LHiGkXzG8hSg9EdNy97GazNG/aiY1w=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha256-U5ZEeKfGNOja007MMD3YBI0A3OSZOQbeG6z2f2Y0hu8=" crossorigin="anonymous"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.7.1/css/all.min.css" integrity="sha256-nAmazAk6vS34Xqo0BSrTb+abbtFlgsFK7NKSi6o7Y78=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.7.1/css/v4-shims.min.css" integrity="sha256-6qHlizsOWFskGlwVOKuns+D1nB6ssZrHQrNj1wGplHc=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.4/clipboard.min.js" integrity="sha256-FiZwavyI2V6+EXO1U+xzLG3IKldpiTFf3153ea9zikQ=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.9.4/headroom.min.js" integrity="sha256-DJFC1kqIhelURkuza0AvYal5RxMtpzLjFhsnVIeuk+U=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.9.4/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="pkgdown.css" rel="stylesheet">
<script src="pkgdown.js"></script><meta property="og:title" content="Classifier for Single-cell RNA-seq Using Cell Clusters">
<meta property="og:description" content="Package designed to aid in classifying cells from single-cell RNA sequencing data
    using external reference data (e.g., bulk RNA-seq, scRNA-seq, microarray, gene lists). A 
    variety of correlation based methods and gene list enrichment methods are provided to assist cell
    type assignment.">
<meta property="og:image" content="http://rnabioco.github.io/clustifyr/logo.png">
<meta name="twitter:card" content="summary">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <div class="container template-home">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="index.html">clustifyr</a>
        <span class="version label label-danger" data-toggle="tooltip" data-placement="bottom" title="Unreleased version">0.99.6</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="index.html">
    <span class="fas fa fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="articles/clustifyR.html">Introduction to clustifyr</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/rnabioco/clustifyr">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="contents col-md-9">
<div id="clustifyr" class="section level1">
<div class="page-header"><h1 class="hasAnchor">
<a href="#clustifyr" class="anchor"></a>clustifyr <img src="reference/figures/logo.png" align="right">
</h1></div>

<p>clustifyr classifies cells and clusters in single-cell RNA sequencing experiments using reference bulk RNA-seq data sets, sorted microarray expression data, single-cell gene signatures, or lists of marker genes.</p>
<p>Single cell transcriptomes are difficult to annotate without knowledge of the underlying biology. Even with this knowledge, accurate identification can be challenging due to the lack of detectable expression of common marker genes. clustifyr solves this problem by automatically annotating single cells or clusters of cells using single-cell RNA-seq, bulk RNA-seq data, microarray, or marker gene lists. Additional functions enable exploratory analysis of similarities between single cell RNA-seq datasets and reference data.</p>
<div id="installation" class="section level2">
<h2 class="hasAnchor">
<a href="#installation" class="anchor"></a>Installation</h2>
<div class="sourceCode" id="cb1"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb1-1" data-line-number="1"><span class="co"># install.packages("devtools")</span></a>
<a class="sourceLine" id="cb1-2" data-line-number="2">devtools<span class="op">::</span><span class="kw">install_github</span>(<span class="st">"rnabioco/clustifyr"</span>, <span class="dt">ref =</span> <span class="st">"0.99.4"</span>)</a></code></pre></div>
</div>
<div id="example-usage" class="section level2">
<h2 class="hasAnchor">
<a href="#example-usage" class="anchor"></a>Example usage</h2>
<p>In this example we use the following built-in input data:</p>
<ul>
<li>an expression matrix of single cell RNA-seq data (<code>pbmc_matrix_small</code>)</li>
<li>a metadata data.frame (<code>pbmc_meta</code>), with cluster information stored (<code>"classified"</code>)</li>
<li>a vector of variable genes (<code>pbmc_vargenes</code>)</li>
<li>a matrix of mean normalized scRNA-seq UMI counts by cell type (<code>cbmc_ref</code>):</li>
</ul>
<p>We then calculate correlation coefficients and plot them on a pre-calculated projection (stored in <code>pbmc_meta</code>).</p>
<div class="sourceCode" id="cb2"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb2-1" data-line-number="1"><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span>(clustifyr)</a>
<a class="sourceLine" id="cb2-2" data-line-number="2"></a>
<a class="sourceLine" id="cb2-3" data-line-number="3"><span class="co"># calculate correlation</span></a>
<a class="sourceLine" id="cb2-4" data-line-number="4">res &lt;-<span class="st"> </span><span class="kw"><a href="reference/clustify.html">clustify</a></span>(</a>
<a class="sourceLine" id="cb2-5" data-line-number="5">  <span class="dt">input =</span> pbmc_matrix_small,</a>
<a class="sourceLine" id="cb2-6" data-line-number="6">  <span class="dt">metadata =</span> pbmc_meta<span class="op">$</span>classified,</a>
<a class="sourceLine" id="cb2-7" data-line-number="7">  <span class="dt">ref_mat =</span> cbmc_ref,</a>
<a class="sourceLine" id="cb2-8" data-line-number="8">  <span class="dt">query_genes =</span> pbmc_vargenes</a>
<a class="sourceLine" id="cb2-9" data-line-number="9">)</a>
<a class="sourceLine" id="cb2-10" data-line-number="10"></a>
<a class="sourceLine" id="cb2-11" data-line-number="11"><span class="co"># print assignments</span></a>
<a class="sourceLine" id="cb2-12" data-line-number="12"><span class="kw"><a href="reference/cor_to_call.html">cor_to_call</a></span>(res)</a>
<a class="sourceLine" id="cb2-13" data-line-number="13"><span class="co">#&gt; # A tibble: 9 x 3</span></a>
<a class="sourceLine" id="cb2-14" data-line-number="14"><span class="co">#&gt; # Groups:   cluster [9]</span></a>
<a class="sourceLine" id="cb2-15" data-line-number="15"><span class="co">#&gt;   cluster      type           r</span></a>
<a class="sourceLine" id="cb2-16" data-line-number="16"><span class="co">#&gt;   &lt;chr&gt;        &lt;chr&gt;      &lt;dbl&gt;</span></a>
<a class="sourceLine" id="cb2-17" data-line-number="17"><span class="co">#&gt; 1 B            B          0.909</span></a>
<a class="sourceLine" id="cb2-18" data-line-number="18"><span class="co">#&gt; 2 CD14+ Mono   CD14+ Mono 0.915</span></a>
<a class="sourceLine" id="cb2-19" data-line-number="19"><span class="co">#&gt; 3 FCGR3A+ Mono CD16+ Mono 0.929</span></a>
<a class="sourceLine" id="cb2-20" data-line-number="20"><span class="co">#&gt; 4 Memory CD4 T CD4 T      0.861</span></a>
<a class="sourceLine" id="cb2-21" data-line-number="21"><span class="co">#&gt; 5 Naive CD4 T  CD4 T      0.889</span></a>
<a class="sourceLine" id="cb2-22" data-line-number="22"><span class="co">#&gt; 6 DC           DC         0.849</span></a>
<a class="sourceLine" id="cb2-23" data-line-number="23"><span class="co">#&gt; 7 Platelet     Mk         0.732</span></a>
<a class="sourceLine" id="cb2-24" data-line-number="24"><span class="co">#&gt; 8 CD8 T        NK         0.826</span></a>
<a class="sourceLine" id="cb2-25" data-line-number="25"><span class="co">#&gt; 9 NK           NK         0.894</span></a>
<a class="sourceLine" id="cb2-26" data-line-number="26"></a>
<a class="sourceLine" id="cb2-27" data-line-number="27"><span class="co"># plot assignments on a projection</span></a>
<a class="sourceLine" id="cb2-28" data-line-number="28"><span class="kw"><a href="reference/plot_best_call.html">plot_best_call</a></span>(</a>
<a class="sourceLine" id="cb2-29" data-line-number="29">  <span class="dt">cor_mat =</span> res,</a>
<a class="sourceLine" id="cb2-30" data-line-number="30">  <span class="dt">metadata =</span> pbmc_meta,</a>
<a class="sourceLine" id="cb2-31" data-line-number="31">  <span class="dt">cluster_col =</span> <span class="st">"classified"</span></a>
<a class="sourceLine" id="cb2-32" data-line-number="32">)</a></code></pre></div>
<p><img src="reference/figures/readme_example-1.png"><!-- --></p>
<p><code><a href="reference/clustify.html">clustify()</a></code> can also take a clustered <code>SingleCellExperiment</code> or <code>seurat</code> object (both v2 and v3) and assign identities.</p>
<div class="sourceCode" id="cb3"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb3-1" data-line-number="1"><span class="co"># for SingleCellExperiment</span></a>
<a class="sourceLine" id="cb3-2" data-line-number="2"><span class="kw"><a href="reference/clustify.html">clustify</a></span>(</a>
<a class="sourceLine" id="cb3-3" data-line-number="3">  <span class="dt">input =</span> sce_small,          <span class="co"># an SCE object</span></a>
<a class="sourceLine" id="cb3-4" data-line-number="4">  <span class="dt">ref_mat =</span> cbmc_ref,         <span class="co"># matrix of RNA-seq expression data for each cell type</span></a>
<a class="sourceLine" id="cb3-5" data-line-number="5">  <span class="dt">cluster_col =</span> <span class="st">"cell_type1"</span>, <span class="co"># name of column in meta.data containing cell clusters</span></a>
<a class="sourceLine" id="cb3-6" data-line-number="6">  <span class="dt">obj_out =</span> <span class="ot">TRUE</span>              <span class="co"># output SCE object with cell type inserted as "type" column</span></a>
<a class="sourceLine" id="cb3-7" data-line-number="7">) </a>
<a class="sourceLine" id="cb3-8" data-line-number="8"><span class="co">#&gt; class: SingleCellExperiment </span></a>
<a class="sourceLine" id="cb3-9" data-line-number="9"><span class="co">#&gt; dim: 200 200 </span></a>
<a class="sourceLine" id="cb3-10" data-line-number="10"><span class="co">#&gt; metadata(0):</span></a>
<a class="sourceLine" id="cb3-11" data-line-number="11"><span class="co">#&gt; assays(2): counts logcounts</span></a>
<a class="sourceLine" id="cb3-12" data-line-number="12"><span class="co">#&gt; rownames(200): SGIP1 AZIN2 ... TAF12 SNHG3</span></a>
<a class="sourceLine" id="cb3-13" data-line-number="13"><span class="co">#&gt; rowData names(10): feature_symbol is_feature_control ... total_counts</span></a>
<a class="sourceLine" id="cb3-14" data-line-number="14"><span class="co">#&gt;   log10_total_counts</span></a>
<a class="sourceLine" id="cb3-15" data-line-number="15"><span class="co">#&gt; colnames(200): AZ_A1 AZ_A10 ... HP1502401_E18 HP1502401_E19</span></a>
<a class="sourceLine" id="cb3-16" data-line-number="16"><span class="co">#&gt; colData names(35): cell_quality cell_type1 ... type r</span></a>
<a class="sourceLine" id="cb3-17" data-line-number="17"><span class="co">#&gt; reducedDimNames(0):</span></a>
<a class="sourceLine" id="cb3-18" data-line-number="18"><span class="co">#&gt; spikeNames(1): ERCC</span></a>
<a class="sourceLine" id="cb3-19" data-line-number="19"><span class="co">#&gt; altExpNames(0):</span></a>
<a class="sourceLine" id="cb3-20" data-line-number="20"></a>
<a class="sourceLine" id="cb3-21" data-line-number="21"><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span>(Seurat)</a>
<a class="sourceLine" id="cb3-22" data-line-number="22"></a>
<a class="sourceLine" id="cb3-23" data-line-number="23"><span class="co"># for seurat2</span></a>
<a class="sourceLine" id="cb3-24" data-line-number="24"><span class="kw"><a href="reference/clustify.html">clustify</a></span>(</a>
<a class="sourceLine" id="cb3-25" data-line-number="25">  <span class="dt">input =</span> s_small,</a>
<a class="sourceLine" id="cb3-26" data-line-number="26">  <span class="dt">cluster_col =</span> <span class="st">"res.1"</span>,</a>
<a class="sourceLine" id="cb3-27" data-line-number="27">  <span class="dt">ref_mat =</span> cbmc_ref,</a>
<a class="sourceLine" id="cb3-28" data-line-number="28">  <span class="dt">seurat_out =</span> <span class="ot">TRUE</span></a>
<a class="sourceLine" id="cb3-29" data-line-number="29">)</a>
<a class="sourceLine" id="cb3-30" data-line-number="30"><span class="co">#&gt; An old seurat object</span></a>
<a class="sourceLine" id="cb3-31" data-line-number="31"><span class="co">#&gt;  230 genes across 80 samples</span></a>
<a class="sourceLine" id="cb3-32" data-line-number="32"></a>
<a class="sourceLine" id="cb3-33" data-line-number="33"><span class="co"># for Seurat3</span></a>
<a class="sourceLine" id="cb3-34" data-line-number="34"><span class="kw"><a href="reference/clustify.html">clustify</a></span>(</a>
<a class="sourceLine" id="cb3-35" data-line-number="35">  <span class="dt">input =</span> s_small3,</a>
<a class="sourceLine" id="cb3-36" data-line-number="36">  <span class="dt">cluster_col =</span> <span class="st">"RNA_snn_res.1"</span>,</a>
<a class="sourceLine" id="cb3-37" data-line-number="37">  <span class="dt">ref_mat =</span> cbmc_ref,</a>
<a class="sourceLine" id="cb3-38" data-line-number="38">  <span class="dt">seurat_out =</span> <span class="ot">TRUE</span></a>
<a class="sourceLine" id="cb3-39" data-line-number="39">)</a>
<a class="sourceLine" id="cb3-40" data-line-number="40"><span class="co">#&gt; An object of class Seurat </span></a>
<a class="sourceLine" id="cb3-41" data-line-number="41"><span class="co">#&gt; 230 features across 80 samples within 1 assay </span></a>
<a class="sourceLine" id="cb3-42" data-line-number="42"><span class="co">#&gt; Active assay: RNA (230 features)</span></a>
<a class="sourceLine" id="cb3-43" data-line-number="43"><span class="co">#&gt;  2 dimensional reductions calculated: pca, tsne</span></a></code></pre></div>
<p>New reference matrix can be made directly from <code>SingleCellExperiment</code> and <code>seurat</code> objects as well. Other scRNAseq experiment object types are supported as well.</p>
<div class="sourceCode" id="cb4"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb4-1" data-line-number="1"><span class="co"># make reference from SingleCellExperiment objects</span></a>
<a class="sourceLine" id="cb4-2" data-line-number="2">sce_ref &lt;-<span class="st"> </span><span class="kw"><a href="reference/object_ref.html">object_ref</a></span>(</a>
<a class="sourceLine" id="cb4-3" data-line-number="3">  <span class="dt">input =</span> sce_small,               <span class="co"># SCE object</span></a>
<a class="sourceLine" id="cb4-4" data-line-number="4">  <span class="dt">cluster_col =</span> <span class="st">"cell_type1"</span>       <span class="co"># name of column in colData containing cell identities</span></a>
<a class="sourceLine" id="cb4-5" data-line-number="5">)</a>
<a class="sourceLine" id="cb4-6" data-line-number="6"><span class="co">#&gt; recognized object type - SingleCellExperiment</span></a>
<a class="sourceLine" id="cb4-7" data-line-number="7"></a>
<a class="sourceLine" id="cb4-8" data-line-number="8"><span class="co"># make reference from seurat objects</span></a>
<a class="sourceLine" id="cb4-9" data-line-number="9">s_ref &lt;-<span class="st"> </span><span class="kw"><a href="reference/seurat_ref.html">seurat_ref</a></span>(</a>
<a class="sourceLine" id="cb4-10" data-line-number="10">  <span class="dt">seurat_object =</span> s_small,</a>
<a class="sourceLine" id="cb4-11" data-line-number="11">  <span class="dt">cluster_col =</span> <span class="st">"res.1"</span></a>
<a class="sourceLine" id="cb4-12" data-line-number="12">)</a>
<a class="sourceLine" id="cb4-13" data-line-number="13"></a>
<a class="sourceLine" id="cb4-14" data-line-number="14"><span class="kw"><a href="https://rdrr.io/r/utils/head.html">head</a></span>(s_ref)</a>
<a class="sourceLine" id="cb4-15" data-line-number="15"><span class="co">#&gt;                 0        1        2        3</span></a>
<a class="sourceLine" id="cb4-16" data-line-number="16"><span class="co">#&gt; MS4A1    4.517255 3.204766 0.000000 0.000000</span></a>
<a class="sourceLine" id="cb4-17" data-line-number="17"><span class="co">#&gt; CD79B    4.504191 3.549095 2.580662 0.000000</span></a>
<a class="sourceLine" id="cb4-18" data-line-number="18"><span class="co">#&gt; CD79A    4.457349 4.199849 0.000000 0.000000</span></a>
<a class="sourceLine" id="cb4-19" data-line-number="19"><span class="co">#&gt; HLA-DRA  6.211779 6.430463 3.659590 4.169965</span></a>
<a class="sourceLine" id="cb4-20" data-line-number="20"><span class="co">#&gt; TCL1A    4.394310 2.837922 0.000000 0.000000</span></a>
<a class="sourceLine" id="cb4-21" data-line-number="21"><span class="co">#&gt; HLA-DQB1 4.380289 4.325293 0.000000 1.666167</span></a></code></pre></div>
<p><code><a href="reference/clustify_lists.html">clustify_lists()</a></code> handles identity assignment of matrix or <code>SingleCellExperiment</code> and <code>seurat</code> objects based on marker gene lists.</p>
<div class="sourceCode" id="cb5"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb5-1" data-line-number="1"><span class="kw"><a href="reference/clustify_lists.html">clustify_lists</a></span>(</a>
<a class="sourceLine" id="cb5-2" data-line-number="2">  <span class="dt">input =</span> pbmc_matrix_small,</a>
<a class="sourceLine" id="cb5-3" data-line-number="3">  <span class="dt">metadata =</span> pbmc_meta,</a>
<a class="sourceLine" id="cb5-4" data-line-number="4">  <span class="dt">cluster_col =</span> <span class="st">"classified"</span>,</a>
<a class="sourceLine" id="cb5-5" data-line-number="5">  <span class="dt">marker =</span> pbmc_markers,</a>
<a class="sourceLine" id="cb5-6" data-line-number="6">  <span class="dt">marker_inmatrix =</span> <span class="ot">FALSE</span></a>
<a class="sourceLine" id="cb5-7" data-line-number="7">)</a>
<a class="sourceLine" id="cb5-8" data-line-number="8"><span class="co">#&gt;                      0        1        2         3         4        5        6</span></a>
<a class="sourceLine" id="cb5-9" data-line-number="9"><span class="co">#&gt; Naive CD4 T  1.5639055 20.19469 31.77095  8.664074 23.844992 19.06931 19.06931</span></a>
<a class="sourceLine" id="cb5-10" data-line-number="10"><span class="co">#&gt; Memory CD4 T 1.5639055 20.19469 31.77095 10.568007 23.844992 17.97875 19.06931</span></a>
<a class="sourceLine" id="cb5-11" data-line-number="11"><span class="co">#&gt; CD14+ Mono   0.9575077 14.70716 76.21353 17.899569 11.687739 49.86699 16.83210</span></a>
<a class="sourceLine" id="cb5-12" data-line-number="12"><span class="co">#&gt; B            0.6564777 12.70976 31.77095 26.422929 13.536295 20.19469 13.53630</span></a>
<a class="sourceLine" id="cb5-13" data-line-number="13"><span class="co">#&gt; CD8 T        1.0785353 17.97875 31.82210 12.584823 31.822099 22.71234 40.45383</span></a>
<a class="sourceLine" id="cb5-14" data-line-number="14"><span class="co">#&gt; FCGR3A+ Mono 0.6564777 13.63321 72.43684 17.899569  9.726346 56.48245 14.61025</span></a>
<a class="sourceLine" id="cb5-15" data-line-number="15"><span class="co">#&gt; NK           0.6564777 14.61025 31.82210  7.757206 31.822099 22.71234 45.05072</span></a>
<a class="sourceLine" id="cb5-16" data-line-number="16"><span class="co">#&gt; DC           0.6564777 15.80598 63.34978 19.069308 13.758144 40.56298 17.97875</span></a>
<a class="sourceLine" id="cb5-17" data-line-number="17"><span class="co">#&gt; Platelet     0.5428889 13.34769 59.94938 14.215244 15.158755 46.92861 19.49246</span></a>
<a class="sourceLine" id="cb5-18" data-line-number="18"><span class="co">#&gt;                      7          8</span></a>
<a class="sourceLine" id="cb5-19" data-line-number="19"><span class="co">#&gt; Naive CD4 T   6.165348  0.6055118</span></a>
<a class="sourceLine" id="cb5-20" data-line-number="20"><span class="co">#&gt; Memory CD4 T  6.165348  0.9575077</span></a>
<a class="sourceLine" id="cb5-21" data-line-number="21"><span class="co">#&gt; CD14+ Mono   25.181595  1.0785353</span></a>
<a class="sourceLine" id="cb5-22" data-line-number="22"><span class="co">#&gt; B            17.899569  0.1401901</span></a>
<a class="sourceLine" id="cb5-23" data-line-number="23"><span class="co">#&gt; CD8 T         7.882145  0.3309153</span></a>
<a class="sourceLine" id="cb5-24" data-line-number="24"><span class="co">#&gt; FCGR3A+ Mono 21.409177  0.3309153</span></a>
<a class="sourceLine" id="cb5-25" data-line-number="25"><span class="co">#&gt; NK            5.358651  0.3309153</span></a>
<a class="sourceLine" id="cb5-26" data-line-number="26"><span class="co">#&gt; DC           45.101877  0.1401901</span></a>
<a class="sourceLine" id="cb5-27" data-line-number="27"><span class="co">#&gt; Platelet     19.492465 59.9493793</span></a>
<a class="sourceLine" id="cb5-28" data-line-number="28"></a>
<a class="sourceLine" id="cb5-29" data-line-number="29"><span class="kw"><a href="reference/clustify_lists.html">clustify_lists</a></span>(</a>
<a class="sourceLine" id="cb5-30" data-line-number="30">  <span class="dt">input =</span> s_small,</a>
<a class="sourceLine" id="cb5-31" data-line-number="31">  <span class="dt">marker =</span> pbmc_markers,</a>
<a class="sourceLine" id="cb5-32" data-line-number="32">  <span class="dt">marker_inmatrix =</span> <span class="ot">FALSE</span>,</a>
<a class="sourceLine" id="cb5-33" data-line-number="33">  <span class="dt">cluster_col =</span> <span class="st">"res.1"</span>,</a>
<a class="sourceLine" id="cb5-34" data-line-number="34">  <span class="dt">seurat_out =</span> <span class="ot">TRUE</span></a>
<a class="sourceLine" id="cb5-35" data-line-number="35">)</a>
<a class="sourceLine" id="cb5-36" data-line-number="36"><span class="co">#&gt; An old seurat object</span></a>
<a class="sourceLine" id="cb5-37" data-line-number="37"><span class="co">#&gt;  230 genes across 80 samples</span></a></code></pre></div>
</div>
<div id="additional-reference-data" class="section level2">
<h2 class="hasAnchor">
<a href="#additional-reference-data" class="anchor"></a>Additional reference data</h2>
<p>More reference data (including tabula muris, immgen, etc) are available at <a href="https://github.com/rnabioco/clustifyrdata" class="uri">https://github.com/rnabioco/clustifyrdata</a>.</p>
<p>Also see list for individual downloads at <a href="https://rnabioco.github.io/clustifyrdata/articles/download_refs.html" class="uri">https://rnabioco.github.io/clustifyrdata/articles/download_refs.html</a></p>
<p>Additional tutorials at <a href="https://rnabioco.github.io/clustifyrdata/articles/otherformats.html" class="uri">https://rnabioco.github.io/clustifyrdata/articles/otherformats.html</a></p>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">
    <div class="links">
<h2>Links</h2>
<ul class="list-unstyled">
<li>Browse source code at <br><a href="https://github.com/rnabioco/clustifyr">https://​github.com/​rnabioco/​clustifyr</a>
</li>
<li>Report a bug at <br><a href="https://github.com/rnabioco/clustifyr/issues">https://​github.com/​rnabioco/​clustifyr/​issues</a>
</li>
<li>External data sets at <br><a href="http://github.com/rnabioco/clustifyrdata">http://​github.com/​rnabioco/​clustifyrdata</a>
</li>
</ul>
</div>
<div class="license">
<h2>License</h2>
<ul class="list-unstyled">
<li>
<a href="https://opensource.org/licenses/mit-license.php">MIT</a> + file <a href="LICENSE-text.html">LICENSE</a>
</li>
</ul>
</div>
<div class="developers">
<h2>Developers</h2>
<ul class="list-unstyled">
<li>Rui Fu <br><small class="roles"> Author, maintainer </small>  </li>
<li>Kent Riemondy <br><small class="roles"> Author </small>  </li>
<li>Austin Gillen <br><small class="roles"> Author </small>  </li>
<li>Chengzhe Tian <br><small class="roles"> Author </small>  </li>
<li>Jay Hesselberth <br><small class="roles"> Author </small>  </li>
<li><a href="authors.html">All authors...</a></li>
</ul>
</div>

  <div class="dev-status">
<h2>Dev status</h2>
<ul class="list-unstyled">
<li><a href="https://travis-ci.org/rnabioco/clustifyr"><img src="https://travis-ci.org/rnabioco/clustifyr.svg?branch=master" alt="Build Status"></a></li>
<li><a href="https://codecov.io/github/rnabioco/clustifyr?branch=master"><img src="https://codecov.io/gh/rnabioco/clustifyr/branch/master/graph/badge.svg" alt="Coverage status"></a></li>
</ul>
</div>
</div>
</div>


      <footer><div class="copyright">
  <p>Developed by Rui Fu, Kent Riemondy, Austin Gillen, Chengzhe Tian, Jay Hesselberth.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.4.1.</p>
</div>

      </footer>
</div>

  


  </body>
</html>
