---
title: 'Introduction to clustifyR'
date: '`r Sys.Date()`'
output:
  rmarkdown::html_vignette:
    toc: true
    toc_depth: 3
    vignette: >
      %\VignetteIndexEntry{clustifyR-overview}
      %\VignetteEngine{knitr::rmarkdown}
      %\VignetteEncoding{UTF-8}
---

```{r knitr_opts, echo = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>",
  fig.align = "center"
)
```

## Why use `clustifyR`?

Single cell transcriptomes are difficult to annotate without extensive knowledge of the underlying biology of the system in question. Even with this knowledge, accurate identification can be challenging due to the lack of detectable expression of common marker genes defined by bulk RNA-seq, flow cytometry, other single cell RNA-seq platforms, etc.

`clustifyR` solves this problem by providing functions to automatically annotate single cells or clusters using bulk RNA-seq data or marker gene lists (ranked or unranked). Additional functions allow for exploratory analysis of calculated similarities between single cell RNA-seq datasets and reference data.

## A simple example: 10x Genomics PBMCs

In this example, we take an example 10x Genomics 3' scRNA-seq dataset comprised of peripheral blood mononuclear cells from a healthy human donor and annotate the cell clusters (identified using `Seurat`) using bulk RNA-seq data derived from isolated PBMC cell populations. 

```{r fig.height=8, fig.width=12, warning = F, message = F}
library(clustifyR)

# load pacakge data
data("pbmc4k_matrix")
data("pbmc4k_meta")
data("pbmc4k_vargenes")
data("pbmc_bulk_matrix")

# plot tSNE using known identities
plot_tsne(pbmc4k_meta, feature = "classified")
```

`clustifyR` provides the `clustify()` function, which takes as input a matrix of single cell RNA-seq expression values, a metadata table with cluster assignments, and a matrix with bulk RNA seq datasets. 

As the heatmap below shows, most cell types are accurately classified using the default correlation method (pearson) and a list of highly variable genes in the single cell dataset identified using `Seurat`'s VMR method.

```{r fig.height=8, fig.width=12, warning = F, message = F}
# run correlation (pearson by default)
res <- clustify(input = pbmc4k_matrix,
               ref_mat = pbmc_bulk_matrix,
               metadata = pbmc4k_meta,
               query_genes = pbmc4k_vargenes,
               cluster_col = "cluster")

# peek at correlation matrix
res[1:5, 1:5]

# view heatmap of correlation coefficients
library(ComplexHeatmap)
library(circlize)
ComplexHeatmap::Heatmap(res, colorRamp2(c(0, 1), c("white", "black")))
```

### Plot cluster identities and correlation coefficients

`clustifyR` provides functions to plot single cells using pre-calculated tSNE embeddings (or those from any other dimensionality reduction method), and to plot correlation coefficients on those projections. The plots below show that the highest correlations between these bulk RNA-seq data and the test scRNA-seq dataset are restricted to the correct cell clusters.

```{r plot, fig.height=8, fig.width=12, warning = F, message = F}
# plot tSNE using known identities
plot_tsne(pbmc4k_meta, feature = "classified")

# plot correlation coefficients on tSNE for each identity class
plot_cor(res,
         pbmc4k_meta,
         colnames(res)[1],
         cluster_col = "cluster")

# plot called identities
plot_best_call(res, pbmc4k_meta)
```

### Correlation per cell

```{r fig.height=8, fig.width=12, warning = F, message = F}
# run correlation (pearson by default)
res <- clustify(input = pbmc4k_matrix,
               ref_mat = pbmc_bulk_matrix,
               query_genes = pbmc4k_vargenes,
               per_cell = T)

plot_cor(res,
         pbmc4k_meta,
         colnames(res)[c(1, 5)])
```

## Using gene lists to classify cells (ranked or unranked)

```{r fig.height=8, fig.width=12, warning = F, message = F}
# available metrics are "hyper", "jaccard", "spearman", "gsea"
res2 <- clustify_lists(pbmc4k_matrix,
               per_cell = F,
               cluster_info = pbmc4k_meta,
               cluster_col = "cluster",
               marker = cbmc_m,
               marker_inmatrix = T,
               metric = "hyper")

# view as heatmap, or plot_best_call
ComplexHeatmap::Heatmap(res2, 
        cluster_rows = FALSE, 
        cluster_columns = FALSE, 
        heatmap_legend_param = list(title = "-log10(adjP)"))
```


## Direct handling of `seurat` objects, v2 and v3

```{r, eval = F, warning = F, message = F}
# can even returning a new object with `type` added to metadata.
res <- clustify(
  input = s_small,
  cluster_col = "res.1",
  ref_mat = pbmc_bulk_matrix,
  seurat_out = T
)

# or return a correlation matrix
res <- clustify_lists(s_small,
               per_cell = F,
               marker = pbmc4k_markers,
               marker_inmatrix = F,
               cluster_col = "res.1",
               seurat_out = F
)

# and combined ref matrix and marker info
res <- clustify_nudge(input = s_small,
                      ref_mat = cbmc_ref,
                      marker = cbmc_m,
                      cluster_col = "res.1",
                      threshold = 0.8,
                      seurat_out = F)
```
