---
title: 'Introduction to clustifyR'
date: '`r Sys.Date()`'
output:
  rmarkdown::html_vignette:
    toc: true
    toc_depth: 3
    vignette: >
      %\VignetteIndexEntry{clustifyR-overview}
      %\VignetteEngine{knitr::rmarkdown}
      %\VignetteEncoding{UTF-8}
---

```{r knitr_opts, echo = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>",
  fig.align = "center",
  out.width = "100%"
)
```

## Why use clustifyR?

Single cell transcriptomes are difficult to annotate without knowledge of the underlying biology of the system in question. Even with this knowledge, accurate identification can be challenging due to the lack of detectable expression of common marker genes.

`clustifyR` solves this problem by providing functions to automatically annotate single cells or clusters using bulk RNA-seq data or marker gene lists (ranked or unranked). Additional functions allow for exploratory analysis of calculated similarities between single cell RNA-seq datasets and reference data.

## A simple example: 10x Genomics PBMCs

In this example, we take an example 10x Genomics 3' scRNA-seq dataset comprised of peripheral blood mononuclear cells from a healthy human donor and annotate the cell clusters (identified using `Seurat`) using bulk RNA-seq data derived from isolated PBMC cell populations. 

```{r fig.height=8, fig.width=12, fig.cap="tSNE projection of PBMCs from a health human donor", warning = F, message = F}
library(clustifyR)

# plot tSNE using known identities
plot_tsne(pbmc4k_meta, feature = "classified")
```

Bulk RNA-Seq data can be obtained from any input source; here we will obtain a dataset from the [recount2](https://jhubiostatistics.shinyapps.io/recount/) database. This database provides > 2000 human RNA-Seq experiments that have been processed using a consistent pipeline. We have written a wrapper function to download a count matrix from `recount2`, given an SRA ID. 

```{r, warning=F, message=F}
source(system.file("dl_recount.R", package = "clustifyR"))

pbmc_data <- dl_recount("SRP051688")

pbmc_data$read_counts[1:5, 1:5]

pbmc_data$meta_data[1:5, ]
```

`clustifyR` provides the `clustify()` function, which takes as input a matrix of single cell RNA-seq expression values, a metadata table with cluster assignments, and a matrix with bulk RNA seq datasets. 

As the heatmap below shows, most cell types are accurately classified using the default correlation method (pearson) and a list of highly variable genes in the single cell dataset identified using `Seurat`'s VMR method.

```{r fig.height=8, fig.width=12}
# run correlation (pearson by default)
res <- clustify(expr_mat = pbmc4k_matrix,
               metadata = pbmc4k_meta,
               bulk_mat = pbmc_bulk_matrix,
               query_gene_list = pbmc4k_vargenes,
               cluster_col = "classified",
               compute_method = corr_coef)

# peek at correlation matrix
res[1:5, 1:5]

# view heatmap of correlation coefficients
library(ComplexHeatmap)
library(circlize)
ComplexHeatmap::Heatmap(res, colorRamp2(c(0, 1), c("white", "black")))
```

### Plot cluster identities and correlation coefficients

`clustifyR` provides functions to plot single cells uing pre-calculated tSNE embeddings (or those from any other dimensionality reduction method), and to plot correlation coefficients on those projections. The plots below show that the highest correlations between these bulk RNA-seq data and the test scRNA-seq dataset are restricted to the correct cell clusters.

```{r plot, fig.height=8, fig.width=12}
# plot tSNE using known identities
plot_tsne(pbmc4k_meta, feature = "classified")

# plot correlation coefficients on tSNE for each identity class
plot_cor(res,
         pbmc4k_meta,
         colnames(res)[c(1, 5)],
         cluster_col = "classified")
```


### Correlation per cell

```{r fig.height=8, fig.width=12}
# run correlation (pearson by default)
res <- clustify(expr_mat = pbmc4k_matrix,
               metadata = pbmc4k_meta,
               bulk_mat = pbmc_bulk_matrix,
               query_genes = pbmc4k_vargenes,
               cluster_col = "rn",
               compute_method = corr_coef, 
               per_cell = T)

plot_cor(res,
         pbmc4k_meta,
         colnames(res)[c(1, 5)],
         cluster_col = "classified")
```

### Obtain p-value for similarity using a permutation test

In addition to automatically generating a correlation matrix, `clustifyR` can calculate correlation p-values using permutation with the `clustify()` function. As with raw correlation coefficients, plotting a heatmap of -log(p-values) demonstrates that the most significant correlations are between corresponding bulk and single cell clusters.

```{r fig.height = 8, fig.width = 12}
# re-run clustify with if_permute = TRUE, num_permute=1000 and return_full = TRUE
res <- clustify(pbmc4k_matrix,
               pbmc4k_meta,
               pbmc_bulk_matrix,
               pbmc4k_vargenes,
               num_perm = 1000,
               cluster_col = "classified",
               compute_method = corr_coef,
               return_full = TRUE)

res$p_val[1:5, 1:5]

# view heatmap of correlation coefficients
Heatmap(-log(res$p_val+.01,10), colorRamp2(c(0, 2), c("white", "black")))
```

### Using gene lists to classify cells (ranked or unranked)

```{r}
# generate a dataframe of markers from Seurat (a common format)
pbmc4k_mm <- matrixize_markers(pbmc4k_markers)
head(pbmc4k_mm)
```

```{r}
# or just lists of genes
NKm <- c("GNLY", "NKG7", "ZBTB32", "IL2RB")
DCm <- c("FCER1A", "CST3", "ITGAX", "IRF8")
Bm <- c("MS4A1", "CD79A","CD79B","CD37")

markers <- data_frame(Bm, DCm, NKm)
markers

```

Generate binary expression calls from single cell data. 

```{r}
pbmc4k_avgb <- binarize_expr(pbmc4k_avg)
pbmc4k_avgb[1:5, c(1, 2, 10)]
```

#### Classify using a hypergeometric test

Here's an example using ~100 markers per cluster

```{r}
res <- compare_lists(pbmc4k_avgb, pbmc4k_mm)

# rename columns 
colnames(res) <- left_join(data_frame(cluster = colnames(res)),
                           unique(data_frame(cluster = pbmc4k_meta$cluster,
                                             classified =  pbmc4k_meta$classified)), 
                           by = "cluster") %>% 
  pull(classified)

rownames(res) <- colnames(res)

Heatmap(-log10(res + 1), 
        cluster_rows = FALSE,
        cluster_columns = FALSE, 
        heatmap_legend_param = list(title = "-log10(adjP)"))
```


Here's an example using ~100 markers per cluster

```{r}
res2 <- compare_lists(pbmc4k_avgb, markers)

rownames(res2) <- left_join(data_frame(cluster = rownames(res2)),
                           unique(data_frame(cluster = pbmc4k_meta$cluster,
                                             classified =  pbmc4k_meta$classified)), 
                           by = "cluster") %>% 
  pull(classified)

Heatmap(-log10(res2 + 1), 
        cluster_rows = FALSE, 
        cluster_columns = FALSE, 
        heatmap_legend_param = list(title = "-log10(adjP)"))

```
