---
title: 'Cluster classification using mark gene lists'
date: '`r Sys.Date()`'
output:
  rmarkdown::html_vignette:
    toc: true
    toc_depth: 3
    vignette: >
      %\VignetteIndexEntry{clustifyR-genelists}
      %\VignetteEngine{knitr::rmarkdown}
      %\VignetteEncoding{UTF-8}
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
knitr::opts_chunk$set(warning = FALSE)
knitr::opts_chunk$set(message = FALSE)
```

## Unranked list: Preprocessing gene list into dataframe/matrix form

```{r}
library('recount')
library('biomaRt')
library('dplyr')
library('purrr')
library('stringr')
library("ComplexHeatmap")
```

```{r}
library(clustifyR)
# from dataframe generated by Seurat
pbmc4k_mm <- matrixize_markers(pbmc4k_markers)
head(pbmc4k_mm)

# or just lists of genes
NKm <- c("GNLY", "NKG7")
DCm <- c("FCER1A", "CST3")
Bm <- c("MS4A1", "CD79A")
mm2 <- data.frame(Bm, DCm, NKm)
```

## Unranked list: Binarize single cell gene expression data

```{r}
pbmc4k_avgb <- binarize_expr(pbmc4k_avg)
```

## Unranked list: Classify

```{r}
res <- hyperp(pbmc4k_avgb,pbmc4k_mm)
Heatmap(-log10(res), cluster_rows = FALSE, cluster_columns = FALSE, heatmap_legend_param = list(title = "-log10(adjP)"))

res2 <- hyperp(pbmc4k_avgb,mm2)
Heatmap(-log10(res2), cluster_rows = FALSE, cluster_columns = FALSE, heatmap_legend_param = list(title = "-log10(adjP)"))
```

## Ranked list: Preprocessing gene list into dataframe/matrix form

```{r}
pbmc4k_rm <- matrixize_markers(pbmc4k_markers, TRUE)
```

## Ranked list: Classify

```{r}
gene_constraints <- list(rownames(pbmc4k_avg),rownames(pbmc4k_rm))
pbmc4k_avg_d<- select_gene_subset(pbmc4k_avg, gene_constraints)
pbmc4k_rm_d <- select_gene_subset(t, gene_constraints)
out <- lapply(colnames(pbmc4k_avg_d),
              function(x){
                per_col <- lapply(colnames(pbmc4k_rm_d),
                                  function(y){
                                    compute_similarity(pbmc4k_avg_d[,x],
                                                       pbmc4k_rm_d[,y], corr_coef,
                                                       method = "spearman")})
                do.call(cbind, per_col)
              })

res <- do.call(rbind, out)
rownames(res) <- colnames(pbmc4k_avg_d)
colnames(res) <- colnames(pbmc4k_rm_d)

plot_cor(res,
         pbmc4k_meta,
         colnames(res)[1:7])
```
