<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><meta name="description" content="Combined function to compare scRNA-seq data to
bulk RNA-seq data and marker list"><title>Combined function to compare scRNA-seq data to
bulk RNA-seq data and marker list — clustify_nudge • clustifyr</title><!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png"><link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png"><link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png"><link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png"><link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png"><link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png"><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.1.0/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.1.0/bootstrap.bundle.min.js"></script><link href="../deps/_Source%20Sans%20Pro-0.4.1/font.css" rel="stylesheet"><link href="../deps/_Source%20Code%20Pro-0.4.1/font.css" rel="stylesheet"><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- bootstrap-toc --><script src="https://cdn.rawgit.com/afeld/bootstrap-toc/v1.0.1/dist/bootstrap-toc.min.js"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Combined function to compare scRNA-seq data to
bulk RNA-seq data and marker list — clustify_nudge"><meta property="og:description" content="Combined function to compare scRNA-seq data to
bulk RNA-seq data and marker list"><meta property="og:image" content="http://rnabioco.github.io/clustifyr/logo.svg"><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--><script defer data-domain="rnabioco.github.io" src="https://plausible.io/js/plausible.js"></script></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-dark navbar-expand-lg bg-dark"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">clustifyr</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="Released version">1.8.0</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="active nav-item">
  <a class="nav-link" href="../reference/index.html">Reference</a>
</li>
<li class="nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-articles">Articles</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-articles">
    <a class="dropdown-item" href="../articles/clustifyR.html">Introduction to clustifyr</a>
    <a class="dropdown-item" href="../articles/geo-annotations.html">Improving NCBI GEO submissions of scRNA-seq data</a>
  </div>
</li>
      </ul><form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off"></form>

      <ul class="navbar-nav"><li class="nav-item">
  <a class="external-link nav-link" href="https://github.com/rnabioco/clustifyr/" aria-label="github">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul></div>

    
  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="../logo.svg" class="logo" alt=""><h1>Combined function to compare scRNA-seq data to
bulk RNA-seq data and marker list</h1>
      <small class="dont-index">Source: <a href="https://github.com/rnabioco/clustifyr/blob/HEAD/R/utils.R" class="external-link"><code>R/utils.R</code></a></small>
      <div class="d-none name"><code>clustify_nudge.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p>Combined function to compare scRNA-seq data to
bulk RNA-seq data and marker list</p>
    </div>

    <div class="section level2">
    <h2 id="ref-usage">Usage<a class="anchor" aria-label="anchor" href="#ref-usage"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="fu">clustify_nudge</span><span class="op">(</span><span class="va">input</span>, <span class="va">...</span><span class="op">)</span>

<span class="co"># S3 method for default</span>
<span class="fu">clustify_nudge</span><span class="op">(</span>
  <span class="va">input</span>,
  <span class="va">ref_mat</span>,
  <span class="va">marker</span>,
  metadata <span class="op">=</span> <span class="cn">NULL</span>,
  cluster_col <span class="op">=</span> <span class="cn">NULL</span>,
  query_genes <span class="op">=</span> <span class="cn">NULL</span>,
  compute_method <span class="op">=</span> <span class="st">"spearman"</span>,
  weight <span class="op">=</span> <span class="fl">1</span>,
  seurat_out <span class="op">=</span> <span class="cn">FALSE</span>,
  threshold <span class="op">=</span> <span class="op">-</span><span class="cn">Inf</span>,
  dr <span class="op">=</span> <span class="st">"umap"</span>,
  norm <span class="op">=</span> <span class="st">"diff"</span>,
  call <span class="op">=</span> <span class="cn">TRUE</span>,
  marker_inmatrix <span class="op">=</span> <span class="cn">TRUE</span>,
  mode <span class="op">=</span> <span class="st">"rank"</span>,
  obj_out <span class="op">=</span> <span class="cn">FALSE</span>,
  rename_prefix <span class="op">=</span> <span class="cn">NULL</span>,
  lookuptable <span class="op">=</span> <span class="cn">NULL</span>,
  <span class="va">...</span>
<span class="op">)</span>

<span class="co"># S3 method for Seurat</span>
<span class="fu">clustify_nudge</span><span class="op">(</span>
  <span class="va">input</span>,
  <span class="va">ref_mat</span>,
  <span class="va">marker</span>,
  cluster_col <span class="op">=</span> <span class="cn">NULL</span>,
  query_genes <span class="op">=</span> <span class="cn">NULL</span>,
  compute_method <span class="op">=</span> <span class="st">"spearman"</span>,
  weight <span class="op">=</span> <span class="fl">1</span>,
  seurat_out <span class="op">=</span> <span class="cn">TRUE</span>,
  obj_out <span class="op">=</span> <span class="cn">FALSE</span>,
  threshold <span class="op">=</span> <span class="op">-</span><span class="cn">Inf</span>,
  dr <span class="op">=</span> <span class="st">"umap"</span>,
  norm <span class="op">=</span> <span class="st">"diff"</span>,
  marker_inmatrix <span class="op">=</span> <span class="cn">TRUE</span>,
  mode <span class="op">=</span> <span class="st">"rank"</span>,
  rename_prefix <span class="op">=</span> <span class="cn">NULL</span>,
  <span class="va">...</span>
<span class="op">)</span></code></pre></div>
    </div>

    <div class="section level2">
    <h2 id="arguments">Arguments<a class="anchor" aria-label="anchor" href="#arguments"></a></h2>
    <dl><dt>input</dt>
<dd><p>express matrix or object</p></dd>
<dt>...</dt>
<dd><p>passed to matrixize_markers</p></dd>
<dt>ref_mat</dt>
<dd><p>reference expression matrix</p></dd>
<dt>marker</dt>
<dd><p>matrix of markers</p></dd>
<dt>metadata</dt>
<dd><p>cell cluster assignments, supplied as a vector
or data.frame. If
data.frame is supplied then <code>cluster_col</code> needs to be set.</p></dd>
<dt>cluster_col</dt>
<dd><p>column in metadata that contains cluster ids per cell.
Will default to first
column of metadata if not supplied.
Not required if running correlation per cell.</p></dd>
<dt>query_genes</dt>
<dd><p>A vector of genes of interest to compare.
If NULL, then common genes between
the expr_mat and ref_mat will be used for comparision.</p></dd>
<dt>compute_method</dt>
<dd><p>method(s) for computing similarity scores</p></dd>
<dt>weight</dt>
<dd><p>relative weight for the gene list scores,
when added to correlation score</p></dd>
<dt>seurat_out</dt>
<dd><p>output cor matrix or called seurat object</p></dd>
<dt>threshold</dt>
<dd><p>identity calling minimum score threshold,
only used when obj_out = T</p></dd>
<dt>dr</dt>
<dd><p>stored dimension reduction</p></dd>
<dt>norm</dt>
<dd><p>whether and how the results are normalized</p></dd>
<dt>call</dt>
<dd><p>make call or just return score matrix</p></dd>
<dt>marker_inmatrix</dt>
<dd><p>whether markers genes are already
in preprocessed matrix form</p></dd>
<dt>mode</dt>
<dd><p>use marker expression pct or ranked cor score for nudging</p></dd>
<dt>obj_out</dt>
<dd><p>whether to output object instead of cor matrix</p></dd>
<dt>rename_prefix</dt>
<dd><p>prefix to add to type and r column names</p></dd>
<dt>lookuptable</dt>
<dd><p>if not supplied, will look in built-in
table for object parsing</p></dd>
</dl></div>
    <div class="section level2">
    <h2 id="value">Value<a class="anchor" aria-label="anchor" href="#value"></a></h2>
    <p>single cell object, or matrix of numeric values,
clusters from input as row names, cell types from ref_mat as column names</p>
    </div>

    <div class="section level2">
    <h2 id="ref-examples">Examples<a class="anchor" aria-label="anchor" href="#ref-examples"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"></span>
<span class="r-in"><span class="co"># Seurat3</span></span>
<span class="r-in"><span class="fu">clustify_nudge</span><span class="op">(</span></span>
<span class="r-in">    input <span class="op">=</span> <span class="va">s_small3</span>,</span>
<span class="r-in">    ref_mat <span class="op">=</span> <span class="va">cbmc_ref</span>,</span>
<span class="r-in">    marker <span class="op">=</span> <span class="va">cbmc_m</span>,</span>
<span class="r-in">    cluster_col <span class="op">=</span> <span class="st">"RNA_snn_res.1"</span>,</span>
<span class="r-in">    threshold <span class="op">=</span> <span class="fl">0.8</span>,</span>
<span class="r-in">    seurat_out <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span class="r-in">    mode <span class="op">=</span> <span class="st">"pct"</span>,</span>
<span class="r-in">    dr <span class="op">=</span> <span class="st">"tsne"</span></span>
<span class="r-in"><span class="op">)</span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> object data retrieval complete, moving to similarity computation</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> using # of genes: 19</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> similarity computation completed, matrix of 3 x 13, preparing output</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>             B CD14+ Mono CD16+ Mono      CD34+     CD4 T      CD8 T</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 0 -0.94493203 -0.9156671 -0.2220272 -0.2825193 0.2887232 0.37971905</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 1  0.04838758  0.7219124  0.7316084  0.3238704 0.3472936 0.19775419</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 2  0.62480323 -0.2550175  0.4973646  0.3895667 0.1769981 0.01448416</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>           DC      Eryth Memory CD4 T         Mk         NK Naive CD4 T</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 0 -0.3168822  0.4203598    0.1987308  0.6204476 0.08688022 -0.08992247</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 1  0.5865428  0.1736468    0.5311026 -0.2163932 0.25334722  0.51050337</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 2  0.5146717 -0.1716881    0.1094072 -0.6697973 0.14695522  0.23853307</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>        pDCs</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 0 0.1134067</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 1 0.4370882</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 2 0.3758960</span>
<span class="r-in"></span>
<span class="r-in"><span class="co"># Matrix</span></span>
<span class="r-in"><span class="fu">clustify_nudge</span><span class="op">(</span></span>
<span class="r-in">    input <span class="op">=</span> <span class="va">pbmc_matrix_small</span>,</span>
<span class="r-in">    ref_mat <span class="op">=</span> <span class="va">cbmc_ref</span>,</span>
<span class="r-in">    metadata <span class="op">=</span> <span class="va">pbmc_meta</span>,</span>
<span class="r-in">    marker <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/matrix.html" class="external-link">as.matrix</a></span><span class="op">(</span><span class="va">cbmc_m</span><span class="op">)</span>,</span>
<span class="r-in">    query_genes <span class="op">=</span> <span class="va">pbmc_vargenes</span>,</span>
<span class="r-in">    cluster_col <span class="op">=</span> <span class="st">"classified"</span>,</span>
<span class="r-in">    threshold <span class="op">=</span> <span class="fl">0.8</span>,</span>
<span class="r-in">    call <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span class="r-in">    marker_inmatrix <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span class="r-in">    mode <span class="op">=</span> <span class="st">"pct"</span></span>
<span class="r-in"><span class="op">)</span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> using # of genes: 599</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> similarity computation completed, matrix of 9 x 13, preparing output</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>                       B CD14+ Mono  CD16+ Mono     CD34+       CD4 T</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> B             0.9093577 -0.2452939  0.22341215 0.6248639  0.39898892</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> CD14+ Mono   -0.3184633  0.9145428  0.52579175 0.5103056  0.20913853</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> CD8 T        -0.2866333 -0.2920874  0.17944732 0.5663773  0.64950686</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> DC           -0.1927865  0.3971091  0.41292804 0.5694074  0.31768980</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> FCGR3A+ Mono -0.2924419  0.6387074  0.92949138 0.5331927  0.22686950</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Memory CD4 T -0.2094668 -0.1734003  0.22245854 0.6265186  0.86108263</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> NK           -0.3523479 -0.3353237  0.15173936 0.5350808  0.47524944</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Naive CD4 T  -0.2115884 -0.2355533  0.19645924 0.6480900  0.80013677</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Platelet     -0.6782452 -0.3759687 -0.08362827 0.2455947 -0.05758481</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>                    CD8 T          DC     Eryth Memory CD4 T         Mk</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> B             0.43921857  0.35367232 0.4817829    0.4803367 -0.6748463</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> CD14+ Mono    0.25084659  0.51955311 0.4559350    0.4553454 -0.4660850</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> CD8 T         0.59418188  0.26878200 0.5007841    0.6323214 -0.5987391</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> DC            0.28543102  0.84933995 0.4470039    0.4764929 -0.5578229</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> FCGR3A+ Mono  0.35105547  0.48594006 0.4604308    0.4614474 -0.4760099</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Memory CD4 T  0.65776299  0.30625770 0.5395170    0.7128879 -0.6336165</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> NK            0.69267415  0.23172856 0.4578726    0.5665250 -0.6199834</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Naive CD4 T   0.72313292  0.29335212 0.5528441    0.7277404 -0.6269043</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Platelet     -0.07272109 -0.06019224 0.3068445    0.1457308  0.7319575</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>                      NK Naive CD4 T      pDCs</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> B            -0.3536418   0.5407149 0.4801131</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> CD14+ Mono   -0.2621333   0.4564707 0.3710598</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> CD8 T         0.5582283   0.6036230 0.3519523</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> DC           -0.1659100   0.4475783 0.6825388</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> FCGR3A+ Mono -0.1921798   0.4574654 0.3755701</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Memory CD4 T -0.1208282   0.6871755 0.3668161</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> NK            0.8940904   0.5505877 0.3186506</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Naive CD4 T  -0.1890644   0.6967073 0.3594580</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Platelet     -0.7349565   0.1609963 0.1110947</span>
</code></pre></div>
    </div>
  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p></p><p>An R package developed by the <a href="https://medschool.cuanschutz.edu/rbi" class="external-link">RNA Bioscience Initiative</a></p>
</div>

<div class="pkgdown-footer-right">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.3.</p>
</div>

    </footer></div>

  

  

  </body></html>

