<!-- Generated by pkgdown: do not edit by hand -->
<!DOCTYPE html>
<html lang="en">
  <head>
  <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<title>Combined function to compare scRNA-seq data to
bulk RNA-seq data and marker list — clustify_nudge • clustifyr</title>

<!-- favicons -->
<link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png">
<link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png">
<link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png" />
<link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png" />
<link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png" />
<link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png" />

<!-- jquery -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script>
<!-- Bootstrap -->

<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous" />

<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script>

<!-- bootstrap-toc -->
<link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script>

<!-- Font Awesome icons -->
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous" />
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous" />

<!-- clipboard.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script>

<!-- headroom.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script>

<!-- pkgdown -->
<link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script>




<meta property="og:title" content="Combined function to compare scRNA-seq data to
bulk RNA-seq data and marker list — clustify_nudge" />
<meta property="og:description" content="Combined function to compare scRNA-seq data to
bulk RNA-seq data and marker list" />
<meta property="og:image" content="http://rnabioco.github.io/clustifyr/logo.svg" />




<!-- mathjax -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script>

<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->



  </head>

  <body data-spy="scroll" data-target="#toc">
    <div class="container template-reference-topic">
      <header>
      <div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">clustifyr</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">1.5.1</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="../index.html">
    <span class="fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="../articles/clustifyR.html">Introduction to clustifyr</a>
    </li>
    <li>
      <a href="../articles/geo-annotations.html">Improving NCBI GEO submissions of scRNA-seq data</a>
    </li>
  </ul>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/rnabioco/clustifyr/">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
      
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header>

<div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Combined function to compare scRNA-seq data to
bulk RNA-seq data and marker list</h1>
    <small class="dont-index">Source: <a href='https://github.com/rnabioco/clustifyr/blob/master/R/utils.R'><code>R/utils.R</code></a></small>
    <div class="hidden name"><code>clustify_nudge.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>Combined function to compare scRNA-seq data to
bulk RNA-seq data and marker list</p>
    </div>

    <pre class="usage"><span class='fu'>clustify_nudge</span><span class='op'>(</span><span class='va'>input</span>, <span class='va'>...</span><span class='op'>)</span>

<span class='co'># S3 method for default</span>
<span class='fu'>clustify_nudge</span><span class='op'>(</span>
  <span class='va'>input</span>,
  <span class='va'>ref_mat</span>,
  <span class='va'>marker</span>,
  metadata <span class='op'>=</span> <span class='cn'>NULL</span>,
  cluster_col <span class='op'>=</span> <span class='cn'>NULL</span>,
  query_genes <span class='op'>=</span> <span class='cn'>NULL</span>,
  compute_method <span class='op'>=</span> <span class='st'>"spearman"</span>,
  weight <span class='op'>=</span> <span class='fl'>1</span>,
  seurat_out <span class='op'>=</span> <span class='cn'>FALSE</span>,
  threshold <span class='op'>=</span> <span class='op'>-</span><span class='cn'>Inf</span>,
  dr <span class='op'>=</span> <span class='st'>"umap"</span>,
  norm <span class='op'>=</span> <span class='st'>"diff"</span>,
  call <span class='op'>=</span> <span class='cn'>TRUE</span>,
  marker_inmatrix <span class='op'>=</span> <span class='cn'>TRUE</span>,
  mode <span class='op'>=</span> <span class='st'>"rank"</span>,
  obj_out <span class='op'>=</span> <span class='cn'>FALSE</span>,
  rename_prefix <span class='op'>=</span> <span class='cn'>NULL</span>,
  lookuptable <span class='op'>=</span> <span class='cn'>NULL</span>,
  <span class='va'>...</span>
<span class='op'>)</span>

<span class='co'># S3 method for Seurat</span>
<span class='fu'>clustify_nudge</span><span class='op'>(</span>
  <span class='va'>input</span>,
  <span class='va'>ref_mat</span>,
  <span class='va'>marker</span>,
  cluster_col <span class='op'>=</span> <span class='cn'>NULL</span>,
  query_genes <span class='op'>=</span> <span class='cn'>NULL</span>,
  compute_method <span class='op'>=</span> <span class='st'>"spearman"</span>,
  weight <span class='op'>=</span> <span class='fl'>1</span>,
  seurat_out <span class='op'>=</span> <span class='cn'>TRUE</span>,
  obj_out <span class='op'>=</span> <span class='cn'>FALSE</span>,
  threshold <span class='op'>=</span> <span class='op'>-</span><span class='cn'>Inf</span>,
  dr <span class='op'>=</span> <span class='st'>"umap"</span>,
  norm <span class='op'>=</span> <span class='st'>"diff"</span>,
  marker_inmatrix <span class='op'>=</span> <span class='cn'>TRUE</span>,
  mode <span class='op'>=</span> <span class='st'>"rank"</span>,
  rename_prefix <span class='op'>=</span> <span class='cn'>NULL</span>,
  <span class='va'>...</span>
<span class='op'>)</span></pre>

    <h2 class="hasAnchor" id="arguments"><a class="anchor" href="#arguments"></a>Arguments</h2>
    <table class="ref-arguments">
    <colgroup><col class="name" /><col class="desc" /></colgroup>
    <tr>
      <th>input</th>
      <td><p>express matrix or object</p></td>
    </tr>
    <tr>
      <th>...</th>
      <td><p>passed to matrixize_markers</p></td>
    </tr>
    <tr>
      <th>ref_mat</th>
      <td><p>reference expression matrix</p></td>
    </tr>
    <tr>
      <th>marker</th>
      <td><p>matrix of markers</p></td>
    </tr>
    <tr>
      <th>metadata</th>
      <td><p>cell cluster assignments, supplied as a vector
or data.frame. If
data.frame is supplied then <code>cluster_col</code> needs to be set.</p></td>
    </tr>
    <tr>
      <th>cluster_col</th>
      <td><p>column in metadata that contains cluster ids per cell.
Will default to first
column of metadata if not supplied.
Not required if running correlation per cell.</p></td>
    </tr>
    <tr>
      <th>query_genes</th>
      <td><p>A vector of genes of interest to compare.
If NULL, then common genes between
the expr_mat and ref_mat will be used for comparision.</p></td>
    </tr>
    <tr>
      <th>compute_method</th>
      <td><p>method(s) for computing similarity scores</p></td>
    </tr>
    <tr>
      <th>weight</th>
      <td><p>relative weight for the gene list scores,
when added to correlation score</p></td>
    </tr>
    <tr>
      <th>seurat_out</th>
      <td><p>output cor matrix or called seurat object</p></td>
    </tr>
    <tr>
      <th>threshold</th>
      <td><p>identity calling minimum score threshold,
only used when obj_out = T</p></td>
    </tr>
    <tr>
      <th>dr</th>
      <td><p>stored dimension reduction</p></td>
    </tr>
    <tr>
      <th>norm</th>
      <td><p>whether and how the results are normalized</p></td>
    </tr>
    <tr>
      <th>call</th>
      <td><p>make call or just return score matrix</p></td>
    </tr>
    <tr>
      <th>marker_inmatrix</th>
      <td><p>whether markers genes are already
in preprocessed matrix form</p></td>
    </tr>
    <tr>
      <th>mode</th>
      <td><p>use marker expression pct or ranked cor score for nudging</p></td>
    </tr>
    <tr>
      <th>obj_out</th>
      <td><p>whether to output object instead of cor matrix</p></td>
    </tr>
    <tr>
      <th>rename_prefix</th>
      <td><p>prefix to add to type and r column names</p></td>
    </tr>
    <tr>
      <th>lookuptable</th>
      <td><p>if not supplied, will look in built-in
table for object parsing</p></td>
    </tr>
    </table>

    <h2 class="hasAnchor" id="value"><a class="anchor" href="#value"></a>Value</h2>

    <p>single cell object, or matrix of numeric values,
clusters from input as row names, cell types from ref_mat as column names</p>

    <h2 class="hasAnchor" id="examples"><a class="anchor" href="#examples"></a>Examples</h2>
    <pre class="examples"><span class='r-in'></span>
<span class='r-in'><span class='co'># Seurat3</span></span>
<span class='r-in'><span class='fu'>clustify_nudge</span><span class='op'>(</span></span>
<span class='r-in'>    input <span class='op'>=</span> <span class='va'>s_small3</span>,</span>
<span class='r-in'>    ref_mat <span class='op'>=</span> <span class='va'>cbmc_ref</span>,</span>
<span class='r-in'>    marker <span class='op'>=</span> <span class='va'>cbmc_m</span>,</span>
<span class='r-in'>    cluster_col <span class='op'>=</span> <span class='st'>"RNA_snn_res.1"</span>,</span>
<span class='r-in'>    threshold <span class='op'>=</span> <span class='fl'>0.8</span>,</span>
<span class='r-in'>    seurat_out <span class='op'>=</span> <span class='cn'>FALSE</span>,</span>
<span class='r-in'>    mode <span class='op'>=</span> <span class='st'>"pct"</span>,</span>
<span class='r-in'>    dr <span class='op'>=</span> <span class='st'>"tsne"</span></span>
<span class='r-in'><span class='op'>)</span></span>
<span class='r-msg co'><span class='r-pr'>#&gt;</span> object data retrieval complete, moving to similarity computation</span>
<span class='r-msg co'><span class='r-pr'>#&gt;</span> using # of genes: 19</span>
<span class='r-msg co'><span class='r-pr'>#&gt;</span> similarity computation completed, matrix of 3 x 13, preparing output</span>
<span class='r-out co'><span class='r-pr'>#&gt;</span>             B CD14+ Mono CD16+ Mono      CD34+     CD4 T      CD8 T         DC</span>
<span class='r-out co'><span class='r-pr'>#&gt;</span> 0 -0.94493203 -0.9156671 -0.2220272 -0.2825193 0.2887232 0.37971905 -0.3168822</span>
<span class='r-out co'><span class='r-pr'>#&gt;</span> 1  0.04838758  0.7219124  0.7316084  0.3238704 0.3472936 0.19775419  0.5865428</span>
<span class='r-out co'><span class='r-pr'>#&gt;</span> 2  0.62480323 -0.2550175  0.4973646  0.3895667 0.1769981 0.01448416  0.5146717</span>
<span class='r-out co'><span class='r-pr'>#&gt;</span>        Eryth Memory CD4 T         Mk Naive CD4 T         NK      pDCs</span>
<span class='r-out co'><span class='r-pr'>#&gt;</span> 0  0.4203598    0.1987308  0.6204476 -0.08992247 0.08688022 0.1134067</span>
<span class='r-out co'><span class='r-pr'>#&gt;</span> 1  0.1736468    0.5311026 -0.2163932  0.51050337 0.25334722 0.4370882</span>
<span class='r-out co'><span class='r-pr'>#&gt;</span> 2 -0.1716881    0.1094072 -0.6697973  0.23853307 0.14695522 0.3758960</span>
<span class='r-in'></span>
<span class='r-in'><span class='co'># Matrix</span></span>
<span class='r-in'><span class='fu'>clustify_nudge</span><span class='op'>(</span></span>
<span class='r-in'>    input <span class='op'>=</span> <span class='va'>pbmc_matrix_small</span>,</span>
<span class='r-in'>    ref_mat <span class='op'>=</span> <span class='va'>cbmc_ref</span>,</span>
<span class='r-in'>    metadata <span class='op'>=</span> <span class='va'>pbmc_meta</span>,</span>
<span class='r-in'>    marker <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/matrix.html'>as.matrix</a></span><span class='op'>(</span><span class='va'>cbmc_m</span><span class='op'>)</span>,</span>
<span class='r-in'>    query_genes <span class='op'>=</span> <span class='va'>pbmc_vargenes</span>,</span>
<span class='r-in'>    cluster_col <span class='op'>=</span> <span class='st'>"classified"</span>,</span>
<span class='r-in'>    threshold <span class='op'>=</span> <span class='fl'>0.8</span>,</span>
<span class='r-in'>    call <span class='op'>=</span> <span class='cn'>FALSE</span>,</span>
<span class='r-in'>    marker_inmatrix <span class='op'>=</span> <span class='cn'>FALSE</span>,</span>
<span class='r-in'>    mode <span class='op'>=</span> <span class='st'>"pct"</span></span>
<span class='r-in'><span class='op'>)</span></span>
<span class='r-msg co'><span class='r-pr'>#&gt;</span> using # of genes: 599</span>
<span class='r-msg co'><span class='r-pr'>#&gt;</span> similarity computation completed, matrix of 9 x 13, preparing output</span>
<span class='r-out co'><span class='r-pr'>#&gt;</span>                       B CD14+ Mono  CD16+ Mono     CD34+       CD4 T</span>
<span class='r-out co'><span class='r-pr'>#&gt;</span> B             0.9093577 -0.2452939  0.22341215 0.6248639  0.39898892</span>
<span class='r-out co'><span class='r-pr'>#&gt;</span> CD14+ Mono   -0.3184633  0.9145428  0.52579175 0.5103056  0.20913853</span>
<span class='r-out co'><span class='r-pr'>#&gt;</span> CD8 T        -0.2866333 -0.2920874  0.17944732 0.5663773  0.64950686</span>
<span class='r-out co'><span class='r-pr'>#&gt;</span> DC           -0.1927865  0.3971091  0.41292804 0.5694074  0.31768980</span>
<span class='r-out co'><span class='r-pr'>#&gt;</span> FCGR3A+ Mono -0.2924419  0.6387074  0.92949138 0.5331927  0.22686950</span>
<span class='r-out co'><span class='r-pr'>#&gt;</span> Memory CD4 T -0.2094668 -0.1734003  0.22245854 0.6265186  0.86108263</span>
<span class='r-out co'><span class='r-pr'>#&gt;</span> Naive CD4 T  -0.2115884 -0.2355533  0.19645924 0.6480900  0.80013677</span>
<span class='r-out co'><span class='r-pr'>#&gt;</span> NK           -0.3523479 -0.3353237  0.15173936 0.5350808  0.47524944</span>
<span class='r-out co'><span class='r-pr'>#&gt;</span> Platelet     -0.6782452 -0.3759687 -0.08362827 0.2455947 -0.05758481</span>
<span class='r-out co'><span class='r-pr'>#&gt;</span>                    CD8 T          DC     Eryth Memory CD4 T         Mk</span>
<span class='r-out co'><span class='r-pr'>#&gt;</span> B             0.43921857  0.35367232 0.4817829    0.4803367 -0.6748463</span>
<span class='r-out co'><span class='r-pr'>#&gt;</span> CD14+ Mono    0.25084659  0.51955311 0.4559350    0.4553454 -0.4660850</span>
<span class='r-out co'><span class='r-pr'>#&gt;</span> CD8 T         0.59418188  0.26878200 0.5007841    0.6323214 -0.5987391</span>
<span class='r-out co'><span class='r-pr'>#&gt;</span> DC            0.28543102  0.84933995 0.4470039    0.4764929 -0.5578229</span>
<span class='r-out co'><span class='r-pr'>#&gt;</span> FCGR3A+ Mono  0.35105547  0.48594006 0.4604308    0.4614474 -0.4760099</span>
<span class='r-out co'><span class='r-pr'>#&gt;</span> Memory CD4 T  0.65776299  0.30625770 0.5395170    0.7128879 -0.6336165</span>
<span class='r-out co'><span class='r-pr'>#&gt;</span> Naive CD4 T   0.72313292  0.29335212 0.5528441    0.7277404 -0.6269043</span>
<span class='r-out co'><span class='r-pr'>#&gt;</span> NK            0.69267415  0.23172856 0.4578726    0.5665250 -0.6199834</span>
<span class='r-out co'><span class='r-pr'>#&gt;</span> Platelet     -0.07272109 -0.06019224 0.3068445    0.1457308  0.7319575</span>
<span class='r-out co'><span class='r-pr'>#&gt;</span>              Naive CD4 T         NK      pDCs</span>
<span class='r-out co'><span class='r-pr'>#&gt;</span> B              0.5407149 -0.3536418 0.4801131</span>
<span class='r-out co'><span class='r-pr'>#&gt;</span> CD14+ Mono     0.4564707 -0.2621333 0.3710598</span>
<span class='r-out co'><span class='r-pr'>#&gt;</span> CD8 T          0.6036230  0.5582283 0.3519523</span>
<span class='r-out co'><span class='r-pr'>#&gt;</span> DC             0.4475783 -0.1659100 0.6825388</span>
<span class='r-out co'><span class='r-pr'>#&gt;</span> FCGR3A+ Mono   0.4574654 -0.1921798 0.3755701</span>
<span class='r-out co'><span class='r-pr'>#&gt;</span> Memory CD4 T   0.6871755 -0.1208282 0.3668161</span>
<span class='r-out co'><span class='r-pr'>#&gt;</span> Naive CD4 T    0.6967073 -0.1890644 0.3594580</span>
<span class='r-out co'><span class='r-pr'>#&gt;</span> NK             0.5505877  0.8940904 0.3186506</span>
<span class='r-out co'><span class='r-pr'>#&gt;</span> Platelet       0.1609963 -0.7349565 0.1110947</span>
</pre>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top">
      <h2 data-toc-skip>Contents</h2>
    </nav>
  </div>
</div>


      <footer>
      <div class="copyright">
  <p>Developed by Rui Fu, Kent Riemondy, <a href='https://medschool.cuanschutz.edu/rbi'>RNA Bioscience Initiative</a>.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.6.1.</p>
</div>

      </footer>
   </div>

  


  </body>
</html>


