<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><meta name="description" content="Compare scRNA-seq data to reference data."><title>Compare scRNA-seq data to reference data. — clustify • clustifyr</title><!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png"><link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png"><link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png"><link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png"><link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png"><link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png"><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.1.3/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.1.3/bootstrap.bundle.min.js"></script><link href="../deps/Source_Sans_Pro-0.4.2/font.css" rel="stylesheet"><link href="../deps/Source_Code_Pro-0.4.2/font.css" rel="stylesheet"><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- bootstrap-toc --><script src="https://cdn.rawgit.com/afeld/bootstrap-toc/v1.0.1/dist/bootstrap-toc.min.js"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Compare scRNA-seq data to reference data. — clustify"><meta property="og:description" content="Compare scRNA-seq data to reference data."><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--><script defer data-domain="rnabioco.github.io" src="https://plausible.io/js/plausible.js"></script></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-dark navbar-expand-lg bg-dark"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">clustifyr</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="Released version">1.11.1</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="nav-item">
  <a class="nav-link" href="../articles/clustifyr.html">Get started</a>
</li>
<li class="active nav-item">
  <a class="nav-link" href="../reference/index.html">Reference</a>
</li>
<li class="nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-articles">Articles</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-articles">
    <a class="dropdown-item" href="../articles/geo-annotations.html">Improving NCBI GEO submissions of scRNA-seq data</a>
  </div>
</li>
      </ul><form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off"></form>

      <ul class="navbar-nav"><li class="nav-item">
  <a class="external-link nav-link" href="https://github.com/rnabioco/clustifyr/" aria-label="github">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul></div>

    
  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="" class="logo" alt=""><h1>Compare scRNA-seq data to reference data.</h1>
      <small class="dont-index">Source: <a href="https://github.com/rnabioco/clustifyr/blob/HEAD/R/main.R" class="external-link"><code>R/main.R</code></a></small>
      <div class="d-none name"><code>clustify.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p>Compare scRNA-seq data to reference data.</p>
    </div>

    <div class="section level2">
    <h2 id="ref-usage">Usage<a class="anchor" aria-label="anchor" href="#ref-usage"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">clustify</span><span class="op">(</span><span class="va">input</span>, <span class="va">...</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># S3 method for default</span></span>
<span><span class="fu">clustify</span><span class="op">(</span></span>
<span>  <span class="va">input</span>,</span>
<span>  <span class="va">ref_mat</span>,</span>
<span>  metadata <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  cluster_col <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  query_genes <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  n_genes <span class="op">=</span> <span class="fl">1000</span>,</span>
<span>  per_cell <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  n_perm <span class="op">=</span> <span class="fl">0</span>,</span>
<span>  compute_method <span class="op">=</span> <span class="st">"spearman"</span>,</span>
<span>  pseudobulk_method <span class="op">=</span> <span class="st">"mean"</span>,</span>
<span>  verbose <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  lookuptable <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  rm0 <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  obj_out <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  seurat_out <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  vec_out <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  rename_prefix <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  threshold <span class="op">=</span> <span class="st">"auto"</span>,</span>
<span>  low_threshold_cell <span class="op">=</span> <span class="fl">0</span>,</span>
<span>  exclude_genes <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="op">)</span>,</span>
<span>  if_log <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  organism <span class="op">=</span> <span class="st">"hsapiens"</span>,</span>
<span>  plot_name <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  rds_name <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  expand_unassigned <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  <span class="va">...</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="co"># S3 method for Seurat</span></span>
<span><span class="fu">clustify</span><span class="op">(</span></span>
<span>  <span class="va">input</span>,</span>
<span>  <span class="va">ref_mat</span>,</span>
<span>  cluster_col <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  query_genes <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  n_genes <span class="op">=</span> <span class="fl">1000</span>,</span>
<span>  per_cell <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  n_perm <span class="op">=</span> <span class="fl">0</span>,</span>
<span>  compute_method <span class="op">=</span> <span class="st">"spearman"</span>,</span>
<span>  pseudobulk_method <span class="op">=</span> <span class="st">"mean"</span>,</span>
<span>  use_var_genes <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  dr <span class="op">=</span> <span class="st">"umap"</span>,</span>
<span>  seurat_out <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  obj_out <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  vec_out <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  threshold <span class="op">=</span> <span class="st">"auto"</span>,</span>
<span>  verbose <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  rm0 <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  rename_prefix <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  exclude_genes <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="op">)</span>,</span>
<span>  metadata <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  organism <span class="op">=</span> <span class="st">"hsapiens"</span>,</span>
<span>  plot_name <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  rds_name <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  expand_unassigned <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  <span class="va">...</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="co"># S3 method for SingleCellExperiment</span></span>
<span><span class="fu">clustify</span><span class="op">(</span></span>
<span>  <span class="va">input</span>,</span>
<span>  <span class="va">ref_mat</span>,</span>
<span>  cluster_col <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  query_genes <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  per_cell <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  n_perm <span class="op">=</span> <span class="fl">0</span>,</span>
<span>  compute_method <span class="op">=</span> <span class="st">"spearman"</span>,</span>
<span>  pseudobulk_method <span class="op">=</span> <span class="st">"mean"</span>,</span>
<span>  use_var_genes <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  dr <span class="op">=</span> <span class="st">"umap"</span>,</span>
<span>  seurat_out <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  obj_out <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  vec_out <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  threshold <span class="op">=</span> <span class="st">"auto"</span>,</span>
<span>  verbose <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  rm0 <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  rename_prefix <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  exclude_genes <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="op">)</span>,</span>
<span>  metadata <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  organism <span class="op">=</span> <span class="st">"hsapiens"</span>,</span>
<span>  plot_name <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  rds_name <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  expand_unassigned <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  <span class="va">...</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div class="section level2">
    <h2 id="arguments">Arguments<a class="anchor" aria-label="anchor" href="#arguments"></a></h2>
    <dl><dt>input</dt>
<dd><p>single-cell expression matrix or Seurat object</p></dd>


<dt>...</dt>
<dd><p>additional arguments to pass to compute_method function</p></dd>


<dt>ref_mat</dt>
<dd><p>reference expression matrix</p></dd>


<dt>metadata</dt>
<dd><p>cell cluster assignments,
supplied as a vector or data.frame.
If data.frame is supplied then <code>cluster_col</code> needs to be set.
Not required if running correlation per cell.</p></dd>


<dt>cluster_col</dt>
<dd><p>column in metadata that contains cluster ids per cell.
Will default to first column of metadata if not supplied.
Not required if running correlation per cell.</p></dd>


<dt>query_genes</dt>
<dd><p>A vector of genes of interest to compare. If NULL, then
common genes between the expr_mat and ref_mat
will be used for comparision.</p></dd>


<dt>n_genes</dt>
<dd><p>number of genes limit for Seurat variable genes, by default 1000,
set to 0 to use all variable genes (generally not recommended)</p></dd>


<dt>per_cell</dt>
<dd><p>if true run per cell, otherwise per cluster.</p></dd>


<dt>n_perm</dt>
<dd><p>number of permutations, set to 0 by default</p></dd>


<dt>compute_method</dt>
<dd><p>method(s) for computing similarity scores</p></dd>


<dt>pseudobulk_method</dt>
<dd><p>method used for summarizing clusters, options are mean (default), median, truncate (10% truncated mean), or trimean, max, min</p></dd>


<dt>verbose</dt>
<dd><p>whether to report certain variables chosen and steps</p></dd>


<dt>lookuptable</dt>
<dd><p>if not supplied, will look in built-in table
for object parsing</p></dd>


<dt>rm0</dt>
<dd><p>consider 0 as missing data, recommended for per_cell</p></dd>


<dt>obj_out</dt>
<dd><p>whether to output object instead of cor matrix</p></dd>


<dt>seurat_out</dt>
<dd><p>output cor matrix or called seurat object
(deprecated, use obj_out instead)</p></dd>


<dt>vec_out</dt>
<dd><p>only output a result vector in the same order as metadata</p></dd>


<dt>rename_prefix</dt>
<dd><p>prefix to add to type and r column names</p></dd>


<dt>threshold</dt>
<dd><p>identity calling minimum correlation score threshold,
only used when obj_out = TRUE</p></dd>


<dt>low_threshold_cell</dt>
<dd><p>option to remove clusters with too few cells</p></dd>


<dt>exclude_genes</dt>
<dd><p>a vector of gene names to throw out of query</p></dd>


<dt>if_log</dt>
<dd><p>input data is natural log,
averaging will be done on unlogged data</p></dd>


<dt>organism</dt>
<dd><p>for GO term analysis, organism name: human - 'hsapiens', mouse - 'mmusculus'</p></dd>


<dt>plot_name</dt>
<dd><p>name for saved pdf, if NULL then no file is written (default)</p></dd>


<dt>rds_name</dt>
<dd><p>name for saved rds of rank_diff, if NULL then no file is written (default)</p></dd>


<dt>expand_unassigned</dt>
<dd><p>test all ref clusters for unassigned results</p></dd>


<dt>use_var_genes</dt>
<dd><p>if providing a seurat object, use the variable genes
(stored in seurat_object@var.genes) as the query_genes.</p></dd>


<dt>dr</dt>
<dd><p>stored dimension reduction</p></dd>

</dl></div>
    <div class="section level2">
    <h2 id="value">Value<a class="anchor" aria-label="anchor" href="#value"></a></h2>
    

<p>single cell object with identity assigned in metadata,
or matrix of correlation values, clusters from input as row names, cell
types from ref_mat as column names</p>
    </div>

    <div class="section level2">
    <h2 id="ref-examples">Examples<a class="anchor" aria-label="anchor" href="#ref-examples"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span><span class="co"># Annotate a matrix and metadata</span></span></span>
<span class="r-in"><span><span class="fu">clustify</span><span class="op">(</span></span></span>
<span class="r-in"><span>    input <span class="op">=</span> <span class="va">pbmc_matrix_small</span>,</span></span>
<span class="r-in"><span>    metadata <span class="op">=</span> <span class="va">pbmc_meta</span>,</span></span>
<span class="r-in"><span>    ref_mat <span class="op">=</span> <span class="va">cbmc_ref</span>,</span></span>
<span class="r-in"><span>    query_genes <span class="op">=</span> <span class="va">pbmc_vargenes</span>,</span></span>
<span class="r-in"><span>    cluster_col <span class="op">=</span> <span class="st">"classified"</span>,</span></span>
<span class="r-in"><span>    verbose <span class="op">=</span> <span class="cn">TRUE</span></span></span>
<span class="r-in"><span><span class="op">)</span></span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> using # of genes: 599</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> similarity computation completed, matrix of 9 x 13, preparing output</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>                      B CD14+ Mono CD16+ Mono     CD34+     CD4 T</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> B            0.9093577  0.5347126  0.5908755 0.6248639 0.6411407</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> CD14+ Mono   0.5279272  0.9145428  0.8919954 0.5103056 0.4636934</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> CD8 T        0.5791953  0.4853982  0.5498177 0.5663773 0.7657468</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> DC           0.6088123  0.7415535  0.7520484 0.5694074 0.4951613</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> FCGR3A+ Mono 0.5380921  0.8429769  0.9294914 0.5331927 0.4857454</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Memory CD4 T 0.6268758  0.5373744  0.5866177 0.6265186 0.8610826</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> NK           0.5042912  0.4399003  0.5092065 0.5350808 0.6755032</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Naive CD4 T  0.6443057  0.5213433  0.5625255 0.6480900 0.8891488</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Platelet     0.1456751  0.2571662  0.2867421 0.2455947 0.1422081</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>                  CD8 T        DC     Eryth Memory CD4 T        Mk</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> B            0.6269530 0.6036723 0.4817829    0.4803367 0.2241017</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> CD14+ Mono   0.4381450 0.7643448 0.4559350    0.4553454 0.4227047</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> CD8 T        0.7692801 0.5187820 0.5007841    0.6323214 0.2961827</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> DC           0.4789794 0.8493399 0.4470039    0.4764929 0.2948557</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> FCGR3A+ Mono 0.4643569 0.7235944 0.4604308    0.4614474 0.3937726</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Memory CD4 T 0.8347482 0.5541873 0.5395170    0.7128879 0.2635539</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> NK           0.6926741 0.4817286 0.4578726    0.5665250 0.2783269</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Naive CD4 T  0.8707703 0.5419174 0.5528441    0.7277404 0.2668581</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Platelet     0.1208273 0.1898078 0.3068445    0.1457308 0.7319575</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>              Naive CD4 T        NK      pDCs</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> B              0.5407149 0.5484213 0.5823418</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> CD14+ Mono     0.4564707 0.5917870 0.4738376</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> CD8 T          0.6036230 0.8257838 0.4536589</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> DC             0.4475783 0.5643991 0.6825388</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> FCGR3A+ Mono   0.4574654 0.6028156 0.4797367</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Memory CD4 T   0.6871755 0.6999959 0.4696025</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> NK             0.5505877 0.8940904 0.4185162</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Naive CD4 T    0.6967073 0.6880495 0.4631464</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Platelet       0.1609963 0.1396979 0.2152614</span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># Annotate using a different method</span></span></span>
<span class="r-in"><span><span class="fu">clustify</span><span class="op">(</span></span></span>
<span class="r-in"><span>    input <span class="op">=</span> <span class="va">pbmc_matrix_small</span>,</span></span>
<span class="r-in"><span>    metadata <span class="op">=</span> <span class="va">pbmc_meta</span>,</span></span>
<span class="r-in"><span>    ref_mat <span class="op">=</span> <span class="va">cbmc_ref</span>,</span></span>
<span class="r-in"><span>    query_genes <span class="op">=</span> <span class="va">pbmc_vargenes</span>,</span></span>
<span class="r-in"><span>    cluster_col <span class="op">=</span> <span class="st">"classified"</span>,</span></span>
<span class="r-in"><span>    compute_method <span class="op">=</span> <span class="st">"cosine"</span></span></span>
<span class="r-in"><span><span class="op">)</span></span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> using # of genes: 599</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> similarity computation completed, matrix of 9 x 13, preparing output</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>                      B CD14+ Mono CD16+ Mono     CD34+     CD4 T</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> B            0.9669867  0.7149343  0.7720125 0.8004738 0.7635724</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> CD14+ Mono   0.7527223  0.9627275  0.9460084 0.7577338 0.7096110</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> CD8 T        0.7510063  0.7070527  0.7467587 0.7677927 0.8562280</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> DC           0.8230905  0.8842235  0.9086642 0.8037140 0.7246647</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> FCGR3A+ Mono 0.7590829  0.9172636  0.9762851 0.7746692 0.7286717</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Memory CD4 T 0.7865337  0.7486511  0.7797697 0.8075621 0.9318837</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> NK           0.6784059  0.6671718  0.7107553 0.7238537 0.7777913</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Naive CD4 T  0.7986227  0.7404425  0.7704809 0.8194404 0.9493149</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Platelet     0.5463488  0.5932935  0.6068613 0.5953661 0.5558004</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>                  CD8 T        DC     Eryth Memory CD4 T        Mk</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> B            0.7657181 0.8062767 0.6514282    0.7123319 0.6594086</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> CD14+ Mono   0.7061112 0.8919577 0.6951238    0.7557225 0.7911713</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> CD8 T        0.8674614 0.7351337 0.6826181    0.8232194 0.7188642</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> DC           0.7260570 0.9560634 0.6874076    0.7519049 0.7448677</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> FCGR3A+ Mono 0.7274581 0.8800476 0.6880012    0.7601400 0.7722167</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Memory CD4 T 0.9235744 0.7594695 0.7101214    0.8855087 0.7122113</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> NK           0.7973584 0.6864934 0.6500106    0.7578212 0.6960715</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Naive CD4 T  0.9444251 0.7508430 0.7181980    0.8877208 0.7125300</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Platelet     0.5516889 0.5730797 0.5898023    0.5655621 0.8775257</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>              Naive CD4 T        NK      pDCs</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> B              0.7494337 0.6957714 0.7863873</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> CD14+ Mono     0.7542713 0.7675894 0.7495702</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> CD8 T          0.8123971 0.9018633 0.7085898</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> DC             0.7491931 0.7595771 0.8503681</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> FCGR3A+ Mono   0.7561998 0.7776798 0.7477358</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Memory CD4 T   0.8731255 0.8102022 0.7241459</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> NK             0.7544227 0.9480188 0.6812692</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Naive CD4 T    0.8847456 0.8097839 0.7246787</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Platelet       0.5455913 0.5569351 0.5756617</span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># Annotate a Seurat object</span></span></span>
<span class="r-in"><span><span class="fu">clustify</span><span class="op">(</span></span></span>
<span class="r-in"><span>    <span class="va">s_small3</span>,</span></span>
<span class="r-in"><span>    <span class="va">cbmc_ref</span>,</span></span>
<span class="r-in"><span>    cluster_col <span class="op">=</span> <span class="st">"RNA_snn_res.1"</span>,</span></span>
<span class="r-in"><span>    obj_out <span class="op">=</span> <span class="cn">TRUE</span>,</span></span>
<span class="r-in"><span>    per_cell <span class="op">=</span> <span class="cn">FALSE</span>,</span></span>
<span class="r-in"><span>    dr <span class="op">=</span> <span class="st">"tsne"</span></span></span>
<span class="r-in"><span><span class="op">)</span></span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> object data retrieval complete, moving to similarity computation</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> using # of genes: 19</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> similarity computation completed, matrix of 3 x 13, preparing output</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> using threshold of 0.55</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> An object of class Seurat </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 230 features across 80 samples within 1 assay </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Active assay: RNA (230 features, 20 variable features)</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  2 dimensional reductions calculated: pca, tsne</span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># Annotate (and return) a Seurat object per-cell</span></span></span>
<span class="r-in"><span><span class="fu">clustify</span><span class="op">(</span></span></span>
<span class="r-in"><span>    input <span class="op">=</span> <span class="va">s_small3</span>,</span></span>
<span class="r-in"><span>    ref_mat <span class="op">=</span> <span class="va">cbmc_ref</span>,</span></span>
<span class="r-in"><span>    cluster_col <span class="op">=</span> <span class="st">"RNA_snn_res.1"</span>,</span></span>
<span class="r-in"><span>    obj_out <span class="op">=</span> <span class="cn">TRUE</span>,</span></span>
<span class="r-in"><span>    per_cell <span class="op">=</span> <span class="cn">TRUE</span>,</span></span>
<span class="r-in"><span>    dr <span class="op">=</span> <span class="st">"tsne"</span></span></span>
<span class="r-in"><span><span class="op">)</span></span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> object data retrieval complete, moving to similarity computation</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> using # of genes: 19</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> similarity computation completed, matrix of 80 x 13, preparing output</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> using threshold of NA</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> An object of class Seurat </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 230 features across 80 samples within 1 assay </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Active assay: RNA (230 features, 20 variable features)</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  2 dimensional reductions calculated: pca, tsne</span>
</code></pre></div>
    </div>
  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p></p><p>An R package developed by the <a href="https://medschool.cuanschutz.edu/rbi" class="external-link">RNA Bioscience Initiative</a></p>
</div>

<div class="pkgdown-footer-right">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.6.</p>
</div>

    </footer></div>

  

  

  </body></html>

