<!-- Generated by pkgdown: do not edit by hand -->
<!DOCTYPE html>
<html lang="en">
  <head>
  <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<title>Main function to compare scRNA-seq data to gene lists. — clustify_lists • clustifyr</title>

<!-- favicons -->
<link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png">
<link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png">
<link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png" />
<link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png" />
<link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png" />
<link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png" />

<!-- jquery -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script>
<!-- Bootstrap -->

<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous" />

<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script>

<!-- bootstrap-toc -->
<link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script>

<!-- Font Awesome icons -->
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous" />
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous" />

<!-- clipboard.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script>

<!-- headroom.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script>

<!-- pkgdown -->
<link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script>




<meta property="og:title" content="Main function to compare scRNA-seq data to gene lists. — clustify_lists" />
<meta property="og:description" content="Main function to compare scRNA-seq data to gene lists." />
<meta property="og:image" content="http://rnabioco.github.io/clustifyr/logo.svg" />




<!-- mathjax -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script>

<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->



  </head>

  <body data-spy="scroll" data-target="#toc">
    <div class="container template-reference-topic">
      <header>
      <div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">clustifyr</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">1.3.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="../articles/clustifyR.html">Introduction to clustifyr</a>
    </li>
    <li>
      <a href="../articles/geo-annotations.html">Improving NCBI GEO submissions of scRNA-seq data</a>
    </li>
  </ul>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/rnabioco/clustifyr/">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
      
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header>

<div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Main function to compare scRNA-seq data to gene lists.</h1>
    <small class="dont-index">Source: <a href='https://github.com/rnabioco/clustifyr/blob/master/R/main.R'><code>R/main.R</code></a></small>
    <div class="hidden name"><code>clustify_lists.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>Main function to compare scRNA-seq data to gene lists.</p>
    </div>

    <pre class="usage"><span class='fu'>clustify_lists</span><span class='op'>(</span><span class='va'>input</span>, <span class='va'>...</span><span class='op'>)</span>

<span class='co'># S3 method for default</span>
<span class='fu'>clustify_lists</span><span class='op'>(</span>
  <span class='va'>input</span>,
  <span class='va'>marker</span>,
  marker_inmatrix <span class='op'>=</span> <span class='cn'>TRUE</span>,
  metadata <span class='op'>=</span> <span class='cn'>NULL</span>,
  cluster_col <span class='op'>=</span> <span class='cn'>NULL</span>,
  if_log <span class='op'>=</span> <span class='cn'>TRUE</span>,
  per_cell <span class='op'>=</span> <span class='cn'>FALSE</span>,
  topn <span class='op'>=</span> <span class='fl'>800</span>,
  cut <span class='op'>=</span> <span class='fl'>0</span>,
  genome_n <span class='op'>=</span> <span class='fl'>30000</span>,
  metric <span class='op'>=</span> <span class='st'>"hyper"</span>,
  output_high <span class='op'>=</span> <span class='cn'>TRUE</span>,
  lookuptable <span class='op'>=</span> <span class='cn'>NULL</span>,
  obj_out <span class='op'>=</span> <span class='cn'>TRUE</span>,
  seurat_out <span class='op'>=</span> <span class='cn'>TRUE</span>,
  rename_prefix <span class='op'>=</span> <span class='cn'>NULL</span>,
  threshold <span class='op'>=</span> <span class='fl'>0</span>,
  low_threshold_cell <span class='op'>=</span> <span class='fl'>0</span>,
  <span class='va'>...</span>
<span class='op'>)</span>

<span class='co'># S3 method for seurat</span>
<span class='fu'>clustify_lists</span><span class='op'>(</span>
  <span class='va'>input</span>,
  metadata <span class='op'>=</span> <span class='cn'>NULL</span>,
  cluster_col <span class='op'>=</span> <span class='cn'>NULL</span>,
  if_log <span class='op'>=</span> <span class='cn'>TRUE</span>,
  per_cell <span class='op'>=</span> <span class='cn'>FALSE</span>,
  topn <span class='op'>=</span> <span class='fl'>800</span>,
  cut <span class='op'>=</span> <span class='fl'>0</span>,
  <span class='va'>marker</span>,
  marker_inmatrix <span class='op'>=</span> <span class='cn'>TRUE</span>,
  genome_n <span class='op'>=</span> <span class='fl'>30000</span>,
  metric <span class='op'>=</span> <span class='st'>"hyper"</span>,
  output_high <span class='op'>=</span> <span class='cn'>TRUE</span>,
  dr <span class='op'>=</span> <span class='st'>"umap"</span>,
  seurat_out <span class='op'>=</span> <span class='cn'>TRUE</span>,
  obj_out <span class='op'>=</span> <span class='cn'>TRUE</span>,
  threshold <span class='op'>=</span> <span class='fl'>0</span>,
  rename_prefix <span class='op'>=</span> <span class='cn'>NULL</span>,
  <span class='va'>...</span>
<span class='op'>)</span>

<span class='co'># S3 method for Seurat</span>
<span class='fu'>clustify_lists</span><span class='op'>(</span>
  <span class='va'>input</span>,
  metadata <span class='op'>=</span> <span class='cn'>NULL</span>,
  cluster_col <span class='op'>=</span> <span class='cn'>NULL</span>,
  if_log <span class='op'>=</span> <span class='cn'>TRUE</span>,
  per_cell <span class='op'>=</span> <span class='cn'>FALSE</span>,
  topn <span class='op'>=</span> <span class='fl'>800</span>,
  cut <span class='op'>=</span> <span class='fl'>0</span>,
  <span class='va'>marker</span>,
  marker_inmatrix <span class='op'>=</span> <span class='cn'>TRUE</span>,
  genome_n <span class='op'>=</span> <span class='fl'>30000</span>,
  metric <span class='op'>=</span> <span class='st'>"hyper"</span>,
  output_high <span class='op'>=</span> <span class='cn'>TRUE</span>,
  dr <span class='op'>=</span> <span class='st'>"umap"</span>,
  seurat_out <span class='op'>=</span> <span class='cn'>TRUE</span>,
  obj_out <span class='op'>=</span> <span class='cn'>TRUE</span>,
  threshold <span class='op'>=</span> <span class='fl'>0</span>,
  rename_prefix <span class='op'>=</span> <span class='cn'>NULL</span>,
  <span class='va'>...</span>
<span class='op'>)</span>

<span class='co'># S3 method for SingleCellExperiment</span>
<span class='fu'>clustify_lists</span><span class='op'>(</span>
  <span class='va'>input</span>,
  metadata <span class='op'>=</span> <span class='cn'>NULL</span>,
  cluster_col <span class='op'>=</span> <span class='cn'>NULL</span>,
  if_log <span class='op'>=</span> <span class='cn'>TRUE</span>,
  per_cell <span class='op'>=</span> <span class='cn'>FALSE</span>,
  topn <span class='op'>=</span> <span class='fl'>800</span>,
  cut <span class='op'>=</span> <span class='fl'>0</span>,
  <span class='va'>marker</span>,
  marker_inmatrix <span class='op'>=</span> <span class='cn'>TRUE</span>,
  genome_n <span class='op'>=</span> <span class='fl'>30000</span>,
  metric <span class='op'>=</span> <span class='st'>"hyper"</span>,
  output_high <span class='op'>=</span> <span class='cn'>TRUE</span>,
  dr <span class='op'>=</span> <span class='st'>"umap"</span>,
  seurat_out <span class='op'>=</span> <span class='cn'>TRUE</span>,
  obj_out <span class='op'>=</span> <span class='cn'>TRUE</span>,
  threshold <span class='op'>=</span> <span class='fl'>0</span>,
  rename_prefix <span class='op'>=</span> <span class='cn'>NULL</span>,
  <span class='va'>...</span>
<span class='op'>)</span></pre>

    <h2 class="hasAnchor" id="arguments"><a class="anchor" href="#arguments"></a>Arguments</h2>
    <table class="ref-arguments">
    <colgroup><col class="name" /><col class="desc" /></colgroup>
    <tr>
      <th>input</th>
      <td><p>single-cell expression matrix or Seurat object</p></td>
    </tr>
    <tr>
      <th>...</th>
      <td><p>passed to matrixize_markers</p></td>
    </tr>
    <tr>
      <th>marker</th>
      <td><p>matrix or dataframe of candidate genes for each cluster</p></td>
    </tr>
    <tr>
      <th>marker_inmatrix</th>
      <td><p>whether markers genes are already in preprocessed
matrix form</p></td>
    </tr>
    <tr>
      <th>metadata</th>
      <td><p>cell cluster assignments,
supplied as a vector or data.frame.
If data.frame is supplied then <code>cluster_col</code> needs to be set.
Not required if running correlation per cell.</p></td>
    </tr>
    <tr>
      <th>cluster_col</th>
      <td><p>column in metadata with cluster number</p></td>
    </tr>
    <tr>
      <th>if_log</th>
      <td><p>input data is natural log, averaging will be done on
unlogged data</p></td>
    </tr>
    <tr>
      <th>per_cell</th>
      <td><p>compare per cell or per cluster</p></td>
    </tr>
    <tr>
      <th>topn</th>
      <td><p>number of top expressing genes to keep from input matrix</p></td>
    </tr>
    <tr>
      <th>cut</th>
      <td><p>expression cut off from input matrix</p></td>
    </tr>
    <tr>
      <th>genome_n</th>
      <td><p>number of genes in the genome</p></td>
    </tr>
    <tr>
      <th>metric</th>
      <td><p>adjusted p-value for hypergeometric test, or jaccard index</p></td>
    </tr>
    <tr>
      <th>output_high</th>
      <td><p>if true (by default to fit with rest of package),
-log10 transform p-value</p></td>
    </tr>
    <tr>
      <th>lookuptable</th>
      <td><p>if not supplied, will look in built-in table
for object parsing</p></td>
    </tr>
    <tr>
      <th>obj_out</th>
      <td><p>whether to output object instead of cor matrix</p></td>
    </tr>
    <tr>
      <th>seurat_out</th>
      <td><p>output cor matrix or called seurat object
(deprecated, use obj_out instead)</p></td>
    </tr>
    <tr>
      <th>rename_prefix</th>
      <td><p>prefix to add to type and r column names</p></td>
    </tr>
    <tr>
      <th>threshold</th>
      <td><p>identity calling minimum correlation score threshold,
only used when obj_out = T</p></td>
    </tr>
    <tr>
      <th>low_threshold_cell</th>
      <td><p>option to remove clusters with too few cells</p></td>
    </tr>
    <tr>
      <th>dr</th>
      <td><p>stored dimension reduction</p></td>
    </tr>
    </table>

    <h2 class="hasAnchor" id="value"><a class="anchor" href="#value"></a>Value</h2>

    <p>matrix of numeric values, clusters from input as row names,
cell types from marker_mat as column names</p>

    <h2 class="hasAnchor" id="examples"><a class="anchor" href="#examples"></a>Examples</h2>
    <pre class="examples"><div class='input'><span class='co'># Annotate a matrix and metadata</span>
<span class='fu'>clustify_lists</span><span class='op'>(</span>
    input <span class='op'>=</span> <span class='va'>pbmc_matrix_small</span>,
    marker <span class='op'>=</span> <span class='va'>cbmc_m</span>,
    metadata <span class='op'>=</span> <span class='va'>pbmc_meta</span>,
    cluster_col <span class='op'>=</span> <span class='st'>"classified"</span>,
    verbose <span class='op'>=</span> <span class='cn'>TRUE</span>
<span class='op'>)</span>
</div><div class='output co'>#&gt;                   CD4 T      CD8 T Memory CD4 T CD14+ Mono Naive CD4 T
#&gt; Naive CD4 T  0.06713589 0.06713589            0   3.609737           0
#&gt; Memory CD4 T 0.06713589 0.00000000            0   3.609737           0
#&gt; CD14+ Mono   0.00000000 0.00000000            0   3.609737           0
#&gt; B            0.00000000 0.00000000            0   3.609737           0
#&gt; CD8 T        0.06713589 0.00000000            0   1.600060           0
#&gt; FCGR3A+ Mono 0.00000000 0.00000000            0   3.609737           0
#&gt; NK           0.00000000 0.06713589            0   3.609737           0
#&gt; DC           0.00000000 0.00000000            0   3.609737           0
#&gt; Platelet     0.12621349 0.00000000            0   3.523004           0
#&gt;                      NK         B CD16+ Mono CD34+ Eryth       Mk        DC
#&gt; Naive CD4 T  3.60973732 0.0000000 0.00000000     0     0 0.000000 0.0000000
#&gt; Memory CD4 T 3.60973732 0.0000000 0.00000000     0     0 0.000000 0.0000000
#&gt; CD14+ Mono   1.60005988 0.0000000 0.00000000     0     0 0.000000 0.0000000
#&gt; B            1.60005988 1.6000599 0.00000000     0     0 0.000000 0.0000000
#&gt; CD8 T        3.60973732 0.0000000 0.00000000     0     0 0.000000 0.0000000
#&gt; FCGR3A+ Mono 0.02934733 0.0000000 0.02934733     0     0 0.000000 0.0000000
#&gt; NK           3.60973732 0.0000000 0.00000000     0     0 0.000000 0.0000000
#&gt; DC           1.60005988 0.1085286 0.00000000     0     0 0.000000 0.1085286
#&gt; Platelet     1.58060198 1.5806020 0.00000000     0     0 3.523004 0.0000000
#&gt;                 pDCs
#&gt; Naive CD4 T  0.00000
#&gt; Memory CD4 T 0.00000
#&gt; CD14+ Mono   0.00000
#&gt; B            0.00000
#&gt; CD8 T        0.00000
#&gt; FCGR3A+ Mono 0.00000
#&gt; NK           0.00000
#&gt; DC           1.60006
#&gt; Platelet     0.00000</div><div class='input'>
<span class='co'># Annotate using a different method</span>
<span class='fu'>clustify_lists</span><span class='op'>(</span>
    input <span class='op'>=</span> <span class='va'>pbmc_matrix_small</span>,
    marker <span class='op'>=</span> <span class='va'>cbmc_m</span>,
    metadata <span class='op'>=</span> <span class='va'>pbmc_meta</span>,
    cluster_col <span class='op'>=</span> <span class='st'>"classified"</span>,
    verbose <span class='op'>=</span> <span class='cn'>TRUE</span>,
    metric <span class='op'>=</span> <span class='st'>"jaccard"</span>
<span class='op'>)</span>
</div><div class='output co'>#&gt;                    CD4 T       CD8 T Memory CD4 T  CD14+ Mono Naive CD4 T
#&gt; Naive CD4 T  0.001246883 0.001246883            0 0.003750000           0
#&gt; Memory CD4 T 0.001246883 0.000000000            0 0.003750000           0
#&gt; CD14+ Mono   0.000000000 0.000000000            0 0.003750000           0
#&gt; B            0.000000000 0.000000000            0 0.003750000           0
#&gt; CD8 T        0.001246883 0.000000000            0 0.002496879           0
#&gt; FCGR3A+ Mono 0.000000000 0.000000000            0 0.003750000           0
#&gt; NK           0.000000000 0.001246883            0 0.003750000           0
#&gt; DC           0.000000000 0.000000000            0 0.003750000           0
#&gt; Platelet     0.001166861 0.000000000            0 0.003508772           0
#&gt;                       NK           B  CD16+ Mono CD34+ Eryth          Mk
#&gt; Naive CD4 T  0.003750000 0.000000000 0.000000000     0     0 0.000000000
#&gt; Memory CD4 T 0.003750000 0.000000000 0.000000000     0     0 0.000000000
#&gt; CD14+ Mono   0.002496879 0.000000000 0.000000000     0     0 0.000000000
#&gt; B            0.002496879 0.002496879 0.000000000     0     0 0.000000000
#&gt; CD8 T        0.003750000 0.000000000 0.000000000     0     0 0.000000000
#&gt; FCGR3A+ Mono 0.001246883 0.000000000 0.001246883     0     0 0.000000000
#&gt; NK           0.003750000 0.000000000 0.000000000     0     0 0.000000000
#&gt; DC           0.002496879 0.001246883 0.000000000     0     0 0.000000000
#&gt; Platelet     0.002336449 0.002336449 0.000000000     0     0 0.003508772
#&gt;                       DC        pDCs
#&gt; Naive CD4 T  0.000000000 0.000000000
#&gt; Memory CD4 T 0.000000000 0.000000000
#&gt; CD14+ Mono   0.000000000 0.000000000
#&gt; B            0.000000000 0.000000000
#&gt; CD8 T        0.000000000 0.000000000
#&gt; FCGR3A+ Mono 0.000000000 0.000000000
#&gt; NK           0.000000000 0.000000000
#&gt; DC           0.001246883 0.002496879
#&gt; Platelet     0.000000000 0.000000000</div></pre>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top">
      <h2 data-toc-skip>Contents</h2>
    </nav>
  </div>
</div>


      <footer>
      <div class="copyright">
  <p>Developed by Rui Fu, Kent Riemondy.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.6.1.9000.</p>
</div>

      </footer>
   </div>

  


  </body>
</html>


