<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><meta name="description" content="Main function to compare scRNA-seq data to gene lists."><title>Main function to compare scRNA-seq data to gene lists. — clustify_lists • clustifyr</title><!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png"><link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png"><link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png"><link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png"><link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png"><link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png"><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.1.0/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.1.0/bootstrap.bundle.min.js"></script><link href="../deps/_Source%20Sans%20Pro-0.4.1/font.css" rel="stylesheet"><link href="../deps/_Source%20Code%20Pro-0.4.1/font.css" rel="stylesheet"><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- bootstrap-toc --><script src="https://cdn.rawgit.com/afeld/bootstrap-toc/v1.0.1/dist/bootstrap-toc.min.js"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Main function to compare scRNA-seq data to gene lists. — clustify_lists"><meta property="og:description" content="Main function to compare scRNA-seq data to gene lists."><meta property="og:image" content="http://rnabioco.github.io/clustifyr/logo.svg"><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--><script defer data-domain="rnabioco.github.io" src="https://plausible.io/js/plausible.js"></script></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-dark navbar-expand-lg bg-dark"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">clustifyr</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="Released version">1.7.3</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="active nav-item">
  <a class="nav-link" href="../reference/index.html">Reference</a>
</li>
<li class="nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-articles">Articles</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-articles">
    <a class="dropdown-item" href="../articles/clustifyR.html">Introduction to clustifyr</a>
    <a class="dropdown-item" href="../articles/geo-annotations.html">Improving NCBI GEO submissions of scRNA-seq data</a>
  </div>
</li>
      </ul><form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off"></form>

      <ul class="navbar-nav"><li class="nav-item">
  <a class="external-link nav-link" href="https://github.com/rnabioco/clustifyr/" aria-label="github">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul></div>

    
  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="../logo.svg" class="logo" alt=""><h1>Main function to compare scRNA-seq data to gene lists.</h1>
      <small class="dont-index">Source: <a href="https://github.com/rnabioco/clustifyr/blob/HEAD/R/main.R" class="external-link"><code>R/main.R</code></a></small>
      <div class="d-none name"><code>clustify_lists.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p>Main function to compare scRNA-seq data to gene lists.</p>
    </div>

    <div class="section level2">
    <h2 id="ref-usage">Usage<a class="anchor" aria-label="anchor" href="#ref-usage"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="fu">clustify_lists</span><span class="op">(</span><span class="va">input</span>, <span class="va">...</span><span class="op">)</span>

<span class="co"># S3 method for default</span>
<span class="fu">clustify_lists</span><span class="op">(</span>
  <span class="va">input</span>,
  <span class="va">marker</span>,
  marker_inmatrix <span class="op">=</span> <span class="cn">TRUE</span>,
  metadata <span class="op">=</span> <span class="cn">NULL</span>,
  cluster_col <span class="op">=</span> <span class="cn">NULL</span>,
  if_log <span class="op">=</span> <span class="cn">TRUE</span>,
  per_cell <span class="op">=</span> <span class="cn">FALSE</span>,
  topn <span class="op">=</span> <span class="fl">800</span>,
  cut <span class="op">=</span> <span class="fl">0</span>,
  genome_n <span class="op">=</span> <span class="fl">30000</span>,
  metric <span class="op">=</span> <span class="st">"hyper"</span>,
  output_high <span class="op">=</span> <span class="cn">TRUE</span>,
  lookuptable <span class="op">=</span> <span class="cn">NULL</span>,
  obj_out <span class="op">=</span> <span class="cn">TRUE</span>,
  seurat_out <span class="op">=</span> <span class="cn">TRUE</span>,
  vec_out <span class="op">=</span> <span class="cn">FALSE</span>,
  rename_prefix <span class="op">=</span> <span class="cn">NULL</span>,
  threshold <span class="op">=</span> <span class="fl">0</span>,
  low_threshold_cell <span class="op">=</span> <span class="fl">0</span>,
  verbose <span class="op">=</span> <span class="cn">TRUE</span>,
  input_markers <span class="op">=</span> <span class="cn">FALSE</span>,
  details_out <span class="op">=</span> <span class="cn">FALSE</span>,
  <span class="va">...</span>
<span class="op">)</span>

<span class="co"># S3 method for Seurat</span>
<span class="fu">clustify_lists</span><span class="op">(</span>
  <span class="va">input</span>,
  metadata <span class="op">=</span> <span class="cn">NULL</span>,
  cluster_col <span class="op">=</span> <span class="cn">NULL</span>,
  if_log <span class="op">=</span> <span class="cn">TRUE</span>,
  per_cell <span class="op">=</span> <span class="cn">FALSE</span>,
  topn <span class="op">=</span> <span class="fl">800</span>,
  cut <span class="op">=</span> <span class="fl">0</span>,
  <span class="va">marker</span>,
  marker_inmatrix <span class="op">=</span> <span class="cn">TRUE</span>,
  genome_n <span class="op">=</span> <span class="fl">30000</span>,
  metric <span class="op">=</span> <span class="st">"hyper"</span>,
  output_high <span class="op">=</span> <span class="cn">TRUE</span>,
  dr <span class="op">=</span> <span class="st">"umap"</span>,
  seurat_out <span class="op">=</span> <span class="cn">TRUE</span>,
  obj_out <span class="op">=</span> <span class="cn">TRUE</span>,
  vec_out <span class="op">=</span> <span class="cn">FALSE</span>,
  threshold <span class="op">=</span> <span class="fl">0</span>,
  rename_prefix <span class="op">=</span> <span class="cn">NULL</span>,
  verbose <span class="op">=</span> <span class="cn">TRUE</span>,
  details_out <span class="op">=</span> <span class="cn">FALSE</span>,
  <span class="va">...</span>
<span class="op">)</span>

<span class="co"># S3 method for SingleCellExperiment</span>
<span class="fu">clustify_lists</span><span class="op">(</span>
  <span class="va">input</span>,
  metadata <span class="op">=</span> <span class="cn">NULL</span>,
  cluster_col <span class="op">=</span> <span class="cn">NULL</span>,
  if_log <span class="op">=</span> <span class="cn">TRUE</span>,
  per_cell <span class="op">=</span> <span class="cn">FALSE</span>,
  topn <span class="op">=</span> <span class="fl">800</span>,
  cut <span class="op">=</span> <span class="fl">0</span>,
  <span class="va">marker</span>,
  marker_inmatrix <span class="op">=</span> <span class="cn">TRUE</span>,
  genome_n <span class="op">=</span> <span class="fl">30000</span>,
  metric <span class="op">=</span> <span class="st">"hyper"</span>,
  output_high <span class="op">=</span> <span class="cn">TRUE</span>,
  dr <span class="op">=</span> <span class="st">"umap"</span>,
  seurat_out <span class="op">=</span> <span class="cn">TRUE</span>,
  obj_out <span class="op">=</span> <span class="cn">TRUE</span>,
  vec_out <span class="op">=</span> <span class="cn">FALSE</span>,
  threshold <span class="op">=</span> <span class="fl">0</span>,
  rename_prefix <span class="op">=</span> <span class="cn">NULL</span>,
  verbose <span class="op">=</span> <span class="cn">TRUE</span>,
  details_out <span class="op">=</span> <span class="cn">FALSE</span>,
  <span class="va">...</span>
<span class="op">)</span></code></pre></div>
    </div>

    <div class="section level2">
    <h2 id="arguments">Arguments<a class="anchor" aria-label="anchor" href="#arguments"></a></h2>
    <dl><dt>input</dt>
<dd><p>single-cell expression matrix or Seurat object</p></dd>
<dt>...</dt>
<dd><p>passed to matrixize_markers</p></dd>
<dt>marker</dt>
<dd><p>matrix or dataframe of candidate genes for each cluster</p></dd>
<dt>marker_inmatrix</dt>
<dd><p>whether markers genes are already in preprocessed
matrix form</p></dd>
<dt>metadata</dt>
<dd><p>cell cluster assignments,
supplied as a vector or data.frame.
If data.frame is supplied then <code>cluster_col</code> needs to be set.
Not required if running correlation per cell.</p></dd>
<dt>cluster_col</dt>
<dd><p>column in metadata with cluster number</p></dd>
<dt>if_log</dt>
<dd><p>input data is natural log, averaging will be done on
unlogged data</p></dd>
<dt>per_cell</dt>
<dd><p>compare per cell or per cluster</p></dd>
<dt>topn</dt>
<dd><p>number of top expressing genes to keep from input matrix</p></dd>
<dt>cut</dt>
<dd><p>expression cut off from input matrix</p></dd>
<dt>genome_n</dt>
<dd><p>number of genes in the genome</p></dd>
<dt>metric</dt>
<dd><p>adjusted p-value for hypergeometric test, or jaccard index</p></dd>
<dt>output_high</dt>
<dd><p>if true (by default to fit with rest of package),
-log10 transform p-value</p></dd>
<dt>lookuptable</dt>
<dd><p>if not supplied, will look in built-in table
for object parsing</p></dd>
<dt>obj_out</dt>
<dd><p>whether to output object instead of cor matrix</p></dd>
<dt>seurat_out</dt>
<dd><p>output cor matrix or called seurat object
(deprecated, use obj_out instead)</p></dd>
<dt>vec_out</dt>
<dd><p>only output a result vector in the same order as metadata</p></dd>
<dt>rename_prefix</dt>
<dd><p>prefix to add to type and r column names</p></dd>
<dt>threshold</dt>
<dd><p>identity calling minimum correlation score threshold,
only used when obj_out = T</p></dd>
<dt>low_threshold_cell</dt>
<dd><p>option to remove clusters with too few cells</p></dd>
<dt>verbose</dt>
<dd><p>whether to report certain variables chosen and steps</p></dd>
<dt>input_markers</dt>
<dd><p>whether input is marker data.frame of 0 and 1s (output of pos_neg_marker), and uses alternate enrichment mode</p></dd>
<dt>details_out</dt>
<dd><p>whether to also output shared gene list from jaccard</p></dd>
<dt>dr</dt>
<dd><p>stored dimension reduction</p></dd>
</dl></div>
    <div class="section level2">
    <h2 id="value">Value<a class="anchor" aria-label="anchor" href="#value"></a></h2>
    <p>matrix of numeric values, clusters from input as row names,
cell types from marker_mat as column names</p>
    </div>

    <div class="section level2">
    <h2 id="ref-examples">Examples<a class="anchor" aria-label="anchor" href="#ref-examples"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span class="co"># Annotate a matrix and metadata</span></span>
<span class="r-in"></span>
<span class="r-in"><span class="co"># Annotate using a different method</span></span>
<span class="r-in"><span class="fu">clustify_lists</span><span class="op">(</span></span>
<span class="r-in">    input <span class="op">=</span> <span class="va">pbmc_matrix_small</span>,</span>
<span class="r-in">    marker <span class="op">=</span> <span class="va">cbmc_m</span>,</span>
<span class="r-in">    metadata <span class="op">=</span> <span class="va">pbmc_meta</span>,</span>
<span class="r-in">    cluster_col <span class="op">=</span> <span class="st">"classified"</span>,</span>
<span class="r-in">    verbose <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span class="r-in">    metric <span class="op">=</span> <span class="st">"jaccard"</span></span>
<span class="r-in"><span class="op">)</span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> list of markers instead of matrix, only supports jaccard</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> similarity computation completed, matrix of 9 x 13, preparing output</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>                    CD4 T       CD8 T Memory CD4 T  CD14+ Mono Naive CD4 T</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Naive CD4 T  0.001246883 0.001246883            0 0.003750000           0</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Memory CD4 T 0.001246883 0.000000000            0 0.003750000           0</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> CD14+ Mono   0.000000000 0.000000000            0 0.003750000           0</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> B            0.000000000 0.000000000            0 0.003750000           0</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> CD8 T        0.001246883 0.000000000            0 0.002496879           0</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> FCGR3A+ Mono 0.000000000 0.000000000            0 0.003750000           0</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> NK           0.000000000 0.001246883            0 0.003750000           0</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> DC           0.000000000 0.000000000            0 0.003750000           0</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Platelet     0.001166861 0.000000000            0 0.003508772           0</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>                       NK           B  CD16+ Mono CD34+ Eryth          Mk</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Naive CD4 T  0.003750000 0.000000000 0.000000000     0     0 0.000000000</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Memory CD4 T 0.003750000 0.000000000 0.000000000     0     0 0.000000000</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> CD14+ Mono   0.002496879 0.000000000 0.000000000     0     0 0.000000000</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> B            0.002496879 0.002496879 0.000000000     0     0 0.000000000</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> CD8 T        0.003750000 0.000000000 0.000000000     0     0 0.000000000</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> FCGR3A+ Mono 0.001246883 0.000000000 0.001246883     0     0 0.000000000</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> NK           0.003750000 0.000000000 0.000000000     0     0 0.000000000</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> DC           0.002496879 0.001246883 0.000000000     0     0 0.000000000</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Platelet     0.002336449 0.002336449 0.000000000     0     0 0.003508772</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>                       DC        pDCs</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Naive CD4 T  0.000000000 0.000000000</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Memory CD4 T 0.000000000 0.000000000</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> CD14+ Mono   0.000000000 0.000000000</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> B            0.000000000 0.000000000</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> CD8 T        0.000000000 0.000000000</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> FCGR3A+ Mono 0.000000000 0.000000000</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> NK           0.000000000 0.000000000</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> DC           0.001246883 0.002496879</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Platelet     0.000000000 0.000000000</span>
</code></pre></div>
    </div>
  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p></p><p>An R package developed by the <a href="https://medschool.cuanschutz.edu/rbi" class="external-link">RNA Bioscience Initiative</a></p>
</div>

<div class="pkgdown-footer-right">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.2.</p>
</div>

    </footer></div>

  

  

  </body></html>

