<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><title>Function reference â€¢ clustifyr</title><!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png"><link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png"><link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png"><link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png"><link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png"><link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png"><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.1.3/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.1.3/bootstrap.bundle.min.js"></script><link href="../deps/Source_Sans_Pro-0.4.2/font.css" rel="stylesheet"><link href="../deps/Source_Code_Pro-0.4.2/font.css" rel="stylesheet"><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- bootstrap-toc --><script src="https://cdn.rawgit.com/afeld/bootstrap-toc/v1.0.1/dist/bootstrap-toc.min.js"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Function reference"><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--><script defer data-domain="rnabioco.github.io" src="https://plausible.io/js/plausible.js"></script></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-dark navbar-expand-lg bg-dark"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">clustifyr</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="Released version">1.11.1</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="nav-item">
  <a class="nav-link" href="../articles/clustifyr.html">Get started</a>
</li>
<li class="active nav-item">
  <a class="nav-link" href="../reference/index.html">Reference</a>
</li>
<li class="nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-articles">Articles</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-articles">
    <a class="dropdown-item" href="../articles/geo-annotations.html">Improving NCBI GEO submissions of scRNA-seq data</a>
  </div>
</li>
      </ul><form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off"></form>

      <ul class="navbar-nav"><li class="nav-item">
  <a class="external-link nav-link" href="https://github.com/rnabioco/clustifyr/" aria-label="github">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul></div>

    
  </div>
</nav><div class="container template-reference-index">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="" class="logo" alt=""><h1>Function reference</h1>
    </div>

    <div class="section level2">
      <h2 id="main-functions">Main functions<a class="anchor" aria-label="anchor" href="#main-functions"></a></h2>
      
      <p class="section-desc"></p><p>Classify cell clusters by transcriptome profiles or gene lists, and building references</p>

      
    </div><div id="" class="section level2">
      
      
      

      <dl><dt>
          
          <code><a href="clustify.html">clustify()</a></code> 
        </dt>
        <dd>Compare scRNA-seq data to reference data.</dd>
      </dl><dl><dt>
          
          <code><a href="clustify_lists.html">clustify_lists()</a></code> 
        </dt>
        <dd>Main function to compare scRNA-seq data to gene lists.</dd>
      </dl><dl><dt>
          
          <code><a href="clustify_nudge.html">clustify_nudge()</a></code> 
        </dt>
        <dd>Combined function to compare scRNA-seq data to
bulk RNA-seq data and marker list</dd>
      </dl><dl><dt>
          
          <code><a href="clustifyr_methods.html">clustifyr_methods</a></code> 
        </dt>
        <dd>Correlation functions available in clustifyr</dd>
      </dl><dl><dt>
          
          <code><a href="average_clusters.html">average_clusters()</a></code> 
        </dt>
        <dd>Average expression values per cluster</dd>
      </dl><dl><dt>
          
          <code><a href="matrixize_markers.html">matrixize_markers()</a></code> 
        </dt>
        <dd>Convert candidate genes list into matrix</dd>
      </dl><dl><dt>
          
          <code><a href="cor_to_call.html">cor_to_call()</a></code> 
        </dt>
        <dd>get best calls for each cluster</dd>
      </dl><dl><dt>
          
          <code><a href="cor_to_call_rank.html">cor_to_call_rank()</a></code> 
        </dt>
        <dd>get ranked calls for each cluster</dd>
      </dl><dl><dt>
          
          <code><a href="cor_to_call_topn.html">cor_to_call_topn()</a></code> 
        </dt>
        <dd>get top calls for each cluster</dd>
      </dl><dl><dt>
          
          <code><a href="call_to_metadata.html">call_to_metadata()</a></code> 
        </dt>
        <dd>Insert called ident results into metadata</dd>
      </dl><dl><dt>
          
          <code><a href="call_consensus.html">call_consensus()</a></code> 
        </dt>
        <dd>get concensus calls for a list of cor calls</dd>
      </dl></div><div class="section level2">
      <h2 id="object-wrappers">Object wrappers<a class="anchor" aria-label="anchor" href="#object-wrappers"></a></h2>
      
      <p class="section-desc"></p><p>Retrieving metadata and reference building from supported object types</p>

      
    </div><div id="" class="section level2">
      
      
      

      <dl><dt>
          
          <code><a href="seurat_meta.html">seurat_meta()</a></code> 
        </dt>
        <dd>Function to convert labelled seurat object to fully prepared metadata</dd>
      </dl><dl><dt>
          
          <code><a href="seurat_ref.html">seurat_ref()</a></code> 
        </dt>
        <dd>Function to convert labelled seurat object to avg expression matrix</dd>
      </dl><dl><dt>
          
          <code><a href="object_ref.html">object_ref()</a></code> 
        </dt>
        <dd>Function to convert labelled object to avg expression matrix</dd>
      </dl><dl><dt>
          
          <code><a href="object_loc_lookup.html">object_loc_lookup</a></code> 
        </dt>
        <dd>lookup table for single cell object structures</dd>
      </dl></div><div class="section level2">
      <h2 id="plotting">Plotting<a class="anchor" aria-label="anchor" href="#plotting"></a></h2>
      
      

      
    </div><div id="" class="section level2">
      
      
      

      <dl><dt>
          
          <code><a href="plot_best_call.html">plot_best_call()</a></code> 
        </dt>
        <dd>Plot best calls for each cluster on a tSNE or umap</dd>
      </dl><dl><dt>
          
          <code><a href="plot_call.html">plot_call()</a></code> 
        </dt>
        <dd>Plot called clusters on a tSNE or umap, for each reference cluster given</dd>
      </dl><dl><dt>
          
          <code><a href="plot_cor.html">plot_cor()</a></code> 
        </dt>
        <dd>Plot similarity measures on a tSNE or umap</dd>
      </dl><dl><dt>
          
          <code><a href="plot_cor_heatmap.html">plot_cor_heatmap()</a></code> 
        </dt>
        <dd>Plot similarity measures on heatmap</dd>
      </dl><dl><dt>
          
          <code><a href="plot_dims.html">plot_dims()</a></code> 
        </dt>
        <dd>Plot a tSNE or umap colored by feature.</dd>
      </dl><dl><dt>
          
          <code><a href="plot_gene.html">plot_gene()</a></code> 
        </dt>
        <dd>Plot gene expression on to tSNE or umap</dd>
      </dl><dl><dt>
          
          <code><a href="plot_pathway_gsea.html">plot_pathway_gsea()</a></code> 
        </dt>
        <dd>plot GSEA pathway scores as heatmap,
returns a list containing results and plot.</dd>
      </dl><dl><dt>
          
          <code><a href="plot_rank_bias.html">plot_rank_bias()</a></code> 
        </dt>
        <dd>Query rank bias results</dd>
      </dl></div><div class="section level2">
      <h2 id="data-sets">Data sets<a class="anchor" aria-label="anchor" href="#data-sets"></a></h2>
      
      

      
    </div><div id="" class="section level2">
      
      
      

      <dl><dt>
          
          <code><a href="cbmc_m.html">cbmc_m</a></code> 
        </dt>
        <dd>reference marker matrix from seurat citeseq CBMC tutorial</dd>
      </dl><dl><dt>
          
          <code><a href="cbmc_ref.html">cbmc_ref</a></code> 
        </dt>
        <dd>reference matrix from seurat citeseq CBMC tutorial</dd>
      </dl><dl><dt>
          
          <code><a href="downrefs.html">downrefs</a></code> 
        </dt>
        <dd>table of references stored in clustifyrdata</dd>
      </dl><dl><dt>
          
          <code><a href="human_genes_10x.html">human_genes_10x</a></code> 
        </dt>
        <dd>Vector of human genes for 10x cellranger pipeline</dd>
      </dl><dl><dt>
          
          <code><a href="mouse_genes_10x.html">mouse_genes_10x</a></code> 
        </dt>
        <dd>Vector of mouse genes for 10x cellranger pipeline</dd>
      </dl><dl><dt>
          
          <code><a href="object_loc_lookup.html">object_loc_lookup</a></code> 
        </dt>
        <dd>lookup table for single cell object structures</dd>
      </dl><dl><dt>
          
          <code><a href="pbmc_markers.html">pbmc_markers</a></code> 
        </dt>
        <dd>Marker genes identified by Seurat from single-cell RNA-seq PBMCs.</dd>
      </dl><dl><dt>
          
          <code><a href="pbmc_markers_M3Drop.html">pbmc_markers_M3Drop</a></code> 
        </dt>
        <dd>Marker genes identified by M3Drop from single-cell RNA-seq PBMCs.</dd>
      </dl><dl><dt>
          
          <code><a href="pbmc_matrix_small.html">pbmc_matrix_small</a></code> 
        </dt>
        <dd>Matrix of single-cell RNA-seq PBMCs.</dd>
      </dl><dl><dt>
          
          <code><a href="pbmc_meta.html">pbmc_meta</a></code> 
        </dt>
        <dd>Meta-data for single-cell RNA-seq PBMCs.</dd>
      </dl><dl><dt>
          
          <code><a href="pbmc_vargenes.html">pbmc_vargenes</a></code> 
        </dt>
        <dd>Variable genes identified by Seurat from single-cell RNA-seq PBMCs.</dd>
      </dl><dl><dt>
          
          <code><a href="s_small.html">s_small</a></code> 
        </dt>
        <dd>Small clustered Seurat2 object</dd>
      </dl><dl><dt>
          
          <code><a href="s_small3.html">s_small3</a></code> 
        </dt>
        <dd>Small clustered Seurat3 object</dd>
      </dl><dl><dt>
          
          <code><a href="sce_small.html">sce_small</a></code> 
        </dt>
        <dd>Small SingleCellExperiment object</dd>
      </dl></div><div class="section level2">
      <h2 id="shiny">Shiny<a class="anchor" aria-label="anchor" href="#shiny"></a></h2>
      
      

      
    </div><div id="" class="section level2">
      
      
      

      <dl><dt>
          
          <code><a href="run_clustifyr_app.html">run_clustifyr_app()</a></code> 
        </dt>
        <dd>Launch Shiny app version of clustifyr,
may need to run install_clustifyr_app() at first time to install packages</dd>
      </dl></div><div class="section level2">
      <h2 id="utilities">Utilities<a class="anchor" aria-label="anchor" href="#utilities"></a></h2>
      
      

      
    </div><div id="" class="section level2">
      
      
      

      <dl><dt>
          
          <code><a href="run_gsea.html">run_gsea()</a></code> 
        </dt>
        <dd>Run GSEA to compare a gene list(s) to per cell or
per cluster expression data</dd>
      </dl><dl><dt>
          
          <code><a href="calculate_pathway_gsea.html">calculate_pathway_gsea()</a></code> 
        </dt>
        <dd>Convert expression matrix to GSEA pathway scores
(would take a similar place in workflow before average_clusters/binarize)</dd>
      </dl><dl><dt>
          
          <code><a href="get_best_match_matrix.html">get_best_match_matrix()</a></code> 
        </dt>
        <dd>Function to make best call from correlation matrix</dd>
      </dl><dl><dt>
          
          <code><a href="get_best_str.html">get_best_str()</a></code> 
        </dt>
        <dd>Function to make call and attach score</dd>
      </dl><dl><dt>
          
          <code><a href="get_common_elements.html">get_common_elements()</a></code> 
        </dt>
        <dd>Find entries shared in all vectors</dd>
      </dl><dl><dt>
          
          <code><a href="get_similarity.html">get_similarity()</a></code> 
        </dt>
        <dd>Compute similarity of matrices</dd>
      </dl><dl><dt>
          
          <code><a href="get_ucsc_reference.html">get_ucsc_reference()</a></code> 
        </dt>
        <dd>Build reference atlases from external UCSC cellbrowsers</dd>
      </dl><dl><dt>
          
          <code><a href="get_unique_column.html">get_unique_column()</a></code> 
        </dt>
        <dd>Generate a unique column id for a dataframe</dd>
      </dl><dl><dt>
          
          <code><a href="get_vargenes.html">get_vargenes()</a></code> 
        </dt>
        <dd>Generate variable gene list from marker matrix</dd>
      </dl><dl><dt>
          
          <code><a href="assess_rank_bias.html">assess_rank_bias()</a></code> 
        </dt>
        <dd>Find rank bias</dd>
      </dl><dl><dt>
          
          <code><a href="binarize_expr.html">binarize_expr()</a></code> 
        </dt>
        <dd>Binarize scRNAseq data</dd>
      </dl><dl><dt>
          
          <code><a href="downsample_matrix.html">downsample_matrix()</a></code> 
        </dt>
        <dd>downsample matrix by cluster or completely random</dd>
      </dl><dl><dt>
          
          <code><a href="feature_select_PCA.html">feature_select_PCA()</a></code> 
        </dt>
        <dd>Returns a list of variable genes based on PCA</dd>
      </dl><dl><dt>
          
          <code><a href="file_marker_parse.html">file_marker_parse()</a></code> 
        </dt>
        <dd>takes files with positive and negative markers, as described in garnett,
and returns list of markers</dd>
      </dl><dl><dt>
          
          <code><a href="find_rank_bias.html">find_rank_bias()</a></code> 
        </dt>
        <dd>Find rank bias</dd>
      </dl><dl><dt>
          
          <code><a href="gmt_to_list.html">gmt_to_list()</a></code> 
        </dt>
        <dd>convert gmt format of pathways to list of vectors</dd>
      </dl><dl><dt>
          
          <code><a href="make_comb_ref.html">make_comb_ref()</a></code> 
        </dt>
        <dd>make combination ref matrix to assess intermixing</dd>
      </dl><dl><dt>
          
          <code><a href="marker_select.html">marker_select()</a></code> 
        </dt>
        <dd>decide for one gene whether it is a marker for a certain cell type</dd>
      </dl><dl><dt>
          
          <code><a href="overcluster_test.html">overcluster_test()</a></code> 
        </dt>
        <dd>compare clustering parameters and classification outcomes</dd>
      </dl><dl><dt>
          
          <code><a href="parse_loc_object.html">parse_loc_object()</a></code> 
        </dt>
        <dd>more flexible parsing of single cell objects</dd>
      </dl><dl><dt>
          
          <code><a href="pos_neg_marker.html">pos_neg_marker()</a></code> 
        </dt>
        <dd>generate pos and negative marker expression matrix from a
list/dataframe of positive markers</dd>
      </dl><dl><dt>
          
          <code><a href="pos_neg_select.html">pos_neg_select()</a></code> 
        </dt>
        <dd>adapt clustify to tweak score for pos and neg markers</dd>
      </dl><dl><dt>
          
          <code><a href="query_rank_bias.html">query_rank_bias()</a></code> 
        </dt>
        <dd>Query rank bias results</dd>
      </dl><dl><dt>
          
          <code><a href="ref_feature_select.html">ref_feature_select()</a></code> 
        </dt>
        <dd>feature select from reference matrix</dd>
      </dl><dl><dt>
          
          <code><a href="ref_marker_select.html">ref_marker_select()</a></code> 
        </dt>
        <dd>marker selection from reference matrix</dd>
      </dl><dl><dt>
          
          <code><a href="reverse_marker_matrix.html">reverse_marker_matrix()</a></code> 
        </dt>
        <dd>generate negative markers from a list of exclusive positive markers</dd>
      </dl><dl><dt>
          
          <code><a href="append_genes.html">append_genes()</a></code> 
        </dt>
        <dd>Given a reference matrix and a list of genes, take the union of
all genes in vector and genes in reference matrix
and insert zero counts for all remaining genes.</dd>
      </dl><dl><dt>
          
          <code><a href="assign_ident.html">assign_ident()</a></code> 
        </dt>
        <dd>manually change idents as needed</dd>
      </dl><dl><dt>
          
          <code><a href="build_atlas.html">build_atlas()</a></code> 
        </dt>
        <dd>Function to combine records into single atlas</dd>
      </dl><dl><dt>
          
          <code><a href="calc_similarity.html">calc_similarity()</a></code> 
        </dt>
        <dd>compute similarity</dd>
      </dl><dl><dt>
          
          <code><a href="calc_distance.html">calc_distance()</a></code> 
        </dt>
        <dd>Distance calculations for spatial coord</dd>
      </dl><dl><dt>
          
          <code><a href="check_raw_counts.html">check_raw_counts()</a></code> 
        </dt>
        <dd>Given a count matrix, determine if the matrix has been either
log-normalized, normalized, or contains raw counts</dd>
      </dl><dl><dt>
          
          <code><a href="collapse_to_cluster.html">collapse_to_cluster()</a></code> 
        </dt>
        <dd>From per-cell calls, take highest freq call in each cluster</dd>
      </dl><dl><dt>
          
          <code><a href="compare_lists.html">compare_lists()</a></code> 
        </dt>
        <dd>Calculate adjusted p-values for hypergeometric test of gene lists
or jaccard index</dd>
      </dl><dl><dt>
          
          <code><a href="cosine.html">cosine()</a></code> 
        </dt>
        <dd>Cosine distance</dd>
      </dl><dl><dt>
          
          <code><a href="gene_pct.html">gene_pct()</a></code> 
        </dt>
        <dd>pct of cells in each cluster that express genelist</dd>
      </dl><dl><dt>
          
          <code><a href="gene_pct_markerm.html">gene_pct_markerm()</a></code> 
        </dt>
        <dd>pct of cells in every cluster that express a series of genelists</dd>
      </dl><dl><dt>
          
          <code><a href="insert_meta_object.html">insert_meta_object()</a></code> 
        </dt>
        <dd>more flexible metadata update of single cell objects</dd>
      </dl><dl><dt>
          
          <code><a href="kl_divergence.html">kl_divergence()</a></code> 
        </dt>
        <dd>KL divergence</dd>
      </dl><dl><dt>
          
          <code><a href="not_pretty_palette.html">not_pretty_palette</a></code> 
        </dt>
        <dd>black and white palette for plotting continous variables</dd>
      </dl><dl><dt>
          
          <code><a href="object_data.html">object_data()</a></code> 
        </dt>
        <dd>Function to access object data</dd>
      </dl><dl><dt>
          
          <code><a href="overcluster.html">overcluster()</a></code> 
        </dt>
        <dd>Overcluster by kmeans per cluster</dd>
      </dl><dl><dt>
          
          <code><a href="percent_clusters.html">percent_clusters()</a></code> 
        </dt>
        <dd>Percentage detected per cluster</dd>
      </dl><dl><dt>
          
          <code><a href="permute_similarity.html">permute_similarity()</a></code> 
        </dt>
        <dd>Compute a p-value for similarity using permutation</dd>
      </dl><dl><dt>
          
          <code><a href="pretty_palette.html">pretty_palette</a></code> 
        </dt>
        <dd>Color palette for plotting continous variables</dd>
      </dl><dl><dt>
          
          <code><a href="pretty_palette2.html">pretty_palette2</a></code> 
        </dt>
        <dd>Color palette for plotting continous variables, starting at gray</dd>
      </dl><dl><dt>
          
          <code><a href="pretty_palette_ramp_d.html">pretty_palette_ramp_d()</a></code> 
        </dt>
        <dd>Expanded color palette ramp for plotting discrete variables</dd>
      </dl><dl><dt>
          
          <code><a href="vector_similarity.html">vector_similarity()</a></code> 
        </dt>
        <dd>Compute similarity between two vectors</dd>
      </dl><dl><dt>
          
          <code><a href="write_meta.html">write_meta()</a></code> 
        </dt>
        <dd>Function to write metadata to object</dd>
      </dl></div>
  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p></p><p>An R package developed by the <a href="https://medschool.cuanschutz.edu/rbi" class="external-link">RNA Bioscience Initiative</a></p>
</div>

<div class="pkgdown-footer-right">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.6.</p>
</div>

    </footer></div>

  

  

  </body></html>

