<!-- Generated by pkgdown: do not edit by hand -->
<!DOCTYPE html>
<html lang="en">
  <head>
  <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<title>Function reference â€¢ clustifyr</title>

<!-- favicons -->
<link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png">
<link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png">
<link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png" />
<link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png" />
<link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png" />
<link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png" />

<!-- jquery -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script>
<!-- Bootstrap -->

<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous" />

<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script>

<!-- bootstrap-toc -->
<link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script>

<!-- Font Awesome icons -->
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous" />
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous" />

<!-- clipboard.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script>

<!-- headroom.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script>

<!-- pkgdown -->
<link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script>




<meta property="og:title" content="Function reference" />
<meta property="og:image" content="http://rnabioco.github.io/clustifyr/logo.svg" />




<!-- mathjax -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script>

<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->



  </head>

  <body data-spy="scroll" data-target="#toc">
    <div class="container template-reference-index">
      <header>
      <div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">clustifyr</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">1.5.1</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="../index.html">
    <span class="fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="../articles/clustifyR.html">Introduction to clustifyr</a>
    </li>
    <li>
      <a href="../articles/geo-annotations.html">Improving NCBI GEO submissions of scRNA-seq data</a>
    </li>
  </ul>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/rnabioco/clustifyr/">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
      
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header>

<div class="row">
  <div class="contents col-md-9">
    <div class="page-header">
      <h1>Reference</h1>
    </div>

    <table class="ref-index">

    <colgroup>
      
      <col class="alias" />
      <col class="title" />
    </colgroup>

    <tbody>
      <tr>
        <th colspan="2">
          <h2 id="section-main-functions" class="hasAnchor"><a href="#section-main-functions" class="anchor"></a>Main functions</h2>
          <p class="section-desc"><p>Classify cell clusters by transcriptome profiles or gene lists, and building references</p></p>
        </th>
      </tr>
      
      
    </tbody><tbody>
      
      
      <tr>
        
        <td>
          <p><code><a href="clustify.html">clustify()</a></code> </p>
        </td>
        <td><p>Compare scRNA-seq data to reference data.</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="clustify_lists.html">clustify_lists()</a></code> </p>
        </td>
        <td><p>Main function to compare scRNA-seq data to gene lists.</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="clustify_nudge.html">clustify_nudge()</a></code> </p>
        </td>
        <td><p>Combined function to compare scRNA-seq data to
bulk RNA-seq data and marker list</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="clustifyr_methods.html">clustifyr_methods</a></code> </p>
        </td>
        <td><p>Correlation functions available in clustifyr</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="average_clusters.html">average_clusters()</a></code> </p>
        </td>
        <td><p>Average expression values per cluster</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="matrixize_markers.html">matrixize_markers()</a></code> </p>
        </td>
        <td><p>Convert candidate genes list into matrix</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="cor_to_call.html">cor_to_call()</a></code> </p>
        </td>
        <td><p>get best calls for each cluster</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="cor_to_call_rank.html">cor_to_call_rank()</a></code> </p>
        </td>
        <td><p>get ranked calls for each cluster</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="cor_to_call_topn.html">cor_to_call_topn()</a></code> </p>
        </td>
        <td><p>get top calls for each cluster</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="call_to_metadata.html">call_to_metadata()</a></code> </p>
        </td>
        <td><p>Insert called ident results into metadata</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="call_consensus.html">call_consensus()</a></code> </p>
        </td>
        <td><p>get concensus calls for a list of cor calls</p></td>
      </tr>
    </tbody><tbody>
      <tr>
        <th colspan="2">
          <h2 id="section-object-wrappers" class="hasAnchor"><a href="#section-object-wrappers" class="anchor"></a>Object wrappers</h2>
          <p class="section-desc"><p>Retrieving metadata and reference building from supported object types</p></p>
        </th>
      </tr>
      
      
    </tbody><tbody>
      
      
      <tr>
        
        <td>
          <p><code><a href="seurat_meta.html">seurat_meta()</a></code> </p>
        </td>
        <td><p>Function to convert labelled seurat object to fully prepared metadata</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="seurat_ref.html">seurat_ref()</a></code> </p>
        </td>
        <td><p>Function to convert labelled seurat object to avg expression matrix</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="object_ref.html">object_ref()</a></code> </p>
        </td>
        <td><p>Function to convert labelled object to avg expression matrix</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="object_loc_lookup.html">object_loc_lookup</a></code> </p>
        </td>
        <td><p>lookup table for single cell object structures</p></td>
      </tr>
    </tbody><tbody>
      <tr>
        <th colspan="2">
          <h2 id="section-plotting" class="hasAnchor"><a href="#section-plotting" class="anchor"></a>Plotting</h2>
          <p class="section-desc"></p>
        </th>
      </tr>
      
      
    </tbody><tbody>
      
      
      <tr>
        
        <td>
          <p><code><a href="plot_best_call.html">plot_best_call()</a></code> </p>
        </td>
        <td><p>Plot best calls for each cluster on a tSNE or umap</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="plot_call.html">plot_call()</a></code> </p>
        </td>
        <td><p>Plot called clusters on a tSNE or umap, for each reference cluster given</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="plot_cor.html">plot_cor()</a></code> </p>
        </td>
        <td><p>Plot similarity measures on a tSNE or umap</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="plot_cor_heatmap.html">plot_cor_heatmap()</a></code> </p>
        </td>
        <td><p>Plot similarity measures on heatmap</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="plot_dims.html">plot_dims()</a></code> </p>
        </td>
        <td><p>Plot a tSNE or umap colored by feature.</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="plot_gene.html">plot_gene()</a></code> </p>
        </td>
        <td><p>Plot gene expression on to tSNE or umap</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="plot_pathway_gsea.html">plot_pathway_gsea()</a></code> </p>
        </td>
        <td><p>plot GSEA pathway scores as heatmap,
returns a list containing results and plot.</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="plot_rank_bias.html">plot_rank_bias()</a></code> </p>
        </td>
        <td><p>Query rank bias results</p></td>
      </tr>
    </tbody><tbody>
      <tr>
        <th colspan="2">
          <h2 id="section-data-sets" class="hasAnchor"><a href="#section-data-sets" class="anchor"></a>Data sets</h2>
          <p class="section-desc"></p>
        </th>
      </tr>
      
      
    </tbody><tbody>
      
      
      <tr>
        
        <td>
          <p><code><a href="cbmc_m.html">cbmc_m</a></code> </p>
        </td>
        <td><p>reference marker matrix from seurat citeseq CBMC tutorial</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="cbmc_ref.html">cbmc_ref</a></code> </p>
        </td>
        <td><p>reference matrix from seurat citeseq CBMC tutorial</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="downrefs.html">downrefs</a></code> </p>
        </td>
        <td><p>table of references stored in clustifyrdata</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="human_genes_10x.html">human_genes_10x</a></code> </p>
        </td>
        <td><p>Vector of human genes for 10x cellranger pipeline</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="mouse_genes_10x.html">mouse_genes_10x</a></code> </p>
        </td>
        <td><p>Vector of mouse genes for 10x cellranger pipeline</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="object_loc_lookup.html">object_loc_lookup</a></code> </p>
        </td>
        <td><p>lookup table for single cell object structures</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="pbmc_markers.html">pbmc_markers</a></code> </p>
        </td>
        <td><p>Marker genes identified by Seurat from single-cell RNA-seq PBMCs.</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="pbmc_markers_M3Drop.html">pbmc_markers_M3Drop</a></code> </p>
        </td>
        <td><p>Marker genes identified by M3Drop from single-cell RNA-seq PBMCs.</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="pbmc_matrix_small.html">pbmc_matrix_small</a></code> </p>
        </td>
        <td><p>Matrix of single-cell RNA-seq PBMCs.</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="pbmc_meta.html">pbmc_meta</a></code> </p>
        </td>
        <td><p>Meta-data for single-cell RNA-seq PBMCs.</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="pbmc_vargenes.html">pbmc_vargenes</a></code> </p>
        </td>
        <td><p>Variable genes identified by Seurat from single-cell RNA-seq PBMCs.</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="s_small.html">s_small</a></code> </p>
        </td>
        <td><p>Small clustered Seurat2 object</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="s_small3.html">s_small3</a></code> </p>
        </td>
        <td><p>Small clustered Seurat3 object</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="sce_small.html">sce_small</a></code> </p>
        </td>
        <td><p>Small SingleCellExperiment object</p></td>
      </tr>
    </tbody><tbody>
      <tr>
        <th colspan="2">
          <h2 id="section-shiny" class="hasAnchor"><a href="#section-shiny" class="anchor"></a>Shiny</h2>
          <p class="section-desc"></p>
        </th>
      </tr>
      
      
    </tbody><tbody>
      
      
      <tr>
        
        <td>
          <p><code><a href="append_genes.html">append_genes()</a></code> </p>
        </td>
        <td><p>Given a reference matrix and a list of genes, take the union of
all genes in vector and genes in reference matrix
and insert zero counts for all remaining genes.</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="assess_rank_bias.html">assess_rank_bias()</a></code> </p>
        </td>
        <td><p>Find rank bias</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="assign_ident.html">assign_ident()</a></code> </p>
        </td>
        <td><p>manually change idents as needed</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="average_clusters.html">average_clusters()</a></code> </p>
        </td>
        <td><p>Average expression values per cluster</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="binarize_expr.html">binarize_expr()</a></code> </p>
        </td>
        <td><p>Binarize scRNAseq data</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="build_atlas.html">build_atlas()</a></code> </p>
        </td>
        <td><p>Function to combine records into single atlas</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="calc_similarity.html">calc_similarity()</a></code> </p>
        </td>
        <td><p>compute similarity</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="calculate_pathway_gsea.html">calculate_pathway_gsea()</a></code> </p>
        </td>
        <td><p>Convert expression matrix to GSEA pathway scores
(would take a similar place in workflow before average_clusters/binarize)</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="call_consensus.html">call_consensus()</a></code> </p>
        </td>
        <td><p>get concensus calls for a list of cor calls</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="call_to_metadata.html">call_to_metadata()</a></code> </p>
        </td>
        <td><p>Insert called ident results into metadata</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="cbmc_m.html">cbmc_m</a></code> </p>
        </td>
        <td><p>reference marker matrix from seurat citeseq CBMC tutorial</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="cbmc_ref.html">cbmc_ref</a></code> </p>
        </td>
        <td><p>reference matrix from seurat citeseq CBMC tutorial</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="check_raw_counts.html">check_raw_counts()</a></code> </p>
        </td>
        <td><p>Given a count matrix, determine if the matrix has been either
log-normalized, normalized, or contains raw counts</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="clustify.html">clustify()</a></code> </p>
        </td>
        <td><p>Compare scRNA-seq data to reference data.</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="clustify_lists.html">clustify_lists()</a></code> </p>
        </td>
        <td><p>Main function to compare scRNA-seq data to gene lists.</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="clustify_nudge.html">clustify_nudge()</a></code> </p>
        </td>
        <td><p>Combined function to compare scRNA-seq data to
bulk RNA-seq data and marker list</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="clustifyr_methods.html">clustifyr_methods</a></code> </p>
        </td>
        <td><p>Correlation functions available in clustifyr</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="collapse_to_cluster.html">collapse_to_cluster()</a></code> </p>
        </td>
        <td><p>From per-cell calls, take highest freq call in each cluster</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="compare_lists.html">compare_lists()</a></code> </p>
        </td>
        <td><p>Calculate adjusted p-values for hypergeometric test of gene lists
or jaccard index</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="cor_to_call.html">cor_to_call()</a></code> </p>
        </td>
        <td><p>get best calls for each cluster</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="cor_to_call_rank.html">cor_to_call_rank()</a></code> </p>
        </td>
        <td><p>get ranked calls for each cluster</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="cor_to_call_topn.html">cor_to_call_topn()</a></code> </p>
        </td>
        <td><p>get top calls for each cluster</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="cosine.html">cosine()</a></code> </p>
        </td>
        <td><p>Cosine distance</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="downrefs.html">downrefs</a></code> </p>
        </td>
        <td><p>table of references stored in clustifyrdata</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="downsample_matrix.html">downsample_matrix()</a></code> </p>
        </td>
        <td><p>downsample matrix by cluster or completely random</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="feature_select_PCA.html">feature_select_PCA()</a></code> </p>
        </td>
        <td><p>Returns a list of variable genes based on PCA</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="file_marker_parse.html">file_marker_parse()</a></code> </p>
        </td>
        <td><p>takes files with positive and negative markers, as described in garnett,
and returns list of markers</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="find_rank_bias.html">find_rank_bias()</a></code> </p>
        </td>
        <td><p>Find rank bias</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="gene_pct.html">gene_pct()</a></code> </p>
        </td>
        <td><p>pct of cells in each cluster that express genelist</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="gene_pct_markerm.html">gene_pct_markerm()</a></code> </p>
        </td>
        <td><p>pct of cells in every cluster that express a series of genelists</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="get_best_match_matrix.html">get_best_match_matrix()</a></code> </p>
        </td>
        <td><p>Function to make best call from correlation matrix</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="get_best_str.html">get_best_str()</a></code> </p>
        </td>
        <td><p>Function to make call and attach score</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="get_common_elements.html">get_common_elements()</a></code> </p>
        </td>
        <td><p>Find entries shared in all vectors</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="get_similarity.html">get_similarity()</a></code> </p>
        </td>
        <td><p>Compute similarity of matrices</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="get_ucsc_reference.html">get_ucsc_reference()</a></code> </p>
        </td>
        <td><p>Build reference atlases from external UCSC cellbrowsers</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="get_unique_column.html">get_unique_column()</a></code> </p>
        </td>
        <td><p>Generate a unique column id for a dataframe</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="get_vargenes.html">get_vargenes()</a></code> </p>
        </td>
        <td><p>Generate variable gene list from marker matrix</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="gmt_to_list.html">gmt_to_list()</a></code> </p>
        </td>
        <td><p>convert gmt format of pathways to list of vectors</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="human_genes_10x.html">human_genes_10x</a></code> </p>
        </td>
        <td><p>Vector of human genes for 10x cellranger pipeline</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="insert_meta_object.html">insert_meta_object()</a></code> </p>
        </td>
        <td><p>more flexible metadata update of single cell objects</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="kl_divergence.html">kl_divergence()</a></code> </p>
        </td>
        <td><p>KL divergence</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="make_comb_ref.html">make_comb_ref()</a></code> </p>
        </td>
        <td><p>make combination ref matrix to assess intermixing</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="marker_select.html">marker_select()</a></code> </p>
        </td>
        <td><p>decide for one gene whether it is a marker for a certain cell type</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="matrixize_markers.html">matrixize_markers()</a></code> </p>
        </td>
        <td><p>Convert candidate genes list into matrix</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="mouse_genes_10x.html">mouse_genes_10x</a></code> </p>
        </td>
        <td><p>Vector of mouse genes for 10x cellranger pipeline</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="not_pretty_palette.html">not_pretty_palette</a></code> </p>
        </td>
        <td><p>black and white palette for plotting continous variables</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="object_data.html">object_data()</a></code> </p>
        </td>
        <td><p>Function to access object data</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="object_loc_lookup.html">object_loc_lookup</a></code> </p>
        </td>
        <td><p>lookup table for single cell object structures</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="object_ref.html">object_ref()</a></code> </p>
        </td>
        <td><p>Function to convert labelled object to avg expression matrix</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="overcluster.html">overcluster()</a></code> </p>
        </td>
        <td><p>Overcluster by kmeans per cluster</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="overcluster_test.html">overcluster_test()</a></code> </p>
        </td>
        <td><p>compare clustering parameters and classification outcomes</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="parse_loc_object.html">parse_loc_object()</a></code> </p>
        </td>
        <td><p>more flexible parsing of single cell objects</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="pbmc_markers.html">pbmc_markers</a></code> </p>
        </td>
        <td><p>Marker genes identified by Seurat from single-cell RNA-seq PBMCs.</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="pbmc_markers_M3Drop.html">pbmc_markers_M3Drop</a></code> </p>
        </td>
        <td><p>Marker genes identified by M3Drop from single-cell RNA-seq PBMCs.</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="pbmc_matrix_small.html">pbmc_matrix_small</a></code> </p>
        </td>
        <td><p>Matrix of single-cell RNA-seq PBMCs.</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="pbmc_meta.html">pbmc_meta</a></code> </p>
        </td>
        <td><p>Meta-data for single-cell RNA-seq PBMCs.</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="pbmc_vargenes.html">pbmc_vargenes</a></code> </p>
        </td>
        <td><p>Variable genes identified by Seurat from single-cell RNA-seq PBMCs.</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="percent_clusters.html">percent_clusters()</a></code> </p>
        </td>
        <td><p>Percentage detected per cluster</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="permute_similarity.html">permute_similarity()</a></code> </p>
        </td>
        <td><p>Compute a p-value for similarity using permutation</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="plot_best_call.html">plot_best_call()</a></code> </p>
        </td>
        <td><p>Plot best calls for each cluster on a tSNE or umap</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="plot_call.html">plot_call()</a></code> </p>
        </td>
        <td><p>Plot called clusters on a tSNE or umap, for each reference cluster given</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="plot_cor.html">plot_cor()</a></code> </p>
        </td>
        <td><p>Plot similarity measures on a tSNE or umap</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="plot_cor_heatmap.html">plot_cor_heatmap()</a></code> </p>
        </td>
        <td><p>Plot similarity measures on heatmap</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="plot_dims.html">plot_dims()</a></code> </p>
        </td>
        <td><p>Plot a tSNE or umap colored by feature.</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="plot_gene.html">plot_gene()</a></code> </p>
        </td>
        <td><p>Plot gene expression on to tSNE or umap</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="plot_pathway_gsea.html">plot_pathway_gsea()</a></code> </p>
        </td>
        <td><p>plot GSEA pathway scores as heatmap,
returns a list containing results and plot.</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="plot_rank_bias.html">plot_rank_bias()</a></code> </p>
        </td>
        <td><p>Query rank bias results</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="pos_neg_marker.html">pos_neg_marker()</a></code> </p>
        </td>
        <td><p>generate pos and negative marker expression matrix from a
list/dataframe of positive markers</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="pos_neg_select.html">pos_neg_select()</a></code> </p>
        </td>
        <td><p>adapt clustify to tweak score for pos and neg markers</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="pretty_palette.html">pretty_palette</a></code> </p>
        </td>
        <td><p>Color palette for plotting continous variables</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="pretty_palette2.html">pretty_palette2</a></code> </p>
        </td>
        <td><p>Color palette for plotting continous variables, starting at gray</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="pretty_palette_ramp_d.html">pretty_palette_ramp_d()</a></code> </p>
        </td>
        <td><p>Expanded color palette ramp for plotting discrete variables</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="query_rank_bias.html">query_rank_bias()</a></code> </p>
        </td>
        <td><p>Query rank bias results</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="ref_feature_select.html">ref_feature_select()</a></code> </p>
        </td>
        <td><p>feature select from reference matrix</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="ref_marker_select.html">ref_marker_select()</a></code> </p>
        </td>
        <td><p>marker selection from reference matrix</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="reverse_marker_matrix.html">reverse_marker_matrix()</a></code> </p>
        </td>
        <td><p>generate negative markers from a list of exclusive positive markers</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="run_clustifyr_app.html">run_clustifyr_app()</a></code> </p>
        </td>
        <td><p>Launch Shiny app version of clustifyr,
may need to run install_clustifyr_app() at first time to install packages</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="run_gsea.html">run_gsea()</a></code> </p>
        </td>
        <td><p>Run GSEA to compare a gene list(s) to per cell or
per cluster expression data</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="s_small.html">s_small</a></code> </p>
        </td>
        <td><p>Small clustered Seurat2 object</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="s_small3.html">s_small3</a></code> </p>
        </td>
        <td><p>Small clustered Seurat3 object</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="sce_small.html">sce_small</a></code> </p>
        </td>
        <td><p>Small SingleCellExperiment object</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="seurat_meta.html">seurat_meta()</a></code> </p>
        </td>
        <td><p>Function to convert labelled seurat object to fully prepared metadata</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="seurat_ref.html">seurat_ref()</a></code> </p>
        </td>
        <td><p>Function to convert labelled seurat object to avg expression matrix</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="vector_similarity.html">vector_similarity()</a></code> </p>
        </td>
        <td><p>Compute similarity between two vectors</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="write_meta.html">write_meta()</a></code> </p>
        </td>
        <td><p>Function to write metadata to object</p></td>
      </tr>
    </tbody><tbody>
      <tr>
        <th colspan="2">
          <h2 id="section-utilities" class="hasAnchor"><a href="#section-utilities" class="anchor"></a>Utilities</h2>
          <p class="section-desc"></p>
        </th>
      </tr>
      
      
    </tbody><tbody>
      
      
      <tr>
        
        <td>
          <p><code><a href="run_gsea.html">run_gsea()</a></code> </p>
        </td>
        <td><p>Run GSEA to compare a gene list(s) to per cell or
per cluster expression data</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="calculate_pathway_gsea.html">calculate_pathway_gsea()</a></code> </p>
        </td>
        <td><p>Convert expression matrix to GSEA pathway scores
(would take a similar place in workflow before average_clusters/binarize)</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="get_best_match_matrix.html">get_best_match_matrix()</a></code> </p>
        </td>
        <td><p>Function to make best call from correlation matrix</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="get_best_str.html">get_best_str()</a></code> </p>
        </td>
        <td><p>Function to make call and attach score</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="get_common_elements.html">get_common_elements()</a></code> </p>
        </td>
        <td><p>Find entries shared in all vectors</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="get_similarity.html">get_similarity()</a></code> </p>
        </td>
        <td><p>Compute similarity of matrices</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="get_ucsc_reference.html">get_ucsc_reference()</a></code> </p>
        </td>
        <td><p>Build reference atlases from external UCSC cellbrowsers</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="get_unique_column.html">get_unique_column()</a></code> </p>
        </td>
        <td><p>Generate a unique column id for a dataframe</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="get_vargenes.html">get_vargenes()</a></code> </p>
        </td>
        <td><p>Generate variable gene list from marker matrix</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="assess_rank_bias.html">assess_rank_bias()</a></code> </p>
        </td>
        <td><p>Find rank bias</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="binarize_expr.html">binarize_expr()</a></code> </p>
        </td>
        <td><p>Binarize scRNAseq data</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="downsample_matrix.html">downsample_matrix()</a></code> </p>
        </td>
        <td><p>downsample matrix by cluster or completely random</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="feature_select_PCA.html">feature_select_PCA()</a></code> </p>
        </td>
        <td><p>Returns a list of variable genes based on PCA</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="file_marker_parse.html">file_marker_parse()</a></code> </p>
        </td>
        <td><p>takes files with positive and negative markers, as described in garnett,
and returns list of markers</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="find_rank_bias.html">find_rank_bias()</a></code> </p>
        </td>
        <td><p>Find rank bias</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="gmt_to_list.html">gmt_to_list()</a></code> </p>
        </td>
        <td><p>convert gmt format of pathways to list of vectors</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="make_comb_ref.html">make_comb_ref()</a></code> </p>
        </td>
        <td><p>make combination ref matrix to assess intermixing</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="marker_select.html">marker_select()</a></code> </p>
        </td>
        <td><p>decide for one gene whether it is a marker for a certain cell type</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="overcluster_test.html">overcluster_test()</a></code> </p>
        </td>
        <td><p>compare clustering parameters and classification outcomes</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="parse_loc_object.html">parse_loc_object()</a></code> </p>
        </td>
        <td><p>more flexible parsing of single cell objects</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="pos_neg_marker.html">pos_neg_marker()</a></code> </p>
        </td>
        <td><p>generate pos and negative marker expression matrix from a
list/dataframe of positive markers</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="pos_neg_select.html">pos_neg_select()</a></code> </p>
        </td>
        <td><p>adapt clustify to tweak score for pos and neg markers</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="query_rank_bias.html">query_rank_bias()</a></code> </p>
        </td>
        <td><p>Query rank bias results</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="ref_feature_select.html">ref_feature_select()</a></code> </p>
        </td>
        <td><p>feature select from reference matrix</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="ref_marker_select.html">ref_marker_select()</a></code> </p>
        </td>
        <td><p>marker selection from reference matrix</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="reverse_marker_matrix.html">reverse_marker_matrix()</a></code> </p>
        </td>
        <td><p>generate negative markers from a list of exclusive positive markers</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="append_genes.html">append_genes()</a></code> </p>
        </td>
        <td><p>Given a reference matrix and a list of genes, take the union of
all genes in vector and genes in reference matrix
and insert zero counts for all remaining genes.</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="assign_ident.html">assign_ident()</a></code> </p>
        </td>
        <td><p>manually change idents as needed</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="build_atlas.html">build_atlas()</a></code> </p>
        </td>
        <td><p>Function to combine records into single atlas</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="calc_similarity.html">calc_similarity()</a></code> </p>
        </td>
        <td><p>compute similarity</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="check_raw_counts.html">check_raw_counts()</a></code> </p>
        </td>
        <td><p>Given a count matrix, determine if the matrix has been either
log-normalized, normalized, or contains raw counts</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="collapse_to_cluster.html">collapse_to_cluster()</a></code> </p>
        </td>
        <td><p>From per-cell calls, take highest freq call in each cluster</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="compare_lists.html">compare_lists()</a></code> </p>
        </td>
        <td><p>Calculate adjusted p-values for hypergeometric test of gene lists
or jaccard index</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="cosine.html">cosine()</a></code> </p>
        </td>
        <td><p>Cosine distance</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="gene_pct.html">gene_pct()</a></code> </p>
        </td>
        <td><p>pct of cells in each cluster that express genelist</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="gene_pct_markerm.html">gene_pct_markerm()</a></code> </p>
        </td>
        <td><p>pct of cells in every cluster that express a series of genelists</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="insert_meta_object.html">insert_meta_object()</a></code> </p>
        </td>
        <td><p>more flexible metadata update of single cell objects</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="kl_divergence.html">kl_divergence()</a></code> </p>
        </td>
        <td><p>KL divergence</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="not_pretty_palette.html">not_pretty_palette</a></code> </p>
        </td>
        <td><p>black and white palette for plotting continous variables</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="object_data.html">object_data()</a></code> </p>
        </td>
        <td><p>Function to access object data</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="overcluster.html">overcluster()</a></code> </p>
        </td>
        <td><p>Overcluster by kmeans per cluster</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="percent_clusters.html">percent_clusters()</a></code> </p>
        </td>
        <td><p>Percentage detected per cluster</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="permute_similarity.html">permute_similarity()</a></code> </p>
        </td>
        <td><p>Compute a p-value for similarity using permutation</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="pretty_palette.html">pretty_palette</a></code> </p>
        </td>
        <td><p>Color palette for plotting continous variables</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="pretty_palette2.html">pretty_palette2</a></code> </p>
        </td>
        <td><p>Color palette for plotting continous variables, starting at gray</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="pretty_palette_ramp_d.html">pretty_palette_ramp_d()</a></code> </p>
        </td>
        <td><p>Expanded color palette ramp for plotting discrete variables</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="vector_similarity.html">vector_similarity()</a></code> </p>
        </td>
        <td><p>Compute similarity between two vectors</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="write_meta.html">write_meta()</a></code> </p>
        </td>
        <td><p>Function to write metadata to object</p></td>
      </tr>
    </tbody>
    </table>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top">
      <h2 data-toc-skip>Contents</h2>
    </nav>
  </div>
</div>


      <footer>
      <div class="copyright">
  <p>Developed by Rui Fu, Kent Riemondy, <a href='https://medschool.cuanschutz.edu/rbi'>RNA Bioscience Initiative</a>.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.6.1.</p>
</div>

      </footer>
   </div>

  


  </body>
</html>


