<!-- Generated by pkgdown: do not edit by hand -->
<!DOCTYPE html>
<html lang="en">
  <head>
  <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<title>Run GSEA to compare a gene list(s) to per cell or
per cluster expression data — run_gsea • clustifyr</title>

<!-- favicons -->
<link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png">
<link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png">
<link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png" />
<link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png" />
<link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png" />
<link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png" />

<!-- jquery -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script>
<!-- Bootstrap -->

<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous" />

<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script>

<!-- bootstrap-toc -->
<link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script>

<!-- Font Awesome icons -->
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous" />
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous" />

<!-- clipboard.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script>

<!-- headroom.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script>

<!-- pkgdown -->
<link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script>




<meta property="og:title" content="Run GSEA to compare a gene list(s) to per cell or
per cluster expression data — run_gsea" />
<meta property="og:description" content="Use fgsea algorithm to compute normalized enrichment
scores and pvalues for gene
set ovelap" />
<meta property="og:image" content="http://rnabioco.github.io/clustifyr/logo.svg" />




<!-- mathjax -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script>

<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->



  </head>

  <body data-spy="scroll" data-target="#toc">
    <div class="container template-reference-topic">
      <header>
      <div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">clustifyr</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">1.1.1</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="../index.html">
    <span class="fas fa fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="../articles/clustifyR.html">Introduction to clustifyr</a>
    </li>
    <li>
      <a href="../articles/geo-annotations.html">Improving NCBI GEO submissions of scRNA-seq data</a>
    </li>
  </ul>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/rnabioco/clustifyr/">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
      
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header>

<div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Run GSEA to compare a gene list(s) to per cell or
per cluster expression data</h1>
    <small class="dont-index">Source: <a href='https://github.com/rnabioco/clustifyr/blob/master/R/run_fgsea.R'><code>R/run_fgsea.R</code></a></small>
    <div class="hidden name"><code>run_gsea.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>Use fgsea algorithm to compute normalized enrichment
scores and pvalues for gene
set ovelap</p>
    </div>

    <pre class="usage"><span class='fu'>run_gsea</span>(
  <span class='kw'>expr_mat</span>,
  <span class='kw'>query_genes</span>,
  cluster_ids = <span class='kw'>NULL</span>,
  n_perm = <span class='fl'>1000</span>,
  per_cell = <span class='fl'>FALSE</span>,
  scale = <span class='fl'>FALSE</span>,
  no_warnings = <span class='fl'>TRUE</span>
)</pre>

    <h2 class="hasAnchor" id="arguments"><a class="anchor" href="#arguments"></a>Arguments</h2>
    <table class="ref-arguments">
    <colgroup><col class="name" /><col class="desc" /></colgroup>
    <tr>
      <th>expr_mat</th>
      <td><p>single-cell expression matrix or Seurat object</p></td>
    </tr>
    <tr>
      <th>query_genes</th>
      <td><p>A vector or named list of vectors of genesets of interest
to compare via GSEA. If supplying a named list, then the gene set names
will appear in the output.</p></td>
    </tr>
    <tr>
      <th>cluster_ids</th>
      <td><p>vector of cell cluster assignments, supplied as a
vector with order that
matches columns in <code>expr_mat</code>. Not required if running per cell.</p></td>
    </tr>
    <tr>
      <th>n_perm</th>
      <td><p>Number of permutation for fgsea function. Defaults to 1000.</p></td>
    </tr>
    <tr>
      <th>per_cell</th>
      <td><p>if true run per cell, otherwise per cluster.</p></td>
    </tr>
    <tr>
      <th>scale</th>
      <td><p>convert expr_mat into zscores prior to running GSEA?,
default = FALSE</p></td>
    </tr>
    <tr>
      <th>no_warnings</th>
      <td><p>suppress warnings from gsea ties</p></td>
    </tr>
    </table>

    <h2 class="hasAnchor" id="value"><a class="anchor" href="#value"></a>Value</h2>

    <p>dataframe of gsea scores (pval, NES), with clusters as rownames</p>

    <h2 class="hasAnchor" id="examples"><a class="anchor" href="#examples"></a>Examples</h2>
    <pre class="examples"><div class='input'><span class='fu'>run_gsea</span>(
    expr_mat = <span class='kw'>pbmc_matrix_small</span>,
    query_genes = <span class='kw'>pbmc_vargenes</span>[<span class='fl'>1</span><span class='op'>:</span><span class='fl'>100</span>],
    n_perm = <span class='fl'>10</span>,
    cluster_ids = <span class='kw'>pbmc_meta</span><span class='op'>$</span><span class='kw'>classified</span>,
    no_warnings = <span class='fl'>TRUE</span>
)
</div><div class='output co'>#&gt;                  pathway       pval      NES
#&gt; B            query_genes 0.09090909 1.376519
#&gt; CD14+ Mono   query_genes 0.09090909 1.380002
#&gt; CD8 T        query_genes 0.09090909 1.343579
#&gt; DC           query_genes 0.09090909 1.410047
#&gt; FCGR3A+ Mono query_genes 0.09090909 1.366121
#&gt; Memory CD4 T query_genes 0.09090909 1.278387
#&gt; Naive CD4 T  query_genes 0.09090909 1.243963
#&gt; NK           query_genes 0.09090909 1.464424
#&gt; Platelet     query_genes 0.09090909 1.171866</div></pre>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top">
      <h2 data-toc-skip>Contents</h2>
    </nav>
  </div>
</div>


      <footer>
      <div class="copyright">
  <p>Developed by Rui Fu, Kent Riemondy.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.6.1.9000.</p>
</div>

      </footer>
   </div>

  


  </body>
</html>


