<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html>
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Compare Methods â€¢ clustifyR</title>
<!-- jquery --><script src="https://code.jquery.com/jquery-3.1.0.min.js" integrity="sha384-nrOSfDHtoPMzJHjVTdCopGqIqeYETSXhZDFyniQ8ZHcVy08QesyHcnOUpMpqnmWq" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/css/bootstrap.min.css" rel="stylesheet" integrity="sha384-BVYiiSIFeK1dGmJRAkycuHAHRg32OmUcww7on3RYdg4Va+PmSTsz/K68vbdEjh4u" crossorigin="anonymous">
<script src="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha384-Tc5IQib027qvyjSMfHjOMaLkfuWVxZxUPnCJA7l2mCWNIpG9mGCD8wGNIcPD7Txa" crossorigin="anonymous"></script><!-- Font Awesome icons --><link href="https://maxcdn.bootstrapcdn.com/font-awesome/4.6.3/css/font-awesome.min.css" rel="stylesheet" integrity="sha384-T8Gy5hrqNKT+hzMclPo118YTQO6cYprQmhrYwIiQ/3axmI1hQomh7Ud2hPOy8SP1" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/1.7.1/clipboard.min.js" integrity="sha384-cV+rhyOuRHc9Ub/91rihWcGmMmCXDeksTtCihMupQHSsi8GIIRDG0ThDc3HGQFJ3" crossorigin="anonymous"></script><!-- sticky kit --><script src="https://cdnjs.cloudflare.com/ajax/libs/sticky-kit/1.1.3/sticky-kit.min.js" integrity="sha256-c4Rlo1ZozqTPE2RLuvbusY3+SU1pQaJC0TjuhygMipw=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Compare Methods">
<meta property="og:description" content="">
<meta name="twitter:card" content="summary">
<!-- mathjax --><script src="https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">clustifyR</a>
        <span class="label label-danger" data-toggle="tooltip" data-placement="bottom" title="Unreleased package">0.0.0.9000</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../index.html">
    <span class="fa fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/compare_methods.html">Compare Methods</a>
    </li>
    <li>
      <a href="../articles/exploratory_pca.html">Exploratory PCA</a>
    </li>
    <li>
      <a href="../articles/gene_list.html">Cluster classification using mark gene lists</a>
    </li>
    <li>
      <a href="../articles/get_recount_data.html">Downloading data from recount</a>
    </li>
    <li>
      <a href="../articles/overview.html">clustifyR overview</a>
    </li>
    <li>
      <a href="../articles/random_forest.html">randomForest model</a>
    </li>
    <li>
      <a href="../articles/refine_cluster.html">Clustering Single Cell RNA-Seq Data Using Reference Bulk RNA-Seq Data</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="http://github.com/NCBI-Hackathons/clustifyR">
    <span class="fa fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      
      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1>Compare Methods</h1>
            
      
      <small class="dont-index">Source: <a href="http://github.com/NCBI-Hackathons/clustifyR/blob/master/vignettes/compare_methods.Rmd"><code>vignettes/compare_methods.Rmd</code></a></small>
      <div class="hidden name"><code>compare_methods.Rmd</code></div>

    </div>

    
    
<p>This vignette illustrates how one can explore how well different feature selection methods work (which gene selection methods yields the most useful correlations). We have 4 different feature selection methods</p>
<ol style="list-style-type: decimal">
<li>Highly variable genes via Seurat</li>
<li>Markers of each cluster</li>
<li>M3Drop identified genes</li>
<li>PCA based method</li>
</ol>
<div id="load-required-example-data-sets" class="section level2">
<h2 class="hasAnchor">
<a href="#load-required-example-data-sets" class="anchor"></a>Load required example data sets</h2>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">library</span>(clustifyR)
<span class="kw">data</span>(<span class="st">"pbmc4k_vargenes"</span>)
<span class="kw">data</span>(<span class="st">"pbmc4k_markers"</span>)
<span class="kw">data</span>(<span class="st">"pbmc4k_markers_M3Drop"</span>)
<span class="kw">data</span>(<span class="st">"pbmc_pca"</span>)
<span class="kw">data</span>(<span class="st">"pbmc4k_avg"</span>)
<span class="kw">data</span>(<span class="st">"pbmc_bulk_matrix"</span>)


<span class="co">#Assign pre-labelled names to single cell clusters</span>
<span class="kw">names</span>(pbmc4k_avg) &lt;-<span class="st"> </span><span class="kw">c</span>(
   <span class="st">"CD4 T cells, 1"</span>,
   <span class="st">"CD14+ Monocytes"</span>,
   <span class="st">"B cells"</span>,
   <span class="st">"CD4 T cells, 2"</span>,
   <span class="st">"CD8 T cells"</span>,
   <span class="st">"NK cells"</span>,
   <span class="st">"Dendritic cells, 1"</span>,
   <span class="st">"FCGR3A+ Monocytes"</span>,
   <span class="st">"Dendritic Cells, 2?"</span>,
   <span class="st">"Megakaryocytes"</span>
)</code></pre></div>
</div>
<div id="function-to-create-correlation-matrix" class="section level2">
<h2 class="hasAnchor">
<a href="#function-to-create-correlation-matrix" class="anchor"></a>Function to create correlation matrix</h2>
<ul>
<li>Row names are the single cell RNA pre-assigned tissue labels.</li>
<li>Column names are the bulk RNA sample names.</li>
<li>Values are the Spearman correlations between the average single cell RNA expression and the bulk RNA sample expression, for a given set of genes (determined by the feature selection methods).</li>
</ul>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">cor_mat &lt;-<span class="st"> </span><span class="kw"><a href="../reference/run_cor.html">run_cor</a></span>(<span class="dt">expr_mat =</span> pbmc4k_matrix,
               <span class="dt">metadata =</span> pbmc4k_meta,
               <span class="dt">bulk_mat =</span> pbmc_bulk_matrix,
               <span class="dt">query_gene_list =</span> pbmc4k_vargenes,
               <span class="dt">cluster_col =</span> <span class="st">"classified"</span>,
               <span class="dt">compute_method =</span> corr_coef)

knitr<span class="op">::</span><span class="kw"><a href="http://www.rdocumentation.org/packages/knitr/topics/kable">kable</a></span>(<span class="kw">round</span>(cor_mat, <span class="dv">2</span>))</code></pre></div>
<table class="table">
<thead><tr class="header">
<th></th>
<th align="right">primary human B cells rep 1</th>
<th align="right">primary human myeloid DC rep 1</th>
<th align="right">primary human monocytes rep 1</th>
<th align="right">primary human neutrophils rep 1</th>
<th align="right">primary human NK cells rep 1</th>
<th align="right">primary human PBMC rep 1</th>
<th align="right">primary human T cells rep 1</th>
<th align="right">primary human B cells rep 2</th>
<th align="right">primary human myeloid DC rep 2</th>
<th align="right">primary human monocytes rep 2</th>
<th align="right">primary human neutrophils rep 2</th>
<th align="right">primary human NK cells rep 2</th>
<th align="right">primary human PBMC rep 2</th>
<th align="right">primary human T cells rep 2</th>
</tr></thead>
<tbody>
<tr class="odd">
<td>B cells</td>
<td align="right">0.61</td>
<td align="right">0.31</td>
<td align="right">0.20</td>
<td align="right">0.10</td>
<td align="right">0.13</td>
<td align="right">0.28</td>
<td align="right">0.17</td>
<td align="right">0.58</td>
<td align="right">0.31</td>
<td align="right">0.21</td>
<td align="right">0.11</td>
<td align="right">0.13</td>
<td align="right">0.30</td>
<td align="right">0.19</td>
</tr>
<tr class="even">
<td>CD14+ Monocytes</td>
<td align="right">0.16</td>
<td align="right">0.56</td>
<td align="right">0.57</td>
<td align="right">0.33</td>
<td align="right">0.15</td>
<td align="right">0.47</td>
<td align="right">0.13</td>
<td align="right">0.14</td>
<td align="right">0.56</td>
<td align="right">0.60</td>
<td align="right">0.37</td>
<td align="right">0.14</td>
<td align="right">0.48</td>
<td align="right">0.15</td>
</tr>
<tr class="odd">
<td>CD4 T cells, 1</td>
<td align="right">0.26</td>
<td align="right">0.32</td>
<td align="right">0.24</td>
<td align="right">0.14</td>
<td align="right">0.28</td>
<td align="right">0.40</td>
<td align="right">0.47</td>
<td align="right">0.22</td>
<td align="right">0.33</td>
<td align="right">0.25</td>
<td align="right">0.15</td>
<td align="right">0.28</td>
<td align="right">0.43</td>
<td align="right">0.49</td>
</tr>
<tr class="even">
<td>CD4 T cells, 2</td>
<td align="right">0.27</td>
<td align="right">0.33</td>
<td align="right">0.24</td>
<td align="right">0.14</td>
<td align="right">0.31</td>
<td align="right">0.43</td>
<td align="right">0.49</td>
<td align="right">0.23</td>
<td align="right">0.34</td>
<td align="right">0.26</td>
<td align="right">0.15</td>
<td align="right">0.30</td>
<td align="right">0.44</td>
<td align="right">0.51</td>
</tr>
<tr class="odd">
<td>CD8 T cells</td>
<td align="right">0.20</td>
<td align="right">0.27</td>
<td align="right">0.19</td>
<td align="right">0.11</td>
<td align="right">0.41</td>
<td align="right">0.36</td>
<td align="right">0.36</td>
<td align="right">0.17</td>
<td align="right">0.27</td>
<td align="right">0.20</td>
<td align="right">0.12</td>
<td align="right">0.41</td>
<td align="right">0.36</td>
<td align="right">0.39</td>
</tr>
<tr class="even">
<td>Dendritic cells, 1</td>
<td align="right">0.22</td>
<td align="right">0.52</td>
<td align="right">0.44</td>
<td align="right">0.25</td>
<td align="right">0.15</td>
<td align="right">0.41</td>
<td align="right">0.13</td>
<td align="right">0.19</td>
<td align="right">0.52</td>
<td align="right">0.46</td>
<td align="right">0.27</td>
<td align="right">0.15</td>
<td align="right">0.41</td>
<td align="right">0.15</td>
</tr>
<tr class="odd">
<td>Dendritic Cells, 2?</td>
<td align="right">0.28</td>
<td align="right">0.33</td>
<td align="right">0.26</td>
<td align="right">0.10</td>
<td align="right">0.14</td>
<td align="right">0.29</td>
<td align="right">0.13</td>
<td align="right">0.25</td>
<td align="right">0.34</td>
<td align="right">0.26</td>
<td align="right">0.11</td>
<td align="right">0.13</td>
<td align="right">0.28</td>
<td align="right">0.14</td>
</tr>
<tr class="even">
<td>FCGR3A+ Monocytes</td>
<td align="right">0.17</td>
<td align="right">0.64</td>
<td align="right">0.49</td>
<td align="right">0.27</td>
<td align="right">0.14</td>
<td align="right">0.44</td>
<td align="right">0.13</td>
<td align="right">0.15</td>
<td align="right">0.65</td>
<td align="right">0.52</td>
<td align="right">0.32</td>
<td align="right">0.14</td>
<td align="right">0.44</td>
<td align="right">0.14</td>
</tr>
<tr class="odd">
<td>Megakaryocytes</td>
<td align="right">0.06</td>
<td align="right">0.26</td>
<td align="right">0.22</td>
<td align="right">0.13</td>
<td align="right">0.06</td>
<td align="right">0.20</td>
<td align="right">0.06</td>
<td align="right">0.06</td>
<td align="right">0.26</td>
<td align="right">0.23</td>
<td align="right">0.15</td>
<td align="right">0.07</td>
<td align="right">0.21</td>
<td align="right">0.07</td>
</tr>
<tr class="even">
<td>NK cells</td>
<td align="right">0.13</td>
<td align="right">0.24</td>
<td align="right">0.17</td>
<td align="right">0.10</td>
<td align="right">0.61</td>
<td align="right">0.34</td>
<td align="right">0.24</td>
<td align="right">0.11</td>
<td align="right">0.24</td>
<td align="right">0.18</td>
<td align="right">0.11</td>
<td align="right">0.59</td>
<td align="right">0.29</td>
<td align="right">0.26</td>
</tr>
</tbody>
</table>
</div>
<div id="function-to-create-best-match-matrix" class="section level2">
<h2 class="hasAnchor">
<a href="#function-to-create-best-match-matrix" class="anchor"></a>Function to create best match matrix</h2>
<p>This function takes a correlation matrix, and determines which bulk RNA sample is the best match (1) for a single cell RNA tissue</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">get_best_match_matrix &lt;-<span class="st"> </span><span class="cf">function</span>(correlation_matrix) {
  
  best_mat &lt;-<span class="st"> </span><span class="kw">as.data.frame</span>(<span class="kw">t</span>(<span class="kw">apply</span>(correlation_matrix, <span class="dv">1</span>, <span class="cf">function</span>(x) x <span class="op">-</span><span class="st"> </span><span class="kw">max</span>(x))))
  best_mat[best_mat<span class="op">==</span><span class="dv">0</span>]=<span class="st">"1"</span>
  best_mat[best_mat<span class="op">!=</span><span class="st">"1"</span>]=<span class="st">"0"</span>

  <span class="kw">return</span> (best_mat)
}

best_match_matrix &lt;-<span class="st"> </span><span class="kw">get_best_match_matrix</span>(cor_mat)

knitr<span class="op">::</span><span class="kw"><a href="http://www.rdocumentation.org/packages/knitr/topics/kable">kable</a></span>(best_match_matrix)</code></pre></div>
<table class="table">
<thead><tr class="header">
<th></th>
<th align="left">primary human B cells rep 1</th>
<th align="left">primary human myeloid DC rep 1</th>
<th align="left">primary human monocytes rep 1</th>
<th align="left">primary human neutrophils rep 1</th>
<th align="left">primary human NK cells rep 1</th>
<th align="left">primary human PBMC rep 1</th>
<th align="left">primary human T cells rep 1</th>
<th align="left">primary human B cells rep 2</th>
<th align="left">primary human myeloid DC rep 2</th>
<th align="left">primary human monocytes rep 2</th>
<th align="left">primary human neutrophils rep 2</th>
<th align="left">primary human NK cells rep 2</th>
<th align="left">primary human PBMC rep 2</th>
<th align="left">primary human T cells rep 2</th>
</tr></thead>
<tbody>
<tr class="odd">
<td>B cells</td>
<td align="left">1</td>
<td align="left">0</td>
<td align="left">0</td>
<td align="left">0</td>
<td align="left">0</td>
<td align="left">0</td>
<td align="left">0</td>
<td align="left">0</td>
<td align="left">0</td>
<td align="left">0</td>
<td align="left">0</td>
<td align="left">0</td>
<td align="left">0</td>
<td align="left">0</td>
</tr>
<tr class="even">
<td>CD14+ Monocytes</td>
<td align="left">0</td>
<td align="left">0</td>
<td align="left">0</td>
<td align="left">0</td>
<td align="left">0</td>
<td align="left">0</td>
<td align="left">0</td>
<td align="left">0</td>
<td align="left">0</td>
<td align="left">1</td>
<td align="left">0</td>
<td align="left">0</td>
<td align="left">0</td>
<td align="left">0</td>
</tr>
<tr class="odd">
<td>CD4 T cells, 1</td>
<td align="left">0</td>
<td align="left">0</td>
<td align="left">0</td>
<td align="left">0</td>
<td align="left">0</td>
<td align="left">0</td>
<td align="left">0</td>
<td align="left">0</td>
<td align="left">0</td>
<td align="left">0</td>
<td align="left">0</td>
<td align="left">0</td>
<td align="left">0</td>
<td align="left">1</td>
</tr>
<tr class="even">
<td>CD4 T cells, 2</td>
<td align="left">0</td>
<td align="left">0</td>
<td align="left">0</td>
<td align="left">0</td>
<td align="left">0</td>
<td align="left">0</td>
<td align="left">0</td>
<td align="left">0</td>
<td align="left">0</td>
<td align="left">0</td>
<td align="left">0</td>
<td align="left">0</td>
<td align="left">0</td>
<td align="left">1</td>
</tr>
<tr class="odd">
<td>CD8 T cells</td>
<td align="left">0</td>
<td align="left">0</td>
<td align="left">0</td>
<td align="left">0</td>
<td align="left">1</td>
<td align="left">0</td>
<td align="left">0</td>
<td align="left">0</td>
<td align="left">0</td>
<td align="left">0</td>
<td align="left">0</td>
<td align="left">0</td>
<td align="left">0</td>
<td align="left">0</td>
</tr>
<tr class="even">
<td>Dendritic cells, 1</td>
<td align="left">0</td>
<td align="left">0</td>
<td align="left">0</td>
<td align="left">0</td>
<td align="left">0</td>
<td align="left">0</td>
<td align="left">0</td>
<td align="left">0</td>
<td align="left">1</td>
<td align="left">0</td>
<td align="left">0</td>
<td align="left">0</td>
<td align="left">0</td>
<td align="left">0</td>
</tr>
<tr class="odd">
<td>Dendritic Cells, 2?</td>
<td align="left">0</td>
<td align="left">0</td>
<td align="left">0</td>
<td align="left">0</td>
<td align="left">0</td>
<td align="left">0</td>
<td align="left">0</td>
<td align="left">0</td>
<td align="left">1</td>
<td align="left">0</td>
<td align="left">0</td>
<td align="left">0</td>
<td align="left">0</td>
<td align="left">0</td>
</tr>
<tr class="even">
<td>FCGR3A+ Monocytes</td>
<td align="left">0</td>
<td align="left">0</td>
<td align="left">0</td>
<td align="left">0</td>
<td align="left">0</td>
<td align="left">0</td>
<td align="left">0</td>
<td align="left">0</td>
<td align="left">1</td>
<td align="left">0</td>
<td align="left">0</td>
<td align="left">0</td>
<td align="left">0</td>
<td align="left">0</td>
</tr>
<tr class="odd">
<td>Megakaryocytes</td>
<td align="left">0</td>
<td align="left">0</td>
<td align="left">0</td>
<td align="left">0</td>
<td align="left">0</td>
<td align="left">0</td>
<td align="left">0</td>
<td align="left">0</td>
<td align="left">1</td>
<td align="left">0</td>
<td align="left">0</td>
<td align="left">0</td>
<td align="left">0</td>
<td align="left">0</td>
</tr>
<tr class="even">
<td>NK cells</td>
<td align="left">0</td>
<td align="left">0</td>
<td align="left">0</td>
<td align="left">0</td>
<td align="left">1</td>
<td align="left">0</td>
<td align="left">0</td>
<td align="left">0</td>
<td align="left">0</td>
<td align="left">0</td>
<td align="left">0</td>
<td align="left">0</td>
<td align="left">0</td>
<td align="left">0</td>
</tr>
</tbody>
</table>
</div>
<div id="summary-of-best-matches" class="section level2">
<h2 class="hasAnchor">
<a href="#summary-of-best-matches" class="anchor"></a>Summary of best matches</h2>
<p>Using the above functions, we can now summarize and compare and contrast the best matches for each of the different feature selection methods.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">vargenes_cor_mat &lt;-<span class="st"> </span><span class="kw"><a href="../reference/run_cor.html">run_cor</a></span>(<span class="dt">expr_mat =</span> pbmc4k_matrix,
               <span class="dt">metadata =</span> pbmc4k_meta,
               <span class="dt">bulk_mat =</span> pbmc_bulk_matrix,
               <span class="dt">query_gene_list =</span> pbmc4k_vargenes,
               <span class="dt">cluster_col =</span> <span class="st">"classified"</span>,
               <span class="dt">compute_method =</span> corr_coef)

vargenes_best_mat &lt;-<span class="st"> </span><span class="kw">get_best_match_matrix</span>(vargenes_cor_mat)

markers_cor_mat &lt;-<span class="st"> </span><span class="kw"><a href="../reference/run_cor.html">run_cor</a></span>(<span class="dt">expr_mat =</span> pbmc4k_matrix,
               <span class="dt">metadata =</span> pbmc4k_meta,
               <span class="dt">bulk_mat =</span> pbmc_bulk_matrix,
               <span class="dt">query_gene_list =</span> pbmc4k_markers<span class="op">$</span>gene,
               <span class="dt">cluster_col =</span> <span class="st">"classified"</span>,
               <span class="dt">compute_method =</span> corr_coef)

markers_best_mat &lt;-<span class="st"> </span><span class="kw">get_best_match_matrix</span>(markers_cor_mat)


m3drop_cor_mat &lt;-<span class="st"> </span><span class="kw"><a href="../reference/run_cor.html">run_cor</a></span>(<span class="dt">expr_mat =</span> pbmc4k_matrix,
               <span class="dt">metadata =</span> pbmc4k_meta,
               <span class="dt">bulk_mat =</span> pbmc_bulk_matrix,
               <span class="dt">query_gene_list =</span> pbmc4k_markers_M3Drop<span class="op">$</span>Gene,
               <span class="dt">cluster_col =</span> <span class="st">"classified"</span>,
               <span class="dt">compute_method =</span> corr_coef)

m3drop_best_mat &lt;-<span class="st"> </span><span class="kw">get_best_match_matrix</span>(m3drop_cor_mat)


pca_cor_mat &lt;-<span class="st"> </span><span class="kw"><a href="../reference/run_cor.html">run_cor</a></span>(<span class="dt">expr_mat =</span> pbmc4k_matrix,
               <span class="dt">metadata =</span> pbmc4k_meta,
               <span class="dt">bulk_mat =</span> pbmc_bulk_matrix,
               <span class="dt">query_gene_list =</span> pbmc_pca,
               <span class="dt">cluster_col =</span> <span class="st">"classified"</span>,
               <span class="dt">compute_method =</span> corr_coef)

pca_best_mat &lt;-<span class="st"> </span><span class="kw">get_best_match_matrix</span>(pca_cor_mat)

get_best_str &lt;-<span class="st"> </span><span class="cf">function</span>(name, best_mat, cor_mat) {
  <span class="cf">if</span> (<span class="kw">sum</span>(<span class="kw">as.numeric</span>(best_mat[,name])) <span class="op">&gt;</span><span class="st"> </span><span class="dv">0</span>) {
      best.names &lt;-<span class="st"> </span><span class="kw">rownames</span>(best_mat)[<span class="kw">which</span>(best_mat[,name]<span class="op">==</span><span class="dv">1</span>)]  
      best.cor &lt;-<span class="st"> </span><span class="kw">round</span>(cor_mat[<span class="kw">which</span>(best_mat[,name]<span class="op">==</span><span class="dv">1</span>),name],<span class="dv">2</span>)
      <span class="cf">for</span> (i <span class="cf">in</span> <span class="dv">1</span><span class="op">:</span><span class="kw">length</span>(best.cor)) {
        <span class="cf">if</span> (i <span class="op">==</span><span class="st"> </span><span class="dv">1</span>) {
          str &lt;-<span class="st"> </span><span class="kw">paste0</span>(best.names[i], <span class="st">" ("</span>, best.cor[i], <span class="st">") "</span>) 
        } <span class="cf">else</span> {
            str &lt;-<span class="st"> </span><span class="kw">paste0</span>(str, <span class="st">"; "</span>, best.names[i], <span class="st">" ("</span>, best.cor[i], <span class="st">") "</span>) 
        }
      }
  } <span class="cf">else</span> {
    str &lt;-<span class="st"> ""</span>
  }
  <span class="kw">return</span>(str)
}

summary &lt;-<span class="st"> </span><span class="kw">data.frame</span>()
bulk_rna_sample_names &lt;-<span class="st"> </span><span class="kw">colnames</span>(pbmc_bulk_matrix)
<span class="cf">for</span> (name <span class="cf">in</span> bulk_rna_sample_names) {
  variable_genes &lt;-<span class="st"> </span><span class="kw">get_best_str</span>(name, vargenes_best_mat, vargenes_cor_mat)
  marker_genes &lt;-<span class="st"> </span><span class="kw">get_best_str</span>(name, markers_best_mat, markers_cor_mat)
  m3drop_genes &lt;-<span class="st"> </span><span class="kw">get_best_str</span>(name, m3drop_best_mat, m3drop_cor_mat)
  pca_genes &lt;-<span class="st"> </span><span class="kw">get_best_str</span>(name, pca_best_mat, pca_cor_mat)
  <span class="cf">if</span> (<span class="kw">paste0</span>(variable_genes, marker_genes, m3drop_genes, pca_genes) <span class="op">!=</span><span class="st"> ""</span>) {
      row &lt;-<span class="st"> </span><span class="kw">data.frame</span>(
        variable_genes,
        marker_genes,
        m3drop_genes,
        pca_genes
      )
      <span class="kw">rownames</span>(row) &lt;-<span class="st"> </span>name
      summary &lt;-<span class="st"> </span><span class="kw">rbind</span>(summary,row)    
  }
}

knitr<span class="op">::</span><span class="kw"><a href="http://www.rdocumentation.org/packages/knitr/topics/kable">kable</a></span>(summary)</code></pre></div>
<table class="table">
<thead><tr class="header">
<th></th>
<th align="left">variable_genes</th>
<th align="left">marker_genes</th>
<th align="left">m3drop_genes</th>
<th align="left">pca_genes</th>
</tr></thead>
<tbody>
<tr class="odd">
<td>primary human B cells rep 1</td>
<td align="left">B cells (0.61)</td>
<td align="left">B cells (0.51)</td>
<td align="left">B cells (0.72)</td>
<td align="left">B cells (0.7) ; Dendritic Cells, 2? (0.24)</td>
</tr>
<tr class="even">
<td>primary human myeloid DC rep 1</td>
<td align="left"></td>
<td align="left">FCGR3A+ Monocytes (0.51)</td>
<td align="left"></td>
<td align="left"></td>
</tr>
<tr class="odd">
<td>primary human NK cells rep 1</td>
<td align="left">CD8 T cells (0.41) ; NK cells (0.61)</td>
<td align="left"></td>
<td align="left">NK cells (0.61)</td>
<td align="left">NK cells (0.66)</td>
</tr>
<tr class="even">
<td>primary human myeloid DC rep 2</td>
<td align="left">Dendritic cells, 1 (0.52) ; Dendritic Cells, 2? (0.34) ; FCGR3A+ Monocytes (0.65) ; Megakaryocytes (0.26)</td>
<td align="left"></td>
<td align="left">Dendritic cells, 1 (0.59) ; Dendritic Cells, 2? (0.38) ; FCGR3A+ Monocytes (0.68)</td>
<td align="left">FCGR3A+ Monocytes (0.72)</td>
</tr>
<tr class="odd">
<td>primary human monocytes rep 2</td>
<td align="left">CD14+ Monocytes (0.6)</td>
<td align="left">CD14+ Monocytes (0.51)</td>
<td align="left">CD14+ Monocytes (0.62) ; Megakaryocytes (0.27)</td>
<td align="left">CD14+ Monocytes (0.74) ; Dendritic cells, 1 (0.67)</td>
</tr>
<tr class="even">
<td>primary human PBMC rep 2</td>
<td align="left"></td>
<td align="left">Dendritic cells, 1 (0.44) ; Megakaryocytes (0.4)</td>
<td align="left"></td>
<td align="left">Megakaryocytes (0.54)</td>
</tr>
<tr class="odd">
<td>primary human T cells rep 2</td>
<td align="left">CD4 T cells, 1 (0.49) ; CD4 T cells, 2 (0.51)</td>
<td align="left">CD4 T cells, 1 (0.52) ; CD4 T cells, 2 (0.56) ; CD8 T cells (0.54) ; Dendritic Cells, 2? (0.39) ; NK cells (0.51)</td>
<td align="left">CD4 T cells, 1 (0.56) ; CD4 T cells, 2 (0.51) ; CD8 T cells (0.38)</td>
<td align="left">CD4 T cells, 1 (0.61) ; CD4 T cells, 2 (0.62) ; CD8 T cells (0.44)</td>
</tr>
</tbody>
</table>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">
        <div id="tocnav">
      <h2 class="hasAnchor">
<a href="#tocnav" class="anchor"></a>Contents</h2>
      <ul class="nav nav-pills nav-stacked">
<li><a href="#load-required-example-data-sets">Load required example data sets</a></li>
      <li><a href="#function-to-create-correlation-matrix">Function to create correlation matrix</a></li>
      <li><a href="#function-to-create-best-match-matrix">Function to create best match matrix</a></li>
      <li><a href="#summary-of-best-matches">Summary of best matches</a></li>
      </ul>
</div>
      </div>

</div>


      <footer><div class="copyright">
  <p>Developed by Kent Riemondy, Austin Gillen, Rui Fu, Jay Hesselberth, Yue Hao, Chengzhe Tian, M. Day.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="http://pkgdown.r-lib.org/">pkgdown</a>.</p>
</div>

      </footer>
</div>

  

  </body>
</html>
