<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html>
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Cluster classification using mark gene lists â€¢ clustifyR</title>
<!-- jquery --><script src="https://code.jquery.com/jquery-3.1.0.min.js" integrity="sha384-nrOSfDHtoPMzJHjVTdCopGqIqeYETSXhZDFyniQ8ZHcVy08QesyHcnOUpMpqnmWq" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/css/bootstrap.min.css" rel="stylesheet" integrity="sha384-BVYiiSIFeK1dGmJRAkycuHAHRg32OmUcww7on3RYdg4Va+PmSTsz/K68vbdEjh4u" crossorigin="anonymous">
<script src="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha384-Tc5IQib027qvyjSMfHjOMaLkfuWVxZxUPnCJA7l2mCWNIpG9mGCD8wGNIcPD7Txa" crossorigin="anonymous"></script><!-- Font Awesome icons --><link href="https://maxcdn.bootstrapcdn.com/font-awesome/4.6.3/css/font-awesome.min.css" rel="stylesheet" integrity="sha384-T8Gy5hrqNKT+hzMclPo118YTQO6cYprQmhrYwIiQ/3axmI1hQomh7Ud2hPOy8SP1" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/1.7.1/clipboard.min.js" integrity="sha384-cV+rhyOuRHc9Ub/91rihWcGmMmCXDeksTtCihMupQHSsi8GIIRDG0ThDc3HGQFJ3" crossorigin="anonymous"></script><!-- sticky kit --><script src="https://cdnjs.cloudflare.com/ajax/libs/sticky-kit/1.1.3/sticky-kit.min.js" integrity="sha256-c4Rlo1ZozqTPE2RLuvbusY3+SU1pQaJC0TjuhygMipw=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Cluster classification using mark gene lists">
<meta property="og:description" content="">
<meta property="og:image" content="http://rnabioco.github.io/clustifyR/logo.png">
<meta name="twitter:card" content="summary">
<!-- mathjax --><script src="https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">clustifyR</a>
        <span class="label label-danger" data-toggle="tooltip" data-placement="bottom" title="Unreleased package">0.0.0.9000</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../index.html">
    <span class="fa fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../articles/clustifyR.html">Get started</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/compare_methods.html">Compare Methods</a>
    </li>
    <li>
      <a href="../articles/exploratory_pca.html">Exploratory PCA</a>
    </li>
    <li>
      <a href="../articles/gene_list.html">Cluster classification using mark gene lists</a>
    </li>
    <li>
      <a href="../articles/get_recount_data.html">Downloading data from recount</a>
    </li>
    <li>
      <a href="../articles/random_forest.html">randomForest model</a>
    </li>
    <li>
      <a href="../articles/refine_cluster.html">Clustering Single Cell RNA-Seq Data Using Reference Bulk RNA-Seq Data</a>
    </li>
  </ul>
</li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/rnabioco/clustifyR">
    <span class="fa fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      
      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1>Cluster classification using mark gene lists</h1>
            
            <h4 class="date">2018-09-13</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/rnabioco/clustifyR/blob/master/vignettes/gene_list.Rmd"><code>vignettes/gene_list.Rmd</code></a></small>
      <div class="hidden name"><code>gene_list.Rmd</code></div>

    </div>

    
    
<p>This vignette demonstrates how to apply ranked or unranked gene lists, based on prior knowledge of the cells/samples, to cluster classification. Here, average single cell RNA expression data for each cluster is used as input. The approach does work with full single cell expression matrix as well.</p>
<div id="unranked-list-preprocessing-gene-list-into-dataframematrix-form" class="section level2">
<h2 class="hasAnchor">
<a href="#unranked-list-preprocessing-gene-list-into-dataframematrix-form" class="anchor"></a>Unranked list: Preprocessing gene list into dataframe/matrix form</h2>
<p>An unranked gene list can be converted from Seurat::FindAllMarkers(), or assembled from vectors of genes, into a format as shown below:</p>
<div class="sourceCode" id="cb1"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb1-1" data-line-number="1"><span class="kw">library</span>(dplyr)</a>
<a class="sourceLine" id="cb1-2" data-line-number="2"><span class="kw">library</span>(purrr)</a>
<a class="sourceLine" id="cb1-3" data-line-number="3"><span class="kw">library</span>(stringr)</a>
<a class="sourceLine" id="cb1-4" data-line-number="4"><span class="kw">library</span>(ComplexHeatmap)</a></code></pre></div>
<div class="sourceCode" id="cb2"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb2-1" data-line-number="1"><span class="kw">library</span>(clustifyR)</a>
<a class="sourceLine" id="cb2-2" data-line-number="2"><span class="co"># from dataframe generated by Seurat</span></a>
<a class="sourceLine" id="cb2-3" data-line-number="3">pbmc4k_mm &lt;-<span class="st"> </span><span class="kw"><a href="../reference/matrixize_markers.html">matrixize_markers</a></span>(pbmc4k_markers)</a>
<a class="sourceLine" id="cb2-4" data-line-number="4"><span class="kw">head</span>(pbmc4k_mm)</a></code></pre></div>
<pre><code>##       0             1     2     3     4      5       6              7
## 1 RPL31        S100A8 CD79A  IL7R  GZMK FGFBP2  FCER1A           HES4
## 2  RPS6        S100A9  IGHM ITGB1  CCL5   PRF1 CLEC10A         CDKN1C
## 3 RPS25           LYZ CD79B  IL32 TRGC2   HOPX    ENHO RP11-1008C21.1
## 4 RPS12       S100A12  IGHD  AQP3  CST7   GZMB   CLIC2         FCGR3A
## 5 RPS27 RP11-1143G9.4 TCL1A  TRAC  GZMA  KLRD1    FLT3         TCF7L2
## 6 RPL32          FCN1 MS4A1   LTB  LYAR  SPON2    PKIB            CKB
##        8      9
## 1 LILRA4    PF4
## 2 LRRC26   SDPR
## 3   TPM2  TUBB1
## 4    SCT  ACRBP
## 5  PTPRS    GP9
## 6 CLEC4C TMEM40</code></pre>
<div class="sourceCode" id="cb4"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb4-1" data-line-number="1"><span class="co"># or just lists of genes</span></a>
<a class="sourceLine" id="cb4-2" data-line-number="2">NKm &lt;-<span class="st"> </span><span class="kw">c</span>(<span class="st">"GNLY"</span>, <span class="st">"NKG7"</span>, <span class="st">"ZBTB32"</span>, <span class="st">"IL2RB"</span>)</a>
<a class="sourceLine" id="cb4-3" data-line-number="3">DCm &lt;-<span class="st"> </span><span class="kw">c</span>(<span class="st">"FCER1A"</span>, <span class="st">"CST3"</span>, <span class="st">"ITGAX"</span>, <span class="st">"IRF8"</span>)</a>
<a class="sourceLine" id="cb4-4" data-line-number="4">Bm &lt;-<span class="st"> </span><span class="kw">c</span>(<span class="st">"MS4A1"</span>, <span class="st">"CD79A"</span>,<span class="st">"CD79B"</span>,<span class="st">"CD37"</span>)</a>
<a class="sourceLine" id="cb4-5" data-line-number="5">mm2 &lt;-<span class="st"> </span><span class="kw">data.frame</span>(Bm, DCm, NKm)</a></code></pre></div>
</div>
<div id="unranked-list-binarize-single-cell-gene-expression-data" class="section level2">
<h2 class="hasAnchor">
<a href="#unranked-list-binarize-single-cell-gene-expression-data" class="anchor"></a>Unranked list: Binarize single cell gene expression data</h2>
<p>Cluster average gene expression is transformed into binary 0 or 1s (defaults to keeping the top1000 genes by expression)</p>
<div class="sourceCode" id="cb5"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb5-1" data-line-number="1">pbmc4k_avgb &lt;-<span class="st"> </span><span class="kw"><a href="../reference/binarize_expr.html">binarize_expr</a></span>(pbmc4k_avg)</a></code></pre></div>
</div>
<div id="unranked-list-classify" class="section level2">
<h2 class="hasAnchor">
<a href="#unranked-list-classify" class="anchor"></a>Unranked list: Classify</h2>
<p>Applying the gene lists to binary expression data ouputs a matrix of -log10 of adjusted P values for hypergeometric test, or jaccard index. Correct cluster identity is called when only 4 marker genes is provided in PBMCs.</p>
<div class="sourceCode" id="cb6"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb6-1" data-line-number="1">res &lt;-<span class="st"> </span><span class="kw"><a href="../reference/compare_lists.html">compare_lists</a></span>(pbmc4k_avgb,pbmc4k_mm)</a>
<a class="sourceLine" id="cb6-2" data-line-number="2"><span class="kw"><a href="http://www.rdocumentation.org/packages/ComplexHeatmap/topics/Heatmap">Heatmap</a></span>(res, <span class="dt">cluster_rows =</span> <span class="ot">FALSE</span>, <span class="dt">cluster_columns =</span> <span class="ot">FALSE</span>, <span class="dt">heatmap_legend_param =</span> <span class="kw">list</span>(<span class="dt">title =</span> <span class="st">"-log10(adjP)"</span>))</a></code></pre></div>
<p><img src="gene_list_files/figure-html/unnamed-chunk-4-1.png" width="700"></p>
<div class="sourceCode" id="cb7"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb7-1" data-line-number="1">res2 &lt;-<span class="st"> </span><span class="kw"><a href="../reference/compare_lists.html">compare_lists</a></span>(pbmc4k_avgb,mm2)</a>
<a class="sourceLine" id="cb7-2" data-line-number="2"><span class="kw"><a href="http://www.rdocumentation.org/packages/ComplexHeatmap/topics/Heatmap">Heatmap</a></span>(res2, <span class="dt">cluster_rows =</span> <span class="ot">FALSE</span>, <span class="dt">cluster_columns =</span> <span class="ot">FALSE</span>, <span class="dt">heatmap_legend_param =</span> <span class="kw">list</span>(<span class="dt">title =</span> <span class="st">"-log10(adjP)"</span>))</a></code></pre></div>
<p><img src="gene_list_files/figure-html/unnamed-chunk-4-2.png" width="700"></p>
</div>
<div id="ranked-list-preprocessing-gene-list-into-dataframematrix-form" class="section level2">
<h2 class="hasAnchor">
<a href="#ranked-list-preprocessing-gene-list-into-dataframematrix-form" class="anchor"></a>Ranked list: Preprocessing gene list into dataframe/matrix form</h2>
<p>Arbitrary pseudo expression numbers are assigned when a ranked gene list is provided.</p>
<div class="sourceCode" id="cb8"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb8-1" data-line-number="1">pbmc4k_rm &lt;-<span class="st"> </span><span class="kw"><a href="../reference/matrixize_markers.html">matrixize_markers</a></span>(pbmc4k_markers, <span class="ot">TRUE</span>)</a></code></pre></div>
</div>
<div id="ranked-list-classify" class="section level2">
<h2 class="hasAnchor">
<a href="#ranked-list-classify" class="anchor"></a>Ranked list: Classify</h2>
<p>Spearman correlation or other approaches implemented in the package can be applied.</p>
<div class="sourceCode" id="cb9"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb9-1" data-line-number="1">gene_constraints &lt;-<span class="st"> </span><span class="kw"><a href="../reference/get_common_elements.html">get_common_elements</a></span>(<span class="kw">rownames</span>(pbmc4k_avg),</a>
<a class="sourceLine" id="cb9-2" data-line-number="2">                                        <span class="kw">rownames</span>(pbmc4k_rm))</a>
<a class="sourceLine" id="cb9-3" data-line-number="3">pbmc4k_avg_d&lt;-<span class="st"> </span>pbmc4k_avg[gene_constraints,]</a>
<a class="sourceLine" id="cb9-4" data-line-number="4">pbmc4k_rm_d &lt;-<span class="st"> </span>pbmc4k_rm[gene_constraints, ]</a>
<a class="sourceLine" id="cb9-5" data-line-number="5">out &lt;-<span class="st"> </span><span class="kw">lapply</span>(<span class="kw">colnames</span>(pbmc4k_avg_d),</a>
<a class="sourceLine" id="cb9-6" data-line-number="6">              <span class="cf">function</span>(x){</a>
<a class="sourceLine" id="cb9-7" data-line-number="7">                per_col &lt;-<span class="st"> </span><span class="kw">lapply</span>(<span class="kw">colnames</span>(pbmc4k_rm_d),</a>
<a class="sourceLine" id="cb9-8" data-line-number="8">                                  <span class="cf">function</span>(y){</a>
<a class="sourceLine" id="cb9-9" data-line-number="9">                                    <span class="kw">cor</span>(pbmc4k_avg_d[,x],</a>
<a class="sourceLine" id="cb9-10" data-line-number="10">                                        pbmc4k_rm_d[,y],</a>
<a class="sourceLine" id="cb9-11" data-line-number="11">                                        <span class="dt">method =</span> <span class="st">"spearman"</span>)})</a>
<a class="sourceLine" id="cb9-12" data-line-number="12">                <span class="kw">do.call</span>(cbind, per_col)</a>
<a class="sourceLine" id="cb9-13" data-line-number="13">              })</a>
<a class="sourceLine" id="cb9-14" data-line-number="14"></a>
<a class="sourceLine" id="cb9-15" data-line-number="15">res &lt;-<span class="st"> </span><span class="kw">do.call</span>(rbind, out)</a>
<a class="sourceLine" id="cb9-16" data-line-number="16"><span class="kw">rownames</span>(res) &lt;-<span class="st"> </span><span class="kw">colnames</span>(pbmc4k_avg_d)</a>
<a class="sourceLine" id="cb9-17" data-line-number="17"><span class="kw">colnames</span>(res) &lt;-<span class="st"> </span><span class="kw">colnames</span>(pbmc4k_rm_d)</a>
<a class="sourceLine" id="cb9-18" data-line-number="18"></a>
<a class="sourceLine" id="cb9-19" data-line-number="19"><span class="kw"><a href="../reference/plot_cor.html">plot_cor</a></span>(res,</a>
<a class="sourceLine" id="cb9-20" data-line-number="20">         pbmc4k_meta,</a>
<a class="sourceLine" id="cb9-21" data-line-number="21">         <span class="kw">colnames</span>(res)[<span class="dv">1</span><span class="op">:</span><span class="dv">7</span>],</a>
<a class="sourceLine" id="cb9-22" data-line-number="22">         <span class="dt">cluster_col =</span> <span class="st">"cluster"</span>)</a></code></pre></div>
<pre><code>## [[1]]</code></pre>
<p><img src="gene_list_files/figure-html/unnamed-chunk-6-1.png" width="700"></p>
<pre><code>## 
## [[2]]</code></pre>
<p><img src="gene_list_files/figure-html/unnamed-chunk-6-2.png" width="700"></p>
<pre><code>## 
## [[3]]</code></pre>
<p><img src="gene_list_files/figure-html/unnamed-chunk-6-3.png" width="700"></p>
<pre><code>## 
## [[4]]</code></pre>
<p><img src="gene_list_files/figure-html/unnamed-chunk-6-4.png" width="700"></p>
<pre><code>## 
## [[5]]</code></pre>
<p><img src="gene_list_files/figure-html/unnamed-chunk-6-5.png" width="700"></p>
<pre><code>## 
## [[6]]</code></pre>
<p><img src="gene_list_files/figure-html/unnamed-chunk-6-6.png" width="700"></p>
<pre><code>## 
## [[7]]</code></pre>
<p><img src="gene_list_files/figure-html/unnamed-chunk-6-7.png" width="700"></p>
</div>
<div id="pancreas-data-from-indrop" class="section level2">
<h2 class="hasAnchor">
<a href="#pancreas-data-from-indrop" class="anchor"></a>Pancreas data from Indrop</h2>
<p>Another example using pancreas single RNA seq data from Indrop method.</p>
<div class="sourceCode" id="cb17"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb17-1" data-line-number="1"><span class="co"># from dataframe generated by Seurat via SCE objects</span></a>
<a class="sourceLine" id="cb17-2" data-line-number="2">betam &lt;-<span class="st"> </span><span class="kw">c</span>(<span class="st">"INS"</span>, <span class="st">"IGF2"</span>, <span class="st">"IAPP"</span>,<span class="st">"MAFA"</span>,<span class="st">"NPTX2"</span>)</a>
<a class="sourceLine" id="cb17-3" data-line-number="3">alpham &lt;-<span class="st"> </span><span class="kw">c</span>(<span class="st">"GCG"</span>, <span class="st">"PDK4"</span>,<span class="st">"LOXL2"</span>,<span class="st">"IRX2"</span>,<span class="st">"GC"</span>)</a>
<a class="sourceLine" id="cb17-4" data-line-number="4">deltam &lt;-<span class="st"> </span><span class="kw">c</span>(<span class="st">"SST"</span>, <span class="st">"RBP4"</span>,<span class="st">"HHEX"</span>,<span class="st">"PCSK1"</span>,<span class="st">"LEPR"</span>)</a>
<a class="sourceLine" id="cb17-5" data-line-number="5">mm2 &lt;-<span class="st"> </span><span class="kw">data.frame</span>(alpham, betam, deltam)</a>
<a class="sourceLine" id="cb17-6" data-line-number="6"></a>
<a class="sourceLine" id="cb17-7" data-line-number="7">b &lt;-<span class="st"> </span><span class="kw"><a href="../reference/binarize_expr.html">binarize_expr</a></span>(pan_indrop_avg)</a>
<a class="sourceLine" id="cb17-8" data-line-number="8">res &lt;-<span class="st"> </span><span class="kw"><a href="../reference/compare_lists.html">compare_lists</a></span>(b,mm2)</a>
<a class="sourceLine" id="cb17-9" data-line-number="9"><span class="kw"><a href="http://www.rdocumentation.org/packages/ComplexHeatmap/topics/Heatmap">Heatmap</a></span>(res, <span class="dt">cluster_rows =</span> <span class="ot">FALSE</span>, <span class="dt">cluster_columns =</span> <span class="ot">FALSE</span>, <span class="dt">heatmap_legend_param =</span> <span class="kw">list</span>(<span class="dt">title =</span> <span class="st">"-log10(adjP)"</span>))</a>
<a class="sourceLine" id="cb17-10" data-line-number="10"></a>
<a class="sourceLine" id="cb17-11" data-line-number="11">res2 &lt;-<span class="st"> </span><span class="kw"><a href="../reference/compare_lists.html">compare_lists</a></span>(b,mm2,<span class="dt">metric =</span> <span class="st">"jaccard"</span>)</a>
<a class="sourceLine" id="cb17-12" data-line-number="12"><span class="kw"><a href="http://www.rdocumentation.org/packages/ComplexHeatmap/topics/Heatmap">Heatmap</a></span>(res2, <span class="dt">cluster_rows =</span> <span class="ot">FALSE</span>, <span class="dt">cluster_columns =</span> <span class="ot">FALSE</span>, <span class="dt">heatmap_legend_param =</span> <span class="kw">list</span>(<span class="dt">title =</span> <span class="st">"jaccard index"</span>))</a>
<a class="sourceLine" id="cb17-13" data-line-number="13"></a>
<a class="sourceLine" id="cb17-14" data-line-number="14"><span class="co"># not very good on per cell basis</span></a>
<a class="sourceLine" id="cb17-15" data-line-number="15"><span class="co"># b2 &lt;- binarize_expr(pan_indrop_matrix)</span></a>
<a class="sourceLine" id="cb17-16" data-line-number="16"><span class="co"># res2 &lt;- compare_lists(b2, mm2)</span></a>
<a class="sourceLine" id="cb17-17" data-line-number="17"><span class="co"># plot_cor(res2,</span></a>
<a class="sourceLine" id="cb17-18" data-line-number="18"><span class="co">#          pan_indrop_meta,</span></a>
<a class="sourceLine" id="cb17-19" data-line-number="19"><span class="co">#          colnames(res2)[1:7])</span></a></code></pre></div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">
        <div id="tocnav">
      <h2 class="hasAnchor">
<a href="#tocnav" class="anchor"></a>Contents</h2>
      <ul class="nav nav-pills nav-stacked">
<li><a href="#unranked-list-preprocessing-gene-list-into-dataframematrix-form">Unranked list: Preprocessing gene list into dataframe/matrix form</a></li>
      <li><a href="#unranked-list-binarize-single-cell-gene-expression-data">Unranked list: Binarize single cell gene expression data</a></li>
      <li><a href="#unranked-list-classify">Unranked list: Classify</a></li>
      <li><a href="#ranked-list-preprocessing-gene-list-into-dataframematrix-form">Ranked list: Preprocessing gene list into dataframe/matrix form</a></li>
      <li><a href="#ranked-list-classify">Ranked list: Classify</a></li>
      <li><a href="#pancreas-data-from-indrop">Pancreas data from Indrop</a></li>
      </ul>
</div>
      </div>

</div>


      <footer><div class="copyright">
  <p>Developed by Kent Riemondy, Austin Gillen, Rui Fu, Jay Hesselberth, Yue Hao, Chengzhe Tian, Michelle Daya.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="http://pkgdown.r-lib.org/">pkgdown</a>.</p>
</div>

      </footer>
</div>

  

  </body>
</html>
