<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html>
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>clustifyR overview • clustifyR</title>
<!-- jquery --><script src="https://code.jquery.com/jquery-3.1.0.min.js" integrity="sha384-nrOSfDHtoPMzJHjVTdCopGqIqeYETSXhZDFyniQ8ZHcVy08QesyHcnOUpMpqnmWq" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/css/bootstrap.min.css" rel="stylesheet" integrity="sha384-BVYiiSIFeK1dGmJRAkycuHAHRg32OmUcww7on3RYdg4Va+PmSTsz/K68vbdEjh4u" crossorigin="anonymous">
<script src="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha384-Tc5IQib027qvyjSMfHjOMaLkfuWVxZxUPnCJA7l2mCWNIpG9mGCD8wGNIcPD7Txa" crossorigin="anonymous"></script><!-- Font Awesome icons --><link href="https://maxcdn.bootstrapcdn.com/font-awesome/4.6.3/css/font-awesome.min.css" rel="stylesheet" integrity="sha384-T8Gy5hrqNKT+hzMclPo118YTQO6cYprQmhrYwIiQ/3axmI1hQomh7Ud2hPOy8SP1" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/1.7.1/clipboard.min.js" integrity="sha384-cV+rhyOuRHc9Ub/91rihWcGmMmCXDeksTtCihMupQHSsi8GIIRDG0ThDc3HGQFJ3" crossorigin="anonymous"></script><!-- sticky kit --><script src="https://cdnjs.cloudflare.com/ajax/libs/sticky-kit/1.1.3/sticky-kit.min.js" integrity="sha256-c4Rlo1ZozqTPE2RLuvbusY3+SU1pQaJC0TjuhygMipw=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="clustifyR overview">
<meta property="og:description" content="">
<meta name="twitter:card" content="summary">
<!-- mathjax --><script src="https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">clustifyR</a>
        <span class="label label-danger" data-toggle="tooltip" data-placement="bottom" title="Unreleased package">0.0.0.9000</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../index.html">
    <span class="fa fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/compare_methods.html">Compare Methods</a>
    </li>
    <li>
      <a href="../articles/exploratory_pca.html">Exploratory PCA</a>
    </li>
    <li>
      <a href="../articles/gene_list.html">Cluster classification using mark gene lists</a>
    </li>
    <li>
      <a href="../articles/get_recount_data.html">Downloading data from recount</a>
    </li>
    <li>
      <a href="../articles/overview.html">clustifyR overview</a>
    </li>
    <li>
      <a href="../articles/random_forest.html">randomForest model</a>
    </li>
    <li>
      <a href="../articles/refine_cluster.html">Clustering Single Cell RNA-Seq Data Using Reference Bulk RNA-Seq Data</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="http://github.com/NCBI-Hackathons/clustifyR">
    <span class="fa fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      
      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1>clustifyR overview</h1>
            
            <h4 class="date">2018-06-23</h4>
      
      <small class="dont-index">Source: <a href="http://github.com/NCBI-Hackathons/clustifyR/blob/master/vignettes/overview.Rmd"><code>vignettes/overview.Rmd</code></a></small>
      <div class="hidden name"><code>overview.Rmd</code></div>

    </div>

    
    
<div id="why-use-clustifyr" class="section level2">
<h2 class="hasAnchor">
<a href="#why-use-clustifyr" class="anchor"></a>Why use <code>clustifyR</code>?</h2>
<p>Single cell transcriptomes are difficult to annotate without extensive knowledge of the underlying biology of the system in question. Even with this knowledge, accurate identification can be challenging due to the lack of detectable expression of common marker genes defined by bulk RNA-seq, flow cytometry, etc.</p>
<p><code>clustifyR</code> solves this problem by providing functions to automatically annotate single cells or clusters using bulk RNA-seq data or marker gene lists (ranked or unranked). Additional functions allow for exploratory analysis of calculated similarities between single cell RNA-seq datasets and reference data.</p>
</div>
<div id="a-simple-example-10x-genomics-pbmcs" class="section level2">
<h2 class="hasAnchor">
<a href="#a-simple-example-10x-genomics-pbmcs" class="anchor"></a>A simple example: 10x Genomics PBMCs</h2>
<p>In this example, we take an example 10x Genomics 3’ scRNA-seq dataset comprised of peripheral blood mononuclear cells from a healthy human donor and annotate the cell clusters (identified using <code>Seurat</code>) using bulk RNA-seq data derived from isolated PBMC cell populations.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">library</span>(clustifyR)
<span class="kw">library</span>(ComplexHeatmap)
<span class="co">#&gt; Loading required package: grid</span>
<span class="co">#&gt; ========================================</span>
<span class="co">#&gt; ComplexHeatmap version 1.17.1</span>
<span class="co">#&gt; Bioconductor page: http://bioconductor.org/packages/ComplexHeatmap/</span>
<span class="co">#&gt; Github page: https://github.com/jokergoo/ComplexHeatmap</span>
<span class="co">#&gt; Documentation: http://bioconductor.org/packages/ComplexHeatmap/</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; If you use it in published research, please cite:</span>
<span class="co">#&gt; Gu, Z. Complex heatmaps reveal patterns and correlations in multidimensional </span>
<span class="co">#&gt;   genomic data. Bioinformatics 2016.</span>
<span class="co">#&gt; ========================================</span>
<span class="kw">library</span>(circlize)
<span class="co">#&gt; Warning: package 'circlize' was built under R version 3.4.3</span>
<span class="co">#&gt; ========================================</span>
<span class="co">#&gt; circlize version 0.4.3</span>
<span class="co">#&gt; CRAN page: https://cran.r-project.org/package=circlize</span>
<span class="co">#&gt; Github page: https://github.com/jokergoo/circlize</span>
<span class="co">#&gt; Documentation: http://jokergoo.github.io/circlize_book/book/</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; If you use it in published research, please cite:</span>
<span class="co">#&gt; Gu, Z. circlize implements and enhances circular visualization </span>
<span class="co">#&gt;   in R. Bioinformatics 2014.</span>
<span class="co">#&gt; ========================================</span>

<span class="co"># load pacakge data</span>
<span class="kw">data</span>(<span class="st">"pbmc4k_matrix"</span>)
<span class="kw">data</span>(<span class="st">"pbmc4k_meta"</span>)
<span class="kw">data</span>(<span class="st">"pbmc4k_vargenes"</span>)
<span class="kw">data</span>(<span class="st">"pbmc_bulk_matrix"</span>)

<span class="co"># plot tSNE using known identities</span>
<span class="kw"><a href="../reference/plot_tsne.html">plot_tsne</a></span>(pbmc4k_meta, <span class="dt">feature =</span> <span class="st">"classified"</span>)</code></pre></div>
<div class="figure" style="text-align: center">
<img src="overview_files/figure-html/unnamed-chunk-1-1.png" alt="tSNE projection of PBMCs from a health human donor" width="1152"><p class="caption">
tSNE projection of PBMCs from a health human donor
</p>
</div>
<p>Bulk RNA-Seq data can be obtained from any input source, here we will obtain a dataset from the <a href="https://jhubiostatistics.shinyapps.io/recount/">recount2</a> database. This database provides &gt; 2000 human RNA-Seq experiments that have been processed using a consistent pipeline. We have written a wrapper function to download a count matrix from <code>recount2</code>, given an SRA ID.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">source</span>(<span class="kw">system.file</span>(<span class="st">"dl_recount.R"</span>, 
                   <span class="dt">package =</span> <span class="st">"clustifyR"</span>))
<span class="co">#&gt; Warning: package 'dplyr' was built under R version 3.4.4</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; Attaching package: 'dplyr'</span>
<span class="co">#&gt; The following objects are masked from 'package:stats':</span>
<span class="co">#&gt; </span>
<span class="co">#&gt;     filter, lag</span>
<span class="co">#&gt; The following objects are masked from 'package:base':</span>
<span class="co">#&gt; </span>
<span class="co">#&gt;     intersect, setdiff, setequal, union</span>
<span class="co">#&gt; Warning: package 'purrr' was built under R version 3.4.4</span>
<span class="co">#&gt; Warning: package 'tidyr' was built under R version 3.4.4</span>
<span class="co">#&gt; Warning: package 'stringr' was built under R version 3.4.4</span>
<span class="co">#&gt; Warning: package 'recount' was built under R version 3.4.3</span>
<span class="co">#&gt; Loading required package: SummarizedExperiment</span>
<span class="co">#&gt; Warning: package 'SummarizedExperiment' was built under R version 3.4.3</span>
<span class="co">#&gt; Loading required package: GenomicRanges</span>
<span class="co">#&gt; Warning: package 'GenomicRanges' was built under R version 3.4.3</span>
<span class="co">#&gt; Loading required package: stats4</span>
<span class="co">#&gt; Loading required package: BiocGenerics</span>
<span class="co">#&gt; Loading required package: parallel</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; Attaching package: 'BiocGenerics'</span>
<span class="co">#&gt; The following objects are masked from 'package:parallel':</span>
<span class="co">#&gt; </span>
<span class="co">#&gt;     clusterApply, clusterApplyLB, clusterCall, clusterEvalQ,</span>
<span class="co">#&gt;     clusterExport, clusterMap, parApply, parCapply, parLapply,</span>
<span class="co">#&gt;     parLapplyLB, parRapply, parSapply, parSapplyLB</span>
<span class="co">#&gt; The following objects are masked from 'package:dplyr':</span>
<span class="co">#&gt; </span>
<span class="co">#&gt;     combine, intersect, setdiff, union</span>
<span class="co">#&gt; The following objects are masked from 'package:stats':</span>
<span class="co">#&gt; </span>
<span class="co">#&gt;     IQR, mad, sd, var, xtabs</span>
<span class="co">#&gt; The following objects are masked from 'package:base':</span>
<span class="co">#&gt; </span>
<span class="co">#&gt;     anyDuplicated, append, as.data.frame, cbind, colMeans,</span>
<span class="co">#&gt;     colnames, colSums, do.call, duplicated, eval, evalq, Filter,</span>
<span class="co">#&gt;     Find, get, grep, grepl, intersect, is.unsorted, lapply,</span>
<span class="co">#&gt;     lengths, Map, mapply, match, mget, order, paste, pmax,</span>
<span class="co">#&gt;     pmax.int, pmin, pmin.int, Position, rank, rbind, Reduce,</span>
<span class="co">#&gt;     rowMeans, rownames, rowSums, sapply, setdiff, sort, table,</span>
<span class="co">#&gt;     tapply, union, unique, unsplit, which, which.max, which.min</span>
<span class="co">#&gt; Loading required package: S4Vectors</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; Attaching package: 'S4Vectors'</span>
<span class="co">#&gt; The following object is masked from 'package:tidyr':</span>
<span class="co">#&gt; </span>
<span class="co">#&gt;     expand</span>
<span class="co">#&gt; The following objects are masked from 'package:dplyr':</span>
<span class="co">#&gt; </span>
<span class="co">#&gt;     first, rename</span>
<span class="co">#&gt; The following object is masked from 'package:base':</span>
<span class="co">#&gt; </span>
<span class="co">#&gt;     expand.grid</span>
<span class="co">#&gt; Loading required package: IRanges</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; Attaching package: 'IRanges'</span>
<span class="co">#&gt; The following object is masked from 'package:purrr':</span>
<span class="co">#&gt; </span>
<span class="co">#&gt;     reduce</span>
<span class="co">#&gt; The following objects are masked from 'package:dplyr':</span>
<span class="co">#&gt; </span>
<span class="co">#&gt;     collapse, desc, slice</span>
<span class="co">#&gt; Loading required package: GenomeInfoDb</span>
<span class="co">#&gt; Loading required package: Biobase</span>
<span class="co">#&gt; Welcome to Bioconductor</span>
<span class="co">#&gt; </span>
<span class="co">#&gt;     Vignettes contain introductory material; view with</span>
<span class="co">#&gt;     'browseVignettes()'. To cite Bioconductor, see</span>
<span class="co">#&gt;     'citation("Biobase")', and for packages 'citation("pkgname")'.</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; Attaching package: 'Biobase'</span>
<span class="co">#&gt; The following object is masked from 'package:dplyr':</span>
<span class="co">#&gt; </span>
<span class="co">#&gt;     exprs</span>
<span class="co">#&gt; Loading required package: DelayedArray</span>
<span class="co">#&gt; Loading required package: matrixStats</span>
<span class="co">#&gt; Warning: package 'matrixStats' was built under R version 3.4.3</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; Attaching package: 'matrixStats'</span>
<span class="co">#&gt; The following objects are masked from 'package:Biobase':</span>
<span class="co">#&gt; </span>
<span class="co">#&gt;     anyMissing, rowMedians</span>
<span class="co">#&gt; The following object is masked from 'package:dplyr':</span>
<span class="co">#&gt; </span>
<span class="co">#&gt;     count</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; Attaching package: 'DelayedArray'</span>
<span class="co">#&gt; The following objects are masked from 'package:matrixStats':</span>
<span class="co">#&gt; </span>
<span class="co">#&gt;     colMaxs, colMins, colRanges, rowMaxs, rowMins, rowRanges</span>
<span class="co">#&gt; The following object is masked from 'package:base':</span>
<span class="co">#&gt; </span>
<span class="co">#&gt;     apply</span>
<span class="co">#&gt; Setting options('download.file.method.GEOquery'='auto')</span>
<span class="co">#&gt; Setting options('GEOquery.inmemory.gpl'=FALSE)</span>

pbmc_data &lt;-<span class="st"> </span><span class="kw">dl_recount</span>(<span class="st">"SRP051688"</span>)
<span class="co">#&gt; 2018-06-23 14:17:33 downloading file rse_gene.Rdata to SRP051688</span>
<span class="co">#&gt; Warning: package 'bindrcpp' was built under R version 3.4.4</span>

pbmc_data<span class="op">$</span>read_counts[<span class="dv">1</span><span class="op">:</span><span class="dv">5</span>, <span class="dv">1</span><span class="op">:</span><span class="dv">5</span>]
<span class="co">#&gt;          SRR1740034 SRR1740035 SRR1740036 SRR1740037 SRR1740038</span>
<span class="co">#&gt; A1BG            317        369        319        273        148</span>
<span class="co">#&gt; A1BG-AS1        421        442        405        409        109</span>
<span class="co">#&gt; A1CF              3          8          6          5          1</span>
<span class="co">#&gt; A2M               3          5          1          6         21</span>
<span class="co">#&gt; A2M-AS1           2          0          3          3         21</span>

pbmc_data<span class="op">$</span>meta_data[<span class="dv">1</span><span class="op">:</span><span class="dv">5</span>, ]
<span class="co">#&gt; # A tibble: 5 x 5</span>
<span class="co">#&gt;   run        time  `cell type`              assay   donor      </span>
<span class="co">#&gt;   &lt;chr&gt;      &lt;chr&gt; &lt;chr&gt;                    &lt;chr&gt;   &lt;chr&gt;      </span>
<span class="co">#&gt; 1 SRR1740034 0 d   primary human B cells    RNA-seq Donor: HD30</span>
<span class="co">#&gt; 2 SRR1740035 1 d   primary human B cells    RNA-seq Donor: HD30</span>
<span class="co">#&gt; 3 SRR1740036 3 d   primary human B cells    RNA-seq Donor: HD30</span>
<span class="co">#&gt; 4 SRR1740037 7 d   primary human B cells    RNA-seq Donor: HD30</span>
<span class="co">#&gt; 5 SRR1740038 0 d   primary human myeloid DC RNA-seq Donor: HD30</span></code></pre></div>
<p>We provide a vignette that demonstrates the few steps necessary to format the bulk rna-seq ma <code>clustifyR</code> provides the <code><a href="../reference/run_cor.html">run_cor()</a></code> function, which takes as input a matrix of single cell RNA-seq expression values, a metadata table with cluster assignments, and a matrix with bulk RNA seq datasets.</p>
<p>As the heatmap below shows, most cell types are accurately classified using the default correlation method (pearson) and a list of highly variable genes in the single cell dataset identified using <code>Seurat</code>’s VMR method.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># run correlation (pearson by default)</span>
res &lt;-<span class="st"> </span><span class="kw"><a href="../reference/run_cor.html">run_cor</a></span>(<span class="dt">expr_mat =</span> pbmc4k_matrix,
               <span class="dt">metadata =</span> pbmc4k_meta,
               <span class="dt">bulk_mat =</span> pbmc_bulk_matrix,
               <span class="dt">query_gene_list =</span> pbmc4k_vargenes,
               <span class="dt">cluster_col =</span> <span class="st">"classified"</span>,
               <span class="dt">compute_method =</span> corr_coef)

<span class="co"># peek at correlation matrix</span>
res[<span class="dv">1</span><span class="op">:</span><span class="dv">5</span>, <span class="dv">1</span><span class="op">:</span><span class="dv">5</span>]
<span class="co">#&gt;                 primary human B cells rep 1 primary human myeloid DC rep 1</span>
<span class="co">#&gt; B cells                           0.6141671                      0.3091946</span>
<span class="co">#&gt; CD14+ Monocytes                   0.1586617                      0.5616840</span>
<span class="co">#&gt; CD4 T cells, 1                    0.2607140                      0.3241628</span>
<span class="co">#&gt; CD4 T cells, 2                    0.2701179                      0.3312665</span>
<span class="co">#&gt; CD8 T cells                       0.2017300                      0.2694282</span>
<span class="co">#&gt;                 primary human monocytes rep 1</span>
<span class="co">#&gt; B cells                             0.2010505</span>
<span class="co">#&gt; CD14+ Monocytes                     0.5664662</span>
<span class="co">#&gt; CD4 T cells, 1                      0.2373057</span>
<span class="co">#&gt; CD4 T cells, 2                      0.2438347</span>
<span class="co">#&gt; CD8 T cells                         0.1899286</span>
<span class="co">#&gt;                 primary human neutrophils rep 1</span>
<span class="co">#&gt; B cells                               0.0952484</span>
<span class="co">#&gt; CD14+ Monocytes                       0.3287754</span>
<span class="co">#&gt; CD4 T cells, 1                        0.1369353</span>
<span class="co">#&gt; CD4 T cells, 2                        0.1352250</span>
<span class="co">#&gt; CD8 T cells                           0.1053600</span>
<span class="co">#&gt;                 primary human NK cells rep 1</span>
<span class="co">#&gt; B cells                            0.1292641</span>
<span class="co">#&gt; CD14+ Monocytes                    0.1456926</span>
<span class="co">#&gt; CD4 T cells, 1                     0.2840439</span>
<span class="co">#&gt; CD4 T cells, 2                     0.3055039</span>
<span class="co">#&gt; CD8 T cells                        0.4130706</span>

<span class="co"># view heatmap of correlation coefficients</span>
<span class="kw"><a href="http://www.rdocumentation.org/packages/ComplexHeatmap/topics/Heatmap">Heatmap</a></span>(res, <span class="kw"><a href="http://www.rdocumentation.org/packages/circlize/topics/colorRamp2">colorRamp2</a></span>(<span class="kw">c</span>(<span class="dv">0</span>, <span class="dv">1</span>), <span class="kw">c</span>(<span class="st">"white"</span>, <span class="st">"black"</span>)))</code></pre></div>
<p><img src="overview_files/figure-html/unnamed-chunk-3-1.png" width="1152" style="display: block; margin: auto;"></p>
<div id="plot-cluster-identities-and-correlation-coefficients" class="section level3">
<h3 class="hasAnchor">
<a href="#plot-cluster-identities-and-correlation-coefficients" class="anchor"></a>Plot cluster identities and correlation coefficients</h3>
<p><code>clustifyR</code> provides functions to plot single cells uing pre-calculated tSNE embeddings (or those from any other dimensionality reduction method), and to plot correlation coefficients on those projections. The plots below again show that the highest correlations between these bulk RNA-seq data and the test scRNA-seq dataset are restricted to the correct cell clusters.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># plot tSNE using known identities</span>
<span class="kw"><a href="../reference/plot_tsne.html">plot_tsne</a></span>(pbmc4k_meta, <span class="dt">feature =</span> <span class="st">"classified"</span>)</code></pre></div>
<p><img src="overview_files/figure-html/plot-1.png" width="1152" style="display: block; margin: auto;"></p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">
<span class="co"># plot correlation coefficients on tSNE for each identity class</span>
<span class="kw"><a href="../reference/plot_cor.html">plot_cor</a></span>(res,
         pbmc4k_meta,
         <span class="kw">colnames</span>(res)[<span class="kw">c</span>(<span class="dv">1</span>, <span class="dv">5</span>)],
         <span class="dt">cluster_col =</span> <span class="st">"classified"</span>)
<span class="co">#&gt; [[1]]</span></code></pre></div>
<p><img src="overview_files/figure-html/plot-2.png" width="1152" style="display: block; margin: auto;"></p>
<pre><code>#&gt; 
#&gt; [[2]]</code></pre>
<p><img src="overview_files/figure-html/plot-3.png" width="1152" style="display: block; margin: auto;"></p>
</div>
<div id="obtain-p-value-for-similarity-using-a-permutation-test" class="section level3">
<h3 class="hasAnchor">
<a href="#obtain-p-value-for-similarity-using-a-permutation-test" class="anchor"></a>Obtain p-value for similarity using a permutation test</h3>
<p>In addition to automatically generating a correlation matrix, <code>clustifyR</code> can calculate correlation p-values using permutation with the <code><a href="../reference/run_cor.html">run_cor()</a></code> function. As with raw correlation coefficients, plotting a heatmap of -log(p-values) demonstrates that the most significant correlations are between corresponding bulk and single cell clusters.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># re-run run_cor with if_permute = TRUE, num_permute=1000 and return_full = TRUE</span>
res &lt;-<span class="st"> </span><span class="kw"><a href="../reference/run_cor.html">run_cor</a></span>(pbmc4k_matrix,
               pbmc4k_meta,
               pbmc_bulk_matrix,
               pbmc4k_vargenes,
               <span class="dt">num_perm =</span> <span class="dv">1000</span>,
               <span class="dt">cluster_col =</span> <span class="st">"classified"</span>,
               <span class="dt">compute_method =</span> corr_coef,
               <span class="dt">return_full =</span> <span class="ot">TRUE</span>)

res<span class="op">$</span>p_val[<span class="dv">1</span><span class="op">:</span><span class="dv">5</span>, <span class="dv">1</span><span class="op">:</span><span class="dv">5</span>]
<span class="co">#&gt;                 primary human B cells rep 1 primary human myeloid DC rep 1</span>
<span class="co">#&gt; B cells                                   0                              1</span>
<span class="co">#&gt; CD14+ Monocytes                           1                              0</span>
<span class="co">#&gt; CD4 T cells, 1                            1                              1</span>
<span class="co">#&gt; CD4 T cells, 2                            1                              1</span>
<span class="co">#&gt; CD8 T cells                               1                              1</span>
<span class="co">#&gt;                 primary human monocytes rep 1</span>
<span class="co">#&gt; B cells                                     1</span>
<span class="co">#&gt; CD14+ Monocytes                             0</span>
<span class="co">#&gt; CD4 T cells, 1                              1</span>
<span class="co">#&gt; CD4 T cells, 2                              1</span>
<span class="co">#&gt; CD8 T cells                                 1</span>
<span class="co">#&gt;                 primary human neutrophils rep 1</span>
<span class="co">#&gt; B cells                                       1</span>
<span class="co">#&gt; CD14+ Monocytes                               0</span>
<span class="co">#&gt; CD4 T cells, 1                                1</span>
<span class="co">#&gt; CD4 T cells, 2                                1</span>
<span class="co">#&gt; CD8 T cells                                   1</span>
<span class="co">#&gt;                 primary human NK cells rep 1</span>
<span class="co">#&gt; B cells                                1.000</span>
<span class="co">#&gt; CD14+ Monocytes                        1.000</span>
<span class="co">#&gt; CD4 T cells, 1                         1.000</span>
<span class="co">#&gt; CD4 T cells, 2                         0.712</span>
<span class="co">#&gt; CD8 T cells                            0.000</span>

<span class="co"># view heatmap of correlation coefficients</span>
<span class="kw"><a href="http://www.rdocumentation.org/packages/ComplexHeatmap/topics/Heatmap">Heatmap</a></span>(<span class="op">-</span><span class="kw">log</span>(res<span class="op">$</span>p_val<span class="op">+</span>.<span class="dv">01</span>,<span class="dv">10</span>), <span class="kw"><a href="http://www.rdocumentation.org/packages/circlize/topics/colorRamp2">colorRamp2</a></span>(<span class="kw">c</span>(<span class="dv">0</span>, <span class="dv">2</span>), <span class="kw">c</span>(<span class="st">"white"</span>, <span class="st">"black"</span>)))</code></pre></div>
<p><img src="overview_files/figure-html/unnamed-chunk-4-1.png" width="1152" style="display: block; margin: auto;"></p>
</div>
<div id="using-gene-lists-to-classify-cells-ranked-or-unranked" class="section level3">
<h3 class="hasAnchor">
<a href="#using-gene-lists-to-classify-cells-ranked-or-unranked" class="anchor"></a>Using gene lists to classify cells (ranked or unranked)</h3>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># generate a dataframe of markers from Seurat (a common format)</span>
pbmc4k_mm &lt;-<span class="st"> </span><span class="kw"><a href="../reference/matrixize_markers.html">matrixize_markers</a></span>(pbmc4k_markers)
<span class="kw">head</span>(pbmc4k_mm)
<span class="co">#&gt;       0             1     2     3     4      5       6              7</span>
<span class="co">#&gt; 1 RPL31        S100A8 CD79A  IL7R  GZMK FGFBP2  FCER1A           HES4</span>
<span class="co">#&gt; 2  RPS6        S100A9  IGHM ITGB1  CCL5   PRF1 CLEC10A         CDKN1C</span>
<span class="co">#&gt; 3 RPS25           LYZ CD79B  IL32 TRGC2   HOPX    ENHO RP11-1008C21.1</span>
<span class="co">#&gt; 4 RPS12       S100A12  IGHD  AQP3  CST7   GZMB   CLIC2         FCGR3A</span>
<span class="co">#&gt; 5 RPS27 RP11-1143G9.4 TCL1A  TRAC  GZMA  KLRD1    FLT3         TCF7L2</span>
<span class="co">#&gt; 6 RPL32          FCN1 MS4A1   LTB  LYAR  SPON2    PKIB            CKB</span>
<span class="co">#&gt;        8      9</span>
<span class="co">#&gt; 1 LILRA4    PF4</span>
<span class="co">#&gt; 2 LRRC26   SDPR</span>
<span class="co">#&gt; 3   TPM2  TUBB1</span>
<span class="co">#&gt; 4    SCT  ACRBP</span>
<span class="co">#&gt; 5  PTPRS    GP9</span>
<span class="co">#&gt; 6 CLEC4C TMEM40</span></code></pre></div>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># or just lists of genes</span>
NKm &lt;-<span class="st"> </span><span class="kw">c</span>(<span class="st">"GNLY"</span>, <span class="st">"NKG7"</span>, <span class="st">"ZBTB32"</span>, <span class="st">"IL2RB"</span>)
DCm &lt;-<span class="st"> </span><span class="kw">c</span>(<span class="st">"FCER1A"</span>, <span class="st">"CST3"</span>, <span class="st">"ITGAX"</span>, <span class="st">"IRF8"</span>)
Bm &lt;-<span class="st"> </span><span class="kw">c</span>(<span class="st">"MS4A1"</span>, <span class="st">"CD79A"</span>,<span class="st">"CD79B"</span>,<span class="st">"CD37"</span>)
markers &lt;-<span class="st"> </span><span class="kw">data.frame</span>(Bm, DCm, NKm)
markers
<span class="co">#&gt;      Bm    DCm    NKm</span>
<span class="co">#&gt; 1 MS4A1 FCER1A   GNLY</span>
<span class="co">#&gt; 2 CD79A   CST3   NKG7</span>
<span class="co">#&gt; 3 CD79B  ITGAX ZBTB32</span>
<span class="co">#&gt; 4  CD37   IRF8  IL2RB</span></code></pre></div>
<p>Generate binary expression calls from single cell data.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">pbmc4k_avgb &lt;-<span class="st"> </span><span class="kw"><a href="../reference/binarize_expr.html">binarize_expr</a></span>(pbmc4k_avg)
pbmc4k_avgb[<span class="dv">1</span><span class="op">:</span><span class="dv">5</span>, <span class="kw">c</span>(<span class="dv">1</span>, <span class="dv">2</span>, <span class="dv">10</span>)]
<span class="co">#&gt;               0 1 9</span>
<span class="co">#&gt; RP11-34P13.7  0 0 0</span>
<span class="co">#&gt; FO538757.2    0 0 0</span>
<span class="co">#&gt; AP006222.2    0 0 1</span>
<span class="co">#&gt; RP4-669L17.10 0 0 0</span>
<span class="co">#&gt; RP11-206L10.9 0 0 0</span></code></pre></div>
<div id="classify-using-a-hypergeometric-test" class="section level4">
<h4 class="hasAnchor">
<a href="#classify-using-a-hypergeometric-test" class="anchor"></a>Classify using a hypergeometric test</h4>
<p>Here’s an example using ~100 markers per cluster</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">res &lt;-<span class="st"> </span><span class="kw"><a href="../reference/compare_lists.html">compare_lists</a></span>(pbmc4k_avgb, pbmc4k_mm)

<span class="co"># rename columns </span>
<span class="kw">colnames</span>(res) &lt;-<span class="st"> </span><span class="kw">left_join</span>(<span class="kw">data_frame</span>(<span class="dt">cluster =</span> <span class="kw">colnames</span>(res)),
                           <span class="kw">unique</span>(<span class="kw">data_frame</span>(<span class="dt">cluster =</span> pbmc4k_meta<span class="op">$</span>cluster,
                                             <span class="dt">classified =</span>  pbmc4k_meta<span class="op">$</span>classified)), 
                           <span class="dt">by =</span> <span class="st">"cluster"</span>) <span class="op">%&gt;%</span><span class="st"> </span>
<span class="st">  </span><span class="kw">pull</span>(classified)

<span class="kw"><a href="http://www.rdocumentation.org/packages/ComplexHeatmap/topics/Heatmap">Heatmap</a></span>(<span class="op">-</span><span class="kw">log10</span>(res), 
        <span class="dt">cluster_rows =</span> <span class="ot">FALSE</span>,
        <span class="dt">cluster_columns =</span> <span class="ot">FALSE</span>, 
        <span class="dt">heatmap_legend_param =</span> <span class="kw">list</span>(<span class="dt">title =</span> <span class="st">"-log10(adjP)"</span>))</code></pre></div>
<p><img src="overview_files/figure-html/unnamed-chunk-8-1.png" width="700" style="display: block; margin: auto;"></p>
<p>Here’s an example using ~100 markers per cluster</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">res2 &lt;-<span class="st"> </span><span class="kw"><a href="../reference/compare_lists.html">compare_lists</a></span>(pbmc4k_avgb, markers)
<span class="kw"><a href="http://www.rdocumentation.org/packages/ComplexHeatmap/topics/Heatmap">Heatmap</a></span>(<span class="op">-</span><span class="kw">log10</span>(res2), 
        <span class="dt">cluster_rows =</span> <span class="ot">FALSE</span>, 
        <span class="dt">cluster_columns =</span> <span class="ot">FALSE</span>, 
        <span class="dt">heatmap_legend_param =</span> <span class="kw">list</span>(<span class="dt">title =</span> <span class="st">"-log10(adjP)"</span>))</code></pre></div>
<p><img src="overview_files/figure-html/unnamed-chunk-9-1.png" width="700" style="display: block; margin: auto;"></p>
</div>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">
        <div id="tocnav">
      <h2 class="hasAnchor">
<a href="#tocnav" class="anchor"></a>Contents</h2>
      <ul class="nav nav-pills nav-stacked">
<li><a href="#why-use-clustifyr">Why use <code>clustifyR</code>?</a></li>
      <li><a href="#a-simple-example-10x-genomics-pbmcs">A simple example: 10x Genomics PBMCs</a></li>
      </ul>
</div>
      </div>

</div>


      <footer><div class="copyright">
  <p>Developed by Kent Riemondy, Austin Gillen, Rui Fu, Jay Hesselberth, Yue Hao, Chengzhe Tian, M. Day.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="http://pkgdown.r-lib.org/">pkgdown</a>.</p>
</div>

      </footer>
</div>

  

  </body>
</html>
