<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html>
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>randomForest model â€¢ clustifyR</title>
<!-- jquery --><script src="https://code.jquery.com/jquery-3.1.0.min.js" integrity="sha384-nrOSfDHtoPMzJHjVTdCopGqIqeYETSXhZDFyniQ8ZHcVy08QesyHcnOUpMpqnmWq" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/css/bootstrap.min.css" rel="stylesheet" integrity="sha384-BVYiiSIFeK1dGmJRAkycuHAHRg32OmUcww7on3RYdg4Va+PmSTsz/K68vbdEjh4u" crossorigin="anonymous">
<script src="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha384-Tc5IQib027qvyjSMfHjOMaLkfuWVxZxUPnCJA7l2mCWNIpG9mGCD8wGNIcPD7Txa" crossorigin="anonymous"></script><!-- Font Awesome icons --><link href="https://maxcdn.bootstrapcdn.com/font-awesome/4.6.3/css/font-awesome.min.css" rel="stylesheet" integrity="sha384-T8Gy5hrqNKT+hzMclPo118YTQO6cYprQmhrYwIiQ/3axmI1hQomh7Ud2hPOy8SP1" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/1.7.1/clipboard.min.js" integrity="sha384-cV+rhyOuRHc9Ub/91rihWcGmMmCXDeksTtCihMupQHSsi8GIIRDG0ThDc3HGQFJ3" crossorigin="anonymous"></script><!-- sticky kit --><script src="https://cdnjs.cloudflare.com/ajax/libs/sticky-kit/1.1.3/sticky-kit.min.js" integrity="sha256-c4Rlo1ZozqTPE2RLuvbusY3+SU1pQaJC0TjuhygMipw=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="randomForest model">
<meta property="og:description" content="">
<meta name="twitter:card" content="summary">
<!-- mathjax --><script src="https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">clustifyR</a>
        <span class="label label-danger" data-toggle="tooltip" data-placement="bottom" title="Unreleased package">0.0.0.9000</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../index.html">
    <span class="fa fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/compare_methods.html">Compare Methods</a>
    </li>
    <li>
      <a href="../articles/exploratory_pca.html">Exploratory PCA</a>
    </li>
    <li>
      <a href="../articles/gene_list.html">Cluster classification using mark gene lists</a>
    </li>
    <li>
      <a href="../articles/get_recount_data.html">Downloading data from recount</a>
    </li>
    <li>
      <a href="../articles/overview.html">clustifyR overview</a>
    </li>
    <li>
      <a href="../articles/random_forest.html">randomForest model</a>
    </li>
    <li>
      <a href="../articles/refine_cluster.html">Clustering Single Cell RNA-Seq Data Using Reference Bulk RNA-Seq Data</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="http://github.com/NCBI-Hackathons/clustifyR">
    <span class="fa fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      
      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1>randomForest model</h1>
            
            <h4 class="date">2018-06-23</h4>
      
      <small class="dont-index">Source: <a href="http://github.com/NCBI-Hackathons/clustifyR/blob/master/vignettes/random_forest.Rmd"><code>vignettes/random_forest.Rmd</code></a></small>
      <div class="hidden name"><code>random_forest.Rmd</code></div>

    </div>

    
    
<div id="cell-type-prediction-using-random-forest-model" class="section level2">
<h2 class="hasAnchor">
<a href="#cell-type-prediction-using-random-forest-model" class="anchor"></a>Cell type prediction using Random Forest model</h2>
<p>The goal is to build a random forest model using single-cell RNA-seq expression data to predict the cell type in each cluster. The training set contains an expression matrix with normalized readcounts per gene per cell, clustering information and identified cell types. The test set contains only the expression and clusters.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># Reduce the expression matrix to only highly variable genes</span>
<span class="co"># source('reduce_matrix.R')</span>

<span class="co"># Input files for random forest</span>
<span class="co"># Training set</span>
pbmc4 &lt;-<span class="st"> </span><span class="kw"><a href="../reference/reduce_expr_matrix.html">reduce_expr_matrix</a></span>(pbmc4k_matrix, pbmc4k_meta, pbmc4k_vargenes)
<span class="co"># Test set</span>
pbmc5 &lt;-<span class="st"> </span><span class="kw"><a href="../reference/reduce_expr_matrix.html">reduce_expr_matrix</a></span>(pbmc5_matrix, pbmc5_meta, pbmc4k_vargenes)

<span class="co"># Make sure the listed genes (predictors) are the same in both dataset</span>
<span class="kw">names</span>(pbmc5)[<span class="kw">names</span>(pbmc5) <span class="op">==</span><span class="st"> "cluster"</span>] &lt;-<span class="st"> "classified"</span>
pbmc4.sub &lt;-<span class="st"> </span>pbmc4[, <span class="kw">colnames</span>(pbmc4) <span class="op">%in%</span><span class="st"> </span><span class="kw">colnames</span>(pbmc5)]
pbmc5.sub &lt;-<span class="st"> </span>pbmc5[, <span class="kw">colnames</span>(pbmc5) <span class="op">%in%</span><span class="st"> </span><span class="kw">colnames</span>(pbmc4.sub)]</code></pre></div>
<div id="plot-training-data" class="section level3">
<h3 class="hasAnchor">
<a href="#plot-training-data" class="anchor"></a>Plot training data</h3>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># plot tsne using known identities</span>
<span class="kw">qplot</span>(tSNE_<span class="dv">1</span>, tSNE_<span class="dv">2</span>, <span class="dt">colour =</span> classified,  <span class="dt">data =</span> pbmc4k_meta)</code></pre></div>
<p><img src="random_forest_files/figure-html/plot-1.png" width="768" style="display: block; margin: auto;"></p>
</div>
<div id="model-parameter-tuning" class="section level3">
<h3 class="hasAnchor">
<a href="#model-parameter-tuning" class="anchor"></a>Model parameter tuning</h3>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># Algorithm Tune (tuneRF)</span>
seed &lt;-<span class="st"> </span><span class="dv">100</span>
<span class="kw">set.seed</span>(seed)
<span class="co"># initial mtry &lt;- sqrt(ncol(pbmc4))</span>
x &lt;-<span class="st"> </span>pbmc4[,<span class="dv">2</span><span class="op">:</span><span class="kw">ncol</span>(pbmc4)]
y &lt;-<span class="st"> </span>pbmc4[,<span class="dv">1</span>]
bestmtry &lt;-<span class="st"> </span><span class="kw">tuneRF</span>(x, y, <span class="dt">stepFactor=</span><span class="fl">1.5</span>, <span class="dt">improve=</span><span class="fl">1e-5</span>, <span class="dt">ntree=</span><span class="dv">500</span>)
<span class="kw">print</span>(bestmtry)
<span class="co"># In this case bestmtry is 256 when ntree is 500</span>

<span class="co"># Tune model using caret</span>
<span class="co"># Custom tuning</span>
<span class="co"># https://machinelearningmastery.com/tune-machine-learning-algorithms-in-r/</span>
customRF &lt;-<span class="st"> </span><span class="kw">list</span>(<span class="dt">type =</span> <span class="st">"Classification"</span>, <span class="dt">library =</span> <span class="st">"randomForest"</span>, <span class="dt">loop =</span> <span class="ot">NULL</span>)
customRF<span class="op">$</span>parameters &lt;-<span class="st"> </span><span class="kw">data.frame</span>(<span class="dt">parameter =</span> <span class="kw">c</span>(<span class="st">"mtry"</span>, <span class="st">"ntree"</span>), <span class="dt">class =</span> <span class="kw">rep</span>(<span class="st">"numeric"</span>, <span class="dv">2</span>), <span class="dt">label =</span> <span class="kw">c</span>(<span class="st">"mtry"</span>, <span class="st">"ntree"</span>))
customRF<span class="op">$</span>grid &lt;-<span class="st"> </span><span class="cf">function</span>(x, y, <span class="dt">len =</span> <span class="ot">NULL</span>, <span class="dt">search =</span> <span class="st">"grid"</span>) {}
customRF<span class="op">$</span>fit &lt;-<span class="st"> </span><span class="cf">function</span>(x, y, wts, param, lev, last, weights, classProbs, ...) {
  <span class="kw">randomForest</span>(x, y, <span class="dt">mtry =</span> param<span class="op">$</span>mtry, <span class="dt">ntree=</span>param<span class="op">$</span>ntree, ...)
}
customRF<span class="op">$</span>predict &lt;-<span class="st"> </span><span class="cf">function</span>(modelFit, newdata, <span class="dt">preProc =</span> <span class="ot">NULL</span>, <span class="dt">submodels =</span> <span class="ot">NULL</span>)
  <span class="kw">predict</span>(modelFit, newdata)
customRF<span class="op">$</span>prob &lt;-<span class="st"> </span><span class="cf">function</span>(modelFit, newdata, <span class="dt">preProc =</span> <span class="ot">NULL</span>, <span class="dt">submodels =</span> <span class="ot">NULL</span>)
  <span class="kw">predict</span>(modelFit, newdata, <span class="dt">type =</span> <span class="st">"prob"</span>)
customRF<span class="op">$</span>sort &lt;-<span class="st"> </span><span class="cf">function</span>(x) x[<span class="kw">order</span>(x[,<span class="dv">1</span>]),]
customRF<span class="op">$</span>levels &lt;-<span class="st"> </span><span class="cf">function</span>(x) x<span class="op">$</span>classes

metric &lt;-<span class="st"> "Accuracy"</span>
control &lt;-<span class="st"> </span><span class="kw">trainControl</span>(<span class="dt">method=</span><span class="st">"repeatedcv"</span>, <span class="dt">number=</span><span class="dv">10</span>, <span class="dt">repeats=</span><span class="dv">3</span>)
tunegrid &lt;-<span class="st"> </span><span class="kw">expand.grid</span>(<span class="dt">.mtry=</span><span class="kw">c</span>(<span class="dv">250</span><span class="op">:</span><span class="dv">260</span>), <span class="dt">.ntree=</span><span class="kw">c</span>(<span class="dv">500</span>, <span class="dv">1000</span>, <span class="dv">1500</span>, <span class="dv">2000</span>))
<span class="kw">set.seed</span>(seed)
rf_tune &lt;-<span class="st"> </span><span class="kw">train</span>(classified<span class="op">~</span>., <span class="dt">data=</span>pbmc4, <span class="dt">method=</span>customRF, <span class="dt">metric=</span>metric, <span class="dt">tuneGrid=</span>tunegrid, <span class="dt">trControl=</span>control)
<span class="kw">summary</span>(rf_tune)
<span class="kw">plot</span>(rf_tune)</code></pre></div>
</div>
<div id="build-model-with-optimal-parameters" class="section level3">
<h3 class="hasAnchor">
<a href="#build-model-with-optimal-parameters" class="anchor"></a>Build model with optimal parameters</h3>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co">#Optimal mtry and ntree</span>
bmtry&lt;-<span class="st"> </span><span class="dv">256</span>
bntree&lt;-<span class="st"> </span><span class="dv">1000</span> <span class="co">#for testing purpose</span>
<span class="co">#Build the model</span>
rFmodel.sub &lt;-<span class="st"> </span><span class="kw">randomForest</span>(classified <span class="op">~</span><span class="st"> </span>., <span class="dt">data =</span> pbmc4.sub, <span class="dt">ntree =</span> bntree, <span class="dt">mtry =</span> bmtry, <span class="dt">importance =</span> <span class="ot">TRUE</span>)
<span class="kw">save</span>(rFmodel.sub,<span class="dt">file =</span> <span class="st">"rFmodelsub.RData"</span>)</code></pre></div>
<pre><code>Call:
 randomForest(formula = classified ~ ., data = pbmc4.sub, ntree = bntree,      mtry = bmtry, importance = TRUE) 
               Type of random forest: classification
                     Number of trees: 1000
No. of variables tried at each split: 256

        OOB estimate of  error rate: 8%
Confusion matrix:
                    B cells CD14+ Monocytes CD4 T cells, 1 CD4 T cells, 2
B cells                 642               3              6              0
CD14+ Monocytes           0            1006              6              1
CD4 T cells, 1            2               2           1270             12
CD4 T cells, 2            1               1            146            403
CD8 T cells               1               0             37             20
Dendritic cells, 1        0              15              3              0
Dendritic Cells, 2?       1               0              1              0
FCGR3A+ Monocytes         0              19              0              1
Megakaryocytes            1               8              0              0
NK cells                  1               2              1              0
                    CD8 T cells Dendritic cells, 1 Dendritic Cells, 2?
B cells                       0                  1                   0
CD14+ Monocytes               0                  2                   0
CD4 T cells, 1                1                  0                   0
CD4 T cells, 2               34                  0                   0
CD8 T cells                 502                  1                   0
Dendritic cells, 1            0                110                   0
Dendritic Cells, 2?           0                  0                  36
FCGR3A+ Monocytes             0                  0                   0
Megakaryocytes                0                  0                   0
NK cells                     34                  0                   0
                    FCGR3A+ Monocytes Megakaryocytes NK cells class.error
B cells                             0              0        0  0.01533742
CD14+ Monocytes                     1              0        0  0.00984252
CD4 T cells, 1                      0              0        0  0.01320901
CD4 T cells, 2                      0              0        0  0.31111111
CD8 T cells                         0              0       11  0.12237762
Dendritic cells, 1                  0              0        0  0.14062500
Dendritic Cells, 2?                 0              0        0  0.05263158
FCGR3A+ Monocytes                  93              0        0  0.17699115
Megakaryocytes                      1             25        0  0.28571429
NK cells                            0              0      249  0.13240418</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># To check important variables</span>
<span class="kw">importance</span>(rFmodel)        
<span class="kw">varImpPlot</span>(rFmodel)</code></pre></div>
<div class="figure">
<img src="https://i.imgur.com/h5d0aST.png" alt="Important variables"><p class="caption">Important variables</p>
</div>
</div>
<div id="check-expression-patterns-of-important-genes" class="section level3">
<h3 class="hasAnchor">
<a href="#check-expression-patterns-of-important-genes" class="anchor"></a>Check expression patterns of important genes</h3>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">load</span>(<span class="st">"rFmodelsub.RData"</span>)
important &lt;-<span class="kw">as.data.frame</span>(<span class="kw">importance</span>(rFmodel.sub))
<span class="co"># source('important_heatmap.R')</span>
<span class="kw">dev.print</span>(png, <span class="dt">file =</span> <span class="st">"myplot.png"</span>)
<span class="kw">png</span>(<span class="dt">file =</span> <span class="st">"expression_heatmap.png"</span>, <span class="dt">bg =</span> <span class="st">"white"</span>)
<span class="kw"><a href="../reference/important_heatmap.html">important_heatmap</a></span>(important, pbmc4k_avg, pbmc4k_meta, <span class="dv">20</span>, <span class="dv">30</span>)
<span class="kw">dev.off</span>()</code></pre></div>
<p>With MeanDecreaseAccuracy &gt;=20 and MeanDecreaseGini &gt;=30: <img src="https://i.imgur.com/8BQO3al.png" alt="Expression of best predictors"><br>With MeanDecreaseAccuracy &gt;=15 and MeanDecreaseGini &gt;=20: <img src="https://i.imgur.com/M32CMFE.png" alt="15_20"></p>
</div>
<div id="fit-the-model-with-new-data" class="section level3">
<h3 class="hasAnchor">
<a href="#fit-the-model-with-new-data" class="anchor"></a>Fit the model with new data</h3>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">pred &lt;-<span class="st"> </span><span class="kw">predict</span>(rFmodel.sub, pbmc5.sub, <span class="dt">type =</span> <span class="st">"class"</span>)
<span class="co"># Checking classification accuracy</span>
<span class="kw">table</span>(pred, pbmc5.sub<span class="op">$</span>classified) </code></pre></div>
<p>Column names: cluster number from the pbmc5 dataset (testing set). Row names: predicted cell type</p>
<pre><code>pred                     0    1   10   11   12    2    3    4    5    6    7    8    9
  B cells                0    0    0    2    3 1233    0    1    0    3    0    0    3
  CD14+ Monocytes     1765    0    9   34    2    1    1  555    0    1   88    0  159
  CD4 T cells, 1         0 1626    5    9    4   17  274   60    5    0    5  225    9
  CD4 T cells, 2         0   56    1    5    0    2  838    0    4    1    3  121    1
  CD8 T cells            0    0    0    1    0    1   73    1  450   38    0    5    0
  Dendritic cells, 1     5    0   93    0    0    0    0    5    0    0    0    0    0
  Dendritic Cells, 2?    0    0    0    0   67    1    0    0    0    0    0    0    0
  FCGR3A+ Monocytes      1    0    0    1    0    0    0    1    0    0  286    0    4
  Megakaryocytes         0    0    0   40    0    0    0    0    0    0    0    0    0
  NK cells               0    0    0    1    0    0    0    0    3  347    0    0    0</code></pre>
<p><img src="random_forest_files/figure-html/plot2-1.png" width="768" style="display: block; margin: auto;"></p>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">
        <div id="tocnav">
      <h2 class="hasAnchor">
<a href="#tocnav" class="anchor"></a>Contents</h2>
      <ul class="nav nav-pills nav-stacked">
<li><a href="#cell-type-prediction-using-random-forest-model">Cell type prediction using Random Forest model</a></li>
      </ul>
</div>
      </div>

</div>


      <footer><div class="copyright">
  <p>Developed by Kent Riemondy, Austin Gillen, Rui Fu, Jay Hesselberth, Yue Hao, Chengzhe Tian, M. Day.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="http://pkgdown.r-lib.org/">pkgdown</a>.</p>
</div>

      </footer>
</div>

  

  </body>
</html>
