<!-- Generated by pkgdown: do not edit by hand -->
<!DOCTYPE html>
<html lang="en">
  <head>
  <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<title>Main function to compare scRNA-seq data to gene lists. — clustify_lists • clustifyr</title>

<!-- favicons -->
<link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png">
<link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png">
<link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png" />
<link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png" />
<link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png" />
<link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png" />

<!-- jquery -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.3.1/jquery.min.js" integrity="sha256-FgpCb/KJQlLNfOu91ta32o/NMZxltwRo8QtmkMRdAu8=" crossorigin="anonymous"></script>
<!-- Bootstrap -->

<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/css/bootstrap.min.css" integrity="sha256-916EbMg70RQy9LHiGkXzG8hSg9EdNy97GazNG/aiY1w=" crossorigin="anonymous" />

<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha256-U5ZEeKfGNOja007MMD3YBI0A3OSZOQbeG6z2f2Y0hu8=" crossorigin="anonymous"></script>

<!-- Font Awesome icons -->
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.7.1/css/all.min.css" integrity="sha256-nAmazAk6vS34Xqo0BSrTb+abbtFlgsFK7NKSi6o7Y78=" crossorigin="anonymous" />
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.7.1/css/v4-shims.min.css" integrity="sha256-6qHlizsOWFskGlwVOKuns+D1nB6ssZrHQrNj1wGplHc=" crossorigin="anonymous" />

<!-- clipboard.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.4/clipboard.min.js" integrity="sha256-FiZwavyI2V6+EXO1U+xzLG3IKldpiTFf3153ea9zikQ=" crossorigin="anonymous"></script>

<!-- headroom.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.9.4/headroom.min.js" integrity="sha256-DJFC1kqIhelURkuza0AvYal5RxMtpzLjFhsnVIeuk+U=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.9.4/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script>

<!-- pkgdown -->
<link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script>




<meta property="og:title" content="Main function to compare scRNA-seq data to gene lists. — clustify_lists" />
<meta property="og:description" content="Main function to compare scRNA-seq data to gene lists." />
<meta property="og:image" content="http://rnabioco.github.io/clustifyr/logo.png" />
<meta name="twitter:card" content="summary" />




<!-- mathjax -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script>

<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->



  </head>

  <body>
    <div class="container template-reference-topic">
      <header>
      <div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">clustifyr</a>
        <span class="version label label-danger" data-toggle="tooltip" data-placement="bottom" title="Unreleased version">0.99.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="../index.html">
    <span class="fas fa fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="../articles/clustifyR.html">Introduction to clustifyr</a>
    </li>
  </ul>
</li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/rnabioco/clustifyr">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
      
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header>

<div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Main function to compare scRNA-seq data to gene lists.</h1>
    <small class="dont-index">Source: <a href='https://github.com/rnabioco/clustifyr/blob/master/R/main.R'><code>R/main.R</code></a></small>
    <div class="hidden name"><code>clustify_lists.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>Main function to compare scRNA-seq data to gene lists.</p>
    </div>

    <pre class="usage"><span class='fu'>clustify_lists</span>(<span class='no'>input</span>, <span class='no'>...</span>)

<span class='co'># S3 method for default</span>
<span class='fu'>clustify_lists</span>(
  <span class='no'>input</span>,
  <span class='no'>marker</span>,
  <span class='kw'>marker_inmatrix</span> <span class='kw'>=</span> <span class='fl'>TRUE</span>,
  <span class='kw'>metadata</span> <span class='kw'>=</span> <span class='kw'>NULL</span>,
  <span class='kw'>cluster_col</span> <span class='kw'>=</span> <span class='kw'>NULL</span>,
  <span class='kw'>if_log</span> <span class='kw'>=</span> <span class='fl'>TRUE</span>,
  <span class='kw'>per_cell</span> <span class='kw'>=</span> <span class='fl'>FALSE</span>,
  <span class='kw'>topn</span> <span class='kw'>=</span> <span class='fl'>800</span>,
  <span class='kw'>cut</span> <span class='kw'>=</span> <span class='fl'>0</span>,
  <span class='kw'>genome_n</span> <span class='kw'>=</span> <span class='fl'>30000</span>,
  <span class='kw'>metric</span> <span class='kw'>=</span> <span class='st'>"hyper"</span>,
  <span class='kw'>output_high</span> <span class='kw'>=</span> <span class='fl'>TRUE</span>,
  <span class='kw'>lookuptable</span> <span class='kw'>=</span> <span class='kw'>NULL</span>,
  <span class='kw'>obj_out</span> <span class='kw'>=</span> <span class='fl'>FALSE</span>,
  <span class='kw'>seurat_out</span> <span class='kw'>=</span> <span class='fl'>FALSE</span>,
  <span class='kw'>rename_prefix</span> <span class='kw'>=</span> <span class='kw'>NULL</span>,
  <span class='kw'>threshold</span> <span class='kw'>=</span> <span class='fl'>0</span>,
  <span class='kw'>low_threshold_cell</span> <span class='kw'>=</span> <span class='fl'>10</span>,
  <span class='no'>...</span>
)

<span class='co'># S3 method for seurat</span>
<span class='fu'>clustify_lists</span>(
  <span class='no'>input</span>,
  <span class='kw'>metadata</span> <span class='kw'>=</span> <span class='kw'>NULL</span>,
  <span class='kw'>cluster_col</span> <span class='kw'>=</span> <span class='kw'>NULL</span>,
  <span class='kw'>if_log</span> <span class='kw'>=</span> <span class='fl'>TRUE</span>,
  <span class='kw'>per_cell</span> <span class='kw'>=</span> <span class='fl'>FALSE</span>,
  <span class='kw'>topn</span> <span class='kw'>=</span> <span class='fl'>800</span>,
  <span class='kw'>cut</span> <span class='kw'>=</span> <span class='fl'>0</span>,
  <span class='no'>marker</span>,
  <span class='kw'>marker_inmatrix</span> <span class='kw'>=</span> <span class='fl'>TRUE</span>,
  <span class='kw'>genome_n</span> <span class='kw'>=</span> <span class='fl'>30000</span>,
  <span class='kw'>metric</span> <span class='kw'>=</span> <span class='st'>"hyper"</span>,
  <span class='kw'>output_high</span> <span class='kw'>=</span> <span class='fl'>TRUE</span>,
  <span class='kw'>dr</span> <span class='kw'>=</span> <span class='st'>"umap"</span>,
  <span class='kw'>seurat_out</span> <span class='kw'>=</span> <span class='fl'>TRUE</span>,
  <span class='kw'>obj_out</span> <span class='kw'>=</span> <span class='fl'>FALSE</span>,
  <span class='kw'>threshold</span> <span class='kw'>=</span> <span class='fl'>0</span>,
  <span class='kw'>rename_prefix</span> <span class='kw'>=</span> <span class='kw'>NULL</span>,
  <span class='no'>...</span>
)

<span class='co'># S3 method for Seurat</span>
<span class='fu'>clustify_lists</span>(
  <span class='no'>input</span>,
  <span class='kw'>metadata</span> <span class='kw'>=</span> <span class='kw'>NULL</span>,
  <span class='kw'>cluster_col</span> <span class='kw'>=</span> <span class='kw'>NULL</span>,
  <span class='kw'>if_log</span> <span class='kw'>=</span> <span class='fl'>TRUE</span>,
  <span class='kw'>per_cell</span> <span class='kw'>=</span> <span class='fl'>FALSE</span>,
  <span class='kw'>topn</span> <span class='kw'>=</span> <span class='fl'>800</span>,
  <span class='kw'>cut</span> <span class='kw'>=</span> <span class='fl'>0</span>,
  <span class='no'>marker</span>,
  <span class='kw'>marker_inmatrix</span> <span class='kw'>=</span> <span class='fl'>TRUE</span>,
  <span class='kw'>genome_n</span> <span class='kw'>=</span> <span class='fl'>30000</span>,
  <span class='kw'>metric</span> <span class='kw'>=</span> <span class='st'>"hyper"</span>,
  <span class='kw'>output_high</span> <span class='kw'>=</span> <span class='fl'>TRUE</span>,
  <span class='kw'>dr</span> <span class='kw'>=</span> <span class='st'>"umap"</span>,
  <span class='kw'>seurat_out</span> <span class='kw'>=</span> <span class='fl'>TRUE</span>,
  <span class='kw'>obj_out</span> <span class='kw'>=</span> <span class='fl'>FALSE</span>,
  <span class='kw'>threshold</span> <span class='kw'>=</span> <span class='fl'>0</span>,
  <span class='kw'>rename_prefix</span> <span class='kw'>=</span> <span class='kw'>NULL</span>,
  <span class='no'>...</span>
)

<span class='co'># S3 method for SingleCellExperiment</span>
<span class='fu'>clustify_lists</span>(
  <span class='no'>input</span>,
  <span class='kw'>metadata</span> <span class='kw'>=</span> <span class='kw'>NULL</span>,
  <span class='kw'>cluster_col</span> <span class='kw'>=</span> <span class='kw'>NULL</span>,
  <span class='kw'>if_log</span> <span class='kw'>=</span> <span class='fl'>TRUE</span>,
  <span class='kw'>per_cell</span> <span class='kw'>=</span> <span class='fl'>FALSE</span>,
  <span class='kw'>topn</span> <span class='kw'>=</span> <span class='fl'>800</span>,
  <span class='kw'>cut</span> <span class='kw'>=</span> <span class='fl'>0</span>,
  <span class='no'>marker</span>,
  <span class='kw'>marker_inmatrix</span> <span class='kw'>=</span> <span class='fl'>TRUE</span>,
  <span class='kw'>genome_n</span> <span class='kw'>=</span> <span class='fl'>30000</span>,
  <span class='kw'>metric</span> <span class='kw'>=</span> <span class='st'>"hyper"</span>,
  <span class='kw'>output_high</span> <span class='kw'>=</span> <span class='fl'>TRUE</span>,
  <span class='kw'>dr</span> <span class='kw'>=</span> <span class='st'>"umap"</span>,
  <span class='kw'>seurat_out</span> <span class='kw'>=</span> <span class='fl'>TRUE</span>,
  <span class='kw'>obj_out</span> <span class='kw'>=</span> <span class='fl'>TRUE</span>,
  <span class='kw'>threshold</span> <span class='kw'>=</span> <span class='fl'>0</span>,
  <span class='kw'>rename_prefix</span> <span class='kw'>=</span> <span class='kw'>NULL</span>,
  <span class='no'>...</span>
)</pre>

    <h2 class="hasAnchor" id="arguments"><a class="anchor" href="#arguments"></a>Arguments</h2>
    <table class="ref-arguments">
    <colgroup><col class="name" /><col class="desc" /></colgroup>
    <tr>
      <th>input</th>
      <td><p>single-cell expression matrix or Seurat object</p></td>
    </tr>
    <tr>
      <th>...</th>
      <td><p>passed to matrixize_markers</p></td>
    </tr>
    <tr>
      <th>marker</th>
      <td><p>matrix or dataframe of candidate genes for each cluster</p></td>
    </tr>
    <tr>
      <th>marker_inmatrix</th>
      <td><p>whether markers genes are already in preprocessed matrix form</p></td>
    </tr>
    <tr>
      <th>metadata</th>
      <td><p>cell cluster assignments, supplied as a vector or data.frame.
If data.frame is supplied then <code>cluster_col</code> needs to be set. Not required
if running correlation per cell.</p></td>
    </tr>
    <tr>
      <th>cluster_col</th>
      <td><p>column in metadata with cluster number</p></td>
    </tr>
    <tr>
      <th>if_log</th>
      <td><p>input data is natural log, averaging will be done on unlogged data</p></td>
    </tr>
    <tr>
      <th>per_cell</th>
      <td><p>compare per cell or per cluster</p></td>
    </tr>
    <tr>
      <th>topn</th>
      <td><p>number of top expressing genes to keep from input matrix</p></td>
    </tr>
    <tr>
      <th>cut</th>
      <td><p>expression cut off from input matrix</p></td>
    </tr>
    <tr>
      <th>genome_n</th>
      <td><p>number of genes in the genome</p></td>
    </tr>
    <tr>
      <th>metric</th>
      <td><p>adjusted p-value for hypergeometric test, or jaccard index</p></td>
    </tr>
    <tr>
      <th>output_high</th>
      <td><p>if true (by default to fit with rest of package),
-log10 transform p-value</p></td>
    </tr>
    <tr>
      <th>lookuptable</th>
      <td><p>if not supplied, will look in built-in table for object parsing</p></td>
    </tr>
    <tr>
      <th>obj_out</th>
      <td><p>whether to output object instead of cor matrix</p></td>
    </tr>
    <tr>
      <th>seurat_out</th>
      <td><p>output cor matrix or called seurat object</p></td>
    </tr>
    <tr>
      <th>rename_prefix</th>
      <td><p>prefix to add to type and r column names</p></td>
    </tr>
    <tr>
      <th>threshold</th>
      <td><p>identity calling minimum correlation score threshold, only used when obj_out = T</p></td>
    </tr>
    <tr>
      <th>low_threshold_cell</th>
      <td><p>option to remove clusters with too few cells</p></td>
    </tr>
    <tr>
      <th>dr</th>
      <td><p>stored dimension reduction</p></td>
    </tr>
    </table>

    <h2 class="hasAnchor" id="value"><a class="anchor" href="#value"></a>Value</h2>

    <p>matrix of numeric values, clusters from input as row names, cell types from marker_mat as column names</p>

    <h2 class="hasAnchor" id="examples"><a class="anchor" href="#examples"></a>Examples</h2>
    <pre class="examples"><div class='input'><span class='co'># Annotate a matrix and metadata</span>
<span class='fu'>clustify_lists</span>(
  <span class='kw'>input</span> <span class='kw'>=</span> <span class='no'>pbmc_matrix_small</span>,
  <span class='kw'>marker</span> <span class='kw'>=</span> <span class='no'>cbmc_m</span>,
  <span class='kw'>metadata</span> <span class='kw'>=</span> <span class='no'>pbmc_meta</span>,
  <span class='kw'>cluster_col</span> <span class='kw'>=</span> <span class='st'>"classified"</span>,
  <span class='kw'>verbose</span> <span class='kw'>=</span> <span class='fl'>TRUE</span>
)</div><div class='output co'>#&gt;                   CD4 T      CD8 T Memory CD4 T CD14+ Mono Naive CD4 T
#&gt; Naive CD4 T  0.06713589 0.06713589            0   3.609737           0
#&gt; Memory CD4 T 0.06713589 0.00000000            0   3.609737           0
#&gt; CD14+ Mono   0.00000000 0.00000000            0   3.609737           0
#&gt; B            0.00000000 0.00000000            0   3.609737           0
#&gt; CD8 T        0.06713589 0.00000000            0   1.600060           0
#&gt; FCGR3A+ Mono 0.00000000 0.00000000            0   3.609737           0
#&gt; NK           0.00000000 0.06713589            0   3.609737           0
#&gt; DC           0.00000000 0.00000000            0   3.609737           0
#&gt; Platelet     0.16506077 0.00000000            0   3.783946           0
#&gt;                      NK         B CD16+ Mono CD34+ Eryth       Mk        DC
#&gt; Naive CD4 T  3.60973732 0.0000000 0.00000000     0     0 0.000000 0.0000000
#&gt; Memory CD4 T 3.60973732 0.0000000 0.00000000     0     0 0.000000 0.0000000
#&gt; CD14+ Mono   1.60005988 0.0000000 0.00000000     0     0 0.000000 0.0000000
#&gt; B            1.60005988 1.6000599 0.00000000     0     0 0.000000 0.0000000
#&gt; CD8 T        3.60973732 0.0000000 0.00000000     0     0 0.000000 0.0000000
#&gt; FCGR3A+ Mono 0.02934733 0.0000000 0.02934733     0     0 0.000000 0.0000000
#&gt; NK           3.60973732 0.0000000 0.00000000     0     0 0.000000 0.0000000
#&gt; DC           1.60005988 0.1085286 0.00000000     0     0 0.000000 0.1085286
#&gt; Platelet     1.75292851 0.1650608 0.00000000     0     0 3.783946 0.0000000
#&gt;                 pDCs
#&gt; Naive CD4 T  0.00000
#&gt; Memory CD4 T 0.00000
#&gt; CD14+ Mono   0.00000
#&gt; B            0.00000
#&gt; CD8 T        0.00000
#&gt; FCGR3A+ Mono 0.00000
#&gt; NK           0.00000
#&gt; DC           1.60006
#&gt; Platelet     0.00000</div><div class='input'>
<span class='co'># Annotate using a different method</span>
<span class='fu'>clustify_lists</span>(
  <span class='kw'>input</span> <span class='kw'>=</span> <span class='no'>pbmc_matrix_small</span>,
  <span class='kw'>marker</span> <span class='kw'>=</span> <span class='no'>cbmc_m</span>,
  <span class='kw'>metadata</span> <span class='kw'>=</span> <span class='no'>pbmc_meta</span>,
  <span class='kw'>cluster_col</span> <span class='kw'>=</span> <span class='st'>"classified"</span>,
  <span class='kw'>verbose</span> <span class='kw'>=</span> <span class='fl'>TRUE</span>,
  <span class='kw'>metric</span> <span class='kw'>=</span> <span class='st'>"jaccard"</span>
)</div><div class='output co'>#&gt;                    CD4 T       CD8 T Memory CD4 T  CD14+ Mono Naive CD4 T
#&gt; Naive CD4 T  0.001246883 0.001246883            0 0.003750000           0
#&gt; Memory CD4 T 0.001246883 0.000000000            0 0.003750000           0
#&gt; CD14+ Mono   0.000000000 0.000000000            0 0.003750000           0
#&gt; B            0.000000000 0.000000000            0 0.003750000           0
#&gt; CD8 T        0.001246883 0.000000000            0 0.002496879           0
#&gt; FCGR3A+ Mono 0.000000000 0.000000000            0 0.003750000           0
#&gt; NK           0.000000000 0.001246883            0 0.003750000           0
#&gt; DC           0.000000000 0.000000000            0 0.003750000           0
#&gt; Platelet     0.001424501 0.000000000            0 0.004285714           0
#&gt;                       NK           B  CD16+ Mono CD34+ Eryth          Mk
#&gt; Naive CD4 T  0.003750000 0.000000000 0.000000000     0     0 0.000000000
#&gt; Memory CD4 T 0.003750000 0.000000000 0.000000000     0     0 0.000000000
#&gt; CD14+ Mono   0.002496879 0.000000000 0.000000000     0     0 0.000000000
#&gt; B            0.002496879 0.002496879 0.000000000     0     0 0.000000000
#&gt; CD8 T        0.003750000 0.000000000 0.000000000     0     0 0.000000000
#&gt; FCGR3A+ Mono 0.001246883 0.000000000 0.001246883     0     0 0.000000000
#&gt; NK           0.003750000 0.000000000 0.000000000     0     0 0.000000000
#&gt; DC           0.002496879 0.001246883 0.000000000     0     0 0.000000000
#&gt; Platelet     0.002853067 0.001424501 0.000000000     0     0 0.004285714
#&gt;                       DC        pDCs
#&gt; Naive CD4 T  0.000000000 0.000000000
#&gt; Memory CD4 T 0.000000000 0.000000000
#&gt; CD14+ Mono   0.000000000 0.000000000
#&gt; B            0.000000000 0.000000000
#&gt; CD8 T        0.000000000 0.000000000
#&gt; FCGR3A+ Mono 0.000000000 0.000000000
#&gt; NK           0.000000000 0.000000000
#&gt; DC           0.001246883 0.002496879
#&gt; Platelet     0.000000000 0.000000000</div></pre>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">
    <h2>Contents</h2>
    <ul class="nav nav-pills nav-stacked">
      <li><a href="#arguments">Arguments</a></li>
      <li><a href="#value">Value</a></li>
      <li><a href="#examples">Examples</a></li>
    </ul>

  </div>
</div>


      <footer>
      <div class="copyright">
  <p>Developed by Rui Fu, Kent Riemondy, Austin Gillen, Chengzhe Tian, Jay Hesselberth.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.4.1.</p>
</div>

      </footer>
   </div>

  


  </body>
</html>


