<!-- Generated by pkgdown: do not edit by hand -->
<!DOCTYPE html>
<html lang="en">
  <head>
  <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<title>Combined function to compare scRNA-seq data to bulk RNA-seq data and marker list — clustify_nudge • clustifyr</title>

<!-- favicons -->
<link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png">
<link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png">
<link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png" />
<link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png" />
<link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png" />
<link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png" />

<!-- jquery -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.3.1/jquery.min.js" integrity="sha256-FgpCb/KJQlLNfOu91ta32o/NMZxltwRo8QtmkMRdAu8=" crossorigin="anonymous"></script>
<!-- Bootstrap -->

<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/css/bootstrap.min.css" integrity="sha256-916EbMg70RQy9LHiGkXzG8hSg9EdNy97GazNG/aiY1w=" crossorigin="anonymous" />

<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha256-U5ZEeKfGNOja007MMD3YBI0A3OSZOQbeG6z2f2Y0hu8=" crossorigin="anonymous"></script>

<!-- Font Awesome icons -->
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.7.1/css/all.min.css" integrity="sha256-nAmazAk6vS34Xqo0BSrTb+abbtFlgsFK7NKSi6o7Y78=" crossorigin="anonymous" />
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.7.1/css/v4-shims.min.css" integrity="sha256-6qHlizsOWFskGlwVOKuns+D1nB6ssZrHQrNj1wGplHc=" crossorigin="anonymous" />

<!-- clipboard.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.4/clipboard.min.js" integrity="sha256-FiZwavyI2V6+EXO1U+xzLG3IKldpiTFf3153ea9zikQ=" crossorigin="anonymous"></script>

<!-- headroom.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.9.4/headroom.min.js" integrity="sha256-DJFC1kqIhelURkuza0AvYal5RxMtpzLjFhsnVIeuk+U=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.9.4/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script>

<!-- pkgdown -->
<link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script>




<meta property="og:title" content="Combined function to compare scRNA-seq data to bulk RNA-seq data and marker list — clustify_nudge" />
<meta property="og:description" content="Combined function to compare scRNA-seq data to bulk RNA-seq data and marker list" />
<meta property="og:image" content="http://rnabioco.github.io/clustifyr/logo.png" />
<meta name="twitter:card" content="summary" />




<!-- mathjax -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script>

<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->



  </head>

  <body>
    <div class="container template-reference-topic">
      <header>
      <div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">clustifyr</a>
        <span class="version label label-danger" data-toggle="tooltip" data-placement="bottom" title="Unreleased version">0.99.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="../index.html">
    <span class="fas fa fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="../articles/clustifyR.html">Introduction to clustifyr</a>
    </li>
  </ul>
</li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/rnabioco/clustifyr">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
      
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header>

<div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Combined function to compare scRNA-seq data to bulk RNA-seq data and marker list</h1>
    <small class="dont-index">Source: <a href='https://github.com/rnabioco/clustifyr/blob/master/R/utils.R'><code>R/utils.R</code></a></small>
    <div class="hidden name"><code>clustify_nudge.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>Combined function to compare scRNA-seq data to bulk RNA-seq data and marker list</p>
    </div>

    <pre class="usage"><span class='fu'>clustify_nudge</span>(<span class='no'>input</span>, <span class='no'>...</span>)

<span class='co'># S3 method for default</span>
<span class='fu'>clustify_nudge</span>(
  <span class='no'>input</span>,
  <span class='no'>ref_mat</span>,
  <span class='no'>marker</span>,
  <span class='kw'>metadata</span> <span class='kw'>=</span> <span class='kw'>NULL</span>,
  <span class='kw'>cluster_col</span> <span class='kw'>=</span> <span class='kw'>NULL</span>,
  <span class='kw'>query_genes</span> <span class='kw'>=</span> <span class='kw'>NULL</span>,
  <span class='kw'>compute_method</span> <span class='kw'>=</span> <span class='st'>"spearman"</span>,
  <span class='kw'>weight</span> <span class='kw'>=</span> <span class='fl'>1</span>,
  <span class='kw'>seurat_out</span> <span class='kw'>=</span> <span class='fl'>FALSE</span>,
  <span class='kw'>threshold</span> <span class='kw'>=</span> -<span class='fl'>Inf</span>,
  <span class='kw'>dr</span> <span class='kw'>=</span> <span class='st'>"umap"</span>,
  <span class='kw'>norm</span> <span class='kw'>=</span> <span class='st'>"diff"</span>,
  <span class='kw'>call</span> <span class='kw'>=</span> <span class='fl'>TRUE</span>,
  <span class='kw'>marker_inmatrix</span> <span class='kw'>=</span> <span class='fl'>TRUE</span>,
  <span class='kw'>mode</span> <span class='kw'>=</span> <span class='st'>"rank"</span>,
  <span class='kw'>obj_out</span> <span class='kw'>=</span> <span class='fl'>FALSE</span>,
  <span class='kw'>rename_prefix</span> <span class='kw'>=</span> <span class='kw'>NULL</span>,
  <span class='kw'>lookuptable</span> <span class='kw'>=</span> <span class='kw'>NULL</span>,
  <span class='no'>...</span>
)

<span class='co'># S3 method for seurat</span>
<span class='fu'>clustify_nudge</span>(
  <span class='no'>input</span>,
  <span class='no'>ref_mat</span>,
  <span class='no'>marker</span>,
  <span class='kw'>cluster_col</span> <span class='kw'>=</span> <span class='kw'>NULL</span>,
  <span class='kw'>query_genes</span> <span class='kw'>=</span> <span class='kw'>NULL</span>,
  <span class='kw'>compute_method</span> <span class='kw'>=</span> <span class='st'>"spearman"</span>,
  <span class='kw'>weight</span> <span class='kw'>=</span> <span class='fl'>1</span>,
  <span class='kw'>seurat_out</span> <span class='kw'>=</span> <span class='fl'>TRUE</span>,
  <span class='kw'>obj_out</span> <span class='kw'>=</span> <span class='fl'>FALSE</span>,
  <span class='kw'>threshold</span> <span class='kw'>=</span> -<span class='fl'>Inf</span>,
  <span class='kw'>dr</span> <span class='kw'>=</span> <span class='st'>"umap"</span>,
  <span class='kw'>norm</span> <span class='kw'>=</span> <span class='st'>"diff"</span>,
  <span class='kw'>marker_inmatrix</span> <span class='kw'>=</span> <span class='fl'>TRUE</span>,
  <span class='kw'>mode</span> <span class='kw'>=</span> <span class='st'>"rank"</span>,
  <span class='kw'>rename_prefix</span> <span class='kw'>=</span> <span class='kw'>NULL</span>,
  <span class='no'>...</span>
)

<span class='co'># S3 method for Seurat</span>
<span class='fu'>clustify_nudge</span>(
  <span class='no'>input</span>,
  <span class='no'>ref_mat</span>,
  <span class='no'>marker</span>,
  <span class='kw'>cluster_col</span> <span class='kw'>=</span> <span class='kw'>NULL</span>,
  <span class='kw'>query_genes</span> <span class='kw'>=</span> <span class='kw'>NULL</span>,
  <span class='kw'>compute_method</span> <span class='kw'>=</span> <span class='st'>"spearman"</span>,
  <span class='kw'>weight</span> <span class='kw'>=</span> <span class='fl'>1</span>,
  <span class='kw'>seurat_out</span> <span class='kw'>=</span> <span class='fl'>TRUE</span>,
  <span class='kw'>obj_out</span> <span class='kw'>=</span> <span class='fl'>FALSE</span>,
  <span class='kw'>threshold</span> <span class='kw'>=</span> -<span class='fl'>Inf</span>,
  <span class='kw'>dr</span> <span class='kw'>=</span> <span class='st'>"umap"</span>,
  <span class='kw'>norm</span> <span class='kw'>=</span> <span class='st'>"diff"</span>,
  <span class='kw'>marker_inmatrix</span> <span class='kw'>=</span> <span class='fl'>TRUE</span>,
  <span class='kw'>mode</span> <span class='kw'>=</span> <span class='st'>"rank"</span>,
  <span class='kw'>rename_prefix</span> <span class='kw'>=</span> <span class='kw'>NULL</span>,
  <span class='no'>...</span>
)</pre>

    <h2 class="hasAnchor" id="arguments"><a class="anchor" href="#arguments"></a>Arguments</h2>
    <table class="ref-arguments">
    <colgroup><col class="name" /><col class="desc" /></colgroup>
    <tr>
      <th>input</th>
      <td><p>express matrix or object</p></td>
    </tr>
    <tr>
      <th>...</th>
      <td><p>passed to matrixize_markers</p></td>
    </tr>
    <tr>
      <th>ref_mat</th>
      <td><p>reference expression matrix</p></td>
    </tr>
    <tr>
      <th>marker</th>
      <td><p>matrix of markers</p></td>
    </tr>
    <tr>
      <th>metadata</th>
      <td><p>cell cluster assignments, supplied as a vector or data.frame. If
data.frame is supplied then <code>cluster_col</code> needs to be set.</p></td>
    </tr>
    <tr>
      <th>cluster_col</th>
      <td><p>column in metadata that contains cluster ids per cell. Will default to first
column of metadata if not supplied. Not required if running correlation per cell.</p></td>
    </tr>
    <tr>
      <th>query_genes</th>
      <td><p>A vector of genes of interest to compare. If NULL, then common genes between
the expr_mat and ref_mat will be used for comparision.</p></td>
    </tr>
    <tr>
      <th>compute_method</th>
      <td><p>method(s) for computing similarity scores</p></td>
    </tr>
    <tr>
      <th>weight</th>
      <td><p>relative weight for the gene list scores, when added to correlation score</p></td>
    </tr>
    <tr>
      <th>seurat_out</th>
      <td><p>output cor matrix or called seurat object</p></td>
    </tr>
    <tr>
      <th>threshold</th>
      <td><p>identity calling minimum score threshold, only used when obj_out = T</p></td>
    </tr>
    <tr>
      <th>dr</th>
      <td><p>stored dimension reduction</p></td>
    </tr>
    <tr>
      <th>norm</th>
      <td><p>whether and how the results are normalized</p></td>
    </tr>
    <tr>
      <th>call</th>
      <td><p>make call or just return score matrix</p></td>
    </tr>
    <tr>
      <th>marker_inmatrix</th>
      <td><p>whether markers genes are already in preprocessed matrix form</p></td>
    </tr>
    <tr>
      <th>mode</th>
      <td><p>use marker expression pct or ranked cor score for nudging</p></td>
    </tr>
    <tr>
      <th>obj_out</th>
      <td><p>whether to output object instead of cor matrix</p></td>
    </tr>
    <tr>
      <th>rename_prefix</th>
      <td><p>prefix to add to type and r column names</p></td>
    </tr>
    <tr>
      <th>lookuptable</th>
      <td><p>if not supplied, will look in built-in table for object parsing</p></td>
    </tr>
    </table>

    <h2 class="hasAnchor" id="value"><a class="anchor" href="#value"></a>Value</h2>

    <p>single cell object, or matrix of numeric values, clusters from input as row names, cell types from ref_mat as column names</p>

    <h2 class="hasAnchor" id="examples"><a class="anchor" href="#examples"></a>Examples</h2>
    <pre class="examples"><div class='input'><span class='co'># Seurat2</span>
<span class='fu'>clustify_nudge</span>(
  <span class='kw'>input</span> <span class='kw'>=</span> <span class='no'>s_small</span>,
  <span class='kw'>ref_mat</span> <span class='kw'>=</span> <span class='no'>cbmc_ref</span>,
  <span class='kw'>marker</span> <span class='kw'>=</span> <span class='no'>cbmc_m</span>,
  <span class='kw'>cluster_col</span> <span class='kw'>=</span> <span class='st'>"res.1"</span>,
  <span class='kw'>threshold</span> <span class='kw'>=</span> <span class='fl'>0.8</span>,
  <span class='kw'>seurat_out</span> <span class='kw'>=</span> <span class='fl'>FALSE</span>,
  <span class='kw'>mode</span> <span class='kw'>=</span> <span class='st'>"pct"</span>,
  <span class='kw'>dr</span> <span class='kw'>=</span> <span class='st'>"tsne"</span>
)</div><div class='output co'>#&gt;              B   CD14+ Mono CD16+ Mono      CD34+       CD4 T      CD8 T
#&gt; 0  0.522410963 -0.005572036  0.4119037  0.4172004  0.61062598  0.4431346
#&gt; 1  0.396432770  0.725146450  0.6939952  0.4641469  0.42401600  0.1737119
#&gt; 2  0.007020943  0.211394209  0.6094641  0.3178432  0.46805261  0.4812869
#&gt; 3 -0.400393645 -0.243071872  0.1859978 -0.0854630 -0.08048411 -0.3731234
#&gt;            DC     Eryth Memory CD4 T         Mk Naive CD4 T         NK
#&gt; 0  0.36476066 0.1523530    0.7047302 -0.9967949   0.6398803 -0.2617596
#&gt; 1  0.71146653 0.4359801    0.5349202 -0.7964748   0.4304222 -0.1248677
#&gt; 2  0.67077257 0.4137920    0.6331816 -0.7115560   0.5864320  0.8099383
#&gt; 3 -0.05333518 0.4352437    0.1083785  0.7453407  -0.1097800 -0.8598142
#&gt;        pDCs
#&gt; 0 0.3456669
#&gt; 1 0.5116362
#&gt; 2 0.2859966
#&gt; 3 0.1901284</div><div class='input'>
<span class='co'># Seurat3</span>
 <span class='fu'>clustify_nudge</span>(
  <span class='kw'>input</span> <span class='kw'>=</span> <span class='no'>s_small3</span>,
  <span class='kw'>ref_mat</span> <span class='kw'>=</span> <span class='no'>cbmc_ref</span>,
  <span class='kw'>marker</span> <span class='kw'>=</span> <span class='no'>cbmc_m</span>,
  <span class='kw'>cluster_col</span> <span class='kw'>=</span> <span class='st'>"RNA_snn_res.1"</span>,
  <span class='kw'>threshold</span> <span class='kw'>=</span> <span class='fl'>0.8</span>,
  <span class='kw'>seurat_out</span> <span class='kw'>=</span> <span class='fl'>FALSE</span>,
  <span class='kw'>mode</span> <span class='kw'>=</span> <span class='st'>"pct"</span>,
  <span class='kw'>dr</span> <span class='kw'>=</span> <span class='st'>"tsne"</span>
)</div><div class='output co'>#&gt;            B CD14+ Mono CD16+ Mono      CD34+     CD4 T     CD8 T         DC
#&gt; 0 -0.8386005 -0.8022027 -0.1290444 -0.1811809 0.3775856 0.4804364 -0.1773445
#&gt; 1  0.1116745  0.7647078  0.7930478  0.3569654 0.4509427 0.2469878  0.6437572
#&gt; 2  0.6997539 -0.2133902  0.5339625  0.3773183 0.2712622 0.1166450  0.6264979
#&gt;        Eryth Memory CD4 T         Mk Naive CD4 T        NK      pDCs
#&gt; 0  0.5092140    0.2683011  0.5851824  0.03142658 0.1767582 0.1596988
#&gt; 1  0.2455553    0.5380182 -0.1502676  0.52633608 0.2681751 0.4804654
#&gt; 2 -0.1592523    0.1233831 -0.5265590  0.31191120 0.1667222 0.4075066</div><div class='input'>
<span class='co'># Matrix</span>
<span class='fu'>clustify_nudge</span>(
  <span class='kw'>input</span> <span class='kw'>=</span> <span class='no'>pbmc_matrix_small</span>,
  <span class='kw'>ref_mat</span> <span class='kw'>=</span> <span class='no'>cbmc_ref</span>,
  <span class='kw'>metadata</span> <span class='kw'>=</span> <span class='no'>pbmc_meta</span>,
  <span class='kw'>marker</span> <span class='kw'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/matrix.html'>as.matrix</a></span>(<span class='no'>cbmc_m</span>),
  <span class='kw'>query_genes</span> <span class='kw'>=</span> <span class='no'>pbmc_vargenes</span>,
  <span class='kw'>cluster_col</span> <span class='kw'>=</span> <span class='st'>"classified"</span>,
  <span class='kw'>threshold</span> <span class='kw'>=</span> <span class='fl'>0.8</span>,
  <span class='kw'>call</span> <span class='kw'>=</span> <span class='fl'>FALSE</span>,
  <span class='kw'>marker_inmatrix</span> <span class='kw'>=</span> <span class='fl'>FALSE</span>,
  <span class='kw'>mode</span> <span class='kw'>=</span> <span class='st'>"pct"</span>
)</div><div class='output co'>#&gt;                        B  CD14+ Mono CD16+ Mono     CD34+      CD4 T      CD8 T
#&gt; B             0.89129128 -0.09482820 0.36430264 0.7317217 0.53933823 0.57819315
#&gt; CD14+ Mono   -0.14739603  0.90814717 0.51856965 0.6710434 0.41250211 0.46484177
#&gt; CD8 T        -0.11323147 -0.11229570 0.34225996 0.7136651 0.72384293 0.66175171
#&gt; DC           -0.09890316  0.43792394 0.44916939 0.6814873 0.48842105 0.45893484
#&gt; FCGR3A+ Mono -0.12660135  0.64538033 0.90590926 0.6726128 0.41690109 0.54635896
#&gt; Memory CD4 T -0.06044487 -0.02870830 0.36003221 0.7427243 0.89489774 0.69324349
#&gt; Naive CD4 T  -0.06874533 -0.07764912 0.35150020 0.7548765 0.82685564 0.74596560
#&gt; NK           -0.13741949 -0.12259625 0.33838687 0.7067672 0.59374826 0.79378254
#&gt; Platelet     -0.50747063 -0.25075312 0.02147571 0.3271915 0.08603668 0.08622349
#&gt;                      DC     Eryth Memory CD4 T         Mk Naive CD4 T
#&gt; B            0.47539346 0.6300171    0.5959394 -0.3538824   0.5953671
#&gt; CD14+ Mono   0.56732763 0.6055132    0.5708809 -0.2338858   0.5384300
#&gt; CD8 T        0.43400786 0.6525508    0.6542797 -0.3276053   0.6121911
#&gt; DC           0.83270918 0.6049950    0.5657932 -0.2778113   0.5260392
#&gt; FCGR3A+ Mono 0.54084520 0.6097144    0.5710313 -0.2341562   0.5300822
#&gt; Memory CD4 T 0.44862102 0.6643080    0.6860640 -0.3457256   0.6554955
#&gt; Naive CD4 T  0.44731357 0.6677468    0.6913118 -0.3406683   0.6622109
#&gt; NK           0.42414982 0.6420863    0.6259194 -0.3341079   0.5880876
#&gt; Platelet     0.09763352 0.3599792    0.2945712  0.6304742   0.2899873
#&gt;                        NK      pDCs
#&gt; B            -0.154620036 0.5866840
#&gt; CD14+ Mono   -0.096549233 0.5377967
#&gt; CD8 T         0.598665387 0.5300752
#&gt; DC           -0.011303596 0.7012226
#&gt; FCGR3A+ Mono -0.039743952 0.5332615
#&gt; Memory CD4 T -0.002563329 0.5406439
#&gt; Naive CD4 T  -0.069020452 0.5415440
#&gt; NK            0.896029075 0.5264843
#&gt; Platelet     -0.534417903 0.2308274</div></pre>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">
    <h2>Contents</h2>
    <ul class="nav nav-pills nav-stacked">
      <li><a href="#arguments">Arguments</a></li>
      <li><a href="#value">Value</a></li>
      <li><a href="#examples">Examples</a></li>
    </ul>

  </div>
</div>


      <footer>
      <div class="copyright">
  <p>Developed by Rui Fu, Kent Riemondy, Austin Gillen, Chengzhe Tian, Jay Hesselberth.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.4.1.</p>
</div>

      </footer>
   </div>

  


  </body>
</html>


