<!-- Generated by pkgdown: do not edit by hand -->
<!DOCTYPE html>
<html lang="en">
  <head>
  <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<title>Function reference â€¢ clustifyr</title>

<!-- favicons -->
<link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png">
<link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png">
<link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png" />
<link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png" />
<link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png" />
<link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png" />
<!-- jquery -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.3.1/jquery.min.js" integrity="sha256-FgpCb/KJQlLNfOu91ta32o/NMZxltwRo8QtmkMRdAu8=" crossorigin="anonymous"></script>
<!-- Bootstrap -->

<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/css/bootstrap.min.css" integrity="sha256-916EbMg70RQy9LHiGkXzG8hSg9EdNy97GazNG/aiY1w=" crossorigin="anonymous" />
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha256-U5ZEeKfGNOja007MMD3YBI0A3OSZOQbeG6z2f2Y0hu8=" crossorigin="anonymous"></script>

<!-- Font Awesome icons -->
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.7.1/css/all.min.css" integrity="sha256-nAmazAk6vS34Xqo0BSrTb+abbtFlgsFK7NKSi6o7Y78=" crossorigin="anonymous" />
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.7.1/css/v4-shims.min.css" integrity="sha256-6qHlizsOWFskGlwVOKuns+D1nB6ssZrHQrNj1wGplHc=" crossorigin="anonymous" />

<!-- clipboard.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.4/clipboard.min.js" integrity="sha256-FiZwavyI2V6+EXO1U+xzLG3IKldpiTFf3153ea9zikQ=" crossorigin="anonymous"></script>

<!-- headroom.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.9.4/headroom.min.js" integrity="sha256-DJFC1kqIhelURkuza0AvYal5RxMtpzLjFhsnVIeuk+U=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.9.4/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script>

<!-- pkgdown -->
<link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script>



<meta property="og:title" content="Function reference" />

<meta property="og:image" content="http://rnabioco.github.io/clustifyr/logo.png" />
<meta name="twitter:card" content="summary" />



<!-- mathjax -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script>

<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->



  </head>

  <body>
    <div class="container template-reference-index">
      <header>
      <div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">clustifyr</a>
        <span class="version label label-danger" data-toggle="tooltip" data-placement="bottom" title="Unreleased version">0.0.0.9000</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="../index.html">
    <span class="fas fa fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="../articles/benchmark.html">Benchmarking clustifyr</a>
    </li>
    <li>
      <a href="../articles/clustifyR.html">Introduction to clustifyr</a>
    </li>
    <li>
      <a href="../articles/otherformats.html">Support for various single cell object formats</a>
    </li>
    <li>
      <a href="../articles/othertools.html">Other considerations w/ clustifyr</a>
    </li>
    <li>
      <a href="../articles/references.html">Reference building for clustifyr</a>
    </li>
    <li>
      <a href="../articles/visualizations.html">Visualization of clustifyr results</a>
    </li>
  </ul>
</li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul>
      
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/rnabioco/clustifyr">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
      
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header>

<div class="row">
  <div class="contents col-md-9">
    <div class="page-header">
      <h1>Reference</h1>
    </div>

    <table class="ref-index">

    <colgroup>
      
      <col class="alias" />
      <col class="title" />
    </colgroup>

    <tbody>
      <tr>
        <th colspan="2">
          <h2 id="section-all-functions" class="hasAnchor"><a href="#section-all-functions" class="anchor"></a>All functions</h2>
          <p class="section-desc"></p>
        </th>
      </tr>
      <tr>
        
        <td>
          <p><code><a href="RowVar.html">RowVar()</a></code> </p>
        </td>
        <td><p>get var per row for matrix</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="assign_ident.html">assign_ident()</a></code> </p>
        </td>
        <td><p>manually change idents as needed</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="average_clusters.html">average_clusters()</a></code> </p>
        </td>
        <td><p>#' Average expression values per cluster</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="average_clusters_filter.html">average_clusters_filter()</a></code> </p>
        </td>
        <td><p>Average expression values per cluster, filtered by set parameter, defaults to calculating background</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="binarize_expr.html">binarize_expr()</a></code> </p>
        </td>
        <td><p>Binarize scRNAseq data</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="calc_similarity.html">calc_similarity()</a></code> </p>
        </td>
        <td><p>compute similarity</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="calculate_pathway_gsea.html">calculate_pathway_gsea()</a></code> </p>
        </td>
        <td><p>Convert expression matrix to GSEA pathway scores (would take a similar place in workflow before average_clusters/binarize)</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="call_to_metadata.html">call_to_metadata()</a></code> </p>
        </td>
        <td><p>Insert called ident results into metadata</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="cbmc_m.html">cbmc_m</a></code> </p>
        </td>
        <td><p>reference marker matrix from seurat citeseq CBMC tutorial</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="cbmc_ref.html">cbmc_ref</a></code> </p>
        </td>
        <td><p>reference matrix from seurat citeseq CBMC tutorial</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="clustify.html">clustify()</a></code> </p>
        </td>
        <td><p>Compare scRNA-seq data to bulk RNA-seq data.</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="clustify_intra.html">clustify_intra()</a></code> </p>
        </td>
        <td><p>Intra-experiment cluster projection for one sample/set to the rest</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="clustify_lists.html">clustify_lists()</a></code> </p>
        </td>
        <td><p>Main function to compare scRNA-seq data to gene lists.</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="clustify_nudge.html">clustify_nudge()</a></code> </p>
        </td>
        <td><p>Combined function to compare scRNA-seq data to bulk RNA-seq data and marker list</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="clustifyr_methods.html">clustifyr_methods</a></code> </p>
        </td>
        <td><p>Correlation functions available in clustifyr</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="collapse_to_cluster.html">collapse_to_cluster()</a></code> </p>
        </td>
        <td><p>From per-cell calls, take highest freq call in each cluster</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="compare_lists.html">compare_lists()</a></code> </p>
        </td>
        <td><p>Calculate adjusted p-values for hypergeometric test of gene lists
or jaccard index</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="compute_mean_expr.html">compute_mean_expr()</a></code> </p>
        </td>
        <td><p>compute mean of clusters</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="cor_to_call.html">cor_to_call()</a></code> </p>
        </td>
        <td><p>get best calls for each cluster</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="cor_to_call_topn.html">cor_to_call_topn()</a></code> </p>
        </td>
        <td><p>get top calls for each cluster</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="cosine.html">cosine()</a></code> </p>
        </td>
        <td><p>Cosine distance</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="downsample_matrix.html">downsample_matrix()</a></code> </p>
        </td>
        <td><p>downsample matrix by cluster or completely random</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="feature_select_PCA.html">feature_select_PCA()</a></code> </p>
        </td>
        <td><p>Returns a list of variable genes based on PCA</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="file_marker_parse.html">file_marker_parse()</a></code> </p>
        </td>
        <td><p>takes files with positive and negative markers, as described in garnett, and returns list of markers</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="gene_pct.html">gene_pct()</a></code> </p>
        </td>
        <td><p>pct of cells in each cluster that express genelist</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="gene_pct_markerm.html">gene_pct_markerm()</a></code> </p>
        </td>
        <td><p>pct of cells in every cluster that express a series of genelists</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="get_best_match_matrix.html">get_best_match_matrix()</a></code> </p>
        </td>
        <td><p>Function to make best call from correlation matrix</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="get_best_str.html">get_best_str()</a></code> </p>
        </td>
        <td><p>Function to make call and attach score</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="get_common_elements.html">get_common_elements()</a></code> </p>
        </td>
        <td><p>Find entries shared in all vectors</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="get_similarity.html">get_similarity()</a></code> </p>
        </td>
        <td><p>Compute similarity of matrices</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="get_vargenes.html">get_vargenes()</a></code> </p>
        </td>
        <td><p>Generate variable gene list from marker matrix</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="gmt_to_list.html">gmt_to_list()</a></code> </p>
        </td>
        <td><p>convert gmt format of pathways to list of vectors</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="insert_meta_object.html">insert_meta_object()</a></code> </p>
        </td>
        <td><p>more flexible metadata update of single cell objects</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="kl_divergence.html">kl_divergence()</a></code> </p>
        </td>
        <td><p>KL divergence</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="make_comb_ref.html">make_comb_ref()</a></code> </p>
        </td>
        <td><p>make combination ref matrix to assess intermixing</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="marker_select.html">marker_select()</a></code> </p>
        </td>
        <td><p>decide for one gene whether it is a marker for a certain cell type</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="matrixize_markers.html">matrixize_markers()</a></code> </p>
        </td>
        <td><p>Convert candidate genes list into matrix</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="not_pretty_palette.html">not_pretty_palette</a></code> </p>
        </td>
        <td><p>black and white palette for plotting continous variables</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="object_loc_lookup.html">object_loc_lookup</a></code> </p>
        </td>
        <td><p>lookup table for single cell object structures</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="object_ref.html">object_ref()</a></code> </p>
        </td>
        <td><p>Function to convert labelled object to avg expression matrix</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="overcluster_test.html">overcluster_test()</a></code> </p>
        </td>
        <td><p>compare clustering parameters and classification outcomes</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="parse_loc_object.html">parse_loc_object()</a></code> </p>
        </td>
        <td><p>more flexible parsing of single cell objects</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="pbmc_bulk_matrix.html">pbmc_bulk_matrix</a></code> </p>
        </td>
        <td><p>Bulk RNA-Seq data from sorted populations isolated from PBMCs.</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="pbmc_markers.html">pbmc_markers</a></code> </p>
        </td>
        <td><p>Marker genes identified by Seurat from single-cell RNA-seq PBMCs.</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="pbmc_markers_M3Drop.html">pbmc_markers_M3Drop</a></code> </p>
        </td>
        <td><p>Marker genes identified by M3Drop from single-cell RNA-seq PBMCs.</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="pbmc_matrix_small.html">pbmc_matrix_small</a></code> </p>
        </td>
        <td><p>Matrix of single-cell RNA-seq PBMCs.</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="pbmc_meta.html">pbmc_meta</a></code> </p>
        </td>
        <td><p>Meta-data for single-cell RNA-seq PBMCs.</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="pbmc_vargenes.html">pbmc_vargenes</a></code> </p>
        </td>
        <td><p>Variable genes identified by Seurat from single-cell RNA-seq PBMCs.</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="percent_clusters.html">percent_clusters()</a></code> </p>
        </td>
        <td><p>Percentage detected per cluster</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="permute_similarity.html">permute_similarity()</a></code> </p>
        </td>
        <td><p>Compute a p-value for similarity using permutation</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="plot_best_call.html">plot_best_call()</a></code> </p>
        </td>
        <td><p>Plot best calls for each cluster on a tSNE or umap</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="plot_call.html">plot_call()</a></code> </p>
        </td>
        <td><p>Plot called clusters on a tSNE or umap, for each reference cluster given</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="plot_cols.html">plot_cols()</a></code> </p>
        </td>
        <td><p>Plot variable median per cluster from reference metadata vs new assigned metadata, for visually evaluating classification</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="plot_cor.html">plot_cor()</a></code> </p>
        </td>
        <td><p>Plot similarity measures on a tSNE or umap</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="plot_cor_heatmap.html">plot_cor_heatmap()</a></code> </p>
        </td>
        <td><p>Plot similarity measures on heatmap</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="plot_gene.html">plot_gene()</a></code> </p>
        </td>
        <td><p>Plot gene expression on to tSNE or umap</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="plot_pathway_gsea.html">plot_pathway_gsea()</a></code> </p>
        </td>
        <td><p>plot GSEA pathway scores as heatmap, returns a list containing results and plot.</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="plot_tsne.html">plot_tsne()</a></code> </p>
        </td>
        <td><p>Plot a tSNE or umap colored by feature.</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="pos_neg_select.html">pos_neg_select()</a></code> </p>
        </td>
        <td><p>adapt clustify to tweak score for pos and neg markers</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="pretty_palette.html">pretty_palette</a></code> </p>
        </td>
        <td><p>Color palette for plotting continous variables</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="pretty_palette2.html">pretty_palette2</a></code> </p>
        </td>
        <td><p>Color palette for plotting continous variables, starting at gray</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="pretty_palette_ramp_d.html">pretty_palette_ramp_d()</a></code> </p>
        </td>
        <td><p>Expanded color palette ramp for plotting discrete variables</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="ref_feature_select.html">ref_feature_select()</a></code> </p>
        </td>
        <td><p>feature select from reference matrix</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="ref_marker_select.html">ref_marker_select()</a></code> </p>
        </td>
        <td><p>marker selection from reference matrix</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="remove_background.html">remove_background()</a></code> </p>
        </td>
        <td><p>Remove high background expression genes from matrix</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="reverse_marker_matrix.html">reverse_marker_matrix()</a></code> </p>
        </td>
        <td><p>generate negative markers from a list of exclusive positive markers</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="run_gsea.html">run_gsea()</a></code> </p>
        </td>
        <td><p>Run GSEA to compare a gene list(s) to per cell or per cluster expression data</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="s_small.html">s_small</a></code> </p>
        </td>
        <td><p>Small clustered Seurat2 object</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="s_small3.html">s_small3</a></code> </p>
        </td>
        <td><p>Small clustered Seurat3 object</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="seurat_meta.html">seurat_meta()</a></code> </p>
        </td>
        <td><p>Function to convert labelled seurat object to fully prepared metadata</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="seurat_ref.html">seurat_ref()</a></code> </p>
        </td>
        <td><p>Function to convert labelled seurat object to avg expression matrix</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="vector_similarity.html">vector_similarity()</a></code> </p>
        </td>
        <td><p>Compute similarity between two vectors</p></td>
      </tr>
    </tbody>
    </table>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">
    <h2>Contents</h2>
    <ul class="nav nav-pills nav-stacked">
      <li><a href="#section-all-functions">All functions</a></li>
    </ul>
  </div>
</div>


      <footer>
      <div class="copyright">
  <p>Developed by Rui Fu, Kent Riemondy, Austin Gillen, Chengzhe Tian, Jay Hesselberth.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.4.0.</p>
</div>

      </footer>
   </div>

  


  </body>
</html>


