<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Introduction to clustifyr • clustifyr</title>
<!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png">
<link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png">
<link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png">
<link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png">
<link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png">
<link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png">
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Introduction to clustifyr">
<meta property="og:description" content="clustifyr">
<meta property="og:image" content="http://rnabioco.github.io/clustifyr/logo.svg">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">clustifyr</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">1.5.1</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../index.html">
    <span class="fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/clustifyR.html">Introduction to clustifyr</a>
    </li>
    <li>
      <a href="../articles/geo-annotations.html">Improving NCBI GEO submissions of scRNA-seq data</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/rnabioco/clustifyr/">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><script src="clustifyR_files/accessible-code-block-0.0.1/empty-anchor.js"></script><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Introduction to clustifyr</h1>
                        <h4 class="author">Rui Fu</h4>
            <address class="author_afil">
      RNA Bioscience Initative, University of Colorado School of Medicine<br><h4 class="author">Austin Gillen</h4>
            <address class="author_afil">
      RNA Bioscience Initative, University of Colorado School of Medicine<br><h4 class="author">Ryan Sheridan</h4>
            <address class="author_afil">
      RNA Bioscience Initative, University of Colorado School of Medicine<br><h4 class="author">Chengzhe Tian</h4>
            <address class="author_afil">
      Department of Biochemistry, University of Colorado Boulder<br><h4 class="author">Michelle Daya</h4>
            <address class="author_afil">
      Biomedical Informatics &amp; Personalized Medicine, University of Colorado Anschutz Medical Campus<br><h4 class="author">Yue Hao</h4>
            <address class="author_afil">
      Bioinformatics Research Center, North Carolina State University<br><h4 class="author">Jay Hesselberth</h4>
            <address class="author_afil">
      RNA Bioscience Initative, University of Colorado School of Medicine<br><h4 class="author">Kent Riemondy</h4>
            <address class="author_afil">
      RNA Bioscience Initative, University of Colorado School of Medicine<br><h4 class="date">2021-10-04</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/rnabioco/clustifyr/blob/master/vignettes/clustifyR.Rmd"><code>vignettes/clustifyR.Rmd</code></a></small>
      <div class="hidden name"><code>clustifyR.Rmd</code></div>

    </address>
</address>
</address>
</address>
</address>
</address>
</address>
</address>
</div>

    
    
<div id="introduction-why-use-clustifyr" class="section level1">
<h1 class="hasAnchor">
<a href="#introduction-why-use-clustifyr" class="anchor"></a>Introduction: Why use <code>clustifyr</code>?</h1>
<p>Single cell transcriptomes are difficult to annotate without extensive knowledge of the underlying biology of the system in question. Even with this knowledge, accurate identification can be challenging due to the lack of detectable expression of common marker genes defined by bulk RNA-seq, flow cytometry, other single cell RNA-seq platforms, etc.</p>
<p><code>clustifyr</code> solves this problem by providing functions to automatically annotate single cells or clusters using bulk RNA-seq data or marker gene lists (ranked or unranked). Additional functions allow for exploratory analysis of calculated similarities between single cell RNA-seq datasets and reference data.</p>
</div>
<div id="installation" class="section level1">
<h1 class="hasAnchor">
<a href="#installation" class="anchor"></a>Installation</h1>
<p>To install <code>clustifyr</code> BiocManager must be installed.</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html">install.packages</a></span><span class="op">(</span><span class="st">"BiocManager"</span><span class="op">)</span>

<span class="fu">BiocManager</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/BiocManager/man/install.html">install</a></span><span class="op">(</span><span class="st">"clustifyr"</span><span class="op">)</span></code></pre></div>
</div>
<div id="a-simple-example-10x-genomics-pbmcs" class="section level1">
<h1 class="hasAnchor">
<a href="#a-simple-example-10x-genomics-pbmcs" class="anchor"></a>A simple example: 10x Genomics PBMCs</h1>
<p>In this example, we take a 10x Genomics 3’ scRNA-seq dataset from peripheral blood mononuclear cells (PBMCs) and annotate the cell clusters (identified using <code>Seurat</code>) using scRNA-seq cell clusters assigned from a CITE-seq experiment.</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="http://github.com/rnabioco/clustifyr#readme">clustifyr</a></span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://ggplot2.tidyverse.org">ggplot2</a></span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://wilkelab.org/cowplot/">cowplot</a></span><span class="op">)</span>

<span class="co"># Matrix of normalized single-cell RNA-seq counts</span>
<span class="va">pbmc_matrix</span> <span class="op">&lt;-</span> <span class="fu">clustifyr</span><span class="fu">::</span><span class="va"><a href="../reference/pbmc_matrix_small.html">pbmc_matrix_small</a></span>

<span class="co"># meta.data table containing cluster assignments for each cell</span>
<span class="co"># The table that we are using also contains the known cell identities in the "classified" column</span>
<span class="va">pbmc_meta</span> <span class="op">&lt;-</span> <span class="fu">clustifyr</span><span class="fu">::</span><span class="va"><a href="../reference/pbmc_meta.html">pbmc_meta</a></span></code></pre></div>
</div>
<div id="calculate-correlation-coefficients" class="section level1">
<h1 class="hasAnchor">
<a href="#calculate-correlation-coefficients" class="anchor"></a>Calculate correlation coefficients</h1>
<p>To identify cell types, the <code>clustifyr()</code> function requires several inputs:</p>
<ul>
<li>
<code>input</code>: an SingleCellExperiment or Seurat object or a matrix of normalized single-cell RNA-seq counts</li>
<li>
<code>metadata</code>: a meta.data table containing the cluster assignments for each cell (not required if a Seurat object is given)</li>
<li>
<code>ref_mat</code>: a reference matrix containing RNA-seq expression data for each cell type of interest</li>
<li>
<code>query_genes</code>: a list of genes to use for comparison (optional but recommended)</li>
</ul>
<p>When using a matrix of scRNA-seq counts <code>clustifyr()</code> will return a matrix of correlation coefficients for each cell type and cluster, with the rownames corresponding to the cluster number.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># Calculate correlation coefficients for each cluster (spearman by default)</span>
<span class="va">vargenes</span> <span class="op">&lt;-</span> <span class="va">pbmc_vargenes</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">500</span><span class="op">]</span>

<span class="va">res</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/clustify.html">clustify</a></span><span class="op">(</span>
  input <span class="op">=</span> <span class="va">pbmc_matrix</span>, <span class="co"># matrix of normalized scRNA-seq counts (or SCE/Seurat object)</span>
  metadata <span class="op">=</span> <span class="va">pbmc_meta</span>, <span class="co"># meta.data table containing cell clusters</span>
  cluster_col <span class="op">=</span> <span class="st">"seurat_clusters"</span>, <span class="co"># name of column in meta.data containing cell clusters</span>
  ref_mat <span class="op">=</span> <span class="va">cbmc_ref</span>, <span class="co"># matrix of RNA-seq expression data for each cell type</span>
  query_genes <span class="op">=</span> <span class="va">vargenes</span> <span class="co"># list of highly varible genes identified with Seurat</span>
<span class="op">)</span>

<span class="co"># Peek at correlation matrix</span>
<span class="va">res</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">5</span>, <span class="fl">1</span><span class="op">:</span><span class="fl">5</span><span class="op">]</span>
<span class="co">#&gt;           B CD14+ Mono CD16+ Mono     CD34+     CD4 T</span>
<span class="co">#&gt; 0 0.6563466  0.6454029  0.6485863 0.7089861 0.8804508</span>
<span class="co">#&gt; 1 0.6394363  0.6388404  0.6569401 0.7027430 0.8488750</span>
<span class="co">#&gt; 2 0.5524081  0.9372089  0.8930158 0.5879264 0.5347312</span>
<span class="co">#&gt; 3 0.8945380  0.5801453  0.6146857 0.6955897 0.6566739</span>
<span class="co">#&gt; 4 0.5711643  0.5623870  0.5826233 0.6280913 0.7467347</span>

<span class="co"># Call cell types</span>
<span class="va">res2</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/cor_to_call.html">cor_to_call</a></span><span class="op">(</span>
  cor_mat <span class="op">=</span> <span class="va">res</span>,                  <span class="co"># matrix correlation coefficients</span>
  cluster_col <span class="op">=</span> <span class="st">"seurat_clusters"</span> <span class="co"># name of column in meta.data containing cell clusters</span>
<span class="op">)</span>
<span class="va">res2</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">5</span>, <span class="op">]</span>
<span class="co">#&gt; # A tibble: 5 × 3</span>
<span class="co">#&gt; # Groups:   seurat_clusters [5]</span>
<span class="co">#&gt;   seurat_clusters type           r</span>
<span class="co">#&gt;   &lt;chr&gt;           &lt;chr&gt;      &lt;dbl&gt;</span>
<span class="co">#&gt; 1 3               B          0.895</span>
<span class="co">#&gt; 2 2               CD14+ Mono 0.937</span>
<span class="co">#&gt; 3 5               CD16+ Mono 0.929</span>
<span class="co">#&gt; 4 0               CD4 T      0.880</span>
<span class="co">#&gt; 5 1               CD4 T      0.849</span>

<span class="co"># Insert into original metadata as "type" column</span>
<span class="va">pbmc_meta2</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/call_to_metadata.html">call_to_metadata</a></span><span class="op">(</span>
  res <span class="op">=</span> <span class="va">res2</span>,                     <span class="co"># data.frame of called cell type for each cluster</span>
  metadata <span class="op">=</span> <span class="va">pbmc_meta</span>,           <span class="co"># original meta.data table containing cell clusters</span>
  cluster_col <span class="op">=</span> <span class="st">"seurat_clusters"</span> <span class="co"># name of column in meta.data containing cell clusters</span>
<span class="op">)</span></code></pre></div>
<p>To visualize the <code>clustifyr()</code> results we can use the <code><a href="../reference/plot_cor_heatmap.html">plot_cor_heatmap()</a></code> function to plot the correlation coefficients for each cluster and each cell type.</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># Create heatmap of correlation coefficients using clustifyr() output</span>
<span class="fu"><a href="../reference/plot_cor_heatmap.html">plot_cor_heatmap</a></span><span class="op">(</span>cor_mat <span class="op">=</span> <span class="va">res</span><span class="op">)</span></code></pre></div>
<p><img src="clustifyR_files/figure-html/Create%20correlation%20heatmap-1.png" width="672" style="display: block; margin: auto;"></p>
</div>
<div id="plot-cluster-identities-and-correlation-coefficients" class="section level1">
<h1 class="hasAnchor">
<a href="#plot-cluster-identities-and-correlation-coefficients" class="anchor"></a>Plot cluster identities and correlation coefficients</h1>
<p><code>clustifyr</code> also provides functions to overlay correlation coefficients on pre-calculated tSNE embeddings (or those from any other dimensionality reduction method).</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># Overlay correlation coefficients on UMAPs for the first two cell types</span>
<span class="va">corr_umaps</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/plot_cor.html">plot_cor</a></span><span class="op">(</span>
  cor_mat <span class="op">=</span> <span class="va">res</span>,                     <span class="co"># matrix of correlation coefficients from clustifyr()</span>
  metadata <span class="op">=</span> <span class="va">pbmc_meta</span>,              <span class="co"># meta.data table containing UMAP or tSNE data</span>
  data_to_plot <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html">colnames</a></span><span class="op">(</span><span class="va">res</span><span class="op">)</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">2</span><span class="op">]</span>, <span class="co"># name of cell type(s) to plot correlation coefficients</span>
  cluster_col <span class="op">=</span> <span class="st">"seurat_clusters"</span>    <span class="co"># name of column in meta.data containing cell clusters</span>
<span class="op">)</span>

<span class="fu"><a href="https://wilkelab.org/cowplot/reference/plot_grid.html">plot_grid</a></span><span class="op">(</span>
  plotlist <span class="op">=</span> <span class="va">corr_umaps</span>,
  rel_widths <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">0.47</span>, <span class="fl">0.53</span><span class="op">)</span>
<span class="op">)</span></code></pre></div>
<p><img src="clustifyR_files/figure-html/Overlay%20corr%20coefficients%20on%20UMAP-1.png" width="864" style="display: block; margin: auto;"></p>
<p><br></p>
<p>The <code><a href="../reference/plot_best_call.html">plot_best_call()</a></code> function can be used to label each cluster with the cell type that gives the highest corelation coefficient. Using the <code><a href="../reference/plot_dims.html">plot_dims()</a></code> function, we can also plot the known identities of each cluster, which were stored in the “classified” column of the meta.data table. The plots below show that the highest correlations between the reference RNA-seq data and the 10x Genomics scRNA-seq dataset are restricted to the correct cell clusters.</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># Label clusters with clustifyr cell identities</span>
<span class="va">clustifyr_types</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/plot_best_call.html">plot_best_call</a></span><span class="op">(</span>
  cor_mat <span class="op">=</span> <span class="va">res</span>,          <span class="co"># matrix of correlation coefficients from clustifyr()</span>
  metadata <span class="op">=</span> <span class="va">pbmc_meta</span>,   <span class="co"># meta.data table containing UMAP or tSNE data</span>
  do_label <span class="op">=</span> <span class="cn">TRUE</span>,        <span class="co"># should the feature label be shown on each cluster?</span>
  do_legend <span class="op">=</span> <span class="cn">FALSE</span>,      <span class="co"># should the legend be shown?</span>
  cluster_col <span class="op">=</span> <span class="st">"seurat_clusters"</span>
<span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html">ggtitle</a></span><span class="op">(</span><span class="st">"clustifyr cell types"</span><span class="op">)</span>

<span class="co"># Compare clustifyr results with known cell identities</span>
<span class="va">known_types</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/plot_dims.html">plot_dims</a></span><span class="op">(</span>
  data <span class="op">=</span> <span class="va">pbmc_meta</span>,       <span class="co"># meta.data table containing UMAP or tSNE data</span>
  feature <span class="op">=</span> <span class="st">"classified"</span>, <span class="co"># name of column in meta.data to color clusters by</span>
  do_label <span class="op">=</span> <span class="cn">TRUE</span>,        <span class="co"># should the feature label be shown on each cluster?</span>
  do_legend <span class="op">=</span> <span class="cn">FALSE</span>       <span class="co"># should the legend be shown?</span>
<span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html">ggtitle</a></span><span class="op">(</span><span class="st">"Known cell types"</span><span class="op">)</span>

<span class="fu"><a href="https://wilkelab.org/cowplot/reference/plot_grid.html">plot_grid</a></span><span class="op">(</span><span class="va">known_types</span>, <span class="va">clustifyr_types</span><span class="op">)</span></code></pre></div>
<p><img src="clustifyR_files/figure-html/Label%20clusters-1.png" width="1152" style="display: block; margin: auto;"></p>
</div>
<div id="classify-cells-using-known-marker-genes" class="section level1">
<h1 class="hasAnchor">
<a href="#classify-cells-using-known-marker-genes" class="anchor"></a>Classify cells using known marker genes</h1>
<p>The <code><a href="../reference/clustify_lists.html">clustify_lists()</a></code> function allows cell types to be assigned based on known marker genes. This function requires a table containing markers for each cell type of interest. Cell types can be assigned using several statistical tests including, hypergeometric, Jaccard, Spearman, and GSEA.</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># Take a peek at marker gene table</span>
<span class="va">cbmc_m</span>
<span class="co">#&gt;   CD4 T CD8 T Memory CD4 T CD14+ Mono Naive CD4 T   NK     B CD16+ Mono</span>
<span class="co">#&gt; 1 ITM2A  CD8B        ADCY2     S100A8       CDHR3 GNLY  IGHM     CDKN1C</span>
<span class="co">#&gt; 2 TXNIP  CD8A       PTGDR2     S100A9  DICER1-AS1 NKG7 CD79A       HES4</span>
<span class="co">#&gt; 3   AES S100B      CD200R1        LYZ       RAD9A CST7 MS4A1        CKB</span>
<span class="co">#&gt;      CD34+ Eryth    Mk    DC   pDCs</span>
<span class="co">#&gt; 1   SPINK2   HBM   PF4  ENHO LILRA4</span>
<span class="co">#&gt; 2  C1QTNF4  AHSP  SDPR  CD1E   TPM2</span>
<span class="co">#&gt; 3 KIAA0125   CA1 TUBB1 NDRG2    SCT</span>

<span class="co"># Available metrics include: "hyper", "jaccard", "spearman", "gsea"</span>
<span class="va">list_res</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/clustify_lists.html">clustify_lists</a></span><span class="op">(</span>
  input <span class="op">=</span> <span class="va">pbmc_matrix</span>,             <span class="co"># matrix of normalized single-cell RNA-seq counts</span>
  metadata <span class="op">=</span> <span class="va">pbmc_meta</span>,            <span class="co"># meta.data table containing cell clusters</span>
  cluster_col <span class="op">=</span> <span class="st">"seurat_clusters"</span>, <span class="co"># name of column in meta.data containing cell clusters</span>
  marker <span class="op">=</span> <span class="va">cbmc_m</span>,                 <span class="co"># list of known marker genes</span>
  metric <span class="op">=</span> <span class="st">"pct"</span>                   <span class="co"># test to use for assigning cell types</span>
<span class="op">)</span>

<span class="co"># View as heatmap, or plot_best_call</span>
<span class="fu"><a href="../reference/plot_cor_heatmap.html">plot_cor_heatmap</a></span><span class="op">(</span>
  cor_mat <span class="op">=</span> <span class="va">list_res</span>,              <span class="co"># matrix of correlation coefficients from clustify_lists()</span>
  cluster_rows <span class="op">=</span> <span class="cn">FALSE</span>,            <span class="co"># cluster by row?</span>
  cluster_columns <span class="op">=</span> <span class="cn">FALSE</span>,         <span class="co"># cluster by column?</span>
  legend_title <span class="op">=</span> <span class="st">"% expressed"</span>     <span class="co"># title of heatmap legend</span>
<span class="op">)</span></code></pre></div>
<p><img src="clustifyR_files/figure-html/clustifyr%20with%20gene%20lists-1.png" width="576" style="display: block; margin: auto;"></p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R">
<span class="co"># Downstream functions same as clustify()</span>
<span class="co"># Call cell types</span>
<span class="va">list_res2</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/cor_to_call.html">cor_to_call</a></span><span class="op">(</span>
  cor_mat <span class="op">=</span> <span class="va">list_res</span>,              <span class="co"># matrix correlation coefficients</span>
  cluster_col <span class="op">=</span> <span class="st">"seurat_clusters"</span>  <span class="co"># name of column in meta.data containing cell clusters</span>
<span class="op">)</span>

<span class="co"># Insert into original metadata as "list_type" column</span>
<span class="va">pbmc_meta3</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/call_to_metadata.html">call_to_metadata</a></span><span class="op">(</span>
  res <span class="op">=</span> <span class="va">list_res2</span>,                 <span class="co"># data.frame of called cell type for each cluster</span>
  metadata <span class="op">=</span> <span class="va">pbmc_meta</span>,            <span class="co"># original meta.data table containing cell clusters</span>
  cluster_col <span class="op">=</span> <span class="st">"seurat_clusters"</span>, <span class="co"># name of column in meta.data containing cell clusters</span>
  rename_prefix <span class="op">=</span> <span class="st">"list_"</span>          <span class="co"># set a prefix for the new column</span>
<span class="op">)</span></code></pre></div>
</div>
<div id="direct-handling-of-singlecellexperiment-objects" class="section level1">
<h1 class="hasAnchor">
<a href="#direct-handling-of-singlecellexperiment-objects" class="anchor"></a>Direct handling of <code>SingleCellExperiment</code> objects</h1>
<p><code>clustifyr</code> can also use a <code>SingleCellExperiment</code> object as input and return a new <code>SingleCellExperiment</code> object with the cell types added as a column in the colData.</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">res</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/clustify.html">clustify</a></span><span class="op">(</span>
  input <span class="op">=</span> <span class="va">sce_small</span>,          <span class="co"># an SCE object</span>
  ref_mat <span class="op">=</span> <span class="va">cbmc_ref</span>,         <span class="co"># matrix of RNA-seq expression data for each cell type</span>
  cluster_col <span class="op">=</span> <span class="st">"cell_type1"</span>, <span class="co"># name of column in meta.data containing cell clusters</span>
  obj_out <span class="op">=</span> <span class="cn">TRUE</span>              <span class="co"># output SCE object with cell type inserted as "type" column</span>
<span class="op">)</span>

<span class="fu">SingleCellExperiment</span><span class="fu">::</span><span class="fu">colData</span><span class="op">(</span><span class="va">res</span><span class="op">)</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">10</span>,<span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"type"</span>, <span class="st">"r"</span><span class="op">)</span><span class="op">]</span>
<span class="co">#&gt; DataFrame with 10 rows and 2 columns</span>
<span class="co">#&gt;               type                 r</span>
<span class="co">#&gt;        &lt;character&gt;         &lt;numeric&gt;</span>
<span class="co">#&gt; AZ_A1        CD34+ 0.557678024919381</span>
<span class="co">#&gt; AZ_A10       CD34+ 0.624777701661225</span>
<span class="co">#&gt; AZ_A11       CD4 T 0.695067885340303</span>
<span class="co">#&gt; AZ_A12       CD34+ 0.624777701661225</span>
<span class="co">#&gt; AZ_A2        CD4 T 0.602804908958642</span>
<span class="co">#&gt; AZ_A3        CD34+ 0.557678024919381</span>
<span class="co">#&gt; AZ_A4        CD34+ 0.557678024919381</span>
<span class="co">#&gt; AZ_A5        CD34+ 0.645378073051508</span>
<span class="co">#&gt; AZ_A6        CD4 T 0.695067885340303</span>
<span class="co">#&gt; AZ_A7        CD34+ 0.671644883893203</span></code></pre></div>
</div>
<div id="direct-handling-of-seurat-v2-and-v3-objects" class="section level1">
<h1 class="hasAnchor">
<a href="#direct-handling-of-seurat-v2-and-v3-objects" class="anchor"></a>Direct handling of <code>seurat</code> v2 and v3 objects</h1>
<p><code>clustifyr</code> can also use a <code>Seurat</code> object as input and return a new <code>Seurat</code> object with the cell types added as a column in the meta.data.</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">res</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/clustify.html">clustify</a></span><span class="op">(</span>
  input <span class="op">=</span> <span class="va">s_small3</span>,       <span class="co"># a Seurat object</span>
  ref_mat <span class="op">=</span> <span class="va">cbmc_ref</span>,    <span class="co"># matrix of RNA-seq expression data for each cell type</span>
  cluster_col <span class="op">=</span> <span class="st">"RNA_snn_res.1"</span>, <span class="co"># name of column in meta.data containing cell clusters</span>
  obj_out <span class="op">=</span> <span class="cn">TRUE</span>      <span class="co"># output Seurat object with cell type inserted as "type" column</span>
<span class="op">)</span>

<span class="va">res</span><span class="op">@</span><span class="va">meta.data</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">10</span>, <span class="op">]</span>
<span class="co">#&gt;                nGene nUMI    orig.ident res.0.8 res.1         type         r</span>
<span class="co">#&gt; ATGCCAGAACGACT    47   70 SeuratProject       0     0 Memory CD4 T 0.7047302</span>
<span class="co">#&gt; CATGGCCTGTGCAT    52   85 SeuratProject       0     0 Memory CD4 T 0.7047302</span>
<span class="co">#&gt; GAACCTGATGAACC    50   87 SeuratProject       0     0 Memory CD4 T 0.7047302</span>
<span class="co">#&gt; TGACTGGATTCTCA    56  127 SeuratProject       0     0 Memory CD4 T 0.7047302</span>
<span class="co">#&gt; AGTCAGACTGCACA    53  173 SeuratProject       0     0 Memory CD4 T 0.7047302</span>
<span class="co">#&gt; TCTGATACACGTGT    48   70 SeuratProject       0     0 Memory CD4 T 0.7047302</span>
<span class="co">#&gt; TGGTATCTAAACAG    36   64 SeuratProject       0     0 Memory CD4 T 0.7047302</span>
<span class="co">#&gt; GCAGCTCTGTTTCT    45   72 SeuratProject       0     0 Memory CD4 T 0.7047302</span>
<span class="co">#&gt; GATATAACACGCAT    36   52 SeuratProject       0     0 Memory CD4 T 0.7047302</span>
<span class="co">#&gt; AATGTTGACAGTCA    41  100 SeuratProject       0     0 Memory CD4 T 0.7047302</span></code></pre></div>
</div>
<div id="building-reference-matrix-from-single-cell-expression-matrix" class="section level1">
<h1 class="hasAnchor">
<a href="#building-reference-matrix-from-single-cell-expression-matrix" class="anchor"></a>Building reference matrix from single cell expression matrix</h1>
<p>In its simplest form, a reference matrix is built by averaging expression (also includes an option to take the median) of a single cell RNA-seq expression matrix by cluster. Both log transformed or raw count matrices are supported.</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">new_ref_matrix</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/average_clusters.html">average_clusters</a></span><span class="op">(</span>
  mat <span class="op">=</span> <span class="va">pbmc_matrix</span>,
  metadata <span class="op">=</span> <span class="va">pbmc_meta</span><span class="op">$</span><span class="va">classified</span>, <span class="co"># or use metadata = pbmc_meta, cluster_col = "classified"</span>
  if_log <span class="op">=</span> <span class="cn">TRUE</span>                    <span class="co"># whether the expression matrix is already log transformed</span>
<span class="op">)</span>

<span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span><span class="op">(</span><span class="va">new_ref_matrix</span><span class="op">)</span>
<span class="co">#&gt;                 B CD14+ Mono      CD8 T         DC FCGR3A+ Mono Memory CD4 T</span>
<span class="co">#&gt; PPBP   0.09375021 0.28763857 0.35662599 0.06527347    0.2442300   0.06494743</span>
<span class="co">#&gt; LYZ    1.42699419 5.21550849 1.35146753 4.84714962    3.4034309   1.39466552</span>
<span class="co">#&gt; S100A9 0.62123058 4.91453355 0.58823794 2.53310734    2.6277996   0.58080250</span>
<span class="co">#&gt; IGLL5  2.44576997 0.02434753 0.03284986 0.10986617    0.2581198   0.04826212</span>
<span class="co">#&gt; GNLY   0.37877736 0.53592906 2.53161887 0.46959958    0.2903092   0.41001072</span>
<span class="co">#&gt; FTL    3.66698837 5.86217774 3.37056910 4.21848878    5.9518479   3.31062958</span>
<span class="co">#&gt;        Naive CD4 T         NK  Platelet</span>
<span class="co">#&gt; PPBP    0.04883837 0.00000000 6.0941782</span>
<span class="co">#&gt; LYZ     1.40165143 1.32701580 2.5303912</span>
<span class="co">#&gt; S100A9  0.55679700 0.52098541 1.6775692</span>
<span class="co">#&gt; IGLL5   0.03116080 0.05247669 0.2501642</span>
<span class="co">#&gt; GNLY    0.46041901 4.70481754 0.3845813</span>
<span class="co">#&gt; FTL     3.35611600 3.38471536 4.5508242</span>

<span class="co"># For further convenience, a shortcut function for generating reference matrix from `SingleCellExperiment` or `seurat` object is used.</span>
<span class="va">new_ref_matrix_sce</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/object_ref.html">object_ref</a></span><span class="op">(</span>
  input <span class="op">=</span> <span class="va">sce_small</span>,               <span class="co"># SCE object</span>
  cluster_col <span class="op">=</span> <span class="st">"cell_type1"</span>       <span class="co"># name of column in colData containing cell identities</span>
<span class="op">)</span>

<span class="va">new_ref_matrix_v3</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/seurat_ref.html">seurat_ref</a></span><span class="op">(</span>
  seurat_object <span class="op">=</span> <span class="va">s_small3</span>,        <span class="co"># SeuratV3 object</span>
  cluster_col <span class="op">=</span> <span class="st">"RNA_snn_res.1"</span>    <span class="co"># name of column in meta.data containing cell identities</span>
<span class="op">)</span>

<span class="fu"><a href="https://rdrr.io/r/utils/head.html">tail</a></span><span class="op">(</span><span class="va">new_ref_matrix_v3</span><span class="op">)</span>
<span class="co">#&gt;                 0         1        2</span>
<span class="co">#&gt; C12orf75 3.223784 0.4297757 2.305767</span>
<span class="co">#&gt; RARRES3  4.341699 1.7550597 3.794566</span>
<span class="co">#&gt; PCMT1    4.576310 2.8893691 2.223141</span>
<span class="co">#&gt; LAMP1    3.919129 0.9441022 1.565788</span>
<span class="co">#&gt; SPON2    3.670497 1.3182933 3.200796</span>
<span class="co">#&gt; S100B    3.069210 0.0000000 0.000000</span></code></pre></div>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># There's also the option to pull UCSC cell browser data.</span>
<span class="fu">get_ext_reference</span><span class="op">(</span>cb_url <span class="op">=</span> <span class="st">"http://cells.ucsc.edu/?ds=kidney-atlas%2FFetal_Immune"</span>,
                  cluster_col <span class="op">=</span> <span class="st">"celltype"</span><span class="op">)</span></code></pre></div>
<p>More reference data, including tabula muris, and code used to generate them are available at <a href="https://github.com/rnabioco/clustifyrdata" class="uri">https://github.com/rnabioco/clustifyrdata</a></p>
<p>Also see list for individual downloads at <a href="https://rnabioco.github.io/clustifyr/articles/download_refs.html" class="uri">https://rnabioco.github.io/clustifyr/articles/download_refs.html</a></p>
<p>Additional tutorials at <a href="https://rnabioco.github.io/clustifyrdata/articles/otherformats.html" class="uri">https://rnabioco.github.io/clustifyrdata/articles/otherformats.html</a></p>
</div>
<div id="session-info" class="section level1">
<h1 class="hasAnchor">
<a href="#session-info" class="anchor"></a>Session info</h1>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/utils/sessionInfo.html">sessionInfo</a></span><span class="op">(</span><span class="op">)</span>
<span class="co">#&gt; R version 4.1.1 (2021-08-10)</span>
<span class="co">#&gt; Platform: x86_64-pc-linux-gnu (64-bit)</span>
<span class="co">#&gt; Running under: Ubuntu 18.04.6 LTS</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; Matrix products: default</span>
<span class="co">#&gt; BLAS:   /usr/lib/x86_64-linux-gnu/openblas/libblas.so.3</span>
<span class="co">#&gt; LAPACK: /usr/lib/x86_64-linux-gnu/libopenblasp-r0.2.20.so</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; locale:</span>
<span class="co">#&gt;  [1] LC_CTYPE=C.UTF-8       LC_NUMERIC=C           LC_TIME=C.UTF-8       </span>
<span class="co">#&gt;  [4] LC_COLLATE=C.UTF-8     LC_MONETARY=C.UTF-8    LC_MESSAGES=C.UTF-8   </span>
<span class="co">#&gt;  [7] LC_PAPER=C.UTF-8       LC_NAME=C              LC_ADDRESS=C          </span>
<span class="co">#&gt; [10] LC_TELEPHONE=C         LC_MEASUREMENT=C.UTF-8 LC_IDENTIFICATION=C   </span>
<span class="co">#&gt; </span>
<span class="co">#&gt; attached base packages:</span>
<span class="co">#&gt; [1] stats     graphics  grDevices utils     datasets  methods   base     </span>
<span class="co">#&gt; </span>
<span class="co">#&gt; other attached packages:</span>
<span class="co">#&gt; [1] cowplot_1.1.1    ggplot2_3.3.5    clustifyr_1.5.1  BiocStyle_2.20.2</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; loaded via a namespace (and not attached):</span>
<span class="co">#&gt;  [1] bitops_1.0-7                matrixStats_0.61.0         </span>
<span class="co">#&gt;  [3] fs_1.5.0                    doParallel_1.0.16          </span>
<span class="co">#&gt;  [5] RColorBrewer_1.1-2          httr_1.4.2                 </span>
<span class="co">#&gt;  [7] rprojroot_2.0.2             GenomeInfoDb_1.28.4        </span>
<span class="co">#&gt;  [9] tools_4.1.1                 bslib_0.3.0                </span>
<span class="co">#&gt; [11] utf8_1.2.2                  R6_2.5.1                   </span>
<span class="co">#&gt; [13] BiocGenerics_0.38.0         colorspace_2.0-2           </span>
<span class="co">#&gt; [15] GetoptLong_1.0.5            withr_2.4.2                </span>
<span class="co">#&gt; [17] tidyselect_1.1.1            gridExtra_2.3              </span>
<span class="co">#&gt; [19] compiler_4.1.1              textshaping_0.3.5          </span>
<span class="co">#&gt; [21] cli_3.0.1                   Biobase_2.52.0             </span>
<span class="co">#&gt; [23] Cairo_1.5-12.2              desc_1.4.0                 </span>
<span class="co">#&gt; [25] DelayedArray_0.18.0         labeling_0.4.2             </span>
<span class="co">#&gt; [27] entropy_1.3.1               bookdown_0.24              </span>
<span class="co">#&gt; [29] sass_0.4.0                  scales_1.1.1               </span>
<span class="co">#&gt; [31] pkgdown_1.6.1               systemfonts_1.0.2          </span>
<span class="co">#&gt; [33] stringr_1.4.0               digest_0.6.28              </span>
<span class="co">#&gt; [35] rmarkdown_2.11              XVector_0.32.0             </span>
<span class="co">#&gt; [37] pkgconfig_2.0.3             htmltools_0.5.2            </span>
<span class="co">#&gt; [39] MatrixGenerics_1.4.3        highr_0.9                  </span>
<span class="co">#&gt; [41] fastmap_1.1.0               rlang_0.4.11               </span>
<span class="co">#&gt; [43] GlobalOptions_0.1.2         farver_2.1.0               </span>
<span class="co">#&gt; [45] shape_1.4.6                 jquerylib_0.1.4            </span>
<span class="co">#&gt; [47] generics_0.1.0              jsonlite_1.7.2             </span>
<span class="co">#&gt; [49] BiocParallel_1.26.2         dplyr_1.0.7                </span>
<span class="co">#&gt; [51] RCurl_1.98-1.5              magrittr_2.0.1             </span>
<span class="co">#&gt; [53] GenomeInfoDbData_1.2.6      Matrix_1.3-4               </span>
<span class="co">#&gt; [55] Rcpp_1.0.7                  munsell_0.5.0              </span>
<span class="co">#&gt; [57] S4Vectors_0.30.2            fansi_0.5.0                </span>
<span class="co">#&gt; [59] lifecycle_1.0.1             stringi_1.7.5              </span>
<span class="co">#&gt; [61] yaml_2.2.1                  SummarizedExperiment_1.22.0</span>
<span class="co">#&gt; [63] zlibbioc_1.38.0             grid_4.1.1                 </span>
<span class="co">#&gt; [65] ggrepel_0.9.1               parallel_4.1.1             </span>
<span class="co">#&gt; [67] crayon_1.4.1                lattice_0.20-45            </span>
<span class="co">#&gt; [69] circlize_0.4.13             knitr_1.36                 </span>
<span class="co">#&gt; [71] ComplexHeatmap_2.8.0        pillar_1.6.3               </span>
<span class="co">#&gt; [73] fgsea_1.18.0                GenomicRanges_1.44.0       </span>
<span class="co">#&gt; [75] rjson_0.2.20                codetools_0.2-18           </span>
<span class="co">#&gt; [77] stats4_4.1.1                fastmatch_1.1-3            </span>
<span class="co">#&gt; [79] glue_1.4.2                  evaluate_0.14              </span>
<span class="co">#&gt; [81] data.table_1.14.2           BiocManager_1.30.16        </span>
<span class="co">#&gt; [83] vctrs_0.3.8                 png_0.1-7                  </span>
<span class="co">#&gt; [85] foreach_1.5.1               gtable_0.3.0               </span>
<span class="co">#&gt; [87] purrr_0.3.4                 tidyr_1.1.4                </span>
<span class="co">#&gt; [89] clue_0.3-59                 cachem_1.0.6               </span>
<span class="co">#&gt; [91] xfun_0.26                   ragg_1.1.3                 </span>
<span class="co">#&gt; [93] SingleCellExperiment_1.14.1 tibble_3.1.5               </span>
<span class="co">#&gt; [95] iterators_1.0.13            memoise_2.0.0              </span>
<span class="co">#&gt; [97] IRanges_2.26.0              cluster_2.1.2              </span>
<span class="co">#&gt; [99] ellipsis_0.3.2</span></code></pre></div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p>Developed by Rui Fu, Kent Riemondy, <a href="https://medschool.cuanschutz.edu/rbi">RNA Bioscience Initiative</a>.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.6.1.</p>
</div>

      </footer>
</div>

  


  </body>
</html>
